<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="D G Rossiter">
<meta name="dcterms.date" content="2025-01-08">

<title>quarto-input1e20992c</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="PatternAnalysisWorkshopTutorial_DSM2025_files/libs/clipboard/clipboard.min.js"></script>
<script src="PatternAnalysisWorkshopTutorial_DSM2025_files/libs/quarto-html/quarto.js"></script>
<script src="PatternAnalysisWorkshopTutorial_DSM2025_files/libs/quarto-html/popper.min.js"></script>
<script src="PatternAnalysisWorkshopTutorial_DSM2025_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="PatternAnalysisWorkshopTutorial_DSM2025_files/libs/quarto-html/anchor.min.js"></script>
<link href="PatternAnalysisWorkshopTutorial_DSM2025_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="PatternAnalysisWorkshopTutorial_DSM2025_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="PatternAnalysisWorkshopTutorial_DSM2025_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="PatternAnalysisWorkshopTutorial_DSM2025_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="PatternAnalysisWorkshopTutorial_DSM2025_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

<link href="PatternAnalysisWorkshopTutorial_DSM2025_files/libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="PatternAnalysisWorkshopTutorial_DSM2025_files/libs/pagedtable-1.1/js/pagedtable.js"></script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#abstract" id="toc-abstract" class="nav-link active" data-scroll-target="#abstract"><span class="header-section-number">1</span> Abstract</a></li>
  <li><a href="#motivation" id="toc-motivation" class="nav-link" data-scroll-target="#motivation"><span class="header-section-number">2</span> Motivation</a></li>
  <li><a href="#setup" id="toc-setup" class="nav-link" data-scroll-target="#setup"><span class="header-section-number">3</span> Setup</a>
  <ul class="collapse">
  <li><a href="#packages" id="toc-packages" class="nav-link" data-scroll-target="#packages"><span class="header-section-number">3.1</span> Packages</a></li>
  <li><a href="#directories" id="toc-directories" class="nav-link" data-scroll-target="#directories"><span class="header-section-number">3.2</span> Directories</a></li>
  <li><a href="#dsm-product-to-evaluate" id="toc-dsm-product-to-evaluate" class="nav-link" data-scroll-target="#dsm-product-to-evaluate"><span class="header-section-number">3.3</span> DSM product to evaluate</a></li>
  <li><a href="#crop-to-a-test-area" id="toc-crop-to-a-test-area" class="nav-link" data-scroll-target="#crop-to-a-test-area"><span class="header-section-number">3.4</span> Crop to a test area</a></li>
  <li><a href="#transform-to-a-metric-crs" id="toc-transform-to-a-metric-crs" class="nav-link" data-scroll-target="#transform-to-a-metric-crs"><span class="header-section-number">3.5</span> Transform to a metric CRS</a></li>
  </ul></li>
  <li><a href="#characterizing-patterns" id="toc-characterizing-patterns" class="nav-link" data-scroll-target="#characterizing-patterns"><span class="header-section-number">4</span> Characterizing patterns</a></li>
  <li><a href="#sec-continuous" id="toc-sec-continuous" class="nav-link" data-scroll-target="#sec-continuous"><span class="header-section-number">5</span> Characterizing patterns – Continuous</a>
  <ul class="collapse">
  <li><a href="#sec-vgm" id="toc-sec-vgm" class="nav-link" data-scroll-target="#sec-vgm"><span class="header-section-number">5.1</span> The global variogram</a></li>
  <li><a href="#sec-mwla" id="toc-sec-mwla" class="nav-link" data-scroll-target="#sec-mwla"><span class="header-section-number">5.2</span> Moving-window local association</a></li>
  <li><a href="#sec-glcm" id="toc-sec-glcm" class="nav-link" data-scroll-target="#sec-glcm"><span class="header-section-number">5.3</span> Grey Level Co-occurrence Matrix (GLCM)</a>
  <ul class="collapse">
  <li><a href="#sec-glcm-quant" id="toc-sec-glcm-quant" class="nav-link" data-scroll-target="#sec-glcm-quant"><span class="header-section-number">5.3.1</span> Quantization</a></li>
  <li><a href="#constructing-a-glcm" id="toc-constructing-a-glcm" class="nav-link" data-scroll-target="#constructing-a-glcm"><span class="header-section-number">5.3.2</span> Constructing a GLCM</a></li>
  <li><a href="#sec-glcm-comp" id="toc-sec-glcm-comp" class="nav-link" data-scroll-target="#sec-glcm-comp"><span class="header-section-number">5.3.3</span> Computation of GLCM texture measures</a></li>
  <li><a href="#sec-glcm-interp" id="toc-sec-glcm-interp" class="nav-link" data-scroll-target="#sec-glcm-interp"><span class="header-section-number">5.3.4</span> Interpretation</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#sec-classified" id="toc-sec-classified" class="nav-link" data-scroll-target="#sec-classified"><span class="header-section-number">6</span> Characterizing patterns – Classified</a>
  <ul class="collapse">
  <li><a href="#sec-hist-equal" id="toc-sec-hist-equal" class="nav-link" data-scroll-target="#sec-hist-equal"><span class="header-section-number">6.1</span> Classifying by histogram equalization</a></li>
  <li><a href="#sec-mean-limit" id="toc-sec-mean-limit" class="nav-link" data-scroll-target="#sec-mean-limit"><span class="header-section-number">6.2</span> Classifying by meaningful limits</a></li>
  <li><a href="#sec-coma" id="toc-sec-coma" class="nav-link" data-scroll-target="#sec-coma"><span class="header-section-number">6.3</span> Co-occurrence matrices</a></li>
  <li><a href="#sec-cove" id="toc-sec-cove" class="nav-link" data-scroll-target="#sec-cove"><span class="header-section-number">6.4</span> Co-occurrence vectors</a></li>
  <li><a href="#incove" id="toc-incove" class="nav-link" data-scroll-target="#incove"><span class="header-section-number">6.5</span> Integrated co-occurrence vector</a></li>
  <li><a href="#clustering-pattern-differences" id="toc-clustering-pattern-differences" class="nav-link" data-scroll-target="#clustering-pattern-differences"><span class="header-section-number">6.6</span> Clustering pattern differences</a></li>
  <li><a href="#sec-lsm" id="toc-sec-lsm" class="nav-link" data-scroll-target="#sec-lsm"><span class="header-section-number">6.7</span> Landscape metrics</a>
  <ul class="collapse">
  <li><a href="#landscape-level-metrics" id="toc-landscape-level-metrics" class="nav-link" data-scroll-target="#landscape-level-metrics"><span class="header-section-number">6.7.1</span> Landscape-level metrics</a></li>
  <li><a href="#compute-sec-lsm" id="toc-compute-sec-lsm" class="nav-link" data-scroll-target="#compute-sec-lsm"><span class="header-section-number">6.7.2</span> Computing landscape-level metrics</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references"><span class="header-section-number">7</span> References</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="PatternAnalysisWorkshopTutorial_DSM2025.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li><li><a href="PatternAnalysisWorkshopTutorial_DSM2025.docx"><i class="bi bi-file-word"></i>MS Word</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><p>Workshop – Digital Soil Mapping 2025<br>
Pattern Analysis for Evaluating Soil Maps</p></h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author"><a href="https://www.css.cornell.edu/faculty/dgr2/pubs/index.html">D G Rossiter</a> <a href="https://orcid.org/0000-0003-4558-1286" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            ISRIC-World Soil Information
          </p>
        <p class="affiliation">
            Section of Soil &amp; Crop Sciences, Cornell University
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 8, 2025</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="abstract" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Abstract</h1>
<p>This tutorial presents methods to evaluate the spatial patterns of the spatial distribution of soil properties and map units as shown in gridded maps produced by digital soil mapping (DSM). Methods include whole-map statistics, visually identifiable landscape features, level of detail, range and strength of spatial autocorrelation, landscape metrics (Shannon diversity and evenness, shape, aggregation, mean fractal dimension, and co-occurrence vectors), and spatial patterns of property maps classified by histogram equalization or user-defined cutpoints. The tutorial also shows how to aggregate raster maps into “supercells” to find landscape elements..</p>
<p>This workshop uses an examples from SoilGrids v2.0, but the methods are applicable to any gridded DSM product or polygon map of soil classes.</p>
</section>
<section id="motivation" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Motivation</h1>
<p>Digital soil maps are usually evaluated by point-wise “validation statistics” <span class="citation" data-cites="piikkiPerspectivesValidationDigital2021">(<a href="#ref-piikkiPerspectivesValidationDigital2021" role="doc-biblioref">Piikki et al., 2021</a>)</span>. This evaluation is quite limited from both the mapper’s and map user’s perspectives.</p>
<p><em>Internally</em>, from the mapper’s perspective:</p>
<ol type="1">
<li>The evaluation is based on a necessarily limited number of observations, far fewer than the number of predictions (grid cells, pixels).</li>
<li>The evaluation points are very rarely from an independent probability sample <span class="citation" data-cites="Brus.etal2011">(<a href="#ref-Brus.etal2011" role="doc-biblioref">Brus et al., 2011</a>)</span>.</li>
<li>Cross-validation and data-splitting approaches rely on a biased point set. Note that so-called “spatial cross-validation” does not solve the problem of biased sampling, just cross-validation biases caused by clustered spatial sampling <span class="citation" data-cites="mahoneyAssessingPerformanceSpatial2023">(<a href="#ref-mahoneyAssessingPerformanceSpatial2023" role="doc-biblioref">Mahoney et al., 2023</a>)</span>.</li>
<li>Evidence has shown that widely different DSM approaches can result in maps with quite similar “validation statistics” but obviously different spatial patterns.</li>
</ol>
<p><em>Externally</em>, from the map user’s perspective:</p>
<ol type="1">
<li>Soils are managed as units, not point-wise.</li>
<li>Land-surface models often rely on 2D or 3D connectivity between grid cells.</li>
<li>More than a century of fieldwork has shown that soils occur in more-or-less homogeneous patches of various sizes, not as isolated pedons <span class="citation" data-cites="johnsonPedonPolypedon1963 Fridland1974 boulaineRemarquesQuelquesNotions1982">(<a href="#ref-boulaineRemarquesQuelquesNotions1982" role="doc-biblioref">Boulaine, 1982</a>; <a href="#ref-Fridland1974" role="doc-biblioref">Fridland, 1974</a>; <a href="#ref-johnsonPedonPolypedon1963" role="doc-biblioref">Johnson, 1963</a>)</span>.</li>
<li>The map user may confuse <em>artefacts</em> of the mapping process with real soil patterns.</li>
</ol>
</section>
<section id="setup" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Setup</h1>
<section id="packages" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="packages"><span class="header-section-number">3.1</span> Packages</h2>
<p>These R packages will be used in the analysis. They must be pre-installed.</p>
<p>First, packages in common use for many applications.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">warn =</span> <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># data wrangling</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># colour palettes for graphics</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot graphics</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># multiple graphics in one plot</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gridExtra, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Second, packages in common use for spatial analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Robert Hijmans raster and vector data; also replaces `raster`</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(terra, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>) </span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot with terra SpatRaster objects</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyterra, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># older package still needed to convert to `sp` objects</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(raster, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>) </span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Pebesma et al. spatio-temporal data</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Simple Features</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sf, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)          </span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="co"># `sp` spatial classes -- still needed for conversions</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sp, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Third, packages specific to the pattern analysis in this workshop:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># variogram modelling</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gstat, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Co-occurrence vectors</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(motif, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>) </span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># multivariate distance metrics</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(philentropy, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>) </span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># FRAGSTATS-style metrics</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># this package is in active development, maybe use the development version</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages("remotes")</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github("r-spatialecology/landscapemetrics")</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(landscapemetrics, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>) </span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="co"># utility functions for raster* landscape objects)</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(landscapetools, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="co"># aggreagate maps with supercells</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="co"># this package is in active development, maybe use the development version</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages("supercells", repos = "https://nowosad.r-universe.dev")</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(supercells, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Gray Level Co-occurence Matrices (GLCM)</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glcm, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GLCMTextures, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="directories" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="directories"><span class="header-section-number">3.2</span> Directories</h2>
<p><em>Task:</em> Set up the base directory.</p>
<p>This is on my system, change to wherever you store your DSM GeoTIFF. Note that in Unix-alike systems the <code>~</code> symbol refers to the user’s home directory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>file.dir <span class="ot">&lt;-</span> <span class="fu">path.expand</span>(<span class="st">"~/ds_reference/DSM2025/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="dsm-product-to-evaluate" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="dsm-product-to-evaluate"><span class="header-section-number">3.3</span> DSM product to evaluate</h2>
<p>The output of a DSM prediction can be saved as a GeoTIFF <span class="citation" data-cites="OGCGeoTIFFStandard2023">(<a href="#ref-OGCGeoTIFFStandard2023" role="doc-biblioref">Open Geospatial Consortium, 2023</a>)</span>.</p>
<p>Here we provide an example: (<span class="math inline">\(1^\circ\)</span>~longitude x <span class="math inline">\(1^\circ\)</span>~latitude) tiles of the SoilGrids v2.0 product <span class="citation" data-cites="Poggio.etal2021a">(<a href="#ref-Poggio.etal2021a" role="doc-biblioref">Poggio et al., 2021</a>)</span>, with a set of soil properties at six standard depth slices. The example tile is from Dindigul District, Tamil Nadu State (India). It was selected for this workshop because it has a good contrast of many soil properties within the tile.</p>
<p>You can create a similar files as GeoTIFF raster stack for a tile of your preference; see the scripts <code>SoilGrids250_WCS_import.Rmd</code>, <code>GetTiles.R</code>, and <code>SoilGrids250_MakeRasterStack.Rmd</code>.</p>
<p>Here is a map of the sample study area, obviously yours will be different.</p>
<div id="fig-localmoran" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="./figs/OSM_7710_7811.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;1: Sample study area: 77-78E, 10-11N</figcaption>
</figure>
</div>
<p>We process the raster stack in R with the <code>terra</code> package, which has the advantage that it only loads into computer memory as needed, and can load lower resolution automatically if that’s appropriate.</p>
<p><em>Task:</em> Import the raster stack as <code>terra::SpatRaster</code> objects.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># the GeoTIFF file name</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>sg.fn <span class="ot">&lt;-</span> <span class="st">"lat1011_lon7778_stack.tif"</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>(sg <span class="ot">&lt;-</span> <span class="fu">rast</span>(<span class="fu">paste0</span>(file.dir, sg.fn)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class       : SpatRaster 
dimensions  : 476, 476, 42  (nrow, ncol, nlyr)
resolution  : 0.002100326, 0.002100326  (x, y)
extent      : 77.00086, 78.00062, 10.00124, 11.00099  (xmin, xmax, ymin, ymax)
coord. ref. : lon/lat WGS 84 (EPSG:4326) 
source      : lat1011_lon7778_stack.tif 
names       : bdod_~_mean, bdod_~_mean, bdod_~_mean, bdod_~_mean, bdod_~_mean, bdod_~_mean, ... 
min values  :    83.34045,    103.1984,    90.17023,     94.4779,    88.24824,    100.3863, ... 
max values  :   154.86685,    155.7222,   161.71574,    158.0000,   155.93663,    157.8423, ... </code></pre>
</div>
</div>
<p>The properties and depth slices in this raster stack:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># layers of the raster stack</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>layer.names <span class="ot">&lt;-</span> <span class="fu">names</span>(sg)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(layer.names, <span class="st">"_"</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>(property.names <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">unlist</span>(<span class="fu">lapply</span>(tmp, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="dv">1</span>]))))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "bdod"  "cec"   "cfvo"  "clay"  "phh2o" "silt"  "soc"  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>(depth.names <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">unlist</span>(<span class="fu">lapply</span>(tmp, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="dv">2</span>]))))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "0-5cm"     "100-200cm" "15-30cm"   "30-60cm"   "5-15cm"    "60-100cm" </code></pre>
</div>
</div>
<p>The raster stack has 42 layers, this is six depth slices for each of 7</p>
<p><em>Task:</em> Plot one layers of all the properties.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>to.plot <span class="ot">&lt;-</span> <span class="fu">grep</span>(depth.names[<span class="dv">1</span>], layer.names, <span class="at">fixed =</span> <span class="cn">TRUE</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">4</span>,<span class="dv">2</span>))</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">plot</span>(sg[[to.plot]])</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-layer1-all-properties" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/fig-layer1-all-properties-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Figure&nbsp;2: All properties, surface layer</figcaption>
</figure>
</div>
</div>
</div>
<p>We see a wide range of values and patterns.</p>
<p><em>Task:</em> Plot all layers of one property.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>to.plot <span class="ot">&lt;-</span> <span class="fu">grep</span>(property.names[<span class="dv">3</span>], <span class="fu">names</span>(sg), <span class="at">fixed =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>r.max <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(<span class="fu">max</span>(<span class="fu">global</span>(sg[[to.plot]], <span class="at">fun =</span> <span class="st">"max"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)))</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>r.min <span class="ot">&lt;-</span> <span class="fu">floor</span>(<span class="fu">min</span>(<span class="fu">global</span>(sg[[to.plot]], <span class="at">fun =</span> <span class="st">"min"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)))</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>))</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">plot</span>(sg[[to.plot]], <span class="at">range =</span> <span class="fu">c</span>(r.min, r.max))</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-all-layers" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/fig-all-layers-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Figure&nbsp;3: One property, all layers</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="crop-to-a-test-area" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="crop-to-a-test-area"><span class="header-section-number">3.4</span> Crop to a test area</h2>
<p>For quicker computation, we restrict the maps (<span class="math inline">\(1^\circ\)</span> x 1^) to a quarter-map (<span class="math inline">\(0.25^\circ\)</span> x <span class="math inline">\(0.25^\circ\)</span>), centred to show some interesting patterns.</p>
<p><em>Task:</em> Crop the raster stack to a quarter-map.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>test.tile.size <span class="ot">&lt;-</span> <span class="fl">0.25</span>  <span class="co"># degrees</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>test.tile.x.offset <span class="ot">&lt;-</span> <span class="fl">0.25</span> <span class="co"># lrc west from right edge</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>test.tile.y.offset <span class="ot">&lt;-</span> <span class="fl">0.25</span>  <span class="co"># lrc north from bottom edge</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>ext.crop <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">as.vector</span>(<span class="fu">ext</span>(sg)),<span class="dv">2</span>) <span class="co"># line up to .00 decimal degrees</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>ext.crop[<span class="st">"xmax"</span>] <span class="ot">&lt;-</span> ext.crop[<span class="st">"xmax"</span>] <span class="sc">-</span> test.tile.x.offset</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>ext.crop[<span class="st">"xmin"</span>] <span class="ot">&lt;-</span> ext.crop[<span class="st">"xmax"</span>] <span class="sc">-</span> test.tile.size</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>ext.crop[<span class="st">"ymin"</span>] <span class="ot">&lt;-</span> ext.crop[<span class="st">"ymin"</span>] <span class="sc">+</span> test.tile.y.offset</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>ext.crop[<span class="st">"ymax"</span>] <span class="ot">&lt;-</span> ext.crop[<span class="st">"ymin"</span>] <span class="sc">+</span> test.tile.size</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ext</span>(ext.crop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SpatExtent : 77.5, 77.75, 10.25, 10.5 (xmin, xmax, ymin, ymax)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>sg4 <span class="ot">&lt;-</span> <span class="fu">crop</span>(sg, <span class="fu">ext</span>(ext.crop))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Task:</em> Repeat the plots, but just for the quarter-tile.</p>
<p><em>Task:</em> Plot one layers of all the properties.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>to.plot <span class="ot">&lt;-</span> <span class="fu">grep</span>(depth.names[<span class="dv">1</span>], layer.names, <span class="at">fixed =</span> <span class="cn">TRUE</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">3</span>))</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">plot</span>(sg4[[to.plot]])</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/fig-layer1-properties-1/4-1.png" class="img-fluid">{#fig-layer1-properties-1/4 width=1152}</p>
</div>
</div>
<p>We see a wide range of values and patterns.</p>
<p><em>Task:</em> Plot all layers of one property.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>to.plot <span class="ot">&lt;-</span> <span class="fu">grep</span>(property.names[<span class="dv">3</span>], layer.names, <span class="at">fixed =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>r.max <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(<span class="fu">max</span>(<span class="fu">global</span>(sg4[[to.plot]], <span class="at">fun =</span> <span class="st">"max"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)))</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>r.min <span class="ot">&lt;-</span> <span class="fu">floor</span>(<span class="fu">min</span>(<span class="fu">global</span>(sg4[[to.plot]], <span class="at">fun =</span> <span class="st">"min"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)))</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>))</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">plot</span>(sg4[[to.plot]], <span class="at">range =</span> <span class="fu">c</span>(r.min, r.max))</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-silt-layers" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/fig-silt-layers-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Figure&nbsp;4: <strong>?(caption)</strong></figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="transform-to-a-metric-crs" class="level2" data-number="3.5">
<h2 data-number="3.5" class="anchored" data-anchor-id="transform-to-a-metric-crs"><span class="header-section-number">3.5</span> Transform to a metric CRS</h2>
<p>Landscape metrics require approximately <em>equal-area</em> grid cells, so the raster stack, currently in a geographic Coördinate Reference System (CRS), must be projected to a metric system. CRS in R are most easily expressed by their EPSG code.</p>
<p>CRS definitions and EPSG codes can be found at at the <a href="https://epsg.org">EPSG Geodetic Parameter Dataset</a>. A reasonable choice for areas narrower (longitude) than about <span class="math inline">\(6^\circ\)</span> is the Universal Transmercator (UTM) system, which covers a <span class="math inline">\(6^\circ\)</span>-wide latitude range with about a <span class="math inline">\(0.5^\circ\)</span> buffer on each edge. Since our test area is <span class="math inline">\(1^\circ\)</span>-wide this is a good choice.</p>
<p>Several datums (forms of the Earth, Earth centre origin) can serve as the basis for the UTM CRS. A common choice is the WGS84 datum. This CRS us used by the Global Positioning System (GPS). It is accurate to within 1 m within each <span class="math inline">\(6^\circ\)</span> UTM slice, of which there are 60.</p>
<p>The EPSG codes for these have the format <code>326xx</code>, where <code>xx</code> is the UTM zone number.</p>
<p>Determine the UTM zone from the longitude of the central meridian of the raster stack. Use this to determine the corresponding EPSG code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># a function to find the correct UTM zome</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>long2UTM <span class="ot">&lt;-</span> <span class="cf">function</span>(long) { (<span class="fu">floor</span>((long <span class="sc">+</span> <span class="dv">180</span>)<span class="sc">/</span><span class="dv">6</span>) <span class="sc">%%</span> <span class="dv">60</span>) <span class="sc">+</span> <span class="dv">1</span> }</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co"># find the zone from the central meridian</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>utm.zone <span class="ot">&lt;-</span> <span class="fu">long2UTM</span>(<span class="fu">st_bbox</span>(sg)<span class="sc">$</span>xmin <span class="sc">+</span> </span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>                       <span class="fl">0.5</span><span class="sc">*</span>(<span class="fu">st_bbox</span>(sg)<span class="sc">$</span>xmax <span class="sc">-</span> <span class="fu">st_bbox</span>(sg)<span class="sc">$</span>xmin))</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">paste</span>(<span class="st">"UTM Zone"</span>, utm.zone))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>UTM Zone 43</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>epsg.utm <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"epsg:326"</span>, utm.zone)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">paste</span>(<span class="st">"CRS code:"</span>, epsg.utm))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>CRS code: epsg:32643</code></pre>
</div>
</div>
<p><em>Task:</em> Resample the maps to the UTM projection, at nominal 250 m grid cell resolution.</p>
<p>Notes:</p>
<ol type="1">
<li>The interpolation method used by <code>terra::project</code> is, by default, bilinear. This is appropriate for continuous-valued maps.</li>
<li>Specify the grid cell size with the <code>res</code> argument to <code>terra::project</code>. SoilGrids maps are nominally at this scale, although presented in geographical coördinates and the Homosoline projection.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">st_bbox</span>(sg4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    xmin     ymin     xmax     ymax 
77.50074 10.24908 77.75068 10.49902 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>sg4.utm <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">project</span>(sg4, epsg.utm, </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">res =</span> <span class="fu">c</span>(<span class="dv">250</span>, <span class="dv">250</span>), <span class="at">method =</span> <span class="st">"bilinear"</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="fu">st_bbox</span>(sg4.utm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   xmin    ymin    xmax    ymax 
 773723 1133915  801223 1161915 </code></pre>
</div>
</div>
</section>
</section>
<section id="characterizing-patterns" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Characterizing patterns</h1>
<p>A first step is to characterize maps by statistical measures. This gives objective information about their spatial patterns.</p>
<p>The methods to characterize patterns are different for maps of <em>continuous</em> variables (<a href="#sec-continuous">Section&nbsp;5</a>) and <em>classified</em> (categorical) variables (<a href="#sec-classified">Section&nbsp;6</a>).</p>
</section>
<section id="sec-continuous" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Characterizing patterns – Continuous</h1>
<p>These are methods that require continuous values on at least an interval scale, and usually a ratio scale (with a true zero). Some properties, e.g., pH, do not have a true zero, so they are an interval scale. Other properties such as coarse fragment volume have a true zero, and one can speak of one location being “twice as stony” than another, for example.</p>
<section id="sec-vgm" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="sec-vgm"><span class="header-section-number">5.1</span> The global variogram</h2>
<p>The variogram (or a correlogram) can be used to characterize the degree of spatial continuity and the “roughness” of a continuous property map, averaged across the entire map. Note that this depends on the grid cell size in two ways:</p>
<ol type="1">
<li>Any pattern at finer resolutions has been removed;</li>
<li>The values in grid cells may be produced by punctual or block methods. Block methods smooth values, so that the variogram sill will necessarily be lower than for punctual predictions. Also, the range may be longer.</li>
</ol>
<p>In this section we compute short-range variograms. These reveal local structure. In DSM maps the variogram is typically unbounded, but we don’t care about the long-range structure when we are evaluating patterns. The parameters of the local structure characterize the fine-scale variability.</p>
<p>Note: Variograms are typically produced separately for each mapped soil property. To characterize an inherent landscape scale, a number of properties can be combined by principal component analysis (PCA) and the first component (PC1) can be characterized.</p>
<p><em>Task</em>: Convert the <code>terra::SpatRaster</code> raster stack to an <code>sp::SpatialPointsDataFrame</code>.</p>
<p>and then to an <code>sf:sf</code> stack in order to compute variograms. The <code>gstat::variogram</code> method must be applied to an object of class <code>sp</code> or <code>sf</code>, not directly to a <code>terra::SpatRaster</code>.</p>
<p>Note: There is (so far) no direct conversion from <code>terra::SpatRaster</code> objects to <code>sf:sf</code> objects.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sg4.utm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 112 110  42</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># keep the coordinates in the data frame</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>sg4.sp <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(sg4.utm, <span class="at">xy =</span> <span class="cn">TRUE</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to SpatialPointsDataFrame by indentifying the fields that are coordinates</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="fu">coordinates</span>(sg4.sp) <span class="ot">&lt;-</span> <span class="er">~</span> x <span class="sc">+</span> y</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(sg4.sp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SpatialPointsDataFrame"
attr(,"package")
[1] "sp"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sg4.sp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 11948    42</code></pre>
</div>
</div>
<p>Now convert to Simple Features (<code>sf</code>) objects, one per property and depth:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>sg4.sf <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(sg4.sp)[<span class="dv">2</span>], <span class="cf">function</span>(i) {</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>   <span class="fu">st_as_sf</span>(sg4.sp[,i])</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(sg4.sf); <span class="fu">length</span>(sg4.sf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "list"</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 42</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># examine one `sf` object in the list</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sg4.sf[[<span class="dv">1</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["bdod_0-5cm_mean"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["geometry"],"name":[2],"type":["sf_POINT"],"align":["right"]}],"data":[{"1":"143.3866","2":"<sf_POINT>","_rn_":"57"},{"1":"143.1640","2":"<sf_POINT>","_rn_":"58"},{"1":"144.0793","2":"<sf_POINT>","_rn_":"59"},{"1":"144.0159","2":"<sf_POINT>","_rn_":"60"},{"1":"143.0872","2":"<sf_POINT>","_rn_":"61"},{"1":"141.8891","2":"<sf_POINT>","_rn_":"62"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Each item in the list <code>sg4.sf</code> is a Simple Features points object.</p>
<p><em>Task:</em> Set the initial parameters for empirical variogram as the resolution.</p>
<p>If the bin width is the resolution, we get one-grid-cell spatial correlations.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>range.init <span class="ot">&lt;-</span> <span class="dv">1000</span>  <span class="co"># estimated range, m </span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>cutoff.init <span class="ot">&lt;-</span> range.init<span class="sc">*</span><span class="dv">5</span>  <span class="co"># cutoff for empirical variogram, m</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>width.init <span class="ot">&lt;-</span> <span class="dv">250</span>   <span class="co"># bin width</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Task:</em> Compute and display the empirical variograms for some properties and layers.</p>
<p>Here is an example with the first layer of the raster stack, accessed by the <code>[[1]]</code> syntax. You can substitute any property and layer, according to your interest. You can also use one of the layer names to specify the raster layer to analyse, e.g.&nbsp;<code>[["cfvo_5-15cm_mean"]]</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(layer.names)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "bdod_0-5cm_mean"      "bdod_100-200cm_mean"  "bdod_15-30cm_mean"   
 [4] "bdod_30-60cm_mean"    "bdod_5-15cm_mean"     "bdod_60-100cm_mean"  
 [7] "cec_0-5cm_mean"       "cec_100-200cm_mean"   "cec_15-30cm_mean"    
[10] "cec_30-60cm_mean"     "cec_5-15cm_mean"      "cec_60-100cm_mean"   
[13] "cfvo_0-5cm_mean"      "cfvo_100-200cm_mean"  "cfvo_15-30cm_mean"   
[16] "cfvo_30-60cm_mean"    "cfvo_5-15cm_mean"     "cfvo_60-100cm_mean"  
[19] "clay_0-5cm_mean"      "clay_100-200cm_mean"  "clay_15-30cm_mean"   
[22] "clay_30-60cm_mean"    "clay_5-15cm_mean"     "clay_60-100cm_mean"  
[25] "phh2o_0-5cm_mean"     "phh2o_100-200cm_mean" "phh2o_15-30cm_mean"  
[28] "phh2o_30-60cm_mean"   "phh2o_5-15cm_mean"    "phh2o_60-100cm_mean" 
[31] "silt_0-5cm_mean"      "silt_100-200cm_mean"  "silt_15-30cm_mean"   
[34] "silt_30-60cm_mean"    "silt_5-15cm_mean"     "silt_60-100cm_mean"  
[37] "soc_0-5cm_mean"       "soc_100-200cm_mean"   "soc_15-30cm_mean"    
[40] "soc_30-60cm_mean"     "soc_5-15cm_mean"      "soc_60-100cm_mean"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sg4.sf[[<span class="dv">1</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "bdod_0-5cm_mean" "geometry"       </code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># give the `sf` object a simple name, also the target variable</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>var <span class="ot">&lt;-</span> sg4.sf[[<span class="dv">1</span>]]</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(var)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"z"</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>v.sg <span class="ot">&lt;-</span> <span class="fu">variogram</span>(z <span class="sc">~</span> <span class="dv">1</span>, <span class="at">loc =</span> var, </span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">cutoff=</span>cutoff.init, <span class="at">width=</span>width.init)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(v.sg, <span class="at">main =</span> layer.names[<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/variogram_empirical-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p><em>Task:</em> Fit a variogram model to the empirical variogram.</p>
<p>The differences can be quantified by the parameters of a fitted variogram model. We try an exponential model because (1) it has the simplest theory, and (2) we expect to not reach a sill within the short range investigated.</p>
<p>We use the <code>fit.variogram</code> method to adjust an initial estimate by weighted least squares (linear in the number of point-pairs and inverse squared in the separation disatance, i.e., the default <code>gstat</code> method 7). The estimated sill is the maximum <span class="math inline">\(\gamma\)</span> in the empirical variogram.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>vm.sg <span class="ot">&lt;-</span> <span class="fu">vgm</span>(<span class="fl">0.8</span><span class="sc">*</span><span class="fu">max</span>(v.sg<span class="sc">$</span>gamma), <span class="st">"Exp"</span>, range.init, <span class="dv">0</span>)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(vmf.sg <span class="ot">&lt;-</span> <span class="fu">fit.variogram</span>(v.sg, <span class="at">model=</span>vm.sg))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  model   psill    range
1   Nug    0.00      0.0
2   Exp 5528.71 483595.9</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(v.sg, <span class="at">model=</span>vmf.sg, <span class="at">main =</span> layer.names[<span class="dv">1</span>], </span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"separation m"</span>, <span class="at">ylab =</span> <span class="fu">expression</span>(<span class="fu">paste</span>(Delta, <span class="fu">plain</span>(pH)<span class="sc">^</span><span class="dv">2</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/fitted_variogram_model-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Q: How well does the fitted model match the empirical variogram? If the fit has some problems, what could be a solution?</p>
</section>
<section id="sec-mwla" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="sec-mwla"><span class="header-section-number">5.2</span> Moving-window local association</h2>
<p>The local spatial structure may not be consistent across the mapped area – that is, the assumption of second-order stationarity may be (and often is) false. This means that the average variogram, computed over that area, is misleading.</p>
<p>The gridded maps have so many cells that it’s possible to compute <strong>moving-window variograms</strong>, as in the VESPER program <span class="citation" data-cites="minasnyVESPERVariogramEstimation2005">(<a href="#ref-minasnyVESPERVariogramEstimation2005" role="doc-biblioref">Minasny et al., 2005</a>)</span> developed for precision agriculture applications. This will show if the local spatial association is consistent across the map. This also allows maps to be compared window-by-window. I have not (yet?) implemented this in R, so we must use another method to assess moving-window local spatial association.</p>
<p>A quick way to see the local degree of autocorrelation is with Moran’s I applied to a window of appropriate size around each grid cell, using the <code>terra::autocor</code> function.</p>
<p>Moran’s I is defined as:</p>
<p><span class="math display">\[  I = \frac{n}{\sum_i \sum_j w_{ij}} \frac{\sum_i \sum_j w_{ij}(y_i - \bar{y})(y_j - \bar{y})}{\sum_i (y_i - \bar{y})^2}\]</span></p>
<p>where <span class="math inline">\(y_i\)</span> is the value of the variable in the <span class="math inline">\(i\)</span>th of <span class="math inline">\(n\)</span> neighbouring grid cells, <span class="math inline">\(\bar{y}\)</span> is the global mean of the variable, <span class="math inline">\(w_{ij}\)</span> is the spatial <strong>weight</strong> of the link between the target cell <span class="math inline">\(i\)</span> and its neighbour cell <span class="math inline">\(j\)</span>. The expected value of Moran’s I is <span class="math inline">\(-1/(n-1)\)</span> if the pattern of the response variable is random, i.e., no spatial correlation. So for a <span class="math inline">\(5 \times 5\)</span> neighbourhood the expected value is <span class="math inline">\(-1/24 = -0.041\bar{6} \approx 0\)</span>.</p>
<p>The second term numerator is the weighted covariance. Its denominator normalizes by the variance. The first term normalizes by the sum of all weights, so that the test is comparable among tests with different numbers of neighbours and using different weightings.</p>
<p><em>Task:</em> Construct a weights matrix for local Moran’s I, for a <span class="math inline">\(5 \times 5\)</span> grid cell neighbourhood, i.e., up to <span class="math inline">\(\pm \; 500 \; m\)</span> in the N/S directions and <span class="math inline">\(\pm \; 500\times\sqrt{2} \approx 707 \; m\)</span> along the diagonals.</p>
<p>We determine the weights matrix for Moran’s I from the fitted global variogram of the previous section and the grid cell size. Weights are the one minus the semivariance at each cell distance, so that the centre pixel receives the maximum weight.</p>
<p>Here is a function to make an odd-sized square window (default 5 x 5) with weights taken from the variogram model, scaled to the resolution.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>make.weights <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">n =</span> <span class="dv">5</span>, <span class="at">res =</span> <span class="dv">250</span>, vgm) {</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  m <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="at">nrow =</span> n, <span class="at">ncol =</span> n)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  center <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(n <span class="sc">/</span> <span class="dv">2</span>)</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>n) {</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>n) {</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>      <span class="co"># distance in cell units, multipled by the grid resolution</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>      m[i, j] <span class="ot">&lt;-</span> <span class="fu">sqrt</span>((i <span class="sc">-</span> center)<span class="sc">^</span><span class="dv">2</span> <span class="sc">+</span> (j <span class="sc">-</span> center)<span class="sc">^</span><span class="dv">2</span>)<span class="sc">*</span><span class="dv">250</span></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>  w <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fu">variogramLine</span>(vm.sg, <span class="at">dist_vector =</span> m)</span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(w)</span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><a href="#fig-localmoran">Figure&nbsp;5</a> shows the Euclidean distance weights in a 5 x 5 window.</p>
<div id="fig-localmoran" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="./figs/localMoranWeights.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;5: Computation of local Moran’s neighbour weights (credit: Diana Collazo, ISRIC)</figcaption>
</figure>
</div>
<p>Here is a function to use this to compute and display the moving-window autocorrelation for any odd window size. This uses the <code>terra::autocor</code> method, applied to a weighted window.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>show.autocor <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">n =</span> <span class="dv">5</span>) {</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  sg.utm.autocor <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">autocor</span>(sg4.utm[[<span class="dv">1</span>]], </span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">w=</span><span class="fu">make.weights</span>(n, <span class="fu">res</span>(sg4.utm)[<span class="dv">1</span>], vmf.sg), </span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">method=</span><span class="st">"moran"</span>, <span class="at">global =</span> <span class="cn">FALSE</span>)</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>  terra<span class="sc">::</span><span class="fu">plot</span>(sg.utm.autocor, <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">"SG250, Moran's I"</span>, n, <span class="st">"x"</span>, n), </span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">col =</span> <span class="fu">rev</span>(<span class="fu">hcl.colors</span>(<span class="dv">32</span>, <span class="at">palette =</span> <span class="st">"RdYlGn"</span>)))</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Task:</em> Compute and display the moving-window autocorrelation, for a 5 x 5 window, in this case 1250 x 1250 m; a 7 x 7 window (1500 x 1500); and a 9 x 9 window (1750 x 1750).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show.autocor</span>(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/moving-window-1.png" class="img-fluid" width="1152"></p>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show.autocor</span>(<span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/moving-window-2.png" class="img-fluid" width="1152"></p>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show.autocor</span>(<span class="dv">9</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/moving-window-3.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>These are all very far from the random value <span class="math inline">\(-0.041\bar{6}\)</span>. Both maps show hot spots with much larger local autocorrelation than the map average. Some areas have almost none or even more dispersed than random (negative values).</p>
<p>To appreciate the local Moran’s I values, here is the global Moran’s I with the same weights matrix. These are the averages of all the local (window) Moran’s I.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>global.moran <span class="ot">&lt;-</span> <span class="cf">function</span>(n) {</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"SG:"</span>, <span class="fu">round</span>(terra<span class="sc">::</span><span class="fu">autocor</span>(sg4.utm[[<span class="dv">1</span>]], </span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">w=</span><span class="fu">make.weights</span>(n, <span class="fu">res</span>(sg4.utm)[<span class="dv">1</span>], vmf.sg),</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">method=</span><span class="st">"moran"</span>, <span class="at">global =</span> <span class="cn">TRUE</span>), <span class="dv">3</span>)))</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a><span class="fu">global.moran</span>(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SG: 0.969"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">global.moran</span>(<span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SG: 0.954"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">global.moran</span>(<span class="dv">9</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SG: 0.939"</code></pre>
</div>
</div>
<p>Q: Is the pattern of local autocorrelation the same across the map?</p>
<p>Q: How does this change as the window size increases?</p>
</section>
<section id="sec-glcm" class="level2" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="sec-glcm"><span class="header-section-number">5.3</span> Grey Level Co-occurrence Matrix (GLCM)</h2>
<p>The idea of characterizing the “texture” of an image has a long history in image processing <span class="citation" data-cites="haralickTexturalFeaturesImage1973">Haralick et al. (<a href="#ref-haralickTexturalFeaturesImage1973" role="doc-biblioref">1973</a>)</span>. One method for this is the <strong>Grey Level Co-occurrence Matrix</strong>. Here the “grey levels” (GL) refer to pixel values – in our context, the values of the soil property, typically quantized (sliced) to some precision. The “co-occurrence” (C) refers to the statistical properties within some window, either isotropic or weighted in some direction. The GLCM shows how often different combinations of values (“grey levels”) occur over local windows within the map. These local textures can be related to landscape ecology, in our case the local spatial structure of the values of a soil property. Many statistics can then be computed to characterize this matrix.</p>
<p>GLCM statistics, in the context of DSM, show the <strong>local</strong> statistical properties of a window as it moves across the map. These can be interpreted as, for example, homogeneity or contrast within a window, thereby revealing areas of the map with different spatial structure.</p>
<p>See <span class="citation" data-cites="hall-beyerGLCMTextureTutorial2017">Hall-Beyer (<a href="#ref-hall-beyerGLCMTextureTutorial2017" role="doc-biblioref">2017a</a>)</span> for a tutorial introduction to the construction, use, and interpretation of GLCM-based textures, and <span class="citation" data-cites="hall-beyerPracticalGuidelinesChoosing2017">Hall-Beyer (<a href="#ref-hall-beyerPracticalGuidelinesChoosing2017" role="doc-biblioref">2017b</a>)</span> for guidelines on choosing appropriate GLCM-based textures in the context of land cover classification.</p>
<section id="sec-glcm-quant" class="level3" data-number="5.3.1">
<h3 data-number="5.3.1" class="anchored" data-anchor-id="sec-glcm-quant"><span class="header-section-number">5.3.1</span> Quantization</h3>
<p>The GLCM is constructed from a moving-window analysis of the map, with the (odd-sized) window considered as a matrix of grid cells.</p>
<p>Before analysis the original map is quantized into a fixed number of levels, by analogy with remote sensing image processing, typically from 16 to 64 levels. Quantization is computed by slicing the value range into equal intervals and replacing the original values with the integer level number.</p>
<p>The GLCM approximates the joint probability distribution of the levels of two pixels separated by the specified shift(s), that is, how likely it is that these two levels occur together in the window. We would like to avoid zero probabilities. If there are too many levels, many pairs of will not occur. So we should pick a number of levels for quantization which avoids this.</p>
<p>The following code shows how to quantize the SoilGrids map into 16 levels. This will be done automatically by the <code>glcm</code> function, see below, here we show how this process works. In the actual computation of statistics we will use more levels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">range</span>(<span class="fu">values</span>(sg4.utm[[<span class="dv">1</span>]], <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  95.51815 148.71246</code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>sg4.quant <span class="ot">&lt;-</span> <span class="fu">cut</span>(<span class="fu">values</span>(sg4.utm[[<span class="dv">1</span>]]), <span class="at">breaks =</span> <span class="dv">16</span>, <span class="at">labels =</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">15</span>, <span class="at">include.lowest =</span> <span class="cn">TRUE</span>)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sg4.quant)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>sg4.quant
   0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15 
  75  707 1645 1739 1336  772  600  469  397  459  693  519  588  713 1068  168 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show the breakpoints</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(<span class="fu">cut</span>(<span class="fu">values</span>(sg4.utm[[<span class="dv">1</span>]]), <span class="at">breaks =</span> <span class="dv">16</span>, <span class="at">include.lowest =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "[95.5,98.8]" "(98.8,102]"  "(102,105]"   "(105,109]"   "(109,112]"  
 [6] "(112,115]"   "(115,119]"   "(119,122]"   "(122,125]"   "(125,129]"  
[11] "(129,132]"   "(132,135]"   "(135,139]"   "(139,142]"   "(142,145]"  
[16] "(145,149]"  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>sg4.utm.quant <span class="ot">&lt;-</span> sg4.utm[[<span class="dv">1</span>]]</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="fu">values</span>(sg4.utm.quant) <span class="ot">&lt;-</span> sg4.quant</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sg4.utm.quant, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">16</span>), <span class="at">main =</span> <span class="fu">paste</span>(layer.names[<span class="dv">1</span>], <span class="st">", 16 levels"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/cut.matrix-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>It is difficult to see just from this map if the GLCM will have too many zeroes, or if a finer quantization could be supported.</p>
</section>
<section id="constructing-a-glcm" class="level3" data-number="5.3.2">
<h3 data-number="5.3.2" class="anchored" data-anchor-id="constructing-a-glcm"><span class="header-section-number">5.3.2</span> Constructing a GLCM</h3>
<p>This section shows how a GLCM is constructed. We take a simple example of an 8-class quantization and a 5x5 window near the middle of the map, and a one-cell rightward shift.</p>
<p>The <code>make_glcm</code> method is provided by a different GLCM package: <code>GLCMTextures</code>.</p>
<div class="cell" data-fig.caption="Example GLCM">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># obtain the bounding box of the test area from cell numbers</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>xy <span class="ot">&lt;-</span> <span class="fu">xyFromCell</span>(sg4.utm[[<span class="dv">1</span>]], <span class="fu">cellFromRowCol</span>(sg4.utm[[<span class="dv">1</span>]], <span class="dv">40</span><span class="sc">:</span><span class="dv">45</span>, <span class="dv">40</span><span class="sc">:</span><span class="dv">45</span>))</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="co"># crop to this box</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>w.sg <span class="ot">&lt;-</span> <span class="fu">crop</span>(sg4.utm[[<span class="dv">1</span>]], xy)</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>test.quant <span class="ot">&lt;-</span> <span class="fu">cut</span>(<span class="fu">values</span>(w.sg), <span class="at">breaks =</span> <span class="dv">8</span>, </span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">labels =</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">7</span>, <span class="at">include.lowest =</span> <span class="cn">TRUE</span>)</span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a><span class="co"># the classes of the cut</span></span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>(l<span class="fl">.8</span> <span class="ot">&lt;-</span> <span class="fu">levels</span>(<span class="fu">cut</span>(<span class="fu">values</span>(w.sg), <span class="at">breaks =</span> <span class="dv">8</span>, <span class="at">include.lowest =</span> <span class="cn">TRUE</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "[118,120]" "(120,122]" "(122,124]" "(124,126]" "(126,128]" "(128,129]"
[7] "(129,131]" "(131,133]"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add the class labels to the test map</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>w.sg<span class="fl">.8</span> <span class="ot">&lt;-</span> w.sg; <span class="fu">values</span>(w.sg<span class="fl">.8</span>) <span class="ot">&lt;-</span> test.quant</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a><span class="co"># show the property and the derived grey levels together</span></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(w.sg, <span class="at">main =</span> <span class="st">"property value"</span>)</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(w.sg<span class="fl">.8</span>, <span class="at">main =</span> <span class="st">"grey level"</span>, <span class="at">col =</span> <span class="fu">grey.colors</span>(<span class="dv">8</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/example-glcm-1.png" class="img-fluid" width="1152"></p>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="co"># set up the matrix on which to compute the GLCM</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>(test.matrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(w.sg<span class="fl">.8</span>, <span class="at">wide =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2] [,3] [,4] [,5]
[1,]    8    8    8    8    7
[2,]    8    8    8    8    6
[3,]    8    8    8    7    5
[4,]    8    8    6    5    3
[5,]    7    6    4    2    1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>glcm <span class="ot">&lt;-</span> GLCMTextures<span class="sc">::</span><span class="fu">make_glcm</span>(test.matrix,  </span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">n_levels =</span> <span class="dv">9</span>, <span class="at">shift =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">0</span>), <span class="co"># shift one cell to the right</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">normalize =</span> <span class="cn">FALSE</span> )</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(glcm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9]
 [1,]    0    0    0    0    0    0    0    0    0
 [2,]    0    0    1    0    0    0    0    0    0
 [3,]    0    1    0    0    1    0    0    0    0
 [4,]    0    0    0    0    0    1    0    0    0
 [5,]    0    0    1    0    0    0    1    0    0
 [6,]    0    0    0    1    0    0    1    1    0
 [7,]    0    0    0    0    1    1    0    1    2
 [8,]    0    0    0    0    0    1    1    0    2
 [9,]    0    0    0    0    0    0    2    2   18</code></pre>
</div>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">diag</span>(glcm))<span class="sc">/</span><span class="fu">sum</span>(glcm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.45</code></pre>
</div>
</div>
<p>The original matrix is 5 x 5 cells; the GLCM is 9 x 9 levels.</p>
<p>In this example 0.45 of the adjacenies are on the GLCM diagonal, i.e., with no change in level based on the 8-level GLCM. The off-diagonals show how many shifts in class, the large the more abrupt the difference.</p>
</section>
<section id="sec-glcm-comp" class="level3" data-number="5.3.3">
<h3 data-number="5.3.3" class="anchored" data-anchor-id="sec-glcm-comp"><span class="header-section-number">5.3.3</span> Computation of GLCM texture measures</h3>
<p>From the quantized matrix, the GLCM can be constructed for one or more specified offsets, called a <strong>shift</strong>. These can be either along the row, column, or diagonal, as specified by the analyst. Each element at position <span class="math inline">\((i,j)\)</span> in the GLCM counts how many times a pixel with value <span class="math inline">\(i\)</span> and a value <span class="math inline">\(j\)</span> occur together with the specified offset. So for example a map quantized with 32 levels will have a 32 x 32 GLCM.</p>
<p>If multiple shifts are specified, the texture statistics are computed for all the specified shifts, with the result for a pixel being the mean of these statistics for each pixel.</p>
<p>The GLCM describes the spatial relationships of (quantized) values in the map; this can be considered “texture”. Many statistics can be computed on the GLCM. Among the relevant statistics for pattern analysis are the mean, variance, homogeneity, contrast, entropy, dissimilarity, second moment, and correlation of the GLCM.</p>
<p>The R <code>glcm</code> package computes these metrics. It requires an object in the older <code>raster</code> package format.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to the older `raster` format</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>sg4.utm.raster <span class="ot">&lt;-</span> <span class="fu">raster</span>(sg4.utm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We choose to compute the mean statistics for four shifts: one pixel by row, column, and both diagonals. If there is orientation (anisotropy) evident in the map, just one shift could be used to characterize the shifts in that orientation.</p>
<p>We choose to compute on a 5 x 5 window (both dimensions must be odd). Since the resolution is already coarse (250 m) this will characterize the texture in <span class="math inline">\(1.5625 ~ \mathrm{km}^2\)</span> squares</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>stat.list <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"mean"</span>,<span class="st">"variance"</span>,<span class="st">"homogeneity"</span>,<span class="st">"contrast"</span>,</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>               <span class="st">"entropy"</span>,<span class="st">"dissimilarity"</span>,<span class="st">"second_moment"</span>,</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>               <span class="st">"correlation"</span>)</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>glcm.sg <span class="ot">&lt;-</span> <span class="fu">rast</span>(<span class="fu">glcm</span>(sg4.utm.raster,</span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">window =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">5</span>),</span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">n_grey =</span> <span class="dv">32</span>, <span class="co"># number of levels in the GLCM</span></span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">shift=</span><span class="fu">list</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>), <span class="fu">c</span>(<span class="dv">1</span>,<span class="sc">-</span><span class="dv">1</span>)), <span class="co"># all directions</span></span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">na_opt =</span> <span class="st">"ignore"</span>,</span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">statistics =</span> stat.list))</span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(glcm.sg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SpatRaster"
attr(,"package")
[1] "terra"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(glcm.sg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   glcm_mean      glcm_variance    glcm_homogeneity glcm_contrast    
 Min.   :0.0000   Min.   :  0.00   Min.   :0.0000   Min.   :  0.000  
 1st Qu.:0.2219   1st Qu.: 49.14   1st Qu.:0.5431   1st Qu.:  0.760  
 Median :0.3468   Median :123.28   Median :0.6372   Median :  1.370  
 Mean   :0.4461   Mean   :266.69   Mean   :0.6297   Mean   :  6.156  
 3rd Qu.:0.6918   3rd Qu.:474.09   3rd Qu.:0.7240   3rd Qu.:  2.620  
 Max.   :0.9891   Max.   :947.12   Max.   :1.0000   Max.   :257.050  
 NA's   :1192     NA's   :1192     NA's   :1192     NA's   :1192     
  glcm_entropy   glcm_dissimilarity glcm_second_moment glcm_correlation
 Min.   :0.000   Min.   :0.000      Min.   :0.0000     Min.   :  -Inf  
 1st Qu.:1.879   1st Qu.:0.600      1st Qu.:0.0856     1st Qu.:0.4877  
 Median :2.263   Median :0.840      Median :0.1232     Median :0.6575  
 Mean   :2.191   Mean   :1.049      Mean   :0.1521     Mean   :  -Inf  
 3rd Qu.:2.572   3rd Qu.:1.190      3rd Qu.:0.1848     3rd Qu.:0.8043  
 Max.   :3.150   Max.   :9.530      Max.   :1.0000     Max.   :6.7768  
 NA's   :1192    NA's   :1192       NA's   :1192       NA's   :1243    </code></pre>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(glcm.sg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/measures-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
</section>
<section id="sec-glcm-interp" class="level3" data-number="5.3.4">
<h3 data-number="5.3.4" class="anchored" data-anchor-id="sec-glcm-interp"><span class="header-section-number">5.3.4</span> Interpretation</h3>
<p>Each of the texture metrics quantifies some aspect of the texture. For a thorough explanation see <span class="citation" data-cites="hall-beyerGLCMTextureTutorial2017">Hall-Beyer (<a href="#ref-hall-beyerGLCMTextureTutorial2017" role="doc-biblioref">2017a</a>)</span> and <span class="citation" data-cites="hall-beyerPracticalGuidelinesChoosing2017">Hall-Beyer (<a href="#ref-hall-beyerPracticalGuidelinesChoosing2017" role="doc-biblioref">2017b</a>)</span>. Here we examine a few of them.</p>
<p><strong>Mean</strong> and <strong>Variance</strong> represent the overall inhomogeneity of the window. The mean is the mean change in the selected shift(s) and the variance is how variable are the changes.</p>
<p><span class="math display">\[\mu = \sum_{i,j = 0}^{N-1} i \cdot P_{i,j}\]</span></p>
<div class="cell" data-fig.caption="GLCM mean, 5 x 5 grid cells">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(glcm.sg[[<span class="st">"glcm_mean"</span>]], <span class="at">main =</span> <span class="st">"GLCM mean"</span>, </span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">col=</span>(sp<span class="sc">::</span><span class="fu">bpy.colors</span>(<span class="dv">32</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>Areas with the higher values have more and/or larger differences between neighbours.</p>
<p><span class="math display">\[\sigma^2 = \sum_{i,j = 0}^{N-1} P_{i,j}\cdot (i - \mu)^2\]</span></p>
<div class="cell" data-fig.caption="GLCM variance, 5 x 5 grid cells">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(glcm.sg[[<span class="st">"glcm_variance"</span>]], <span class="at">main =</span> <span class="st">"GLCM variance"</span>, </span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">col=</span>(<span class="fu">cm.colors</span>(<span class="dv">32</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p><strong>Contrast</strong> is the amount of local variation in a window, with emphasis (squared distance) on the off-diagonals of the GLCM, i.e., larger changes in the quanta level.</p>
<p><span class="math display">\[\sum_{i,j = 0}^{N-1} P_{i,j}\cdot (i - j)^2\]</span></p>
<p>where <span class="math inline">\(P_{i,j}\)</span> is the proportion of the class <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> co-occurrence in the window.</p>
<div class="cell" data-fig.caption="GLCM contrast, 5 x 5 grid cells">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(glcm.sg[[<span class="st">"glcm_contrast"</span>]], <span class="at">main =</span> <span class="st">"Contrast"</span>, </span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">col=</span>(<span class="fu">topo.colors</span>(<span class="dv">32</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>There are “hot spots” of high contrast, i.e., areas in the map with a relatively wide range of propertu values. Note that this shows that the assumption of second-order stationarity used in the variogram analysis <a href="#sec-vgm">Section&nbsp;5.1</a> is definitely not correct.</p>
<p>A variant is the <strong>dissimilarity</strong>, where the weights are linear away from the diagonal, rather than quadratic:</p>
<p><span class="math display">\[\sum_{i,j = 0}^{N-1} P_{i,j}\cdot |i - j|\]</span></p>
<div class="cell" data-fig.caption="GLCM dissimilarity, 5 x 5 grid cells">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(glcm.sg[[<span class="st">"glcm_dissimilarity"</span>]], <span class="at">main =</span> <span class="st">"Dissimilarity"</span>, </span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">col=</span>(<span class="fu">topo.colors</span>(<span class="dv">32</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p><strong>Entropy</strong> is a measure of information within a window. It accounts for the number of different levels in the window (the others will have “probability” zero) and their relative frequencies. More classes and more even distribution of classes results in increased entropy. This can be thought of as “lack of information”.</p>
<p><span class="math display">\[\sum_{i,j = 0}^{N-1} P_{i,j}\cdot -\ln(P_{i,j})\]</span></p>
<div class="cell" data-fig.caption="GLCM entropy, 5 x 5 grid cells">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(glcm.sg[[<span class="st">"glcm_entropy"</span>]], <span class="at">main =</span> <span class="st">"Entropy"</span>, </span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">col=</span>(<span class="fu">topo.colors</span>(<span class="dv">32</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p><em>Challenge</em>: compute the GLCM statistics for different window sizes.</p>
</section>
</section>
</section>
<section id="sec-classified" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Characterizing patterns – Classified</h1>
<p>The spatial unit of conventional (legacy) maps is the polygon, not the grid cell. These maps show a discrete number of legend entries (classes), each with one to many polygons. In the soil survey context these are called <strong>mapping units</strong>, and generally are soil classes, possibly with some landscape features (e.g., erosion class, slope class) as part of the definition. Some mapping units may represent water bodies and various other kinds of non-soil.</p>
<p>Here we continue with the continuous property maps of a single property. To use these techniques on continuous property maps, the maps must be <strong>sliced</strong> (discretized) into classes. There are several choices:</p>
<ul>
<li>meaningful limits, matching some thresholds known to be important for a soil function;</li>
<li>equal intervals;</li>
<li>histogram equalization.</li>
</ul>
<p>For equal intervals or histogram equalization, the cutpoints should be the same for all maps, and therefore derived from their combined distribution of values. We illustrate the process here, but do not use it for the landscape metrics examples later on in the tutorial.</p>
<section id="sec-hist-equal" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="sec-hist-equal"><span class="header-section-number">6.1</span> Classifying by histogram equalization</h2>
<p>This section shows how to classify by histogram equalization; the results will not be used later in the tutorial. Instead, we will use meaningful limits (see <a href="#sec-mean-limit">Section&nbsp;6.2</a>) to slice the map.</p>
<p><em>Task:</em> Slice the map by histogram equalization</p>
<p>First, compute the histogram equalization and display the limits on a histogram plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>n.class <span class="ot">&lt;-</span> <span class="dv">8</span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a><span class="co"># combined values</span></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>values.sort <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">values</span>(sg4.utm[[<span class="dv">1</span>]]))</span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a><span class="fu">range</span>(values.sort)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  95.51815 148.71246</code></pre>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co"># number of pixels not NA</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>n.nna <span class="ot">&lt;-</span> <span class="fu">length</span>(values.sort) <span class="sc">-</span> <span class="fu">sum</span>(<span class="fu">is.na</span>(values.sort))</span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="co"># how many pixels in each bin</span></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>(cut.positions <span class="ot">&lt;-</span> <span class="fu">round</span>(n.nna<span class="sc">/</span>n.class))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1494</code></pre>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># the cut positions</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>(cuts <span class="ot">&lt;-</span> values.sort[cut.positions <span class="sc">*</span> <span class="dv">1</span><span class="sc">:</span>(n.class<span class="dv">-1</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 103.6808 106.5202 109.4846 113.9927 123.1471 132.4583 141.3245</code></pre>
</div>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="co"># integer values for the cuts</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>cuts[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="fu">floor</span>(cuts[<span class="dv">1</span>]); cuts[n.class<span class="dv">-1</span>] <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(cuts[n.class<span class="dv">-1</span>])</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>cuts[<span class="dv">2</span><span class="sc">:</span>n.class<span class="dv">-2</span>] <span class="ot">&lt;-</span> <span class="fu">round</span>(cuts[<span class="dv">2</span><span class="sc">:</span>n.class<span class="dv">-2</span>])</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(cuts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 103 107 109 114 123 132 142</code></pre>
</div>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(values.sort, <span class="at">breaks=</span><span class="dv">36</span>, <span class="at">main=</span><span class="st">"Histogram equalization"</span>,</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> layer.names[<span class="dv">1</span>])</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span>cuts, <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">lwd=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/histo-equal-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>In this plot each slice has the same number of pixels.</p>
<p><em>Task:</em> slice the map with histogram equalizatioj and display the result.</p>
<p>Slice the map:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co"># `rcl` is a vector with the lowest limit 0, the cuts, and the maximum valie</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a><span class="co">#       so that all values are classified</span></span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>sg4.class <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">classify</span>(sg4.utm[[<span class="dv">1</span>]], </span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>                             <span class="at">rcl=</span> <span class="fu">c</span>(<span class="dv">0</span>, cuts, <span class="fu">ceiling</span>(<span class="fu">max</span>(values.sort))))</span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">values</span>(sg4.class))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   0    1    2    3    4    5    6    7 
1156 2095 1014 1714 1470 1420 1815 1264 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sg4.class) <span class="ot">&lt;-</span> <span class="st">"class"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Display the classified map:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">plot</span>(sg4.class,</span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">type=</span><span class="st">"classes"</span>,</span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">main=</span>layer.names[<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/show.classified-hist-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>Q: Describe the patterns of the map.</p>
<p>Q: How would these change with different class numbers or limits?</p>
</section>
<section id="sec-mean-limit" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="sec-mean-limit"><span class="header-section-number">6.2</span> Classifying by meaningful limits</h2>
<p>For soil properties we usually have limits that correspond to approximate thresholds in land use. For example, in the case of pH, we can refer to extension or crop consultant publications, or environmental models. Unlike in histogram equalization, the number of classes depends on the user requirements.</p>
<p>For example, the <a href="https://www.nnyagdev.org/PDF/SoilpH.pdf">Cornell pH test kit</a> has a “Wide Range Kit” measuring the soil pH over the range of 4.0–8.6, in increments of 0.2 for an experienced user. Here we will be somewhat less precise, and slice the map in increments of 0.4 pH.</p>
<p><em>Task:</em> slice the map of surface soil pH and display with a common colour ramp.</p>
<p>Find the combined range and divide into classes of 0.4 pH, starting and ending on even units of 0.4.</p>
<p>Set up the cut points.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co"># find the layer number for this property</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>(ix.ph05 <span class="ot">&lt;-</span> <span class="fu">which</span>(layer.names <span class="sc">==</span> <span class="st">"phh2o_0-5cm_mean"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 25</code></pre>
</div>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>(cuts <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">floor</span>(<span class="fu">min</span>(<span class="fu">values</span>(sg4.utm[[ix.ph05]], <span class="at">na.rm =</span> <span class="cn">TRUE</span>))), </span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>             <span class="fu">ceiling</span>(<span class="fu">max</span>(<span class="fu">values</span>(sg4.utm[[ix.ph05]], <span class="at">na.rm =</span> <span class="cn">TRUE</span>))), </span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">by =</span> <span class="fl">0.4</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 53.0 53.4 53.8 54.2 54.6 55.0 55.4 55.8 56.2 56.6 57.0 57.4 57.8 58.2 58.6
[16] 59.0 59.4 59.8 60.2 60.6 61.0 61.4 61.8 62.2 62.6 63.0 63.4 63.8 64.2 64.6
[31] 65.0 65.4 65.8 66.2 66.6 67.0 67.4 67.8 68.2 68.6 69.0 69.4 69.8 70.2 70.6
[46] 71.0 71.4 71.8 72.2 72.6 73.0 73.4 73.8</code></pre>
</div>
</div>
<p>Slice the map of surface soil pH:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>sg.ph.class <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">classify</span>(sg4.utm[[ix.ph05]], <span class="at">rcl=</span> cuts)</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">values</span>(sg.ph.class))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19 
 11  14  58  87  95 100 146 260 294 420 685 723 810 615 565 448 360 344 264 233 
 20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36  37  38  39 
175 151 145 132 123 130 124 123 127 144 155 163 211 148 219 238 201 294 216 216 
 40  41  42  43  44  45  46  47  48  49  50  51 
304 254 502 281 335 111  99  54  14  18   7   1 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sg.ph.class) <span class="ot">&lt;-</span> <span class="st">"class"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Display it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">plot</span>(sg.ph.class,</span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">col=</span>sp<span class="sc">::</span><span class="fu">bpy.colors</span>(<span class="fu">length</span>(cuts)), <span class="at">type=</span><span class="st">"classes"</span>,</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">main=</span><span class="st">"SG2 pH 0.2 units"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/show.classified-cuts-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>Q: Describe the pattern of the map.</p>
<p>Q: How would the maps change with wider or narrower class intervals? You are welcome to experiment!</p>
</section>
<section id="sec-coma" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="sec-coma"><span class="header-section-number">6.3</span> Co-occurrence matrices</h2>
<p>One question for a classified map is which classes tend to be adjacent to each other. In the case of the pH map, we might expect adjacent classes to be in the pH sequence, but maybe not – there may be abrupt transitions of parent materials, for example.</p>
<p>A co-occurrence <em>matrix</em> counts all the pairs of adjacent cells for each category in a local landscape, as a cross-classification matrix.</p>
<p><em>Task:</em> Compute the co-occurrence <em>matrices</em>, using Queen’s Case neighbours (i.e., diagonal links are considered).</p>
<p>Co-occurrence vectors are computed with the <code>lsp_signature</code> function of the <code>motif</code> package, specifyin <code>coma</code> = co-occurrence matrix as the signature.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>coma.ph <span class="ot">&lt;-</span> <span class="fu">lsp_signature</span>(sg.ph.class, <span class="at">type=</span><span class="st">"coma"</span>, <span class="at">neighbourhood =</span> <span class="dv">8</span>)</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(coma.ph.matrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(coma.ph<span class="sc">$</span>signature)[[<span class="dv">1</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   1  2   3   4   5   6   7   8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
1 28 23  24   8   2   1   1   1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
2 23 16  31  27  12   2   0   1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
3 24 31 144 136  65  15  23   7  6  6  1  0  0  0  0  0  0  0  0  0  0  0  0  0
4  8 27 136 168 153  79  61  37 11  4  4  2  0  0  0  0  0  0  0  0  0  0  0  0
5  2 12  65 153 142 135 117  61 35 20  5  3  1  0  0  0  0  0  0  0  0  0  0  0
6  1  2  15  79 135 142 161 132 58 31 17  5  4  0  2  0  0  0  0  0  0  0  0  0
  25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50
1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
2  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
3  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
4  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
5  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
6  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
  51 52
1  0  0
2  0  0
3  0  0
4  0  0
5  0  0
6  0  0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="co"># proportion with adjacent of the same class</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">diag</span>(coma.ph.matrix))<span class="sc">/</span><span class="fu">sum</span>(coma.ph.matrix)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.2531368</code></pre>
</div>
</div>
<p>The proportion of neighbour pixels with the same class as the corresponding centre pixel is 0.25.</p>
<p>Q: Describe the co-occurrence structure. What does this imply for the spatial pattern?</p>
</section>
<section id="sec-cove" class="level2" data-number="6.4">
<h2 data-number="6.4" class="anchored" data-anchor-id="sec-cove"><span class="header-section-number">6.4</span> Co-occurrence vectors</h2>
<p>The <strong>Co-occurrence vector</strong> “COVE” proposed by <span class="citation" data-cites="nowosadSpatialAssociationRegionalizations2018">Nowosad &amp; Stepinski (<a href="#ref-nowosadSpatialAssociationRegionalizations2018" role="doc-biblioref">2018</a>)</span> summarizes the <em>entire adjacency structure</em> of a map and can be used to compare map structures. This is a normalized form of the co-occurrence matrix (see the previous section). Normalization means the matrix sums to 1, and so is independent of the number of grid cells in the map. Therefore this vector can be considered as a probability vector for the co-occurrence of different classes.</p>
<p><em>Task:</em> Compute the co-occurrence <em>vectors</em>, using Queen’s Case neighbours.</p>
<p>Co-occurrence vectors are computed with the <code>lsp_signature</code> function of the <code>motif</code> package, specifying <code>cove</code> (normalized co-occurrence vector) as the signature.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co"># normalized co-occurence vector 8 x 8</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(cove.ph <span class="ot">&lt;-</span> <span class="fu">lsp_signature</span>(sg.ph.class, <span class="at">type=</span><span class="st">"cove"</span>, <span class="at">neighbourhood =</span> <span class="dv">8</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
     id na_prop signature        
* &lt;int&gt;   &lt;dbl&gt; &lt;list&gt;           
1     1  0.0303 &lt;dbl [1 × 1,378]&gt;</code></pre>
</div>
</div>
</section>
<section id="incove" class="level2" data-number="6.5">
<h2 data-number="6.5" class="anchored" data-anchor-id="incove"><span class="header-section-number">6.5</span> Integrated co-occurrence vector</h2>
<p>An <em>integrated</em> co-occurrence vector considers <em>several input layers</em>, for example representing different soil properties of the same area.</p>
<p>To examine this we need another soil property map. Let’s use silt of the 0–5~cm layer. We process this as we did for the pH map. Here the “meaningful limits” for silt content are 5% intervals. Since the SG2 map is expressed in <span class="math inline">\(\mathrm{g} \; \mathrm{kg}^{-1}\)</span>, these are intervals of 50 <span class="math inline">\(\mathrm{g} \; \mathrm{kg}^{-1}\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>(ix.silt05 <span class="ot">&lt;-</span> <span class="fu">which</span>(layer.names <span class="sc">==</span> <span class="st">"silt_0-5cm_mean"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 31</code></pre>
</div>
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(sg4.utm[[ix.silt05]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> silt_0.5cm_mean
 Min.   :213.7  
 1st Qu.:261.6  
 Median :276.7  
 Mean   :278.9  
 3rd Qu.:295.4  
 Max.   :366.6  
 NA's   :372    </code></pre>
</div>
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>(cuts <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">floor</span>(<span class="fu">min</span>(<span class="fu">values</span>(sg4.utm[[ix.silt05]]<span class="sc">-</span><span class="dv">50</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))), </span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>             <span class="fu">ceiling</span>(<span class="fu">max</span>(<span class="fu">values</span>(sg4.utm[[ix.silt05]]<span class="sc">+</span><span class="dv">50</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))), </span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">by =</span> <span class="dv">50</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 163 213 263 313 363 413</code></pre>
</div>
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>sg.silt.class <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">classify</span>(sg4.utm[[ix.silt05]], <span class="at">rcl=</span> cuts)</span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">values</span>(sg.silt.class))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   1    2    3    4 
3261 7743  939    5 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sg.silt.class) <span class="ot">&lt;-</span> <span class="st">"class"</span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sg.silt.class, <span class="at">col =</span> <span class="fu">topo.colors</span>(<span class="dv">11</span>),</span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> layer.names[ix.silt05])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/silt-map-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>This map has much larger homogeneous areas than the pH map.</p>
<p>Examine this single map’s co-occurrence matrix and vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="co">#|.label: coma-cove</span></span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>coma.silt <span class="ot">&lt;-</span> <span class="fu">lsp_signature</span>(sg.silt.class, <span class="at">type=</span><span class="st">"coma"</span>, <span class="at">neighbourhood =</span> <span class="dv">8</span>)</span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(coma.silt.matrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(coma.silt<span class="sc">$</span>signature)[[<span class="dv">1</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      2     3    4  5
2 21978  3406    0  0
3  3406 56252 1451  0
4     0  1451 5896 23
5     0     0   23  6</code></pre>
</div>
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">diag</span>(coma.silt.matrix))<span class="sc">/</span><span class="fu">sum</span>(coma.silt.matrix)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.8960508</code></pre>
</div>
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="co"># the co-occurrence vector</span></span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a>(cove.silt <span class="ot">&lt;-</span> <span class="fu">lsp_signature</span>(sg.silt.class, <span class="at">type=</span><span class="st">"cove"</span>, <span class="at">neighbourhood =</span> <span class="dv">8</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["id"],"name":[1],"type":["int"],"align":["right"]},{"label":["na_prop"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["signature"],"name":[3],"type":["list"],"align":["right"]}],"data":[{"1":"1","2":"0.03019481","3":"<dbl[,10] [1 × 10]>","_rn_":"1"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Most of the adjacencies are to the same class, or the adjacent class.</p>
<p><em>Task:</em> Compute the distance between the co-occurrence vectors for pH and silt:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>cove.df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(cove.ph)<span class="sc">$</span>signature[[<span class="dv">1</span>]][<span class="dv">1</span>,]</span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a>cove.df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(cove.df, cove.silt<span class="sc">$</span>signature[[<span class="dv">1</span>]][<span class="dv">1</span>,])</span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a>cove.dists <span class="ot">&lt;-</span> <span class="fu">round</span>(</span>
<span id="cb124-4"><a href="#cb124-4" aria-hidden="true" tabindex="-1"></a>  philentropy<span class="sc">::</span><span class="fu">distance</span>(cove.df, <span class="at">method =</span> <span class="st">"jensen-shannon"</span>, </span>
<span id="cb124-5"><a href="#cb124-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">use.row.names =</span><span class="cn">TRUE</span>, </span>
<span id="cb124-6"><a href="#cb124-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">as.dist.obj =</span> <span class="cn">FALSE</span>,</span>
<span id="cb124-7"><a href="#cb124-7" aria-hidden="true" tabindex="-1"></a>                        <span class="at">diag =</span> <span class="cn">FALSE</span>) ,<span class="dv">4</span>)</span>
<span id="cb124-8"><a href="#cb124-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(cove.dists)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>jensen-shannon 
       47.2178 </code></pre>
</div>
</div>
</section>
<section id="clustering-pattern-differences" class="level2" data-number="6.6">
<h2 data-number="6.6" class="anchored" data-anchor-id="clustering-pattern-differences"><span class="header-section-number">6.6</span> Clustering pattern differences</h2>
<p>Once a pattern metric is shown across a map, a natural question is whether different areas of the map have different patterns. We illustrate this with the pattern of the integrated co-occurrence vectors.</p>
<p>Any size window can be used. If too small the result is erratic, if too large, local differences may be missed.</p>
<p><em>Task:</em> Identify which parts of the SG2 map have similar <em>integrated co-occurrence</em> pattern differences, considering both properties. For this we use 4 x 4 km windows, i.e., 16 x 16 grid cells.</p>
<p>Again we use <code>lsp_signature</code>, type <code>"incove"</code>, but now specifying a <code>window</code> size within which to compute the pattern.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>sg.ph.silt.class <span class="ot">&lt;-</span> <span class="fu">c</span>(sg.ph.class, sg.silt.class)</span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a>incove.sg <span class="ot">&lt;-</span> <span class="fu">lsp_signature</span>(sg.ph.silt.class,</span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">type =</span> <span class="st">"incove"</span>,</span>
<span id="cb126-4"><a href="#cb126-4" aria-hidden="true" tabindex="-1"></a>                              <span class="at">neighbourhood =</span> <span class="dv">8</span>,</span>
<span id="cb126-5"><a href="#cb126-5" aria-hidden="true" tabindex="-1"></a>                              <span class="at">ordered =</span> <span class="cn">TRUE</span>,  <span class="co"># the pH classes are ordered</span></span>
<span id="cb126-6"><a href="#cb126-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">window =</span> <span class="dv">16</span>,</span>
<span id="cb126-7"><a href="#cb126-7" aria-hidden="true" tabindex="-1"></a>                              <span class="at">normalization =</span> <span class="st">"pdf"</span>)  <span class="co">#sum to one</span></span>
<span id="cb126-8"><a href="#cb126-8" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(incove.sg.dist <span class="ot">&lt;-</span> <span class="fu">lsp_to_dist</span>(incove.sg,</span>
<span id="cb126-9"><a href="#cb126-9" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">dist_fun =</span> <span class="st">"jensen-shannon"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
0.04773 0.37091 0.63975 0.57553 0.75499 0.99690 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(incove.sg.dist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 49 49</code></pre>
</div>
</div>
<p>Here we have defined 49 x 49 distances, i.e., paired distances between each of the windows’ signatures.</p>
<p>Are any of these distances similar? Let’s see with a <em>cluster analysis</em>.</p>
<p><em>Task:</em> Make a hierarchical clustering of the distances between the integrated co-occurrence vectors of the windows.</p>
<p>The <code>hclust</code> function can cluster using many methods to build the dendrogram. Here we use Ward’s D2 method, which aims at finding compact, spherical clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>sg.hclust <span class="ot">&lt;-</span> <span class="fu">hclust</span>(incove.sg.dist, <span class="at">method =</span> <span class="st">"ward.D2"</span>)</span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sg.hclust, <span class="at">main =</span> <span class="st">"clusters of distance between `incove`"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/incove.cluster-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p><em>Task:</em> Define classes of similar distances by cutting the dendrogram.</p>
<p>Examining the dendrogram, it seems that height <code>h = 0.5</code> is a good cutting point, which captures the main differences. Alternatively, a set number of clusters can be requested with the <code>k</code> argument.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>sg.clusters <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(<span class="fu">cutree</span>(sg.hclust, <span class="at">h =</span> <span class="fl">0.5</span>))  <span class="co"># cutpoint by visual inspection</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(sg.clusters)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "1" "2" "3" "4" "5" "6" "7" "8"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a>sg.grid.sf <span class="ot">=</span> <span class="fu">lsp_add_clusters</span>(incove.sg, sg.clusters)</span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a>sg.grid.sf<span class="sc">$</span>clust <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(sg.grid.sf<span class="sc">$</span>clust)</span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a>my.pal <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">8</span>, <span class="st">"Accent"</span>))(<span class="fu">length</span>(<span class="fu">levels</span>(sg.grid.sf<span class="sc">$</span>clust)))</span>
<span id="cb133-4"><a href="#cb133-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> sg.grid.sf) <span class="sc">+</span> </span>
<span id="cb133-5"><a href="#cb133-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> clust), <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb133-6"><a href="#cb133-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_discrete</span>(<span class="at">type =</span> my.pal) <span class="sc">+</span></span>
<span id="cb133-7"><a href="#cb133-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Clusters: distance between integrated co-occurrence vectors"</span>,</span>
<span id="cb133-8"><a href="#cb133-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="st">"cluster"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/incove.cluster.cut-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>This shows which areas of the map have similar integrated co-occurrence patterns. These can be interpreted as similar soils, in the sense that the sum of propertied defines a soil type.</p>
<p>Compare this to a visual inspection of the patterns, next to the 7 x 6 cluster grid.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> sg.grid.sf) <span class="sc">+</span> </span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> clust), <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_discrete</span>(<span class="at">type =</span> my.pal) <span class="sc">+</span></span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"cluster"</span>, <span class="at">title =</span> <span class="st">"INCOVE clusters"</span>)  <span class="sc">+</span></span>
<span id="cb134-5"><a href="#cb134-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>)</span>
<span id="cb134-6"><a href="#cb134-6" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb134-7"><a href="#cb134-7" aria-hidden="true" tabindex="-1"></a>  tidyterra<span class="sc">::</span><span class="fu">geom_spatraster</span>(<span class="at">data =</span> sg.ph.class, <span class="fu">aes</span>(<span class="at">fill =</span> class)) <span class="sc">+</span></span>
<span id="cb134-8"><a href="#cb134-8" aria-hidden="true" tabindex="-1"></a>   <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb134-9"><a href="#cb134-9" aria-hidden="true" tabindex="-1"></a>   <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"pH 0-5 cm classes"</span>)</span>
<span id="cb134-10"><a href="#cb134-10" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb134-11"><a href="#cb134-11" aria-hidden="true" tabindex="-1"></a>  tidyterra<span class="sc">::</span><span class="fu">geom_spatraster</span>(<span class="at">data =</span> sg.silt.class, <span class="fu">aes</span>(<span class="at">fill =</span> class)) <span class="sc">+</span></span>
<span id="cb134-12"><a href="#cb134-12" aria-hidden="true" tabindex="-1"></a>   <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb134-13"><a href="#cb134-13" aria-hidden="true" tabindex="-1"></a>   <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Silt 0-5 cm classes"</span>)</span>
<span id="cb134-14"><a href="#cb134-14" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(p1, p2, p3, <span class="at">nrow=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/incove-clusters-map-3-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>Careful examination reveals that the cluster in the NW corner corresponds to an intricate pattern of pH and mostly one class of silt concentration.</p>
</section>
<section id="sec-lsm" class="level2" data-number="6.7">
<h2 data-number="6.7" class="anchored" data-anchor-id="sec-lsm"><span class="header-section-number">6.7</span> Landscape metrics</h2>
<p>Landscape metrics have a long history of use in landscape ecology <span class="citation" data-cites="Uuemaa.etal2013">(<a href="#ref-Uuemaa.etal2013" role="doc-biblioref">Uuemaa et al., 2013</a>)</span>. A wide variety have been collected in the well-known FRAGSTATS computer program <span class="citation" data-cites="McGarigal.etal2012">(<a href="#ref-McGarigal.etal2012" role="doc-biblioref">McGarigal et al., 2012</a>)</span>. These have been implemented in the R context by the <code>landscapemetrics</code> package<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> <span class="citation" data-cites="Hesselbarth.etal2019 Hesselbarth2021">(<a href="#ref-Hesselbarth.etal2019" role="doc-biblioref">Hesselbarth et al., 2019</a>; <a href="#ref-Hesselbarth2021" role="doc-biblioref">Hesselbarth, 2021</a>)</span>. Although the ecological relevance of FRAGSTATS metrics have been criticized <span class="citation" data-cites="Kupfer2012">(<a href="#ref-Kupfer2012" role="doc-biblioref">Kupfer, 2012</a>)</span>, here we use them to <em>characterize spatial patterns of soil properties</em> or <em>classes</em>, not as inputs to landscape ecology models.</p>
<p>The patterns of soil classes or properties are not expected to have the same characteristics as those for land cover or vegetation types. Land cover is largely controlled by humans, and where it is not, vegetation is mostly placed on the landscape by different mechanisms than are soils. There is a link, however: if the soil property is largely controlled by the <code>o</code> (organism) or <code>h</code> (human) factor, then the patterns on the landscape could be similar to those under it.</p>
<p>There are many metrics, of three levels of detail. We list them here for reference; each has its own help text.</p>
<p>First, the <em>patch-level metrics</em>. These describe every patch, i.e., contiguous cells belonging to the same class.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a>landscapemetrics<span class="sc">::</span><span class="fu">list_lsm</span>(<span class="at">level=</span><span class="st">"patch"</span>) <span class="sc">%&gt;%</span> <span class="fu">print</span>(<span class="at">n =</span> <span class="dv">12</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 × 5
   metric name                                type           level function_name
   &lt;chr&gt;  &lt;chr&gt;                               &lt;chr&gt;          &lt;chr&gt; &lt;chr&gt;        
 1 area   patch area                          area and edge… patch lsm_p_area   
 2 cai    core area index                     core area met… patch lsm_p_cai    
 3 circle related circumscribing circle       shape metric   patch lsm_p_circle 
 4 contig contiguity index                    shape metric   patch lsm_p_contig 
 5 core   core area                           core area met… patch lsm_p_core   
 6 enn    euclidean nearest neighbor distance aggregation m… patch lsm_p_enn    
 7 frac   fractal dimension index             shape metric   patch lsm_p_frac   
 8 gyrate radius of gyration                  area and edge… patch lsm_p_gyrate 
 9 ncore  number of core areas                core area met… patch lsm_p_ncore  
10 para   perimeter-area ratio                shape metric   patch lsm_p_para   
11 perim  patch perimeter                     area and edge… patch lsm_p_perim  
12 shape  shape index                         shape metric   patch lsm_p_shape  </code></pre>
</div>
</div>
<p>Second, the <em>class-level</em> metrics. These describe all patches belonging to a specified class.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>landscapemetrics<span class="sc">::</span><span class="fu">list_lsm</span>(<span class="at">level=</span><span class="st">"class"</span>) <span class="sc">%&gt;%</span> <span class="fu">print</span>(<span class="at">n =</span> <span class="dv">12</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 55 × 5
   metric    name                          type              level function_name
   &lt;chr&gt;     &lt;chr&gt;                         &lt;chr&gt;             &lt;chr&gt; &lt;chr&gt;        
 1 ai        aggregation index             aggregation metr… class lsm_c_ai     
 2 area_cv   patch area                    area and edge me… class lsm_c_area_cv
 3 area_mn   patch area                    area and edge me… class lsm_c_area_mn
 4 area_sd   patch area                    area and edge me… class lsm_c_area_sd
 5 ca        total (class) area            area and edge me… class lsm_c_ca     
 6 cai_cv    core area index               core area metric  class lsm_c_cai_cv 
 7 cai_mn    core area index               core area metric  class lsm_c_cai_mn 
 8 cai_sd    core area index               core area metric  class lsm_c_cai_sd 
 9 circle_cv related circumscribing circle shape metric      class lsm_c_circle…
10 circle_mn related circumscribing circle shape metric      class lsm_c_circle…
11 circle_sd related circumscribing circle shape metric      class lsm_c_circle…
12 clumpy    clumpiness index              aggregation metr… class lsm_c_clumpy 
# ℹ 43 more rows</code></pre>
</div>
</div>
<p>Finally, the <em>landscape-level</em> metrics. These describe the characteristics of the entire landscape, i.e., the assemblage of classes and patches.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>landscapemetrics<span class="sc">::</span><span class="fu">list_lsm</span>(<span class="at">level=</span><span class="st">"landscape"</span>) <span class="sc">%&gt;%</span> <span class="fu">print</span>(<span class="at">n =</span> <span class="dv">12</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 66 × 5
   metric    name                          type              level function_name
   &lt;chr&gt;     &lt;chr&gt;                         &lt;chr&gt;             &lt;chr&gt; &lt;chr&gt;        
 1 ai        aggregation index             aggregation metr… land… lsm_l_ai     
 2 area_cv   patch area                    area and edge me… land… lsm_l_area_cv
 3 area_mn   patch area                    area and edge me… land… lsm_l_area_mn
 4 area_sd   patch area                    area and edge me… land… lsm_l_area_sd
 5 cai_cv    core area index               core area metric  land… lsm_l_cai_cv 
 6 cai_mn    core area index               core area metric  land… lsm_l_cai_mn 
 7 cai_sd    core area index               core area metric  land… lsm_l_cai_sd 
 8 circle_cv related circumscribing circle shape metric      land… lsm_l_circle…
 9 circle_mn related circumscribing circle shape metric      land… lsm_l_circle…
10 circle_sd related circumscribing circle shape metric      land… lsm_l_circle…
11 cohesion  patch cohesion index          aggregation metr… land… lsm_l_cohesi…
12 condent   conditional entropy           complexity metric land… lsm_l_condent
# ℹ 54 more rows</code></pre>
</div>
</div>
<section id="landscape-level-metrics" class="level3" data-number="6.7.1">
<h3 data-number="6.7.1" class="anchored" data-anchor-id="landscape-level-metrics"><span class="header-section-number">6.7.1</span> Landscape-level metrics</h3>
<p>These measures summarize the pattern of the entire map. The following five seem to be most useful for characterizing soil maps.</p>
<ul>
<li><strong>ai</strong>: The <strong>landscape aggregation index</strong> LAI is an ‘Aggregation metric’. This shows how much the classes occur as large units, vs.&nbsp;as scattered patches. It is independent of the number of classes.</li>
</ul>
<p>It equals the number of like adjacencies divided by the theoretical maximum possible number of like adjacencies for that class summed over each class for the entire landscape. The metric is based on the adjacency matrix. It equals 0 for maximally disaggregated and 100 for maximally aggregated classes. <a href="https://r-spatialecology.github.io/landscapemetrics/reference/lsm_l_ai.html">More info</a> <span class="math display">\[\mathrm{LAI} = \Bigg[∑_{i=1}^m \Big( \frac{g_{ii}}{max-g_{ii}} \Big) P_{i} \Bigg](100)\]</span> where <span class="math inline">\(g_{ii}\)</span> is the number of like adjacencies, <span class="math inline">\((\mathrm{max}-g_{ii})\)</span> is the class-wise maximum possible number of like adjacencies of class <span class="math inline">\(i\)</span> (i.e., if all pixels in the class were in one cluster), and <span class="math inline">\(P_{i}\)</span> is the proportion of landscape comprised of class <span class="math inline">\(i\)</span>, to weight the index by class prevalence.</p>
<ul>
<li><strong>frac_mn</strong>: The <strong>mean fractal dimension</strong> FRAC_MN is a ‘Shape metric’. It summarises the landscape as the mean of the fractal dimension index of all patches in the landscape, i.e., the complexity of the map.</li>
</ul>
<p>The fractal dimension index is based on the patch perimeter and the patch area and describes the patch complexity. The Coefficient of variation is scaled to the mean and thus is comparable among different landscapes. <a href="https://r-spatialecology.github.io/landscapemetrics/reference/lsm_l_frac_mn.html">More info</a> <span class="math display">\[\mathrm{FRAC} = \frac{2 * \ln * (0.25 * p_{ij})} {\ln a_{ij}}\]</span> where the patch perimeters are <span class="math inline">\({p_{ij}}\)</span> in linear units and the areas are <span class="math inline">\({a_{ij}}\)</span> in square units.</p>
<ul>
<li><strong>lsi</strong>: <strong>landscape shape index</strong> LSI is an ‘Aggregation metric’. It is the ratio between the actual edge length of class <span class="math inline">\(i\)</span> and the hypothetical minimum edge length of class <span class="math inline">\(i\)</span>. It measures how compact are the classes. For example, long thin classes will have low LSI.</li>
</ul>
<p>The minimum edge length equals the edge length if class i would be maximally aggregated. LSI = 1 when only one square patch is present or all patches are maximally aggregated. Increases, without limit, as the length of the actual edges increases, i.e.&nbsp;the patches become less compact. <a href="https://r-spatialecology.github.io/landscapemetrics/reference/lsm_c_lsi.html?q=lsi">More info</a> <span class="math display">\[   \mathrm{LSI} = \frac{0.25 E'}{\sqrt{A}}\]</span> where <span class="math inline">\(A\)</span> is the total area of the landscape and <span class="math inline">\(E'\)</span> is the total length of edges, including the boundary.</p>
<ul>
<li><strong>shdi</strong>: The <strong>Shannon diversity index</strong> SHDI is a ‘Diversity metric’. It is a widely used metric in biodiversity and ecology and takes both the number of classes and the abundance of each class into account. It is related to the concept of entropy: how much “information” is in the landscape pattern. More classes and more even distribution of their areas implies high information.</li>
</ul>
<p>SHDI = 0 when only one patch is present and increases, without limit, as the number of classes increases while the proportions are equally distributed. <a href="https://r-spatialecology.github.io/landscapemetrics/reference/lsm_l_shdi.html?q=shd">More info</a> <span class="math display">\[ D = - \sum_{i=1}^N p_i \ln p_i\]</span> where <span class="math inline">\(p_i\)</span> is the proportion of pixels of class <span class="math inline">\(i = (1 \ldots N)\)</span>,</p>
<ul>
<li><strong>shei</strong>: The <strong>Shannon evenness index</strong> SHEI is a ‘Diversity metric’. It is the ratio between the Shannon’s diversity index <span class="math inline">\(D\)</span> (see previous) and and the theoretical maximum Shannon diversity index <span class="math inline">\(\ln N\)</span>. It can be understood as a measure of dominance.</li>
</ul>
<p>SHEI = 0 when only one patch present; SHEI = 1 when the proportion of classes is equally distributed. <a href="https://r-spatialecology.github.io/landscapemetrics/reference/lsm_l_shei.html?q=shei">More info</a> <span class="math display">\[E = \frac{D}{\ln N}\]</span></p>
<p>These methods must be applied to classified maps. Continuous soil property maps must first be classified into ranges before analysis, see (<a href="#sec-hist-equal">Section&nbsp;6.1</a>) and (<a href="#sec-mean-limit">Section&nbsp;6.2</a>), above. Different choices of class limits and widths will result in different values of these measures.</p>
</section>
<section id="compute-sec-lsm" class="level3" data-number="6.7.2">
<h3 data-number="6.7.2" class="anchored" data-anchor-id="compute-sec-lsm"><span class="header-section-number">6.7.2</span> Computing landscape-level metrics</h3>
<p>The <code>landscapemetrics</code> package implements a set of metrics as used in ecology and derived from the FRAGSTATS computer program; the metrics are explained in the previous section. Here we compute them for the two maps we are comparing.</p>
<p>To compute landscape metrics:</p>
<ul>
<li>Input is raster map (here, a <code>terra::SpatRaster</code>) with integer values, each of which represents a category, i.e., landscape class.</li>
<li>The map must be in a projected CRS, with distance units in meters;</li>
<li>Results are in meters, square meters or hectares, depending on the function;</li>
</ul>
<p><em>Task:</em> Check that the maps have the proper structure for the landscape metrics.</p>
<p>This is done with the <code>landscapemetrics::check_landscape</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_landscape</span>(sg.ph.class)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["layer"],"name":[1],"type":["int"],"align":["right"]},{"label":["crs"],"name":[2],"type":["chr"],"align":["left"]},{"label":["units"],"name":[3],"type":["chr"],"align":["left"]},{"label":["class"],"name":[4],"type":["chr"],"align":["left"]},{"label":["n_classes"],"name":[5],"type":["int"],"align":["right"]},{"label":["OK"],"name":[6],"type":["chr"],"align":["left"]}],"data":[{"1":"1","2":"projected","3":"m","4":"integer","5":"52","6":"❓"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_landscape</span>(sg.silt.class)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["layer"],"name":[1],"type":["int"],"align":["right"]},{"label":["crs"],"name":[2],"type":["chr"],"align":["left"]},{"label":["units"],"name":[3],"type":["chr"],"align":["left"]},{"label":["class"],"name":[4],"type":["chr"],"align":["left"]},{"label":["n_classes"],"name":[5],"type":["int"],"align":["right"]},{"label":["OK"],"name":[6],"type":["chr"],"align":["left"]}],"data":[{"1":"1","2":"projected","3":"m","4":"integer","5":"4","6":"✔"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p><em>Task:</em> Show the landscapes of each layer, first with all classes on one map, then with the classes separate:</p>
<p><strong>global</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_patches</span>(sg.ph.class, <span class="at">class =</span> <span class="st">"global"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/show.patches.global-1.png" class="img-fluid" width="768"></p>
</div>
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_patches</span>(sg.silt.class, <span class="at">class =</span> <span class="st">"global"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/show.patches.global-2.png" class="img-fluid" width="768"></p>
</div>
</div>
<p><strong>per-class</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_patches</span>(sg.ph.class, <span class="at">class =</span> <span class="st">"all"</span>, <span class="at">nrow =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/show.patches.all-1.png" class="img-fluid" width="1344"></p>
</div>
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_patches</span>(sg.silt.class, <span class="at">class =</span> <span class="st">"all"</span>, <span class="at">nrow =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/show.patches.all-2.png" class="img-fluid" width="1344"></p>
</div>
</div>
<p>Q: Describe the main differences between the patterns. Which map seems more aggregated? More diverse?</p>
<p><em>Task:</em> compute the metrics and tabulate them:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a>lst <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"lsm_l_"</span>, <span class="fu">c</span>(<span class="st">"shdi"</span>, <span class="st">"shei"</span>, <span class="st">"lsi"</span>, <span class="st">"ai"</span>,  <span class="st">"frac_mn"</span>))</span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a>ls.metrics.ph <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(sg.ph.class, <span class="at">what=</span>lst)</span>
<span id="cb151-3"><a href="#cb151-3" aria-hidden="true" tabindex="-1"></a>ls.metrics.silt <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(sg.silt.class, <span class="at">what=</span>lst)</span>
<span id="cb151-4"><a href="#cb151-4" aria-hidden="true" tabindex="-1"></a>metrics.table <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">product=</span><span class="fu">c</span>(<span class="st">"pH"</span>, <span class="st">"silt"</span>),</span>
<span id="cb151-5"><a href="#cb151-5" aria-hidden="true" tabindex="-1"></a>                            <span class="fu">rbind</span>(<span class="fu">round</span>(ls.metrics.ph<span class="sc">$</span>value, <span class="dv">3</span>),</span>
<span id="cb151-6"><a href="#cb151-6" aria-hidden="true" tabindex="-1"></a>                                  <span class="fu">round</span>(ls.metrics.silt<span class="sc">$</span>value, <span class="dv">3</span>)))</span>
<span id="cb151-7"><a href="#cb151-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(metrics.table)[<span class="dv">2</span><span class="sc">:</span><span class="dv">6</span>] <span class="ot">&lt;-</span> ls.metrics.ph<span class="sc">$</span>metric</span>
<span id="cb151-8"><a href="#cb151-8" aria-hidden="true" tabindex="-1"></a>metrics.table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["product"],"name":[1],"type":["chr"],"align":["left"]},{"label":["ai"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["frac_mn"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["lsi"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["shdi"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["shei"],"name":[6],"type":["dbl"],"align":["right"]}],"data":[{"1":"pH","2":"30.385","3":"1.025","4":"39.648","5":"3.643","6":"0.922"},{"1":"silt","2":"91.379","3":"1.033","4":"6.132","5":"0.839","6":"0.605"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Q: Referring to the descriptions of these metrics (above), what are the differences between these maps’ landscape patterns? Where do the maps most differ?</p>
<ul>
<li>Aggregation Index</li>
<li>Mean Fractal Dimension</li>
<li>Landscape Shape Index</li>
<li>Shannon Diversity</li>
<li>Shannon Evenness</li>
</ul>
<p>Question: Which maps in the DSM stack do you expect to have similar landscape metrics?</p>
<p>˙ # Supercells</p>
<p><em>“Superpixels”</em> is a generic name for grouping pixels with similar characteristics into larger assemblages. In the soil map context, the aim is to regionalize into areas with similar values of one or more raster layers.</p>
<p>The <code>supercells::supercells</code> function controls the segmentation: the user can specify the <code>k</code> argument for the number of supercells, and the <code>compactness</code> argument to control shape: larger values lead to more square, less long/twisted shapes. It is also possible to specify a set of initial supercell centres (with an <code>sf</code> POINTS geometry) or a separation between initial centres with the <code>step</code> argument.</p>
<p>This function implements the SLIC algorithm <span class="citation" data-cites="achantaSLICSuperpixelsCompared2012">(<a href="#ref-achantaSLICSuperpixelsCompared2012" role="doc-biblioref">Achanta et al., 2012</a>)</span>.</p>
<p>As an example with the pH map, we divide into about 50 supercells, with low compactness since we don’t expect near-square natural units. Here is the source map:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb152-2"><a href="#cb152-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_spatraster</span>(<span class="at">data=</span>sg4.utm[[ix.ph05]]) <span class="sc">+</span></span>
<span id="cb152-3"><a href="#cb152-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>() <span class="sc">+</span></span>
<span id="cb152-4"><a href="#cb152-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"pH x 10"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/SLIC-source-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>And here are the 50 supercells, with very low compactness, i.e., allowing for irregular and elongated shapes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a>sg4.ph<span class="fl">.50</span> <span class="ot">=</span> <span class="fu">supercells</span>(sg4.utm[[ix.ph05]], <span class="at">k =</span> <span class="dv">50</span>, <span class="at">compactness =</span> <span class="fl">0.1</span>)</span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sg4.ph<span class="fl">.50</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "supercells"       "x"                "y"                "phh2o_0.5cm_mean"
[5] "geometry"        </code></pre>
</div>
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sg4.ph<span class="fl">.50</span>)[<span class="dv">4</span>] <span class="ot">&lt;-</span> <span class="st">"pH_05cm"</span> <span class="co"># `supercells` changes the name -- a bug?</span></span>
<span id="cb155-2"><a href="#cb155-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>sg4.ph<span class="fl">.50</span>) <span class="sc">+</span></span>
<span id="cb155-3"><a href="#cb155-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> pH_05cm)) <span class="sc">+</span></span>
<span id="cb155-4"><a href="#cb155-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>() <span class="sc">+</span> </span>
<span id="cb155-5"><a href="#cb155-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"mean pH"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/supercells-not-compact-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>Try to form more compact supercells:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a>sg4.ph<span class="fl">.50</span> <span class="ot">=</span> <span class="fu">supercells</span>(sg4.utm[[ix.ph05]], <span class="at">k =</span> <span class="dv">50</span>, <span class="at">compactness =</span> <span class="dv">3</span>)</span>
<span id="cb156-2"><a href="#cb156-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sg4.ph<span class="fl">.50</span>)[<span class="dv">4</span>] <span class="ot">&lt;-</span> <span class="st">"pH_05cm"</span> <span class="co"># `supercells` changes the name -- a bug?</span></span>
<span id="cb156-3"><a href="#cb156-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>sg4.ph<span class="fl">.50</span>) <span class="sc">+</span></span>
<span id="cb156-4"><a href="#cb156-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> pH_05cm)) <span class="sc">+</span></span>
<span id="cb156-5"><a href="#cb156-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>() <span class="sc">+</span> </span>
<span id="cb156-6"><a href="#cb156-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"mean pH"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/supercells-compact-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>These do not look realistic.</p>
<p>Try with multiple rasters, here pH and silt concentrations:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a>r <span class="ot">&lt;-</span> <span class="fu">c</span>(sg4.utm[[ix.ph05]], sg4.utm[[ix.silt05]])</span>
<span id="cb157-2"><a href="#cb157-2" aria-hidden="true" tabindex="-1"></a>r<span class="fl">.50</span> <span class="ot">=</span> <span class="fu">supercells</span>(r, <span class="at">k =</span> <span class="dv">50</span>, <span class="at">compactness =</span> <span class="fl">0.1</span>)</span>
<span id="cb157-3"><a href="#cb157-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>r<span class="fl">.50</span>) <span class="sc">+</span></span>
<span id="cb157-4"><a href="#cb157-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> phh2o_0<span class="fl">.5</span>cm_mean)) <span class="sc">+</span></span>
<span id="cb157-5"><a href="#cb157-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"mean pH"</span>) <span class="sc">+</span></span>
<span id="cb157-6"><a href="#cb157-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_continuous</span>(<span class="at">type =</span> <span class="st">"viridis"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/supercells-multiple-1.png" class="img-fluid" width="576"></p>
</div>
<div class="sourceCode cell-code" id="cb158"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>r<span class="fl">.50</span>) <span class="sc">+</span></span>
<span id="cb158-2"><a href="#cb158-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> silt_0<span class="fl">.5</span>cm_mean)) <span class="sc">+</span></span>
<span id="cb158-3"><a href="#cb158-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"silt ppt"</span>) <span class="sc">+</span></span>
<span id="cb158-4"><a href="#cb158-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_continuous</span>(<span class="at">type =</span> <span class="st">"viridis"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_DSM2025_files/figure-html/supercells-multiple-2.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>The segments are the same in the two visualizations.</p>
<p><em>Challenge</em>: Experiment with different <code>compactness</code> and <code>k</code> parameters. Which seem to give a more “realistic” landscape pattern?</p>
</section>
</section>
</section>
<section id="references" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> References</h1>
<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-achantaSLICSuperpixelsCompared2012" class="csl-entry" role="listitem">
Achanta, R., Shaji, A., Smith, K., Lucchi, A., Fua, P., &amp; Süsstrunk, S. (2012). <span>SLIC Superpixels</span> compared to state-of-the-art superpixel methods. <em>IEEE Transactions on Pattern Analysis and Machine Intelligence</em>, <em>34</em>(11), 2274–2282. <span>IEEE Transactions</span> on <span>Pattern Analysis</span> and <span>Machine Intelligence</span>. <a href="https://doi.org/10.1109/TPAMI.2012.120">https://doi.org/10.1109/TPAMI.2012.120</a>
</div>
<div id="ref-boulaineRemarquesQuelquesNotions1982" class="csl-entry" role="listitem">
Boulaine, J. (1982). Remarques sur quelques notions élémentaires de la pédologie”. <em><span>Cahiers O.R.S.T.O.M., Série Pédologie</span></em>, <em>19</em>(1), 29–41.
</div>
<div id="ref-Brus.etal2011" class="csl-entry" role="listitem">
Brus, D. J., Kempen, B., &amp; Heuvelink, G. B. M. (2011). Sampling for validation of digital soil maps. <em>European Journal of Soil Science</em>, <em>62</em>, 394–407. <a href="https://doi.org/10.1111/j.1365-2389.2011.01364.x">https://doi.org/10.1111/j.1365-2389.2011.01364.x</a>
</div>
<div id="ref-Fridland1974" class="csl-entry" role="listitem">
Fridland, V. M. (1974). Structure of the soil mantle. <em>Geoderma</em>, <em>12</em>, 35–42. <a href="https://doi.org/10.1016/0016-7061(74)90036-6">https://doi.org/10.1016/0016-7061(74)90036-6</a>
</div>
<div id="ref-hall-beyerGLCMTextureTutorial2017" class="csl-entry" role="listitem">
Hall-Beyer, M. (2017a). <em><span>GLCM Texture</span>: <span>A Tutorial</span> v. 3.0 <span>March</span> 2017</em>. <a href="http://hdl.handle.net/1880/51900">http://hdl.handle.net/1880/51900</a>
</div>
<div id="ref-hall-beyerPracticalGuidelinesChoosing2017" class="csl-entry" role="listitem">
Hall-Beyer, M. (2017b). Practical guidelines for choosing <span>GLCM</span> textures to use in landscape classification tasks over a range of moderate spatial scales. <em>International Journal of Remote Sensing</em>, <em>38</em>(5), 1312–1338. <a href="https://doi.org/10.1080/01431161.2016.1278314">https://doi.org/10.1080/01431161.2016.1278314</a>
</div>
<div id="ref-haralickTexturalFeaturesImage1973" class="csl-entry" role="listitem">
Haralick, R. M., Shanmugam, K., &amp; Dinstein, I. (1973). Textural features for image classification. <em>IEEE Transactions on Systems, Man, and Cybernetics</em>, <em>SMC-3</em>(6), 610–621. <span>IEEE Transactions</span> on <span>Systems</span>, <span>Man</span>, and <span>Cybernetics</span>. <a href="https://doi.org/10.1109/TSMC.1973.4309314">https://doi.org/10.1109/TSMC.1973.4309314</a>
</div>
<div id="ref-Hesselbarth2021" class="csl-entry" role="listitem">
Hesselbarth, M. H. K. (2021). <em><span>R-Spatialecology/Landscapemetrics</span></em>. r-spatialecology. <a href="https://github.com/r-spatialecology/landscapemetrics">https://github.com/r-spatialecology/landscapemetrics</a>
</div>
<div id="ref-Hesselbarth.etal2019" class="csl-entry" role="listitem">
Hesselbarth, M. H. K., Sciaini, M., With, K. A., Wiegand, K., &amp; Nowosad, J. (2019). Landscapemetrics: An open-source <span>R</span> tool to calculate landscape metrics. <em>Ecography</em>, <em>42</em>, 1648–1657. <a href="https://doi.org/10.1111/ecog.04617">https://doi.org/10.1111/ecog.04617</a>
</div>
<div id="ref-johnsonPedonPolypedon1963" class="csl-entry" role="listitem">
Johnson, W. M. (1963). The <span>Pedon</span> and the <span>Polypedon</span>. <em>Soil Science Society of America Journal</em>, <em>27</em>(2), 212–215. <a href="https://doi.org/10.2136/sssaj1963.03615995002700020034x">https://doi.org/10.2136/sssaj1963.03615995002700020034x</a>
</div>
<div id="ref-Kupfer2012" class="csl-entry" role="listitem">
Kupfer, J. A. (2012). Landscape ecology and biogeography: <span>Rethinking</span> landscape metrics in a post-<span>FRAGSTATS</span> landscape. <em>Progress in Physical Geography-Earth and Environment</em>, <em>36</em>(3), 400–420. <a href="https://doi.org/10.1177/0309133312439594">https://doi.org/10.1177/0309133312439594</a>
</div>
<div id="ref-mahoneyAssessingPerformanceSpatial2023" class="csl-entry" role="listitem">
Mahoney, M. J., Johnson, L. K., Silge, J., Frick, H., Kuhn, M., &amp; Beier, C. M. (2023). <em>Assessing the performance of spatial cross-validation approaches for models of spatially structured data</em> (arXiv:2303.07334). <span>arXiv</span>. <a href="https://doi.org/10.48550/arXiv.2303.07334">https://doi.org/10.48550/arXiv.2303.07334</a>
</div>
<div id="ref-McGarigal.etal2012" class="csl-entry" role="listitem">
McGarigal, K., Cushman, S. A., &amp; Ene, E. (2012). <em><span>FRAGSTATS</span> v4: <span>S</span>patial pattern analysis program for categorical and continuous maps</em>. <span>University of Massachusetts</span>. <a href="http://www.umass.edu/landeco/research/fragstats/fragstats.html">http://www.umass.edu/landeco/research/fragstats/fragstats.html</a>
</div>
<div id="ref-minasnyVESPERVariogramEstimation2005" class="csl-entry" role="listitem">
Minasny, B., McBratney, A. B., &amp; Whelan, B. M. (2005). <em><span>VESPER</span>: <span>Variogram Estimation</span> and <span>Spatial Prediction</span> plus <span>ERror</span> - <span>Australian Centre</span> for <span>Precision Agriculture</span></em>. https://precision-agriculture.sydney.edu.au/resources/software/download-vesper/.
</div>
<div id="ref-nowosadSpatialAssociationRegionalizations2018" class="csl-entry" role="listitem">
Nowosad, J., &amp; Stepinski, T. F. (2018). Spatial association between regionalizations using the information-theoretical <span class="nocase">V-measure</span>. <em>International Journal of Geographical Information Science</em>, <em>32</em>(12), 2386–2401. <a href="https://doi.org/10.1080/13658816.2018.1511794">https://doi.org/10.1080/13658816.2018.1511794</a>
</div>
<div id="ref-OGCGeoTIFFStandard2023" class="csl-entry" role="listitem">
Open Geospatial Consortium. (2023). <span>OGC GeoTIFF Standard</span>. In <em>OGC GeoTIFF Standard</em>. https://www.ogc.org/standard/geotiff/.
</div>
<div id="ref-piikkiPerspectivesValidationDigital2021" class="csl-entry" role="listitem">
Piikki, K., Wetterlind, J., Söderström, M., &amp; Stenberg, B. (2021). Perspectives on validation in digital soil mapping of continuous attributes a review. <em>Soil Use and Management</em>, <em>37</em>(1), 7–21. <a href="https://doi.org/10.1111/sum.12694">https://doi.org/10.1111/sum.12694</a>
</div>
<div id="ref-Poggio.etal2021a" class="csl-entry" role="listitem">
Poggio, L., de Sousa, L. M., Batjes, N. H., Heuvelink, G. B. M., Kempen, B., Ribeiro, E., &amp; Rossiter, D. (2021). <span>SoilGrids</span> 2.0: Producing soil information for the globe with quantified spatial uncertainty. <em>SOIL</em>, <em>7</em>(1), 217–240. <a href="https://doi.org/10.5194/soil-7-217-2021">https://doi.org/10.5194/soil-7-217-2021</a>
</div>
<div id="ref-Uuemaa.etal2013" class="csl-entry" role="listitem">
Uuemaa, E., Mander, U., &amp; Marja, R. (2013). Trends in the use of landscape spatial metrics as landscape indicators: <span>A</span> review. <em>Ecological Indicators</em>, <em>28</em>, 100–106. <a href="https://doi.org/10.1016/j.ecolind.2012.07.018">https://doi.org/10.1016/j.ecolind.2012.07.018</a>
</div>
</div>
</section>


<div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>https://r-spatialecology.github.io/landscapemetrics/<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section><section class="quarto-appendix-contents"><h2 class="anchored quarto-appendix-heading">Reuse</h2><div id="quarto-reuse" class="quarto-appendix-contents"><div>CC-BY</div></div></section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>