<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="D G Rossiter">
<meta name="dcterms.date" content="2024-02-04">

<title>quarto-input1351b1be</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="PatternAnalysisWorkshopTutorial_files/libs/clipboard/clipboard.min.js"></script>
<script src="PatternAnalysisWorkshopTutorial_files/libs/quarto-html/quarto.js"></script>
<script src="PatternAnalysisWorkshopTutorial_files/libs/quarto-html/popper.min.js"></script>
<script src="PatternAnalysisWorkshopTutorial_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="PatternAnalysisWorkshopTutorial_files/libs/quarto-html/anchor.min.js"></script>
<link href="PatternAnalysisWorkshopTutorial_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="PatternAnalysisWorkshopTutorial_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="PatternAnalysisWorkshopTutorial_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="PatternAnalysisWorkshopTutorial_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="PatternAnalysisWorkshopTutorial_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

<link href="PatternAnalysisWorkshopTutorial_files/libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="PatternAnalysisWorkshopTutorial_files/libs/pagedtable-1.1/js/pagedtable.js"></script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#abstract" id="toc-abstract" class="nav-link active" data-scroll-target="#abstract"><span class="header-section-number">1</span> Abstract</a></li>
  <li><a href="#motivation" id="toc-motivation" class="nav-link" data-scroll-target="#motivation"><span class="header-section-number">2</span> Motivation</a></li>
  <li><a href="#setup" id="toc-setup" class="nav-link" data-scroll-target="#setup"><span class="header-section-number">3</span> Setup</a>
  <ul class="collapse">
  <li><a href="#packages" id="toc-packages" class="nav-link" data-scroll-target="#packages"><span class="header-section-number">3.1</span> Packages</a></li>
  <li><a href="#directories" id="toc-directories" class="nav-link" data-scroll-target="#directories"><span class="header-section-number">3.2</span> Directories</a></li>
  <li><a href="#example-dataset-soil-properties" id="toc-example-dataset-soil-properties" class="nav-link" data-scroll-target="#example-dataset-soil-properties"><span class="header-section-number">3.3</span> Example dataset: soil properties</a>
  <ul class="collapse">
  <li><a href="#crop-to-a-smaller-area" id="toc-crop-to-a-smaller-area" class="nav-link" data-scroll-target="#crop-to-a-smaller-area"><span class="header-section-number">3.3.1</span> Crop to a smaller area</a></li>
  <li><a href="#transform-to-a-metric-crs" id="toc-transform-to-a-metric-crs" class="nav-link" data-scroll-target="#transform-to-a-metric-crs"><span class="header-section-number">3.3.2</span> Transform to a metric CRS</a></li>
  <li><a href="#harmonizing-the-mapped-areas" id="toc-harmonizing-the-mapped-areas" class="nav-link" data-scroll-target="#harmonizing-the-mapped-areas"><span class="header-section-number">3.3.3</span> Harmonizing the mapped areas</a></li>
  </ul></li>
  <li><a href="#sec-import-gssurgo" id="toc-sec-import-gssurgo" class="nav-link" data-scroll-target="#sec-import-gssurgo"><span class="header-section-number">3.4</span> Example dataset: Soil survey map unit polygons</a>
  <ul class="collapse">
  <li><a href="#transform-to-a-metric-crs-1" id="toc-transform-to-a-metric-crs-1" class="nav-link" data-scroll-target="#transform-to-a-metric-crs-1"><span class="header-section-number">3.4.1</span> Transform to a metric CRS</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#characterizing-patterns" id="toc-characterizing-patterns" class="nav-link" data-scroll-target="#characterizing-patterns"><span class="header-section-number">4</span> Characterizing patterns</a></li>
  <li><a href="#sec-continuous" id="toc-sec-continuous" class="nav-link" data-scroll-target="#sec-continuous"><span class="header-section-number">5</span> Characterizing patterns – Continuous</a>
  <ul class="collapse">
  <li><a href="#the-global-variogram" id="toc-the-global-variogram" class="nav-link" data-scroll-target="#the-global-variogram"><span class="header-section-number">5.1</span> The global variogram</a></li>
  <li><a href="#moving-window-local-association" id="toc-moving-window-local-association" class="nav-link" data-scroll-target="#moving-window-local-association"><span class="header-section-number">5.2</span> Moving-window local association</a></li>
  </ul></li>
  <li><a href="#sec-classified" id="toc-sec-classified" class="nav-link" data-scroll-target="#sec-classified"><span class="header-section-number">6</span> Characterizing patterns – Classified</a>
  <ul class="collapse">
  <li><a href="#sec-hist-equal" id="toc-sec-hist-equal" class="nav-link" data-scroll-target="#sec-hist-equal"><span class="header-section-number">6.1</span> Classifying by histogram equalization</a></li>
  <li><a href="#sec-mean-limit" id="toc-sec-mean-limit" class="nav-link" data-scroll-target="#sec-mean-limit"><span class="header-section-number">6.2</span> Classifying by meaningful limits</a></li>
  <li><a href="#sec-coma" id="toc-sec-coma" class="nav-link" data-scroll-target="#sec-coma"><span class="header-section-number">6.3</span> Co-occurrence matrices</a></li>
  <li><a href="#sec-cove" id="toc-sec-cove" class="nav-link" data-scroll-target="#sec-cove"><span class="header-section-number">6.4</span> Co-occurrence vectors</a></li>
  <li><a href="#incove" id="toc-incove" class="nav-link" data-scroll-target="#incove"><span class="header-section-number">6.5</span> Integrated co-occurrence vector</a>
  <ul class="collapse">
  <li><a href="#clustering-pattern-differences" id="toc-clustering-pattern-differences" class="nav-link" data-scroll-target="#clustering-pattern-differences"><span class="header-section-number">6.5.1</span> Clustering pattern differences</a></li>
  </ul></li>
  <li><a href="#sec-lsm" id="toc-sec-lsm" class="nav-link" data-scroll-target="#sec-lsm"><span class="header-section-number">6.6</span> Landscape metrics</a>
  <ul class="collapse">
  <li><a href="#landscape-level-metrics" id="toc-landscape-level-metrics" class="nav-link" data-scroll-target="#landscape-level-metrics"><span class="header-section-number">6.6.1</span> Landscape-level metrics</a></li>
  <li><a href="#compute-sec-lsm" id="toc-compute-sec-lsm" class="nav-link" data-scroll-target="#compute-sec-lsm"><span class="header-section-number">6.6.2</span> Computing landscape-level metrics</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#sec-compare-classified" id="toc-sec-compare-classified" class="nav-link" data-scroll-target="#sec-compare-classified"><span class="header-section-number">7</span> Comparing patterns of classified maps</a>
  <ul class="collapse">
  <li><a href="#sec-compare-cove" id="toc-sec-compare-cove" class="nav-link" data-scroll-target="#sec-compare-cove"><span class="header-section-number">7.1</span> Co-occurrence vectors</a></li>
  <li><a href="#vmetrics" id="toc-vmetrics" class="nav-link" data-scroll-target="#vmetrics"><span class="header-section-number">7.2</span> V metrics</a>
  <ul class="collapse">
  <li><a href="#polygonize" id="toc-polygonize" class="nav-link" data-scroll-target="#polygonize"><span class="header-section-number">7.2.1</span> Polygonize</a></li>
  <li><a href="#simple-features" id="toc-simple-features" class="nav-link" data-scroll-target="#simple-features"><span class="header-section-number">7.2.2</span> Simple Features</a></li>
  <li><a href="#topology" id="toc-topology" class="nav-link" data-scroll-target="#topology"><span class="header-section-number">7.2.3</span> Topology</a></li>
  <li><a href="#sec-vmeasure" id="toc-sec-vmeasure" class="nav-link" data-scroll-target="#sec-vmeasure"><span class="header-section-number">7.2.4</span> Compute the V-metrics</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#supercells" id="toc-supercells" class="nav-link" data-scroll-target="#supercells"><span class="header-section-number">8</span> Supercells</a></li>
  <li><a href="#sec-scale-geometry" id="toc-sec-scale-geometry" class="nav-link" data-scroll-target="#sec-scale-geometry"><span class="header-section-number">9</span> Scale issues – geometric</a>
  <ul class="collapse">
  <li><a href="#m-resolution" id="toc-m-resolution" class="nav-link" data-scroll-target="#m-resolution"><span class="header-section-number">9.1</span> 20 m resolution</a>
  <ul class="collapse">
  <li><a href="#landscape-level" id="toc-landscape-level" class="nav-link" data-scroll-target="#landscape-level"><span class="header-section-number">9.1.1</span> Landscape level</a></li>
  <li><a href="#class-level" id="toc-class-level" class="nav-link" data-scroll-target="#class-level"><span class="header-section-number">9.1.2</span> Class level</a></li>
  <li><a href="#patch-level" id="toc-patch-level" class="nav-link" data-scroll-target="#patch-level"><span class="header-section-number">9.1.3</span> Patch level</a></li>
  </ul></li>
  <li><a href="#m-resolution-1" id="toc-m-resolution-1" class="nav-link" data-scroll-target="#m-resolution-1"><span class="header-section-number">9.2</span> 100 m resolution</a>
  <ul class="collapse">
  <li><a href="#landscape-level-1" id="toc-landscape-level-1" class="nav-link" data-scroll-target="#landscape-level-1"><span class="header-section-number">9.2.1</span> Landscape level</a></li>
  <li><a href="#class-level-1" id="toc-class-level-1" class="nav-link" data-scroll-target="#class-level-1"><span class="header-section-number">9.2.2</span> Class level</a></li>
  <li><a href="#patch-level-1" id="toc-patch-level-1" class="nav-link" data-scroll-target="#patch-level-1"><span class="header-section-number">9.2.3</span> Patch level</a></li>
  </ul></li>
  <li><a href="#m-resolution-2" id="toc-m-resolution-2" class="nav-link" data-scroll-target="#m-resolution-2"><span class="header-section-number">9.3</span> 300 m resolution</a>
  <ul class="collapse">
  <li><a href="#landscape-level-2" id="toc-landscape-level-2" class="nav-link" data-scroll-target="#landscape-level-2"><span class="header-section-number">9.3.1</span> Landscape level</a></li>
  <li><a href="#class-level-2" id="toc-class-level-2" class="nav-link" data-scroll-target="#class-level-2"><span class="header-section-number">9.3.2</span> Class level</a></li>
  <li><a href="#patch-level-2" id="toc-patch-level-2" class="nav-link" data-scroll-target="#patch-level-2"><span class="header-section-number">9.3.3</span> Patch level</a></li>
  </ul></li>
  <li><a href="#conclusions-about-geometric-scale-issues" id="toc-conclusions-about-geometric-scale-issues" class="nav-link" data-scroll-target="#conclusions-about-geometric-scale-issues"><span class="header-section-number">9.4</span> Conclusions about geometric scale issues</a></li>
  </ul></li>
  <li><a href="#sec-categorical-generalization" id="toc-sec-categorical-generalization" class="nav-link" data-scroll-target="#sec-categorical-generalization"><span class="header-section-number">10</span> Scale issues – categorical</a>
  <ul class="collapse">
  <li><a href="#landscape-metrics" id="toc-landscape-metrics" class="nav-link" data-scroll-target="#landscape-metrics"><span class="header-section-number">10.0.1</span> Landscape metrics</a></li>
  <li><a href="#conclusions-about-categorical-scale-issues" id="toc-conclusions-about-categorical-scale-issues" class="nav-link" data-scroll-target="#conclusions-about-categorical-scale-issues"><span class="header-section-number">10.1</span> Conclusions about categorical scale issues</a></li>
  </ul></li>
  <li><a href="#comparing-to-reality-or-which-map-is-best" id="toc-comparing-to-reality-or-which-map-is-best" class="nav-link" data-scroll-target="#comparing-to-reality-or-which-map-is-best"><span class="header-section-number">11</span> Comparing to “reality”, or Which map is “best”?</a>
  <ul class="collapse">
  <li><a href="#agreement-with-obvious-landscape-features" id="toc-agreement-with-obvious-landscape-features" class="nav-link" data-scroll-target="#agreement-with-obvious-landscape-features"><span class="header-section-number">11.1</span> Agreement with obvious landscape features</a></li>
  </ul></li>
  <li><a href="#pattern-based-segmentation" id="toc-pattern-based-segmentation" class="nav-link" data-scroll-target="#pattern-based-segmentation"><span class="header-section-number">12</span> Pattern-based segmentation</a>
  <ul class="collapse">
  <li><a href="#pattern-within-a-region" id="toc-pattern-within-a-region" class="nav-link" data-scroll-target="#pattern-within-a-region"><span class="header-section-number">12.1</span> Pattern within a region</a></li>
  <li><a href="#geopat" id="toc-geopat" class="nav-link" data-scroll-target="#geopat"><span class="header-section-number">12.2</span> GeoPAT</a></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references"><span class="header-section-number">13</span> References</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="PatternAnalysisWorkshopTutorial.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><p>Workshop<br>
Pattern Analysis for Evaluating Soil Maps<br>
Pedometrics 2024, Las Cruces NM (USA)</p></h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author"><a href="https://www.css.cornell.edu/faculty/dgr2/pubs/index.html">D G Rossiter</a> <a href="https://orcid.org/0000-0003-4558-1286" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            ISRIC-World Soil Information
          </p>
        <p class="affiliation">
            Section of Soil &amp; Crop Sciences, Cornell University
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 4, 2024</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="abstract" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Abstract</h1>
<p>This tutorial presents methods to evaluate the spatial patterns of the spatial distribution of soil properties and map units as shown in gridded maps produced by digital soil mapping (DSM). It compares patterns from different DSM products to each other, and to spatial patterns known from detailed field surveys or known to local experts but not represented (yet) on maps. Methods include whole-map statistics, visually identifiable landscape features, level of detail, range and strength of spatial autocorrelation, landscape metrics (Shannon diversity and evenness, shape, aggregation, mean fractal dimension, and co-occurrence vectors), and spatial patterns of property maps classified by histogram equalization or user-defined cutpoints. The tutorial also shows how to use patterns within a window to partition a soil landscape into zones with similar patterns. This workshop uses examples from the USA, but the methods are applicable to any gridded DSM product or polygon map of soil classes.</p>
</section>
<section id="motivation" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Motivation</h1>
<p>Digital soil maps are usually evaluated by point-wise “validation statistics” <span class="citation" data-cites="piikkiPerspectivesValidationDigital2021">(<a href="#ref-piikkiPerspectivesValidationDigital2021" role="doc-biblioref">Piikki et al., 2021</a>)</span>. This evaluation is quite limited from both the mapper’s and map user’s perspectives.</p>
<p><em>Internally</em>, from the mapper’s perspective:</p>
<ol type="1">
<li>The evaluation is based on a necessarily limited number of observations, far fewer than the number of predictions (grid cells, pixels).</li>
<li>The evaluation points are very rarely from an independent probability sample <span class="citation" data-cites="Brus.etal2011">(<a href="#ref-Brus.etal2011" role="doc-biblioref">Brus et al., 2011</a>)</span>.</li>
<li>Cross-validation and data-splitting approaches rely on a biased point set. (Note: so-called “spatial cross-validation” does not solve the problem of biased sampling, just cross-validation biases caused by clustered spatial sampling. <span class="citation" data-cites="mahoneyAssessingPerformanceSpatial2023">(<a href="#ref-mahoneyAssessingPerformanceSpatial2023" role="doc-biblioref">Mahoney et al., 2023</a>)</span>)</li>
<li>Evidence has shown that widely different DSM approaches can result in maps with quite similar “validation statistics” but obviously different spatial patterns.</li>
</ol>
<p>See for example <a href="#fig-genova">Figure&nbsp;1</a>, which shows an area near Montréal PQ mapped by convolutional neural networks (CNN) with the same training points and covariates, but with different CNN window sizes. The pointwise evaluation statistics in this example are almost identical.</p>
<div id="fig-genova" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="./figs/GenovaPosterFig1a.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;1: Different methods, different patterns (Giulio Genova, ISRIC)</figcaption>
</figure>
</div>
<p><em>Externally</em>, from the map user’s perspective:</p>
<ol type="1">
<li>Soils are managed as units, not point-wise.</li>
<li>Land-surface models often rely on 2D or 3D connectivity between grid cells.</li>
<li>More than a century of fieldwork has shown that soils occur in more-or-less homogeneous patches of various sizes, not as isolated pedons <span class="citation" data-cites="johnsonPedonPolypedon1963 Fridland1974 boulaineRemarquesQuelquesNotions1982">(<a href="#ref-boulaineRemarquesQuelquesNotions1982" role="doc-biblioref">Boulaine, 1982</a>; <a href="#ref-Fridland1974" role="doc-biblioref">Fridland, 1974</a>; <a href="#ref-johnsonPedonPolypedon1963" role="doc-biblioref">Johnson, 1963</a>)</span>.</li>
<li>The map user may confuse <em>artefacts</em> of the mapping process with real soil patterns.</li>
</ol>
</section>
<section id="setup" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Setup</h1>
<section id="packages" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="packages"><span class="header-section-number">3.1</span> Packages</h2>
<p>These R packages will be used in the analysis. They must be pre-installed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="st">"rgdal_show_exportToProj4_warnings"</span><span class="ot">=</span><span class="st">"none"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">warn =</span> <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Robert Hijmans raster and vector data</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(terra, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>) <span class="co"># replaces `raster`</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># still needed to convert to `sp`</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(raster, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>) </span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Pebesma et al. spation-temporal data</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Simple Features</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sf, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)          </span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co"># `sp` spatial classes -- still needed for conversions</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sp, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co"># variogram modelling</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gstat, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Co-occurrence vectors</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(motif, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>) </span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="co"># multivariate distance metrics</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(philentropy, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>) </span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="co"># compare polygon map spatial structures, V measure</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sabre, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)   </span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="co"># FRAGSTATS-style metrics</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(landscapemetrics, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>) </span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="co"># utility functions for raster* landscape objects)</span></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(landscapetools, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot graphics</span></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a><span class="co"># multiple graphics in one plot</span></span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gridExtra, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot with terra SpatRaster</span></span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyterra, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a><span class="co"># data wrangling</span></span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a><span class="co"># colour palettes for graphics</span></span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a><span class="co"># to access NRCS databases</span></span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(soilDB, <span class="at">warn.conflicts=</span><span class="cn">FALSE</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a><span class="co"># supercells</span></span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages("supercells", repos = "https://nowosad.r-universe.dev")</span></span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(supercells)</span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a><span class="co"># compare two rasters directly -- in development</span></span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a><span class="co"># devtools::install_github("Nowosad/spquery")</span></span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a><span class="co"># library(spquery)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="directories" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="directories"><span class="header-section-number">3.2</span> Directories</h2>
<p>Task: Set up the base directory.</p>
<p>This is on my system, change to wherever you want to store the sample files. Note that in Unix-alike systems the <code>~</code> symbol refers to the user’s home directory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>file.dir <span class="ot">&lt;-</span> <span class="fu">path.expand</span>(<span class="st">"~/ds_reference/Compare_DSM/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="example-dataset-soil-properties" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="example-dataset-soil-properties"><span class="header-section-number">3.3</span> Example dataset: soil properties</h2>
<p>The output of a DSM prediction can be saved as a GeoTiff <span class="citation" data-cites="OGCGeoTIFFStandard2023">(<a href="#ref-OGCGeoTIFFStandard2023" role="doc-biblioref">Open Geospatial Consortium, 2023</a>)</span>.</p>
<p>Here we provide an example: (1x 1) tiles of the soil pH, measured in 1:1 soil:water ratio, over the 0-5 cm depth slice of an area in central NY State (USA):</p>
<ol type="1">
<li>predictive map produced by the SoilGrids v2.0 project <span class="citation" data-cites="Poggio.etal2021a">(<a href="#ref-Poggio.etal2021a" role="doc-biblioref">Poggio et al., 2021</a>)</span>;</li>
<li>a rasterized gNATSGO <span class="citation" data-cites="nrcssoilsGriddedNationalSoil2023">(<a href="#ref-nrcssoilsGriddedNationalSoil2023" role="doc-biblioref">NRCS Soils, 2023</a>)</span> coverage of the same area, downscaled to match the SoilGrids v2.0 resolution.</li>
</ol>
<p>(Further on we examine the effect of <a href="#scale">scale</a> for the gNATSGO product, and the <a href="#incove">co-occurrence of two soil properties</a> from SoilGrids v2.0)</p>
<p>You can download similar files as GeoTiff for an area of your preference; see the scripts <code>gNATSGO_WCS_import.Rmd</code> and <code>SoilGrids v2.0_import.Rmd</code>.</p>
<p>We process these in R with the <code>terra</code> package, which has the advantage that it only loads into computer memory as needed, and can load lower resolution automatically if that’s appropriate.</p>
<p>Task: Import these as <code>terra::SpatRaster</code> objects.</p>
<p>Of course, change the file names if you have downloaded different files (tile, property, and/or depth slice).</p>
<!--# Replace with a New Mexico example? -->
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>file.dir <span class="ot">&lt;-</span> <span class="fu">path.expand</span>(<span class="st">"~/ds_reference/Compare_DSM/"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>(gn <span class="ot">&lt;-</span> <span class="fu">rast</span>(<span class="fu">paste0</span>(file.dir, <span class="st">"gNATSGO/lat4243_lon-77-76/ph1to1h2o_r_05_250.tif"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class       : SpatRaster 
dimensions  : 411, 410, 1  (nrow, ncol, nlyr)
resolution  : 0.002434898, 0.002434898  (x, y)
extent      : -76.99926, -76.00095, 41.99973, 43.00048  (xmin, xmax, ymin, ymax)
coord. ref. : lon/lat WGS 84 (EPSG:4326) 
source      : ph1to1h2o_r_05_250.tif 
name        : ph1to1h2o_r 
min value   :        4.60 
max value   :        7.62 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>(sg <span class="ot">&lt;-</span> <span class="fu">rast</span>(<span class="fu">paste0</span>(file.dir, <span class="st">"SoilGrids250/lat4243_lon-77-76/phh2o_0-5cm_mean.tif"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class       : SpatRaster 
dimensions  : 426, 426, 1  (nrow, ncol, nlyr)
resolution  : 0.002349867, 0.002349867  (x, y)
extent      : -77.00071, -75.99967, 41.99918, 43.00022  (xmin, xmax, ymin, ymax)
coord. ref. : lon/lat WGS 84 (EPSG:4326) 
source      : phh2o_0-5cm_mean.tif 
name        : phh2o_0-5cm_mean 
min value   :         43.11129 
max value   :         71.58004 </code></pre>
</div>
</div>
<p>The SoilGrids map is in units of ph x 10 (to store one decimal place as an integer), so we divide the values by 10 to match the gNATSGO product:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">values</span>(sg) <span class="ot">&lt;-</span> <span class="fu">values</span>(sg)<span class="sc">/</span><span class="dv">10</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Task: Plot the two maps side-by-side, on the same value and colour scale.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>range.sg.gn <span class="ot">&lt;-</span> <span class="fu">range</span>(<span class="fu">range</span>(<span class="fu">values</span>(sg), <span class="at">na.rm =</span> <span class="cn">TRUE</span>), </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>                      <span class="fu">range</span>(<span class="fu">values</span>(gn), <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">plot</span>(sg, <span class="at">main =</span> <span class="st">"SoilGrids v2.0"</span>, </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">range =</span> range.sg.gn, <span class="at">col=</span>(sp<span class="sc">::</span><span class="fu">bpy.colors</span>(<span class="dv">50</span>)))</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">plot</span>(gn, <span class="at">main =</span> <span class="st">"gNATSGO"</span>, </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">range =</span> range.sg.gn, <span class="at">col=</span>(sp<span class="sc">::</span><span class="fu">bpy.colors</span>(<span class="dv">50</span>)))</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-sg-gn" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/fig-sg-gn-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Figure&nbsp;2: pH, 0-5 cm</figcaption>
</figure>
</div>
</div>
</div>
<p>We see a wide range of values, especially in the SoilGrids map, and quite different patterns.</p>
<p>Q: Describe the principal differences between the two maps.</p>
<section id="crop-to-a-smaller-area" class="level3" data-number="3.3.1">
<h3 data-number="3.3.1" class="anchored" data-anchor-id="crop-to-a-smaller-area"><span class="header-section-number">3.3.1</span> Crop to a smaller area</h3>
<p>For quicker computation, we restrict the maps (1x 1) to a quarter-map (0.25x 0.25), centred to show some interesting patterns.</p>
<p>Task: Crop the two maps to a quarter-map.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>test.tile.size <span class="ot">&lt;-</span> <span class="fl">0.25</span>  <span class="co"># degrees</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>test.tile.x.offset <span class="ot">&lt;-</span> <span class="fl">0.2</span> <span class="co"># lrc west from right edge</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>test.tile.y.offset <span class="ot">&lt;-</span> <span class="fl">0.3</span>  <span class="co"># lrc north from bottom edge</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>ext.crop <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">as.vector</span>(<span class="fu">ext</span>(sg)),<span class="dv">2</span>) <span class="co"># line up to .00 decimal degrees</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>ext.crop[<span class="st">"xmax"</span>] <span class="ot">&lt;-</span> ext.crop[<span class="st">"xmax"</span>] <span class="sc">-</span> test.tile.x.offset</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>ext.crop[<span class="st">"xmin"</span>] <span class="ot">&lt;-</span> ext.crop[<span class="st">"xmax"</span>] <span class="sc">-</span> test.tile.size</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>ext.crop[<span class="st">"ymin"</span>] <span class="ot">&lt;-</span> ext.crop[<span class="st">"ymin"</span>] <span class="sc">+</span> test.tile.y.offset</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>ext.crop[<span class="st">"ymax"</span>] <span class="ot">&lt;-</span> ext.crop[<span class="st">"ymin"</span>] <span class="sc">+</span> test.tile.size</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ext</span>(ext.crop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SpatExtent : -76.45, -76.2, 42.3, 42.55 (xmin, xmax, ymin, ymax)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>gn <span class="ot">&lt;-</span> <span class="fu">crop</span>(gn, <span class="fu">ext</span>(ext.crop));sg <span class="ot">&lt;-</span> <span class="fu">crop</span>(sg, <span class="fu">ext</span>(ext.crop))</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ext</span>(gn)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SpatExtent : -76.4489762279447, -76.2006165924792, 42.2992262003736, 42.550020734226 (xmin, xmax, ymin, ymax)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ext</span>(sg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SpatExtent : -76.450841145456, -76.1994053643854, 42.2999613764399, 42.5490472903977 (xmin, xmax, ymin, ymax)</code></pre>
</div>
</div>
<p>Notice that the two extents are not exactly the same because of the different alignments of the pixels in the sources.</p>
<p>Task: Plot the two quarter-tile maps side by side, with a common legend.</p>
<p>The value ranges are ifferent, so we need to set up a common scale for the visualization.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>(range.sg <span class="ot">&lt;-</span> <span class="fu">range</span>(<span class="fu">values</span>(sg), <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4.430727 6.386996</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>(range.gn <span class="ot">&lt;-</span> <span class="fu">range</span>(<span class="fu">values</span>(gn), <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4.80 7.53</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>range.sg.gn <span class="ot">&lt;-</span> <span class="fu">range</span>(<span class="fu">range</span>(<span class="fu">values</span>(sg), <span class="at">na.rm =</span> <span class="cn">TRUE</span>), </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>                     <span class="fu">range</span>(<span class="fu">values</span>(gn), <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sg, <span class="at">main =</span> <span class="st">"SoilGrids v2.0"</span>, </span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">range =</span> range.sg.gn, <span class="at">col=</span>(sp<span class="sc">::</span><span class="fu">bpy.colors</span>(<span class="dv">50</span>)))</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(gn, <span class="at">main =</span> <span class="st">"gNATSGO"</span>, </span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">range =</span> range.sg.gn, <span class="at">col=</span>(sp<span class="sc">::</span><span class="fu">bpy.colors</span>(<span class="dv">50</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/crop.2-1.png" class="img-fluid" width="1152"></p>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Q: Describe the differences between the two quarter-maps, both the values and the patterns.</p>
<p>Q: Will the obvious difference in values affect the analysis of their patterns?</p>
</section>
<section id="transform-to-a-metric-crs" class="level3" data-number="3.3.2">
<h3 data-number="3.3.2" class="anchored" data-anchor-id="transform-to-a-metric-crs"><span class="header-section-number">3.3.2</span> Transform to a metric CRS</h3>
<p>Landscape metrics require approximately <em>equal-area</em> grid cells, so these maps, currently in a geographic Coördinate Reference System (CRS), must be projected to a metric system. A reasonable choice for any small areais the Universal Transmercator (UTM) system, which covers a 6-wide latitude range.</p>
<p>Task: Search for the appropriate EPSG code at the <a href="https://epsg.org">EPSG Geodetic Parameter Dataset</a>, see (<a href="#fig-epsg">Figure&nbsp;3</a>).</p>
<div id="fig-epsg" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="./figs/EPSG32618screen.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;3: EPSG database entry for code 32618</figcaption>
</figure>
</div>
<p>Easiest is to use “Map Search” and further limit the results by the text “UTM”. Several datums can serve as the basis for the UTM CRS; for easiest conversion select WGS84. In the USA these have the format <code>326xx</code>, where <code>xx</code> is the UTM zone number.</p>
<p>Determine the UTM zone and the corresponding EPSG code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># a function to find the correct UTM zome</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>long2UTM <span class="ot">&lt;-</span> <span class="cf">function</span>(long) { (<span class="fu">floor</span>((long <span class="sc">+</span> <span class="dv">180</span>)<span class="sc">/</span><span class="dv">6</span>) <span class="sc">%%</span> <span class="dv">60</span>) <span class="sc">+</span> <span class="dv">1</span> }</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># find the zone from the central meridian</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>utm.zone <span class="ot">&lt;-</span> <span class="fu">long2UTM</span>(<span class="fu">st_bbox</span>(sg)<span class="sc">$</span>xmin <span class="sc">+</span> </span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>                       <span class="fl">0.5</span><span class="sc">*</span>(<span class="fu">st_bbox</span>(sg)<span class="sc">$</span>xmax <span class="sc">-</span> <span class="fu">st_bbox</span>(sg)<span class="sc">$</span>xmin))</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">paste</span>(<span class="st">"UTM Zone"</span>, utm.zone))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>UTM Zone 18</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>epsg.utm <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"epsg:326"</span>, utm.zone)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">paste</span>(<span class="st">"CRS code:"</span>, epsg.utm))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>CRS code: epsg:32618</code></pre>
</div>
</div>
<p>Task: Resample the maps to the UTM projection, at nominal 250 m grid cell resolution.</p>
<p>Notes:</p>
<ol type="1">
<li>The interpolation method used by <code>terra::project</code> is, by default, bilinear. This is appropriate for continuous-valued maps.</li>
<li>Specify the grid cell size with the <code>res</code> argument to <code>terra::project</code>. Both maps were nominally at this scale, although presented in geographical coördinates. If this is omitted, <code>terra::project</code> sets the size as a square with the smaller dimension, here at <span class="math inline">\(\approx 43^{0}N\)</span> this is 229~m.</li>
</ol>
<p>The bounding boxes and resolutions are slightly different for the two products.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">st_bbox</span>(gn)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     xmin      ymin      xmax      ymax 
-76.44898  42.29923 -76.20062  42.55002 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">res</span>(gn)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.002434898 0.002434898</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>gn.utm <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">project</span>(gn, epsg.utm, </span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">res =</span> <span class="fu">c</span>(<span class="dv">250</span>, <span class="dv">250</span>), <span class="at">method =</span> <span class="st">"bilinear"</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="fu">st_bbox</span>(gn.utm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     xmin      ymin      xmax      ymax 
 380561.8 4683745.2  401561.8 4711745.2 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">res</span>(gn.utm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 250 250</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">st_bbox</span>(sg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     xmin      ymin      xmax      ymax 
-76.45084  42.29996 -76.19941  42.54905 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>sg.utm <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">project</span>(sg, epsg.utm, </span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">res =</span> <span class="fu">c</span>(<span class="dv">250</span>, <span class="dv">250</span>), <span class="at">method =</span> <span class="st">"bilinear"</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">st_bbox</span>(sg.utm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     xmin      ymin      xmax      ymax 
 380409.5 4683779.9  401409.5 4711779.9 </code></pre>
</div>
</div>
</section>
<section id="harmonizing-the-mapped-areas" class="level3" data-number="3.3.3">
<h3 data-number="3.3.3" class="anchored" data-anchor-id="harmonizing-the-mapped-areas"><span class="header-section-number">3.3.3</span> Harmonizing the mapped areas</h3>
<p>The two maps have slightly different concepts of areas not mapped: unsurveyed urban areas and water bodies (both sources) and miscellaneous land types such as mines or gravel pits (gNATSGO). SoilGrids v2.0 identified these by remote sensing, whereas gNATSGO used field survey and (for urban areas) survey policy.</p>
<p>The maps also have slightly different extents. These must be made identical before masking. We select one map as a template and resample the other map into that template.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ext</span>(gn.utm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SpatExtent : 380561.847574791, 401561.847574791, 4683745.2384707, 4711745.2384707 (xmin, xmax, ymin, ymax)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ext</span>(sg.utm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SpatExtent : 380409.511975787, 401409.511975787, 4683779.92513894, 4711779.92513894 (xmin, xmax, ymin, ymax)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>sg.utm <span class="ot">&lt;-</span> <span class="fu">resample</span>(sg.utm, gn.utm)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ext</span>(sg.utm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SpatExtent : 380561.847574791, 401561.847574791, 4683745.2384707, 4711745.2384707 (xmin, xmax, ymin, ymax)</code></pre>
</div>
</div>
<p>Notice the small changes in the range: the resampling has slightly lowered the extremes.</p>
<p>For the pattern analysis we want these <code>NA</code> areas to be the same. For this we use a reciprocal mask.</p>
<p>Task: mask each map with the <code>NA</code> areas of the other.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>sg.utm <span class="ot">&lt;-</span> <span class="fu">mask</span>(sg.utm, gn.utm)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="co"># SoilGrids now has some `NA` added from gNATSGO</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>gn.utm <span class="ot">&lt;-</span> <span class="fu">mask</span>(gn.utm, sg.utm)</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="co"># The added `NA` are already in gNATSGO, now it gets `NA` originally on SoilGrids</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Task: Plot the two maps side-by-side.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sg.utm, <span class="at">main =</span> <span class="st">"SoilGrids v2.0"</span>, </span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">range =</span> range.sg.gn, <span class="at">col=</span>(sp<span class="sc">::</span><span class="fu">bpy.colors</span>(<span class="dv">50</span>)))</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(gn.utm, <span class="at">main =</span> <span class="st">"gNATSGO"</span>, </span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">range =</span> range.sg.gn, <span class="at">col=</span>(sp<span class="sc">::</span><span class="fu">bpy.colors</span>(<span class="dv">50</span>)))</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-sg-gn-utm" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/fig-sg-gn-utm-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Figure&nbsp;4: pH, 0-5 cm</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="sec-import-gssurgo" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="sec-import-gssurgo"><span class="header-section-number">3.4</span> Example dataset: Soil survey map unit polygons</h2>
<p>These will be used when examining the effects of scale in (<a href="#sec-scale-geometry">Section&nbsp;9</a>) and (<a href="#sec-categorical-generalization">Section&nbsp;10</a>).</p>
<p>In the USA the polygon maps, as delineated by the soil surveyors and later digitized as vector GIS coverages, are available in the SSURGO and STATSGO databases. These can be accessed with the <code>SDA_spatialQuery()</code> function of the Soil Data Access (SDA) facility of the <code>soilDB</code> package written by NRCS scientists, to allow R access to the NRCS database’s REST/JSON web service. NCSS has written a tutorial on SDA<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>.</p>
<p>We specify the <code>geomIntersection = TRUE</code> argument to clip map unit polygons to the bounding polygon. The bounding box (in WGS84 geographic coordinates) and CRS must be obtained from a spatial object. For this we use SSURGO gridded map <code>gn.utm</code>, a <code>terra::SpatRaster</code> which we created above, as the template. This ensures that the same package is used for the returned object. Since this is a polygon map, it will be a <code>terra::SpatVector</code> object.</p>
<p>Task: Download the map unit polygons for the restricted study area. This may take a few minutes depending on how responsive is the remote server.</p>
<p>The bounding box is specified with the second argument to <code>SDA_spatialQuery</code>; here we have the quarter-degree products from the continuous DSM to serve as the CRS and bounding box. We use the gNATSGO <code>SpatRaster</code> (object <code>gn</code>) for this.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the polygons with their key</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">system.time</span>(</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  mu.poly <span class="ot">&lt;-</span> <span class="fu">SDA_spatialQuery</span>(gn, </span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">what =</span> <span class="st">"mupolygon"</span>, </span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>                            <span class="at">db =</span> <span class="st">"SSURGO"</span>, </span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>                            <span class="at">geomIntersection =</span> <span class="cn">TRUE</span>)</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   user  system elapsed 
  2.276   0.169  19.723 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(mu.poly)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SpatVector"
attr(,"package")
[1] "terra"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">st_crs</span>(mu.poly)<span class="sc">$</span>proj4string</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "+proj=longlat +datum=WGS84 +no_defs"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(mu.poly)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     mukey            area_ac        
 Min.   : 295575   Min.   :  0.0022  
 1st Qu.: 295602   1st Qu.:  2.9416  
 Median : 295649   Median :  6.6646  
 Mean   : 621141   Mean   : 15.1426  
 3rd Qu.: 295687   3rd Qu.: 16.1199  
 Max.   :2760841   Max.   :897.0175  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mu.poly)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["mukey"],"name":[1],"type":["int"],"align":["right"]},{"label":["area_ac"],"name":[2],"type":["dbl"],"align":["right"]}],"data":[{"1":"2723081","2":"1.846781"},{"1":"2723081","2":"4.124984"},{"1":"2723081","2":"19.108964"},{"1":"2723081","2":"7.223155"},{"1":"2723081","2":"1.576274"},{"1":"2723081","2":"6.984551"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot with area in acres</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(mu.poly, <span class="at">y =</span> <span class="st">"area_ac"</span>,</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">type =</span> <span class="st">"continuous"</span>,</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"SSURGO map units, area in acres"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/download-ssurgo-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>What do these map units codes represent? Find the map units from the same geometry.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>mu.key <span class="ot">&lt;-</span> <span class="fu">SDA_spatialQuery</span>(gn.utm, </span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">what =</span> <span class="st">"mukey"</span>, </span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">db =</span> <span class="st">"SSURGO"</span>, </span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">geomIntersection =</span> <span class="cn">TRUE</span>)</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mu.key)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["mukey"],"name":[1],"type":["int"],"align":["right"]},{"label":["muname"],"name":[2],"type":["chr"],"align":["left"]}],"data":[{"1":"295575","2":"Alluvial land","_rn_":"1"},{"1":"295576","2":"Arkport fine sandy loam, 2 to 6 percent slopes","_rn_":"2"},{"1":"295577","2":"Arkport fine sandy loam, 6 to 12 percent slopes","_rn_":"3"},{"1":"295578","2":"Bath channery silt loam, 2 to 5 percent slopes","_rn_":"4"},{"1":"295579","2":"Bath channery silt loam, 5 to 15 percent slopes","_rn_":"5"},{"1":"295580","2":"Bath channery silt loam, 5 to 15 percent slopes, eroded","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>These are named map units.</p>
<p>These have extended site data, which comes from the linked attribute database. For this we use an SQL query and the <code>SDA_query</code> (link to “Soil Data Access”) <a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># format the list of map units for SQL</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>IS <span class="ot">&lt;-</span> soilDB<span class="sc">::</span><span class="fu">format_SQL_in_statement</span>(mu.poly<span class="sc">$</span>mukey)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="co"># query string -- all components</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>ws <span class="ot">&lt;-</span> <span class="fu">sprintf</span>(<span class="st">"mukey IN %s"</span>, IS)</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a><span class="co"># format the SQL query</span></span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>query <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"SELECT * FROM mapunit WHERE"</span>, ws)</span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a><span class="co"># and run it</span></span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>mu.info <span class="ot">&lt;-</span> <span class="fu">SDA_query</span>(query)</span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(mu.info)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 217  26</code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(mu.info)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "musym"         "muname"        "mukind"        "mustatus"     
 [5] "muacres"       "mapunitlfw_l"  "mapunitlfw_r"  "mapunitlfw_h" 
 [9] "mapunitpfa_l"  "mapunitpfa_r"  "mapunitpfa_h"  "farmlndcl"    
[13] "muhelcl"       "muwathelcl"    "muwndhelcl"    "interpfocus"  
[17] "invesintens"   "iacornsr"      "nhiforsoigrp"  "nhspiagr"     
[21] "vtsepticsyscl" "mucertstat"    "lkey"          "mukey"        
[25] "museq"         "nationalmusym"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>series.name <span class="ot">&lt;-</span> <span class="st">"Ovid"</span> <span class="co"># look for a map unit by name</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(ix <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">substr</span>(mu.info<span class="sc">$</span>muname, <span class="dv">1</span>, </span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>                          <span class="fu">nchar</span>(series.name)) <span class="sc">==</span> series.name))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>mu.info[ix, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["musym"],"name":[1],"type":["chr"],"align":["left"]},{"label":["muname"],"name":[2],"type":["chr"],"align":["left"]},{"label":["mukind"],"name":[3],"type":["chr"],"align":["left"]},{"label":["mustatus"],"name":[4],"type":["chr"],"align":["left"]},{"label":["muacres"],"name":[5],"type":["int"],"align":["right"]},{"label":["mapunitlfw_l"],"name":[6],"type":["int"],"align":["right"]},{"label":["mapunitlfw_r"],"name":[7],"type":["int"],"align":["right"]},{"label":["mapunitlfw_h"],"name":[8],"type":["int"],"align":["right"]},{"label":["mapunitpfa_l"],"name":[9],"type":["dbl"],"align":["right"]},{"label":["mapunitpfa_r"],"name":[10],"type":["dbl"],"align":["right"]},{"label":["mapunitpfa_h"],"name":[11],"type":["dbl"],"align":["right"]},{"label":["farmlndcl"],"name":[12],"type":["chr"],"align":["left"]},{"label":["muhelcl"],"name":[13],"type":["chr"],"align":["left"]},{"label":["muwathelcl"],"name":[14],"type":["chr"],"align":["left"]},{"label":["muwndhelcl"],"name":[15],"type":["chr"],"align":["left"]},{"label":["interpfocus"],"name":[16],"type":["chr"],"align":["left"]},{"label":["invesintens"],"name":[17],"type":["chr"],"align":["left"]},{"label":["iacornsr"],"name":[18],"type":["int"],"align":["right"]},{"label":["nhiforsoigrp"],"name":[19],"type":["chr"],"align":["left"]},{"label":["nhspiagr"],"name":[20],"type":["dbl"],"align":["right"]},{"label":["vtsepticsyscl"],"name":[21],"type":["chr"],"align":["left"]},{"label":["mucertstat"],"name":[22],"type":["chr"],"align":["left"]},{"label":["lkey"],"name":[23],"type":["int"],"align":["right"]},{"label":["mukey"],"name":[24],"type":["int"],"align":["right"]},{"label":["museq"],"name":[25],"type":["int"],"align":["right"]},{"label":["nationalmusym"],"name":[26],"type":["chr"],"align":["left"]}],"data":[{"1":"OaA","2":"Ovid silt loam, 0 to 6 percent slopes","3":"Consociation","4":"NA","5":"5157","6":"NA","7":"NA","8":"NA","9":"NA","10":"NA","11":"NA","12":"Prime farmland if drained","13":"NA","14":"NA","15":"NA","16":"NA","17":"NA","18":"NA","19":"NA","20":"NA","21":"NA","22":"NA","23":"12437","24":"295666","25":"87","26":"9xnm","_rn_":"73"},{"1":"OcC3","2":"Ovid silty clay loam, 6 to 12 percent slopes eroded","3":"Consociation","4":"NA","5":"312","6":"NA","7":"NA","8":"NA","9":"NA","10":"NA","11":"NA","12":"Not prime farmland","13":"NA","14":"NA","15":"NA","16":"NA","17":"NA","18":"NA","19":"NA","20":"NA","21":"NA","22":"NA","23":"12437","24":"295667","25":"88","26":"9xnn","_rn_":"74"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<section id="transform-to-a-metric-crs-1" class="level3" data-number="3.4.1">
<h3 data-number="3.4.1" class="anchored" data-anchor-id="transform-to-a-metric-crs-1"><span class="header-section-number">3.4.1</span> Transform to a metric CRS</h3>
<p>Project to the metric CRS we are using in this area. The CRS was defined in the previous section.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">st_crs</span>(mu.poly)<span class="sc">$</span>proj4string</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "+proj=longlat +datum=WGS84 +no_defs"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>mu.poly <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">project</span>(mu.poly, epsg.utm)</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="fu">st_crs</span>(mu.poly)<span class="sc">$</span>proj4string</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "+proj=utm +zone=18 +datum=WGS84 +units=m +no_defs"</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="characterizing-patterns" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Characterizing patterns</h1>
<p>Before comparing patterns of different maps, and trying to evaluate how close they are to “reality”, they first have to be characterized by statistical measures. This gives objective information about their spatial patterns.</p>
<p>The methods to characterize patterns are different for maps of <em>continuous</em> variables (<a href="#sec-continuous">Section&nbsp;5</a>) and <em>classified</em> (categorical) variables (<a href="#sec-classified">Section&nbsp;6</a>).</p>
</section>
<section id="sec-continuous" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Characterizing patterns – Continuous</h1>
<p>These are methods that require continuous values on at least an interval scale, and usually a ratio scale (with a true zero). In the case of the example here, pH does not have a true zero, so it is an interval scale. Other properties such as soil thickness to a restricting layer have a true zero, and one can speak of one location being “twice as thick” than another, for example.</p>
<section id="the-global-variogram" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="the-global-variogram"><span class="header-section-number">5.1</span> The global variogram</h2>
<p>The variogram (or a correlogram) can be used to characterize the degree of spatial continuity and the “roughness” of a continuous property map, averaged across the entire map. Note that this depends on the grid cell size in two ways:</p>
<ol type="1">
<li>Any pattern at finer resolutions has been removed;</li>
<li>The values in grid cells may be produced by punctual or block methods. Block methods smooth values, so that the variogram sill will necessarily be lower than for punctual predictions. Also, the range may be longer.</li>
</ol>
<p>In this section we compute and compare the short-range variograms, these reveal the local structure. In DSM maps the variogram is typically unbounded, but we don’t care about the long-range structure when we are evaluating patterns. The parameters of the local structure characterize the fine-scale variability.</p>
<p>Note: Variograms are typically produced separately for each mapped soil property. To characterize an inherent landscape scale, a number of properties can be combined by principal component analysis (PCA) and the first component (PC1) can be characterized.</p>
<p>Task: Convert the <code>terra::SpatRaster</code> objects to <code>raster::raster</code> and then to <code>sf:sf</code> objects in order to compute variograms.</p>
<p>Note: There is (so far) no direct conversion. The <code>gstat::variogram</code> method must be applied to an object of class <code>sp</code> or <code>sf</code>, not directly to a <code>terra::SpatRaster</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>gn.sp <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">raster</span>(gn.utm), <span class="st">"SpatialPointsDataFrame"</span>)</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>gn.sf <span class="ot">&lt;-</span> <span class="fu">st_as_sf</span>(gn.sp)</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(gn.sf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ph1to1h2o_r" "geometry"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>sg.sp <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">raster</span>(sg.utm), <span class="st">"SpatialPointsDataFrame"</span>)</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>sg.sf <span class="ot">&lt;-</span> <span class="fu">st_as_sf</span>(sg.sp)</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sg.sf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "phh2o_0.5cm_mean" "geometry"        </code></pre>
</div>
</div>
<p>Task: Set the initial parameters for empirical variogram as the resolution.</p>
<p>If the bin width is the resolution, we get one-grid-cell spatial correlations.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>range.init <span class="ot">&lt;-</span> <span class="dv">1000</span>  <span class="co"># estimated range, m </span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>cutoff.init <span class="ot">&lt;-</span> range.init<span class="sc">*</span><span class="dv">5</span>  <span class="co"># cutoff for empirical variogram, m</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>width.init <span class="ot">&lt;-</span> <span class="dv">250</span>   <span class="co"># bin width</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Task: Compute and display the empirical variograms.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>v.sg <span class="ot">&lt;-</span> <span class="fu">variogram</span>(phh2o_0<span class="fl">.5</span>cm_mean <span class="sc">~</span> <span class="dv">1</span>, <span class="at">loc =</span> sg.sf, </span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">cutoff=</span>cutoff.init, <span class="at">width=</span>width.init)</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>v.gn <span class="ot">&lt;-</span> gstat<span class="sc">::</span><span class="fu">variogram</span>(ph1to1h2o_r <span class="sc">~</span> <span class="dv">1</span>, <span class="at">loc =</span> gn.sf, </span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>                         <span class="at">cutoff=</span>cutoff.init, <span class="at">width=</span>width.init)</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>ylim.v <span class="ot">&lt;-</span> <span class="fu">max</span>(v.gn<span class="sc">$</span>gamma, v.sg<span class="sc">$</span>gamma)</span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">plot</span>(v.sg, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, ylim.v), <span class="at">main =</span> <span class="st">"SoilGrids v2.0"</span>)</span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">plot</span>(v.gn, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, ylim.v), <span class="at">main =</span> <span class="st">"gNATSGO"</span>)</span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a><span class="fu">grid.arrange</span>(p1, p2, <span class="at">nrow =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/variogram_empirical-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Q: Describe the empirical differences in spatial structure of the two maps.</p>
<p>Task: Fit a variogram model to the empirical variogram.</p>
<p>The differences can be quantified by the parameters of a fitted variogram model. We try an exponential model because (1) it has the simplest theory, and (2) we expect to not reach a sill within the short range investigated.</p>
<p>We use the <code>fit.variogram</code> method to adjust an initial estimate by weighted least squared. The estimated sill is the maximum <span class="math inline">\(\gamma\)</span> in the empirical variogram.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>vm.gn <span class="ot">&lt;-</span> <span class="fu">vgm</span>(<span class="fl">0.8</span><span class="sc">*</span><span class="fu">max</span>(v.gn<span class="sc">$</span>gamma), <span class="st">"Exp"</span>, range.init, <span class="dv">0</span>)</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>(vmf.gn <span class="ot">&lt;-</span> <span class="fu">fit.variogram</span>(v.gn, <span class="at">model=</span>vm.gn))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["model"],"name":[1],"type":["fct"],"align":["left"]},{"label":["psill"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["range"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["kappa"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["ang1"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["ang2"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["ang3"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["anis1"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["anis2"],"name":[9],"type":["dbl"],"align":["right"]}],"data":[{"1":"Nug","2":"0.0000000","3":"0.0000","4":"0.0","5":"0","6":"0","7":"0","8":"1","9":"1","_rn_":"1"},{"1":"Exp","2":"0.1240257","3":"449.6727","4":"0.5","5":"0","6":"0","7":"0","8":"1","9":"1","_rn_":"2"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>vm.sg <span class="ot">&lt;-</span> <span class="fu">vgm</span>(<span class="fl">0.8</span><span class="sc">*</span><span class="fu">max</span>(v.sg<span class="sc">$</span>gamma), <span class="st">"Exp"</span>, range.init, <span class="dv">0</span>)</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>(vmf.sg <span class="ot">&lt;-</span> <span class="fu">fit.variogram</span>(v.sg, <span class="at">model=</span>vm.sg))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["model"],"name":[1],"type":["fct"],"align":["left"]},{"label":["psill"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["range"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["kappa"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["ang1"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["ang2"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["ang3"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["anis1"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["anis2"],"name":[9],"type":["dbl"],"align":["right"]}],"data":[{"1":"Nug","2":"0.0000000","3":"0.000","4":"0.0","5":"0","6":"0","7":"0","8":"1","9":"1","_rn_":"1"},{"1":"Exp","2":"0.1679165","3":"3960.544","4":"0.5","5":"0","6":"0","7":"0","8":"1","9":"1","_rn_":"2"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">plot</span>(v.sg, <span class="at">model=</span>vmf.sg, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, ylim.v), <span class="at">main =</span> <span class="st">"SoilGrids v2.0"</span>, </span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"separation m"</span>, <span class="at">ylab =</span> <span class="fu">expression</span>(<span class="fu">paste</span>(Delta, <span class="fu">plain</span>(pH)<span class="sc">^</span><span class="dv">2</span>)))</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">plot</span>(v.gn, <span class="at">model=</span>vmf.gn, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, ylim.v), <span class="at">main =</span> <span class="st">"gNATSGO"</span>, </span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"separation m"</span>, <span class="at">ylab =</span> <span class="fu">expression</span>(<span class="fu">paste</span>(Delta, <span class="fu">plain</span>(pH)<span class="sc">^</span><span class="dv">2</span>)))</span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a><span class="fu">grid.arrange</span>(p1, p2, <span class="at">nrow =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/fitted_variogram_model-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Q: How well do the fitted models match the empirical variograms? If the fit has some problems, what could be a solution?</p>
<p>Q: Describe the modelled differences in spatial structure of the two maps (total sill, range).</p>
<p>Q: What is the implication for the utility of the maps?</p>
<p>Q: Is there any way to decide which is “better” in some sense?</p>
</section>
<section id="moving-window-local-association" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="moving-window-local-association"><span class="header-section-number">5.2</span> Moving-window local association</h2>
<p>The local spatial structure may not be consistent across the mapped area, so that the average variogram, computed over that area, can be misleading. With so many values (grid cells) it’s possible to compute moving-window variograms, as in the VESPER program <span class="citation" data-cites="minasnyVESPERVariogramEstimation2005">(<a href="#ref-minasnyVESPERVariogramEstimation2005" role="doc-biblioref">Minasny et al., 2005</a>)</span> developed for precision agriculture applications. This will show if the pattern is consistent across the map, and also allow maps to be compared block-by-block. I have not (yet?) implemented this in R, so we will use another method to assess moving-window local spatial association.</p>
<p>A quick way to see the local degree of autocorrelation is with Moran’s I applied to a window of appropriate size around each grid cell, using the <code>terra::autocor</code> function.</p>
<p>Moran’s I is defined as:</p>
<p><span class="math display">\[  I = \frac{n}{\sum_i \sum_j w_{ij}} \frac{\sum_i \sum_j w_{ij}(y_i - \bar{y})(y_j - \bar{y})}{\sum_i (y_i - \bar{y})^2}\]</span></p>
<p>where <span class="math inline">\(y_i\)</span> is the value of the variable in the <span class="math inline">\(i\)</span>th of <span class="math inline">\(n\)</span> neighbouring grid cells, <span class="math inline">\(\bar{y}\)</span> is the global mean of the variable, <span class="math inline">\(w_{ij}\)</span> is the spatial of the link between the target cell <span class="math inline">\(i\)</span> and its neighbour cell <span class="math inline">\(j\)</span>. The expected value of Moran’s I is <span class="math inline">\(-1/(n-1)\)</span> if the pattern of the response variable is random, i.e., no spatial correlation. So for a 5<sub>x</sub>5 neighbourhood the expected value if random is <span class="math inline">\(-1/24 = -0.041\bar{6} \approx 0\)</span>.</p>
<p>The second term numerator is the weighted covariance; its denominator normalizes by the variance. The first term normalizes by the sum of all weights, so that the test is comparable among tests with different numbers of neighbours and using different weightings.</p>
<p>Task: Construct a weights matrix for local Moran’s I</p>
<p>We determine the weights matrix for Moran’s I from the global variogram analysis and the grid cell size.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for a 5x5 matrix</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="co"># there must be a more elegant way to do this!</span></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>(vl <span class="ot">&lt;-</span> <span class="fu">variogramLine</span>(vm.sg, </span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">dist_vector =</span> <span class="fu">c</span>(<span class="dv">250</span>, <span class="dv">250</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">2</span>), </span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>                                     <span class="dv">500</span>, <span class="dv">250</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">5</span>), </span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>                                     <span class="dv">500</span><span class="sc">*</span><span class="fu">sqrt</span>(<span class="dv">2</span>))))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["dist"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["gamma"],"name":[2],"type":["dbl"],"align":["right"]}],"data":[{"1":"250.0000","2":"0.02045734"},{"1":"353.5534","2":"0.02754274"},{"1":"500.0000","2":"0.03638954"},{"1":"559.0170","2":"0.03960426"},{"1":"707.1068","2":"0.04688293"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>(w.r <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(vl<span class="sc">$</span>gamma <span class="sc">/</span> vl<span class="sc">$</span>gamma[<span class="dv">1</span>]))  <span class="co"># relative weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.0000000 0.7427491 0.5621765 0.5165440 0.4363495</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>(w.m <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(w.r[<span class="dv">5</span>], w.r[<span class="dv">4</span>], w.r[<span class="dv">3</span>], w.r[<span class="dv">4</span>], w.r[<span class="dv">5</span>],</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>                w.r[<span class="dv">5</span>], w.r[<span class="dv">2</span>], w.r[<span class="dv">1</span>], w.r[<span class="dv">2</span>], w.r[<span class="dv">5</span>],</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>                w.r[<span class="dv">3</span>], w.r[<span class="dv">1</span>], <span class="dv">0</span>, w.r[<span class="dv">1</span>], w.r[<span class="dv">3</span>],</span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>                w.r[<span class="dv">5</span>], w.r[<span class="dv">2</span>], w.r[<span class="dv">1</span>], w.r[<span class="dv">2</span>], w.r[<span class="dv">5</span>],</span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>                w.r[<span class="dv">5</span>], w.r[<span class="dv">4</span>], w.r[<span class="dv">3</span>], w.r[<span class="dv">4</span>], w.r[<span class="dv">5</span>]), </span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">nrow =</span> <span class="dv">5</span>, <span class="at">ncol =</span> <span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          [,1]      [,2]      [,3]      [,4]      [,5]
[1,] 0.4363495 0.4363495 0.5621765 0.4363495 0.4363495
[2,] 0.5165440 0.7427491 1.0000000 0.7427491 0.5165440
[3,] 0.5621765 1.0000000 0.0000000 1.0000000 0.5621765
[4,] 0.5165440 0.7427491 1.0000000 0.7427491 0.5165440
[5,] 0.4363495 0.4363495 0.5621765 0.4363495 0.4363495</code></pre>
</div>
</div>
<p>Task: Compute and display the moving-window autocorrelation.</p>
<p>This uses the <code>terra::autocor</code> method, applied to a weighted window.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>sg.utm.autocor <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">autocor</span>(sg.utm, <span class="at">w=</span>w.m, </span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">method=</span><span class="st">"moran"</span>, <span class="at">global =</span> <span class="cn">FALSE</span>)</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>gn.utm.autocor <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">autocor</span>(gn.utm, <span class="at">w=</span>w.m, </span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">method=</span><span class="st">"moran"</span>, <span class="at">global =</span> <span class="cn">FALSE</span>)</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>(range.sg.autocor <span class="ot">&lt;-</span> <span class="fu">range</span>(<span class="fu">values</span>(sg.utm.autocor), <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -0.2621647  6.2650773</code></pre>
</div>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>(range.gn.autocor <span class="ot">&lt;-</span> <span class="fu">range</span>(<span class="fu">values</span>(gn.utm.autocor), <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -1.937685  6.388628</code></pre>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>range.autocor <span class="ot">&lt;-</span> <span class="fu">range</span>(range.sg.autocor, range.gn.autocor)</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="co"># hcl.pals(type = "diverging")</span></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">plot</span>(sg.utm.autocor, <span class="at">main =</span> <span class="st">"SG250, Moran's I, 5x5"</span>, </span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">range =</span> range.autocor, <span class="at">col =</span> <span class="fu">rev</span>(<span class="fu">hcl.colors</span>(<span class="dv">32</span>, <span class="at">palette =</span> <span class="st">"RdYlGn"</span>)))</span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">plot</span>(gn.utm.autocor, <span class="at">main =</span> <span class="st">"gNATSGO, Moran's I, 5x5"</span>, </span>
<span id="cb85-7"><a href="#cb85-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">range =</span> range.autocor, <span class="at">col =</span> <span class="fu">rev</span>(<span class="fu">hcl.colors</span>(<span class="dv">32</span>, <span class="at">palette =</span> <span class="st">"RdYlGn"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/moving-window-5-1.png" class="img-fluid" width="1152"></p>
</div>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To appreciate the local Moran’s I values, here are the global Moran’s I with the same weights matrix. These are the averages of all the local (window) Moran’s I.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">autocor</span>(sg.utm, <span class="at">w=</span>w.m, <span class="at">method=</span><span class="st">"moran"</span>, <span class="at">global =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phh2o_0-5cm_mean 
        1.083992 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">autocor</span>(gn.utm, <span class="at">w=</span>w.m, <span class="at">method=</span><span class="st">"moran"</span>, <span class="at">global =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ph1to1h2o_r 
  0.5343023 </code></pre>
</div>
</div>
<p>These are both very far from the random value <span class="math inline">\(-0.041\bar{6}\)</span>. Both maps show hot spots with much larger local autocorrelation, and some areas with almost none or even more dispersed than random (negative values).</p>
<p>Q: Is the pattern of local autocorrelation the same across the map?</p>
<p>Q: Which map has larger differences?</p>
</section>
</section>
<section id="sec-classified" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Characterizing patterns – Classified</h1>
<p>The spatial unit of conventional (legacy) maps is the polygon, not the grid cell. These maps show a discrete number of legend entries (classes), each with one to many polygons. In the soil survey context these are called <strong>mapping units</strong>, and generally are soil classes, possibly with some landscape features (e.g., erosion class, slope class) as part of the definition. Some mapping units may represent water bodies and various other kinds of non-soil.</p>
<p><a href="#fig-anderson">Figure&nbsp;5</a> shows a typical polygon map from a legacy “land condition” survey <span class="citation" data-cites="soilconservationservicePhysicalLandConditions1951">(<a href="#ref-soilconservationservicePhysicalLandConditions1951" role="doc-biblioref">Soil Conservation Service, 1951</a>)</span>. All polygons with the same label (e.g., “337D22”) refer to a single mapping unit, in this case composed of an erosion class, slope class, and soil type.</p>
<div id="fig-anderson" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="./figs/UpperSavannahSCD_Sheet4_colour_NWcorner.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;5: Land condition, Anderson County SC (USA)</figcaption>
</figure>
</div>
<p>It’s easiest to work with maps already in digital format. The area of the legacy map has been updated and digitized, see <a href="#fig-anderson-wss">Figure&nbsp;6</a>. These polygons can be downloaded in various GIS formats, see <a href="#sec-import-gssurgo">Section&nbsp;3.4</a>, above</p>
<div id="fig-anderson-wss" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="./figs/WebSoilSurvey_AndersonCountyNW.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;6: Anderson County SC (USA)</figcaption>
</figure>
</div>
<p>But here we continue with the continuous property maps of a single property. To use these techniques on continuous property maps, the maps must be <strong>sliced</strong> (discretized) into classes. There are several choices:</p>
<ul>
<li>meaningful limits, matching some thresholds known to be important for a soil function;</li>
<li>equal intervals;</li>
<li>histogram equalization.</li>
</ul>
<p>For equal intervals or histogram equalization, the cutpoints should be the same for all maps, and therefore derived from their combined distribution of values. We illustrate the process here, but do not use it for the landscape metrics examples later on in the tutorial.</p>
<section id="sec-hist-equal" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="sec-hist-equal"><span class="header-section-number">6.1</span> Classifying by histogram equalization</h2>
<p>This section shows how to classify by histogram equalization; the results will not be used later in the tutorial. Instead, we will use meaningful limits (see <a href="#sec-mean-limit">Section&nbsp;6.2</a>) to slice the map.</p>
<p>Task: Slice the two maps by histogram equalization</p>
<p>First, compute the histogram equalization and display the limits on a histogram plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>n.class <span class="ot">&lt;-</span> <span class="dv">8</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="co"># combined values</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>values.all <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">values</span>(gn.utm),</span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>                <span class="fu">values</span>(sg.utm))</span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>values.all.sort <span class="ot">&lt;-</span> <span class="fu">sort</span>(values.all)</span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">length</span>(values.all) <span class="sc">-</span> <span class="fu">sum</span>(<span class="fu">is.na</span>(values.all))</span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a>(cut.positions <span class="ot">&lt;-</span> <span class="fu">round</span>(n<span class="sc">/</span>n.class))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2237</code></pre>
</div>
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>(cuts <span class="ot">&lt;-</span> values.all.sort[cut.positions <span class="sc">*</span> <span class="dv">1</span><span class="sc">:</span>(n.class<span class="dv">-1</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4.969114 5.210513 5.425628 5.573456 5.721922 5.874367 6.078163</code></pre>
</div>
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(values.all, <span class="at">breaks=</span><span class="dv">36</span>, <span class="at">main=</span><span class="st">"Histogram equalization"</span>,</span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"pH"</span>)</span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span>cuts, <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">lwd=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/histo-equal-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Q: How well do these represent the distributions on the two maps individually?</p>
<p>To answer this, compare their histograms with the equalization slices.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">values</span>(gn.utm), <span class="at">breaks=</span><span class="dv">36</span>, <span class="at">main=</span><span class="st">"gNATSGO"</span>,</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"pH"</span>)</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span>cuts, <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">values</span>(sg.utm), <span class="at">breaks=</span><span class="dv">36</span>, <span class="at">main=</span><span class="st">"SoilGrids v2.0"</span>,</span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"pH"</span>)</span>
<span id="cb96-7"><a href="#cb96-7" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span>cuts, <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">lwd=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/histo-equal-one-by-one-1.png" class="img-fluid" width="1344"></p>
</div>
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Task: slice the maps and display with a common colour ramp.</p>
<p>Find the cutpoints and set up the colour ramp:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>(zlim <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">min</span>(values.all, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>                <span class="fu">max</span>(values.all, <span class="at">na.rm=</span><span class="cn">TRUE</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4.462665 7.404359</code></pre>
</div>
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>(cut.names <span class="ot">&lt;-</span> <span class="fu">cut</span>(zlim, <span class="at">breaks=</span><span class="fu">c</span>(zlim[<span class="dv">1</span>], cuts, zlim[<span class="dv">2</span>]),</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ordered_result=</span><span class="cn">TRUE</span>, <span class="at">include.lowest =</span> <span class="cn">TRUE</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] [4.46,4.97] (6.08,7.4] 
8 Levels: [4.46,4.97] &lt; (4.97,5.21] &lt; (5.21,5.43] &lt; ... &lt; (6.08,7.4]</code></pre>
</div>
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make sure lowest value is included</span></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a><span class="co"># common colour ramp</span></span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a>color.ramp <span class="ot">&lt;-</span> <span class="fu">bpy.colors</span>(n.class<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a>(cuts <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">c</span>(zlim[<span class="dv">1</span>], cuts, zlim[<span class="dv">2</span>]),<span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4.46 4.97 5.21 5.43 5.57 5.72 5.87 6.08 7.40</code></pre>
</div>
</div>
<p>Slice the maps:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>gn.class <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">classify</span>(gn.utm, <span class="at">rcl=</span> cuts)</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a><span class="co"># gn.class &lt;- as.factor(gn.class)</span></span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">values</span>(gn.class))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   0    1    2    3    4    5    6    7 
  38  387  775 1057 1216 1507 1828 2138 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(gn.class) <span class="ot">&lt;-</span> <span class="st">"class"</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>sg.class <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">classify</span>(sg.utm, <span class="at">rcl=</span> cuts)</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">values</span>(sg.class))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   0    1    2    3    4    5    6    7 
2205 1839 1524 1065 1058  692  481   83 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sg.class) <span class="ot">&lt;-</span> <span class="st">"class"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Display the maps:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>.l <span class="ot">&lt;-</span> <span class="fu">range</span>(<span class="fu">values</span>(gn.class), <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">plot</span>(gn.class,</span>
<span id="cb109-4"><a href="#cb109-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">col=</span>color.ramp[.l[<span class="dv">1</span>]<span class="sc">:</span>.l[<span class="dv">2</span>]<span class="sc">+</span><span class="dv">1</span>], <span class="at">type=</span><span class="st">"classes"</span>,</span>
<span id="cb109-5"><a href="#cb109-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">main=</span><span class="st">"gNATSGO"</span>)</span>
<span id="cb109-6"><a href="#cb109-6" aria-hidden="true" tabindex="-1"></a>.l <span class="ot">&lt;-</span> <span class="fu">range</span>(<span class="fu">values</span>(sg.class), <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb109-7"><a href="#cb109-7" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">plot</span>(sg.class,</span>
<span id="cb109-8"><a href="#cb109-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">col=</span>color.ramp[.l[<span class="dv">1</span>]<span class="sc">:</span>.l[<span class="dv">2</span>]<span class="sc">+</span><span class="dv">1</span>], <span class="at">type=</span><span class="st">"classes"</span>,</span>
<span id="cb109-9"><a href="#cb109-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">main=</span><span class="st">"SG2"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show.classified-hist-1.png" class="img-fluid" width="1152"></p>
</div>
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Q: Describe the patterns of the two maps</p>
<p>Q: How would these change with different class numbers or limits?</p>
<p>Q: If classifying by histogram equalization, should the two maps be compared with the same limits or each with their own limits? You are welcome to experiment.</p>
</section>
<section id="sec-mean-limit" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="sec-mean-limit"><span class="header-section-number">6.2</span> Classifying by meaningful limits</h2>
<p>For soil properties we usually have limits that correspond to approximate thresholds in land use. In the case of pH, we can refer to extension or crop consultant publications, or environmental models. Unlike in histogram equalization, the number of classes depends on the user requirements.</p>
<p>For example, the <a href="https://www.nnyagdev.org/PDF/SoilpH.pdf">Cornell pH test kit</a> has a “Wide Range Kit” measuring the soil pH over the range of 4.0–8.6, in increments of 0.2 for an experienced user. So, here we will slice the map in increments of 0.2 pH.</p>
<p>Task: slice the maps and display with a common colour ramp.</p>
<p>Find the combined range and divide into one-decimal classes of 0.2 pH, starting and ending on even units of 0.2.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>range.all <span class="ot">&lt;-</span> <span class="fu">range</span>(<span class="fu">values</span>(gn.utm),</span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">values</span>(sg.utm),</span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb111-4"><a href="#cb111-4" aria-hidden="true" tabindex="-1"></a>lim.low <span class="ot">&lt;-</span> <span class="fu">floor</span>(<span class="dv">10</span><span class="sc">*</span>range.all[<span class="dv">1</span>])<span class="sc">/</span><span class="dv">10</span></span>
<span id="cb111-5"><a href="#cb111-5" aria-hidden="true" tabindex="-1"></a>lim.low <span class="ot">&lt;-</span> <span class="fu">ifelse</span>((lim.low <span class="sc">%%</span> .<span class="dv">2</span>) <span class="sc">!=</span> <span class="dv">0</span>, lim.low <span class="sc">-</span> <span class="fl">0.1</span>, lim.low)</span>
<span id="cb111-6"><a href="#cb111-6" aria-hidden="true" tabindex="-1"></a>lim.high <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(<span class="dv">10</span><span class="sc">*</span>range.all[<span class="dv">2</span>])<span class="sc">/</span><span class="dv">10</span></span>
<span id="cb111-7"><a href="#cb111-7" aria-hidden="true" tabindex="-1"></a>lim.high <span class="ot">&lt;-</span> <span class="fu">ifelse</span>((lim.high <span class="sc">%%</span> .<span class="dv">2</span>) <span class="sc">!=</span> <span class="dv">0</span>, lim.high <span class="sc">+</span> <span class="fl">0.1</span>, lim.high)</span>
<span id="cb111-8"><a href="#cb111-8" aria-hidden="true" tabindex="-1"></a>(cuts <span class="ot">&lt;-</span> <span class="fu">seq</span>(lim.low, lim.high, <span class="at">by =</span> <span class="fl">0.2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 4.4 4.6 4.8 5.0 5.2 5.4 5.6 5.8 6.0 6.2 6.4 6.6 6.8 7.0 7.2 7.4 7.6</code></pre>
</div>
</div>
<p>Slice the maps:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>gn.class <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">classify</span>(gn.utm, <span class="at">rcl=</span> cuts)</span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a><span class="co"># gn.class &lt;- as.factor(gn.class)</span></span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">values</span>(gn.class))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   2    3    4    5    6    7    8    9   10   11   12   13   15 
  76  329  629 1452 1727 1992 1348  681  402  253   53    4    1 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(gn.class) <span class="ot">&lt;-</span> <span class="st">"class"</span></span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>sg.class <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">classify</span>(sg.utm, <span class="at">rcl=</span> cuts)</span>
<span id="cb115-3"><a href="#cb115-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">values</span>(sg.class))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   0    1    2    3    4    5    6    7    8    9 
 167  902 1376 1531 1362 1534 1235  653  171   16 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sg.class) <span class="ot">&lt;-</span> <span class="st">"class"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Display them:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>color.ramp <span class="ot">&lt;-</span> <span class="fu">bpy.colors</span>(<span class="fu">length</span>(cuts))</span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>.l <span class="ot">&lt;-</span> <span class="fu">range</span>(<span class="fu">values</span>(gn.class), <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">plot</span>(gn.class,</span>
<span id="cb118-5"><a href="#cb118-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">col=</span>color.ramp[.l[<span class="dv">1</span>]<span class="sc">:</span>.l[<span class="dv">2</span>]<span class="sc">+</span><span class="dv">1</span>], <span class="at">type=</span><span class="st">"classes"</span>,</span>
<span id="cb118-6"><a href="#cb118-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">main=</span><span class="st">"gNATSGO pH 0.2 units"</span>)</span>
<span id="cb118-7"><a href="#cb118-7" aria-hidden="true" tabindex="-1"></a>.l <span class="ot">&lt;-</span> <span class="fu">range</span>(<span class="fu">values</span>(sg.class), <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb118-8"><a href="#cb118-8" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">plot</span>(sg.class,</span>
<span id="cb118-9"><a href="#cb118-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">col=</span>color.ramp[.l[<span class="dv">1</span>]<span class="sc">:</span>.l[<span class="dv">2</span>]<span class="sc">+</span><span class="dv">1</span>], <span class="at">type=</span><span class="st">"classes"</span>,</span>
<span id="cb118-10"><a href="#cb118-10" aria-hidden="true" tabindex="-1"></a>            <span class="at">main=</span><span class="st">"SG2 pH 0.2 units"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show.classified-cuts-1.png" class="img-fluid" width="1152"></p>
</div>
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>These maps are now showing meaningful landscape units, from the point of view of land use and soil processes, on the same scale.</p>
<p>Q: Describe the patterns of the two maps</p>
<p>Q: How would these maps change with wider or narrower class intervals? You are welcome to experiment!</p>
</section>
<section id="sec-coma" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="sec-coma"><span class="header-section-number">6.3</span> Co-occurrence matrices</h2>
<p>One question for a classified map is which classes tend to be adjacent to each other. In the case of the pH map, we might expect adjacent classes to be in the pH sequence, but maybe not – there may be abrupt transitions of parent materials, for example.</p>
<p>A co-occurrence <em>matrix</em> counts all the pairs of adjacent cells for each category in a local landscape, as a cross-classification matrix.</p>
<p>Task: Compute the co-occurrence <em>matrices</em>, using Queen’s Case neighbours (i.e., diagonal links are considered).</p>
<p>Co-occurrence vectors are computed with the <code>lsp_signature</code> function of the <code>motif</code> package, specifyin <code>coma</code> = co-occurrence matrix as the signature.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>coma.gn <span class="ot">&lt;-</span> <span class="fu">lsp_signature</span>(gn.class, <span class="at">type=</span><span class="st">"coma"</span>, <span class="at">neighbourhood =</span> <span class="dv">8</span>)</span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(coma.gn.matrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(coma.gn<span class="sc">$</span>signature)[[<span class="dv">1</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     3   4    5    6    7    8    9   10  11  12  13 14 16
3  148 132  103   76   52   38   13   10   0   0   0  0  0
4  132 608  547  539  368  233   79   30   7   0   0  0  0
5  103 547 1110 1316  893  571  214  114  35   7   0  0  0
6   76 539 1316 3780 2651 1647  784  338 155  61   8  2  0
7   52 368  893 2651 3560 3337 1613  662 274  95   9  8  0
8   38 233  571 1647 3337 5224 2902 1018 470 156  33  8  2
9   13  79  214  784 1613 2902 2764 1210 670 234  41  2  1
10  10  30  114  338  662 1018 1210  948 585 295  63  2  2
11   0   7   35  155  274  470  670  585 522 326  46  5  2
12   0   0    7   61   95  156  234  295 326 642 122  2  1
13   0   0    0    8    9   33   41   63  46 122  78  1  0
14   0   0    0    2    8    8    2    2   5   2   1  2  0
16   0   0    0    0    0    2    1    2   2   1   0  0  0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">diag</span>(coma.gn.matrix))<span class="sc">/</span><span class="fu">sum</span>(coma.gn.matrix)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.277633</code></pre>
</div>
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>coma.sg <span class="ot">&lt;-</span> <span class="fu">lsp_signature</span>(sg.class, <span class="at">type=</span><span class="st">"coma"</span>, <span class="at">neighbourhood =</span> <span class="dv">8</span>)</span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(coma.sg.matrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(coma.sg<span class="sc">$</span>signature)[[<span class="dv">1</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     1    2    3    4    5    6    7    8   9 10
1  904  376   14    0    0    0    0    0   0  0
2  376 5330 1269  139   18    0    0    0   0  0
3   14 1269 7088 2222  287   18    0    0   0  0
4    0  139 2222 6912 2400  378   19    0   0  0
5    0   18  287 2400 5008 2570  284   17   0  0
6    0    0   18  378 2570 6330 2396  118   6  0
7    0    0    0   19  284 2396 5616 1192  45  0
8    0    0    0    0   17  118 1192 3262 464  2
9    0    0    0    0    0    6   45  464 714 79
10   0    0    0    0    0    0    0    2  79 36</code></pre>
</div>
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">diag</span>(coma.sg.matrix))<span class="sc">/</span><span class="fu">sum</span>(coma.sg.matrix)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.5900381</code></pre>
</div>
</div>
<p>Q: Describe the differences in the co-occurrence structure. What does this imply for the spatial pattern?</p>
<p>We see that indeed in this case most adjacencies are within one or at most two classes. The gNATSGO map has more multiple-class adjacencies than does SoilGrids, due to its finer spatial pattern. The SoilGrids map has well over half of the adjacencies in the same class, whereas gNATSGO has about a quarter.</p>
</section>
<section id="sec-cove" class="level2" data-number="6.4">
<h2 data-number="6.4" class="anchored" data-anchor-id="sec-cove"><span class="header-section-number">6.4</span> Co-occurrence vectors</h2>
<p>The <strong>Co-occurrence vector</strong> “COVE” proposed by <span class="citation" data-cites="nowosadSpatialAssociationRegionalizations2018">Nowosad &amp; Stepinski (<a href="#ref-nowosadSpatialAssociationRegionalizations2018" role="doc-biblioref">2018a</a>)</span> summarizes the <em>entire adjacency structure</em> of a map and can be used to compare map structures. This is a normalized form of the co-occurrence matrix (see the previous section). Normalization means the matrix sums to 1, and so is independent of the number of grid cells in the map. Therefore this vector can be considered as a probability vector for the co-occurrence of different classes.</p>
<p>Task: Compute the co-occurrence <em>vectors</em>, using Queen’s Case neighbours.</p>
<p>Co-occurrence vectors are computed with the <code>lsp_signature</code> function of the <code>motif</code> package, specifying <code>cove</code> (normalized co-occurrence vector) as the signature. These will be used to <a href="@compare-cove">compare the maps</a>, below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="co"># normalized co-occurence vector 8 x 8</span></span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>cove.gn <span class="ot">&lt;-</span> <span class="fu">lsp_signature</span>(gn.class, <span class="at">type=</span><span class="st">"cove"</span>, <span class="at">neighbourhood =</span> <span class="dv">8</span>)</span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>cove.sg <span class="ot">&lt;-</span> <span class="fu">lsp_signature</span>(sg.class, <span class="at">type=</span><span class="st">"cove"</span>, <span class="at">neighbourhood =</span> <span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="incove" class="level2" data-number="6.5">
<h2 data-number="6.5" class="anchored" data-anchor-id="incove"><span class="header-section-number">6.5</span> Integrated co-occurrence vector</h2>
<p>An <em>integrated</em> co-occurrence vector considers <em>several input layers</em>, for example, representing different soil properties of the same area.</p>
<p>To examine this we need another soil property map. Let’s use SG2 silt of the 0–5~cm layer. We process this as we did for the pH map. Here the “meaningful limits” for silt content are 5% intervals. Since the SG2 map is expressed in <span class="math inline">\(\mathrm{g} \; \mathrm{kg}^{-1}\)</span>, these are intervals of 50 <span class="math inline">\(\mathrm{g} \; \mathrm{kg}^{-1}\)</span>.</p>
<p>Task: Import and process the silt concentration 0-5 cm SoilGrids product.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>(sg.silt <span class="ot">&lt;-</span> <span class="fu">rast</span>(<span class="fu">paste0</span>(file.dir,</span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"SoilGrids250/lat4243_lon-77-76/silt_0-5cm_mean.tif"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class       : SpatRaster 
dimensions  : 426, 426, 1  (nrow, ncol, nlyr)
resolution  : 0.002349867, 0.002349867  (x, y)
extent      : -77.00071, -75.99967, 41.99918, 43.00022  (xmin, xmax, ymin, ymax)
coord. ref. : lon/lat WGS 84 (EPSG:4326) 
source      : silt_0-5cm_mean.tif 
name        : silt_0-5cm_mean 
min value   :          0.0000 
max value   :        752.4755 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>sg.silt <span class="ot">&lt;-</span> <span class="fu">crop</span>(sg.silt, <span class="fu">ext</span>(ext.crop))</span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a><span class="fu">values</span>(sg.silt) <span class="ot">&lt;-</span> <span class="fu">values</span>(sg.silt)<span class="sc">/</span><span class="dv">10</span> <span class="co"># convert from ppt to %</span></span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a>sg.silt.utm <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">project</span>(sg.silt, epsg.utm, </span>
<span id="cb131-4"><a href="#cb131-4" aria-hidden="true" tabindex="-1"></a>                         <span class="at">res =</span> <span class="fu">c</span>(<span class="dv">250</span>, <span class="dv">250</span>), <span class="at">method =</span> <span class="st">"bilinear"</span>)</span>
<span id="cb131-5"><a href="#cb131-5" aria-hidden="true" tabindex="-1"></a>sg.silt.utm <span class="ot">&lt;-</span> <span class="fu">resample</span>(sg.silt.utm, sg.utm) <span class="co"># make extents identical</span></span>
<span id="cb131-6"><a href="#cb131-6" aria-hidden="true" tabindex="-1"></a>cuts <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">10</span>, <span class="dv">90</span>, <span class="at">by =</span> <span class="dv">5</span>) </span>
<span id="cb131-7"><a href="#cb131-7" aria-hidden="true" tabindex="-1"></a>sg.silt.class <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">classify</span>(sg.silt.utm, <span class="at">rcl=</span> cuts)</span>
<span id="cb131-8"><a href="#cb131-8" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">values</span>(sg.silt.class))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   0    1    2    3    4    5    6    7    8    9   10 
   2    1    5   12   11   12  166 2254 5330 1331    2 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sg.silt.class) <span class="ot">&lt;-</span> <span class="st">"class"</span></span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sg.silt.class, <span class="at">col =</span> <span class="fu">topo.colors</span>(<span class="dv">11</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/silt-map-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>This map has much larger homogeneous areas than the SG2 pH map.</p>
<p>Examine this single map’s co-occurrence matrix and vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="co">#|.label: coma-cove</span></span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a>coma.sg.silt <span class="ot">&lt;-</span> <span class="fu">lsp_signature</span>(sg.silt.class, <span class="at">type=</span><span class="st">"coma"</span>, <span class="at">neighbourhood =</span> <span class="dv">8</span>)</span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(coma.sg.silt.matrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(coma.sg.silt<span class="sc">$</span>signature)[[<span class="dv">1</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   1 2 3  4  5  6   7     8     9   10 11
1  2 0 5  3  4  1   1     0     0    0  0
2  0 0 1  0  2  2   0     2     0    0  0
3  5 1 2  9  6  4   4     7     1    0  0
4  3 0 9 16  9 13  21    20     3    0  0
5  4 2 6  9  8  9  15    24     9    0  0
6  1 2 4 13  9  6   8    25    26    1  0
7  1 0 4 21 15  8 828   416    29    5  0
8  0 2 7 20 24 25 416 14128  3024   10  0
9  0 0 1  3  9 26  29  3024 36196 2651  1
10 0 0 0  0  0  1   5    10  2651 7892 10
11 0 0 0  0  0  0   0     0     1   10  2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">diag</span>(coma.sg.silt.matrix))<span class="sc">/</span><span class="fu">sum</span>(coma.sg.silt.matrix)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.8223602</code></pre>
</div>
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>cove.sg.silt <span class="ot">&lt;-</span> <span class="fu">lsp_signature</span>(sg.silt.class, <span class="at">type=</span><span class="st">"cove"</span>, <span class="at">neighbourhood =</span> <span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Most of the adjacencies are to the same class, or the adjacent class.</p>
<p>Task: Compute the distance between the co-occurrence vectors for pH and silt:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>cove.df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(cove.sg)<span class="sc">$</span>signature[[<span class="dv">1</span>]][<span class="dv">1</span>,]</span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>cove.df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(cove.df, cove.sg.silt<span class="sc">$</span>signature[[<span class="dv">1</span>]][<span class="dv">1</span>,])</span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>cove.dists <span class="ot">&lt;-</span> <span class="fu">round</span>(</span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>  philentropy<span class="sc">::</span><span class="fu">distance</span>(cove.df, <span class="at">method =</span> <span class="st">"jensen-shannon"</span>, </span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">use.row.names =</span><span class="cn">TRUE</span>, </span>
<span id="cb139-6"><a href="#cb139-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">as.dist.obj =</span> <span class="cn">FALSE</span>,</span>
<span id="cb139-7"><a href="#cb139-7" aria-hidden="true" tabindex="-1"></a>                        <span class="at">diag =</span> <span class="cn">FALSE</span>) ,<span class="dv">4</span>)</span>
<span id="cb139-8"><a href="#cb139-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(cove.dists)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>jensen-shannon 
        0.6811 </code></pre>
</div>
</div>
<p>This is a much larger distance than that between SG2 and gNATSGO pH maps co-occurrence vectors.</p>
<section id="clustering-pattern-differences" class="level3" data-number="6.5.1">
<h3 data-number="6.5.1" class="anchored" data-anchor-id="clustering-pattern-differences"><span class="header-section-number">6.5.1</span> Clustering pattern differences</h3>
<p>Once a pattern metric is shown across a map, a natural question is whether different areas of the map have different patterns. We illustrate this with the pattern of the integrated co-occurrence vectors.</p>
<p>Any size window can be used. If too small the result is erratic, if too large, local differences may be missed.</p>
<p>Task: Identify which parts of the SG2 map have similar <em>integrated co-occurrence</em> pattern differences, considering both properties. For this we use 4 x 4 km windows, i.e., 16 x 16 grid cells.</p>
<p>Again we use <code>lsp_signature</code>, type <code>"incove"</code>, but now specifying a <code>window</code> size within which to compute the pattern.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a>sg.ph.silt.class <span class="ot">&lt;-</span> <span class="fu">c</span>(sg.class, sg.silt.class)</span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a>incove.sg <span class="ot">&lt;-</span> <span class="fu">lsp_signature</span>(sg.ph.silt.class,</span>
<span id="cb141-3"><a href="#cb141-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">type =</span> <span class="st">"incove"</span>,</span>
<span id="cb141-4"><a href="#cb141-4" aria-hidden="true" tabindex="-1"></a>                              <span class="at">neighbourhood =</span> <span class="dv">8</span>,</span>
<span id="cb141-5"><a href="#cb141-5" aria-hidden="true" tabindex="-1"></a>                              <span class="at">ordered =</span> <span class="cn">TRUE</span>,  <span class="co"># the pH classes are ordered</span></span>
<span id="cb141-6"><a href="#cb141-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">window =</span> <span class="dv">16</span>,</span>
<span id="cb141-7"><a href="#cb141-7" aria-hidden="true" tabindex="-1"></a>                              <span class="at">normalization =</span> <span class="st">"pdf"</span>)  <span class="co">#sum to one</span></span>
<span id="cb141-8"><a href="#cb141-8" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(incove.sg.dist <span class="ot">&lt;-</span> <span class="fu">lsp_to_dist</span>(incove.sg,</span>
<span id="cb141-9"><a href="#cb141-9" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">dist_fun =</span> <span class="st">"jensen-shannon"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
0.04654 0.28386 0.42389 0.43529 0.59139 0.92259 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(incove.sg.dist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 42 42</code></pre>
</div>
</div>
<p>Here we have defined 42 x 42 distances, i.e., paired distances between each of the windows’ signatures.</p>
<p>Are any of these distances similar? Let’s see with a <em>cluster analysis</em>.</p>
<p>Task: Make a hierarchical clustering of the distances between the integrated co-occurrence vectors of the 42 windows.</p>
<p>The <code>hclust</code> function can cluster using many methods to build the dendrogram. Here we use Ward’s D2 method, which aims at finding compact, spherical clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a>sg.hclust <span class="ot">&lt;-</span> <span class="fu">hclust</span>(incove.sg.dist, <span class="at">method =</span> <span class="st">"ward.D2"</span>)</span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sg.hclust, <span class="at">main =</span> <span class="st">"clusters of distance between `incove`"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/incove.cluster-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>Task: Define classes of similar distances by cutting the dendrogram.</p>
<p>Examining the dendrogram, it seems that height <code>h = 0.5</code> is a good cutting point, which captures the main differences. Alternatively, a set number of clusters can be requested with the <code>k</code> argument.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a>sg.clusters <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(<span class="fu">cutree</span>(sg.hclust, <span class="at">h =</span> <span class="fl">0.5</span>))  <span class="co"># cutpoint by visual inspection</span></span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(sg.clusters)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "1" "2" "3" "4" "5" "6"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a>sg.grid.sf <span class="ot">=</span> <span class="fu">lsp_add_clusters</span>(incove.sg, sg.clusters)</span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a>sg.grid.sf<span class="sc">$</span>clust <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(sg.grid.sf<span class="sc">$</span>clust)</span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a>my.pal <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">8</span>, <span class="st">"Accent"</span>))(<span class="fu">length</span>(<span class="fu">levels</span>(sg.grid.sf<span class="sc">$</span>clust)))</span>
<span id="cb148-4"><a href="#cb148-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> sg.grid.sf) <span class="sc">+</span> </span>
<span id="cb148-5"><a href="#cb148-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> clust), <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb148-6"><a href="#cb148-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_discrete</span>(<span class="at">type =</span> my.pal) <span class="sc">+</span></span>
<span id="cb148-7"><a href="#cb148-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Clusters: distance between integrated co-occurrence vectors"</span>,</span>
<span id="cb148-8"><a href="#cb148-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="st">"cluster"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/incove.cluster.cut-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>This shows which areas of the map have similar integrated co-occurrence patterns. These can be interpreted as similar soils, in the sense that the sum of propertied defines a soil type.</p>
<p>Compare this to a visual inspection of the patterns, next to the 7 x 6 cluster grid.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> sg.grid.sf) <span class="sc">+</span> </span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> clust), <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb149-3"><a href="#cb149-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_discrete</span>(<span class="at">type =</span> my.pal) <span class="sc">+</span></span>
<span id="cb149-4"><a href="#cb149-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"cluster"</span>)  <span class="sc">+</span></span>
<span id="cb149-5"><a href="#cb149-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>)</span>
<span id="cb149-6"><a href="#cb149-6" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb149-7"><a href="#cb149-7" aria-hidden="true" tabindex="-1"></a>  tidyterra<span class="sc">::</span><span class="fu">geom_spatraster</span>(<span class="at">data =</span> sg.class, <span class="fu">aes</span>(<span class="at">fill =</span> class)) <span class="sc">+</span></span>
<span id="cb149-8"><a href="#cb149-8" aria-hidden="true" tabindex="-1"></a>   <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>)</span>
<span id="cb149-9"><a href="#cb149-9" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb149-10"><a href="#cb149-10" aria-hidden="true" tabindex="-1"></a>  tidyterra<span class="sc">::</span><span class="fu">geom_spatraster</span>(<span class="at">data =</span> sg.silt.class, <span class="fu">aes</span>(<span class="at">fill =</span> class)) <span class="sc">+</span></span>
<span id="cb149-11"><a href="#cb149-11" aria-hidden="true" tabindex="-1"></a>   <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>)</span>
<span id="cb149-12"><a href="#cb149-12" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(p1, p2, p3, <span class="at">nrow=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/incove-clusters-map-3-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>Careful examination reveals that the cluster in the NW corner corresponds to an intricate pattern of pH and mostly one class of silt concentration.</p>
</section>
</section>
<section id="sec-lsm" class="level2" data-number="6.6">
<h2 data-number="6.6" class="anchored" data-anchor-id="sec-lsm"><span class="header-section-number">6.6</span> Landscape metrics</h2>
<p>Landscape metrics have a long history of use in landscape ecology <span class="citation" data-cites="Uuemaa.etal2013">(<a href="#ref-Uuemaa.etal2013" role="doc-biblioref">Uuemaa et al., 2013</a>)</span>. A wide variety have been collected in the well-known FRAGSTATS computer program <span class="citation" data-cites="McGarigal.etal2012">(<a href="#ref-McGarigal.etal2012" role="doc-biblioref">McGarigal et al., 2012</a>)</span>. These have been implemented in the R context by the package<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a> <span class="citation" data-cites="Hesselbarth.etal2019 Hesselbarth2021">(<a href="#ref-Hesselbarth.etal2019" role="doc-biblioref">Hesselbarth et al., 2019</a>; <a href="#ref-Hesselbarth2021" role="doc-biblioref">Hesselbarth, 2021</a>)</span>. Although the ecological relevance of FRAGSTATS metrics have been criticized <span class="citation" data-cites="Kupfer2012">(<a href="#ref-Kupfer2012" role="doc-biblioref">Kupfer, 2012</a>)</span>, here we use them to <em>characterize spatial patterns of soil properties</em> or <em>classes</em>, not as inputs to landscape ecology models.</p>
<p>The patterns of soil classes or properties are not expected to have the same characteristics as those for land cover or vegetation types. Land cover is largely controlled by humans, and where it is not, vegetation is mostly placed on the landscape by different mechanisms than are soils. There is a link, however: if the soil property is largely controlled by the <code>o</code> (organism) or <code>h</code> (human) factor, then the patterns on the landscape could be similar to those under it.</p>
<p>There are many metrics, of three levels of detail. We list them here for reference; each has its own help text.</p>
<p>First, the <em>patch-level metrics</em>. These describe every patch, i.e., contiguous cells belonging to the same class.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a>landscapemetrics<span class="sc">::</span><span class="fu">list_lsm</span>(<span class="at">level=</span><span class="st">"patch"</span>) <span class="sc">%&gt;%</span> <span class="fu">print</span>(<span class="at">n=</span><span class="cn">Inf</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 × 5
   metric name                                type           level function_name
   &lt;chr&gt;  &lt;chr&gt;                               &lt;chr&gt;          &lt;chr&gt; &lt;chr&gt;        
 1 area   patch area                          area and edge… patch lsm_p_area   
 2 cai    core area index                     core area met… patch lsm_p_cai    
 3 circle related circumscribing circle       shape metric   patch lsm_p_circle 
 4 contig contiguity index                    shape metric   patch lsm_p_contig 
 5 core   core area                           core area met… patch lsm_p_core   
 6 enn    euclidean nearest neighbor distance aggregation m… patch lsm_p_enn    
 7 frac   fractal dimension index             shape metric   patch lsm_p_frac   
 8 gyrate radius of gyration                  area and edge… patch lsm_p_gyrate 
 9 ncore  number of core areas                core area met… patch lsm_p_ncore  
10 para   perimeter-area ratio                shape metric   patch lsm_p_para   
11 perim  patch perimeter                     area and edge… patch lsm_p_perim  
12 shape  shape index                         shape metric   patch lsm_p_shape  </code></pre>
</div>
</div>
<p>Second, the <em>class-level</em> metrics. These describe all patches belonging to a specified class.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a>landscapemetrics<span class="sc">::</span><span class="fu">list_lsm</span>(<span class="at">level=</span><span class="st">"class"</span>) <span class="sc">%&gt;%</span> <span class="fu">print</span>(<span class="at">n=</span><span class="cn">Inf</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 55 × 5
   metric    name                                  type      level function_name
   &lt;chr&gt;     &lt;chr&gt;                                 &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt;        
 1 ai        aggregation index                     aggregat… class lsm_c_ai     
 2 area_cv   patch area                            area and… class lsm_c_area_cv
 3 area_mn   patch area                            area and… class lsm_c_area_mn
 4 area_sd   patch area                            area and… class lsm_c_area_sd
 5 ca        total (class) area                    area and… class lsm_c_ca     
 6 cai_cv    core area index                       core are… class lsm_c_cai_cv 
 7 cai_mn    core area index                       core are… class lsm_c_cai_mn 
 8 cai_sd    core area index                       core are… class lsm_c_cai_sd 
 9 circle_cv related circumscribing circle         shape me… class lsm_c_circle…
10 circle_mn related circumscribing circle         shape me… class lsm_c_circle…
11 circle_sd related circumscribing circle         shape me… class lsm_c_circle…
12 clumpy    clumpiness index                      aggregat… class lsm_c_clumpy 
13 cohesion  patch cohesion index                  aggregat… class lsm_c_cohesi…
14 contig_cv contiguity index                      shape me… class lsm_c_contig…
15 contig_mn contiguity index                      shape me… class lsm_c_contig…
16 contig_sd contiguity index                      shape me… class lsm_c_contig…
17 core_cv   core area                             core are… class lsm_c_core_cv
18 core_mn   core area                             core are… class lsm_c_core_mn
19 core_sd   core area                             core are… class lsm_c_core_sd
20 cpland    core area percentage of landscape     core are… class lsm_c_cpland 
21 dcad      disjunct core area density            core are… class lsm_c_dcad   
22 dcore_cv  disjunct core area                    core are… class lsm_c_dcore_…
23 dcore_mn  disjunct core area                    core are… class lsm_c_dcore_…
24 dcore_sd  disjunct core area                    core are… class lsm_c_dcore_…
25 division  division index                        aggregat… class lsm_c_divisi…
26 ed        edge density                          area and… class lsm_c_ed     
27 enn_cv    euclidean nearest neighbor distance   aggregat… class lsm_c_enn_cv 
28 enn_mn    euclidean nearest neighbor distance   aggregat… class lsm_c_enn_mn 
29 enn_sd    euclidean nearest neighbor distance   aggregat… class lsm_c_enn_sd 
30 frac_cv   fractal dimension index               shape me… class lsm_c_frac_cv
31 frac_mn   fractal dimension index               shape me… class lsm_c_frac_mn
32 frac_sd   fractal dimension index               shape me… class lsm_c_frac_sd
33 gyrate_cv radius of gyration                    area and… class lsm_c_gyrate…
34 gyrate_mn radius of gyration                    area and… class lsm_c_gyrate…
35 gyrate_sd radius of gyration                    area and… class lsm_c_gyrate…
36 iji       interspersion and juxtaposition index aggregat… class lsm_c_iji    
37 lpi       largest patch index                   area and… class lsm_c_lpi    
38 lsi       landscape shape index                 aggregat… class lsm_c_lsi    
39 mesh      effective mesh size                   aggregat… class lsm_c_mesh   
40 ndca      number of disjunct core areas         core are… class lsm_c_ndca   
41 nlsi      normalized landscape shape index      aggregat… class lsm_c_nlsi   
42 np        number of patches                     aggregat… class lsm_c_np     
43 pafrac    perimeter-area fractal dimension      shape me… class lsm_c_pafrac 
44 para_cv   perimeter-area ratio                  shape me… class lsm_c_para_cv
45 para_mn   perimeter-area ratio                  shape me… class lsm_c_para_mn
46 para_sd   perimeter-area ratio                  shape me… class lsm_c_para_sd
47 pd        patch density                         aggregat… class lsm_c_pd     
48 pladj     percentage of like adjacencies        aggregat… class lsm_c_pladj  
49 pland     percentage of landscape               area and… class lsm_c_pland  
50 shape_cv  shape index                           shape me… class lsm_c_shape_…
51 shape_mn  shape index                           shape me… class lsm_c_shape_…
52 shape_sd  shape index                           shape me… class lsm_c_shape_…
53 split     splitting index                       aggregat… class lsm_c_split  
54 tca       total core area                       core are… class lsm_c_tca    
55 te        total edge                            area and… class lsm_c_te     </code></pre>
</div>
</div>
<p>Finally, the <em>landscape-level</em> metrics. These describe the characteristics of the entire landscape, i.e., the assemblage of classes and patches.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a>landscapemetrics<span class="sc">::</span><span class="fu">list_lsm</span>(<span class="at">level=</span><span class="st">"landscape"</span>) <span class="sc">%&gt;%</span> <span class="fu">print</span>(<span class="at">n=</span><span class="cn">Inf</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 66 × 5
   metric    name                                  type      level function_name
   &lt;chr&gt;     &lt;chr&gt;                                 &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt;        
 1 ai        aggregation index                     aggregat… land… lsm_l_ai     
 2 area_cv   patch area                            area and… land… lsm_l_area_cv
 3 area_mn   patch area                            area and… land… lsm_l_area_mn
 4 area_sd   patch area                            area and… land… lsm_l_area_sd
 5 cai_cv    core area index                       core are… land… lsm_l_cai_cv 
 6 cai_mn    core area index                       core are… land… lsm_l_cai_mn 
 7 cai_sd    core area index                       core are… land… lsm_l_cai_sd 
 8 circle_cv related circumscribing circle         shape me… land… lsm_l_circle…
 9 circle_mn related circumscribing circle         shape me… land… lsm_l_circle…
10 circle_sd related circumscribing circle         shape me… land… lsm_l_circle…
11 cohesion  patch cohesion index                  aggregat… land… lsm_l_cohesi…
12 condent   conditional entropy                   complexi… land… lsm_l_condent
13 contag    connectance                           aggregat… land… lsm_l_contag 
14 contig_cv contiguity index                      shape me… land… lsm_l_contig…
15 contig_mn contiguity index                      shape me… land… lsm_l_contig…
16 contig_sd contiguity index                      shape me… land… lsm_l_contig…
17 core_cv   core area                             core are… land… lsm_l_core_cv
18 core_mn   core area                             core are… land… lsm_l_core_mn
19 core_sd   core area                             core are… land… lsm_l_core_sd
20 dcad      disjunct core area density            core are… land… lsm_l_dcad   
21 dcore_cv  disjunct core area                    core are… land… lsm_l_dcore_…
22 dcore_mn  disjunct core area                    core are… land… lsm_l_dcore_…
23 dcore_sd  disjunct core area                    core are… land… lsm_l_dcore_…
24 division  division index                        aggregat… land… lsm_l_divisi…
25 ed        edge density                          area and… land… lsm_l_ed     
26 enn_cv    euclidean nearest neighbor distance   aggregat… land… lsm_l_enn_cv 
27 enn_mn    euclidean nearest neighbor distance   aggregat… land… lsm_l_enn_mn 
28 enn_sd    euclidean nearest neighbor distance   aggregat… land… lsm_l_enn_sd 
29 ent       shannon entropy                       complexi… land… lsm_l_ent    
30 frac_cv   fractal dimension index               shape me… land… lsm_l_frac_cv
31 frac_mn   fractal dimension index               shape me… land… lsm_l_frac_mn
32 frac_sd   fractal dimension index               shape me… land… lsm_l_frac_sd
33 gyrate_cv radius of gyration                    area and… land… lsm_l_gyrate…
34 gyrate_mn radius of gyration                    area and… land… lsm_l_gyrate…
35 gyrate_sd radius of gyration                    area and… land… lsm_l_gyrate…
36 iji       interspersion and juxtaposition index aggregat… land… lsm_l_iji    
37 joinent   joint entropy                         complexi… land… lsm_l_joinent
38 lpi       largest patch index                   area and… land… lsm_l_lpi    
39 lsi       landscape shape index                 aggregat… land… lsm_l_lsi    
40 mesh      effective mesh size                   aggregat… land… lsm_l_mesh   
41 msidi     modified simpson's diversity index    diversit… land… lsm_l_msidi  
42 msiei     modified simpson's evenness index     diversit… land… lsm_l_msiei  
43 mutinf    mutual information                    complexi… land… lsm_l_mutinf 
44 ndca      number of disjunct core areas         core are… land… lsm_l_ndca   
45 np        number of patches                     aggregat… land… lsm_l_np     
46 pafrac    perimeter-area fractal dimension      shape me… land… lsm_l_pafrac 
47 para_cv   perimeter-area ratio                  shape me… land… lsm_l_para_cv
48 para_mn   perimeter-area ratio                  shape me… land… lsm_l_para_mn
49 para_sd   perimeter-area ratio                  shape me… land… lsm_l_para_sd
50 pd        patch density                         aggregat… land… lsm_l_pd     
51 pladj     percentage of like adjacencies        aggregat… land… lsm_l_pladj  
52 pr        patch richness                        diversit… land… lsm_l_pr     
53 prd       patch richness density                diversit… land… lsm_l_prd    
54 relmutinf relative mutual information           complexi… land… lsm_l_relmut…
55 rpr       relative patch richness               diversit… land… lsm_l_rpr    
56 shape_cv  shape index                           shape me… land… lsm_l_shape_…
57 shape_mn  shape index                           shape me… land… lsm_l_shape_…
58 shape_sd  shape index                           shape me… land… lsm_l_shape_…
59 shdi      shannon's diversity index             diversit… land… lsm_l_shdi   
60 shei      shannon's evenness index              diversit… land… lsm_l_shei   
61 sidi      simpson's diversity index             diversit… land… lsm_l_sidi   
62 siei      simspon's evenness index              diversit… land… lsm_l_siei   
63 split     splitting index                       aggregat… land… lsm_l_split  
64 ta        total area                            area and… land… lsm_l_ta     
65 tca       total core area                       core are… land… lsm_l_tca    
66 te        total edge                            area and… land… lsm_l_te     </code></pre>
</div>
</div>
<section id="landscape-level-metrics" class="level3" data-number="6.6.1">
<h3 data-number="6.6.1" class="anchored" data-anchor-id="landscape-level-metrics"><span class="header-section-number">6.6.1</span> Landscape-level metrics</h3>
<p>These measures summarize the pattern of the entire map. The following five seem to be most useful for characterizing soil maps.</p>
<ul>
<li><strong>ai</strong>: The <strong>landscape aggregation index</strong> LAI is an ‘Aggregation metric’. This shows how much the classes occur as large units, vs.&nbsp;as scattered patches. It is independent of the number of classes.</li>
</ul>
<p>It equals the number of like adjacencies divided by the theoretical maximum possible number of like adjacencies for that class summed over each class for the entire landscape. The metric is based on the adjacency matrix. It equals 0 for maximally disaggregated and 100 for maximally aggregated classes. <a href="https://r-spatialecology.github.io/landscapemetrics/reference/lsm_l_ai.html">More info</a> <span class="math display">\[\mathrm{LAI} = \Bigg[∑_{i=1}^m \Big( \frac{g_{ii}}{max-g_{ii}} \Big) P_{i} \Bigg](100)\]</span> where <span class="math inline">\(g_{ii}\)</span> is the number of like adjacencies, <span class="math inline">\((\mathrm{max}-g_{ii})\)</span> is the class-wise maximum possible number of like adjacencies of class <span class="math inline">\(i\)</span> (i.e., if all pixels in the class were in one cluster), and <span class="math inline">\(P_{i}\)</span> is the proportion of landscape comprised of class <span class="math inline">\(i\)</span>, to weight the index by class prevalence.</p>
<ul>
<li><strong>frac_mn</strong>: The <strong>mean fractal dimension</strong> FRAC_MN is a ‘Shape metric’. It summarises the landscape as the mean of the fractal dimension index of all patches in the landscape, i.e., the complexity of the map.</li>
</ul>
<p>The fractal dimension index is based on the patch perimeter and the patch area and describes the patch complexity. The Coefficient of variation is scaled to the mean and thus is comparable among different landscapes. <a href="https://r-spatialecology.github.io/landscapemetrics/reference/lsm_l_frac_mn.html">More info</a> <span class="math display">\[\mathrm{FRAC} = \frac{2 * \ln * (0.25 * p_{ij})} {\ln a_{ij}}\]</span> where the patch perimeters are <span class="math inline">\({p_{ij}}\)</span> in linear units and the areas are <span class="math inline">\({a_{ij}}\)</span> in square units.</p>
<ul>
<li><strong>lsi</strong>: <strong>landscape shape index</strong> LSI is an ‘Aggregation metric’. It is the ratio between the actual edge length of class <span class="math inline">\(i\)</span> and the hypothetical minimum edge length of class <span class="math inline">\(i\)</span>. It measures how compact are the classes. For example, long thin classes will have low LSI.</li>
</ul>
<p>The minimum edge length equals the edge length if class i would be maximally aggregated. LSI = 1 when only one square patch is present or all patches are maximally aggregated. Increases, without limit, as the length of the actual edges increases, i.e.&nbsp;the patches become less compact. <a href="https://r-spatialecology.github.io/landscapemetrics/reference/lsm_c_lsi.html?q=lsi">More info</a> <span class="math display">\[   \mathrm{LSI} = \frac{0.25 E'}{\sqrt{A}}\]</span> where <span class="math inline">\(A\)</span> is the total area of the landscape and <span class="math inline">\(E'\)</span> is the total length of edges, including the boundary.</p>
<ul>
<li><strong>shdi</strong>: The <strong>Shannon diversity index</strong> SHDI is a ‘Diversity metric’. It is a widely used metric in biodiversity and ecology and takes both the number of classes and the abundance of each class into account. It is related to the concept of entropy: how much “information” is in the landscape pattern. More classes and more even distribution of their areas implies high information.</li>
</ul>
<p>SHDI = 0 when only one patch is present and increases, without limit, as the number of classes increases while the proportions are equally distributed. <a href="https://r-spatialecology.github.io/landscapemetrics/reference/lsm_l_shdi.html?q=shd">More info</a> <span class="math display">\[ D = - \sum_{i=1}^N p_i \ln p_i\]</span> where <span class="math inline">\(p_i\)</span> is the proportion of pixels of class <span class="math inline">\(i = (1 \ldots N)\)</span>,</p>
<ul>
<li><strong>shei</strong>: The <strong>Shannon evenness index</strong> SHEI is a ‘Diversity metric’. It is the ratio between the Shannon’s diversity index <span class="math inline">\(D\)</span> (see previous) and and the theoretical maximum Shannon diversity index <span class="math inline">\(\ln N\)</span>. It can be understood as a measure of dominance.</li>
</ul>
<p>SHEI = 0 when only one patch present; SHEI = 1 when the proportion of classes is equally distributed. <a href="https://r-spatialecology.github.io/landscapemetrics/reference/lsm_l_shei.html?q=shei">More info</a> <span class="math display">\[E = \frac{D}{\ln N}\]</span></p>
<p>These methods must be applied to classified maps. Continuous soil property maps must first be classified into ranges before analysis, see (<a href="#sec-hist-equal">Section&nbsp;6.1</a>) and (<a href="#sec-mean-limit">Section&nbsp;6.2</a>), above. Different choices of class limits and widths will result in different values of these measures.</p>
</section>
<section id="compute-sec-lsm" class="level3" data-number="6.6.2">
<h3 data-number="6.6.2" class="anchored" data-anchor-id="compute-sec-lsm"><span class="header-section-number">6.6.2</span> Computing landscape-level metrics</h3>
<p>The <code>landscapemetrics</code> package implements a set of metrics as used in ecology and derived from the FRAGSTATS computer program; the metrics are explained in the previous section. Here we compute them for the two maps we are comparing.</p>
<p>To compute landscape metrics:</p>
<ul>
<li>Input is raster map (here, a <code>terra::SpatRaster</code>) with integer values, each of which represents a category, i.e., landscape class.</li>
<li>The map must be in a projected CRS, with distance units in meters;</li>
<li>Results are in meters, square meters or hectares, depending on the function;</li>
</ul>
<p>Task: Check that the maps have the proper structure for the landscape metrics.</p>
<p>This is done with the <code>landscapemetrics::check_landscape</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_landscape</span>(gn.class)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["layer"],"name":[1],"type":["int"],"align":["right"]},{"label":["crs"],"name":[2],"type":["chr"],"align":["left"]},{"label":["units"],"name":[3],"type":["chr"],"align":["left"]},{"label":["class"],"name":[4],"type":["chr"],"align":["left"]},{"label":["n_classes"],"name":[5],"type":["int"],"align":["right"]},{"label":["OK"],"name":[6],"type":["chr"],"align":["left"]}],"data":[{"1":"1","2":"projected","3":"m","4":"integer","5":"13","6":"✔"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_landscape</span>(sg.class)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["layer"],"name":[1],"type":["int"],"align":["right"]},{"label":["crs"],"name":[2],"type":["chr"],"align":["left"]},{"label":["units"],"name":[3],"type":["chr"],"align":["left"]},{"label":["class"],"name":[4],"type":["chr"],"align":["left"]},{"label":["n_classes"],"name":[5],"type":["int"],"align":["right"]},{"label":["OK"],"name":[6],"type":["chr"],"align":["left"]}],"data":[{"1":"1","2":"projected","3":"m","4":"integer","5":"10","6":"✔"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Task: Show the landscapes of each product, first with all classes on one map, then with the classes separate:</p>
<p><strong>global</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb158"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_patches</span>(gn.class, <span class="at">class =</span> <span class="st">"global"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show.patches.global-1.png" class="img-fluid" width="768"></p>
</div>
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_patches</span>(sg.class, <span class="at">class =</span> <span class="st">"global"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show.patches.global-2.png" class="img-fluid" width="768"></p>
</div>
</div>
<p><strong>per-class</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_patches</span>(sg.class, <span class="at">class =</span> <span class="st">"all"</span>, <span class="at">nrow =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show.patches.all-1.png" class="img-fluid" width="1344"></p>
</div>
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_patches</span>(gn.class, <span class="at">class =</span> <span class="st">"all"</span>, <span class="at">nrow =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show.patches.all-2.png" class="img-fluid" width="1344"></p>
</div>
</div>
<p>Q: Describe the main differences between the patterns. Which map seems more aggregated? More diverse?</p>
<p>Task: compute the metrics and tabulate them:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a>lst <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"lsm_l_"</span>, <span class="fu">c</span>(<span class="st">"shdi"</span>, <span class="st">"shei"</span>, <span class="st">"lsi"</span>, <span class="st">"ai"</span>,  <span class="st">"frac_mn"</span>))</span>
<span id="cb166-2"><a href="#cb166-2" aria-hidden="true" tabindex="-1"></a>ls.metrics.gn <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(gn.class, <span class="at">what=</span>lst)</span>
<span id="cb166-3"><a href="#cb166-3" aria-hidden="true" tabindex="-1"></a>ls.metrics.sg <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(sg.class, <span class="at">what=</span>lst)</span>
<span id="cb166-4"><a href="#cb166-4" aria-hidden="true" tabindex="-1"></a>metrics.table <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">product=</span><span class="fu">c</span>(<span class="st">"gNATSGO"</span>, <span class="st">"SG2"</span>),</span>
<span id="cb166-5"><a href="#cb166-5" aria-hidden="true" tabindex="-1"></a>                            <span class="fu">rbind</span>(<span class="fu">round</span>(ls.metrics.gn<span class="sc">$</span>value, <span class="dv">3</span>),</span>
<span id="cb166-6"><a href="#cb166-6" aria-hidden="true" tabindex="-1"></a>                                  <span class="fu">round</span>(ls.metrics.sg<span class="sc">$</span>value, <span class="dv">3</span>)))</span>
<span id="cb166-7"><a href="#cb166-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(metrics.table)[<span class="dv">2</span><span class="sc">:</span><span class="dv">6</span>] <span class="ot">&lt;-</span> ls.metrics.gn<span class="sc">$</span>metric</span>
<span id="cb166-8"><a href="#cb166-8" aria-hidden="true" tabindex="-1"></a>metrics.table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["product"],"name":[1],"type":["chr"],"align":["left"]},{"label":["ai"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["frac_mn"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["lsi"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["shdi"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["shei"],"name":[6],"type":["dbl"],"align":["right"]}],"data":[{"1":"gNATSGO","2":"31.482","3":"1.028","4":"33.605","5":"2.052","6":"0.800"},{"1":"SG2","2":"64.694","3":"1.050","4":"18.450","5":"2.036","6":"0.884"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Q: Referring to the descriptions of these metrics (above), what are the differences between these maps’ landscape patterns? Where do the maps most differ?</p>
<ul>
<li>Aggregation Index</li>
<li>Mean Fractal Dimension</li>
<li>Landscape Shape Index</li>
<li>Shannon Diversity</li>
<li>Shannon Evenness</li>
</ul>
</section>
</section>
</section>
<section id="sec-compare-classified" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> Comparing patterns of classified maps</h1>
<p>Once we have various pattern metrics computed on different maps of the same area, an obvious question is “How much and how do they differ?”. The question of “best” map is not (yet) asked.</p>
<ol type="1">
<li><p>We can directly compare the metrics; see above (<a href="#sec-lsm">Section&nbsp;6.6</a>).</p></li>
<li><p>We can compare the adjacency structures; see above (<a href="#sec-cove">Section&nbsp;6.4</a>) and next (<a href="#sec-compare-cove">Section&nbsp;7.1</a>).</p></li>
<li><p>We can compare the intersections of the maps: use one as a reference and determine how well the other map reproduces the structure of the first; see below (<a href="#sec-vmeasure">Section&nbsp;7.2.4</a>) .</p></li>
</ol>
<section id="sec-compare-cove" class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="sec-compare-cove"><span class="header-section-number">7.1</span> Co-occurrence vectors</h2>
<p>Task: Compute the difference between the co-occurrence patterns of the two maps.</p>
<p>This uses the Jensen-Shannon distance between matrix columns. Each row of the column vector is a co-occurrence metric of two classes. The <code>philentropy</code> (“Similarity and Distance Quantification Between Probability Functions”) package implements this distance metric. This metric is commonly used to compare probability distributions. It computes the entropy of each probability vector (here, the co-occurrence vector) and the entropy of their average and, from these, the distance in entropy space between them:</p>
<p><span class="math display">\[2d = \sum_i \big(P_i \; log\frac{2 P_i}{P_i + Q_i}\big) + \sum_i \big(Q_i \; log\frac{2 Q_i}{P_i + Q_i}\big)\]</span> where <span class="math inline">\(P\)</span> and <span class="math inline">\(Q\)</span> are the two vectors, and <span class="math inline">\(i\)</span> is the row, i.e., single co-occurrence value.</p>
<p>Increasing values indicate increasing dissimilarity in the adjacency patterns, i.e., greater entropy. If the adjacency structures are identical, the distance is zero.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cove.gn)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "id"        "na_prop"   "signature"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a>cove.df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(cove.gn)<span class="sc">$</span>signature[[<span class="dv">1</span>]][<span class="dv">1</span>,]</span>
<span id="cb169-2"><a href="#cb169-2" aria-hidden="true" tabindex="-1"></a>cove.df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(cove.df, cove.sg<span class="sc">$</span>signature[[<span class="dv">1</span>]][<span class="dv">1</span>,])</span>
<span id="cb169-3"><a href="#cb169-3" aria-hidden="true" tabindex="-1"></a>cove.dists <span class="ot">&lt;-</span> <span class="fu">round</span>(</span>
<span id="cb169-4"><a href="#cb169-4" aria-hidden="true" tabindex="-1"></a>  philentropy<span class="sc">::</span><span class="fu">distance</span>(cove.df, <span class="at">method =</span> <span class="st">"jensen-shannon"</span>, </span>
<span id="cb169-5"><a href="#cb169-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">use.row.names =</span><span class="cn">TRUE</span>, </span>
<span id="cb169-6"><a href="#cb169-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">as.dist.obj =</span> <span class="cn">FALSE</span>,</span>
<span id="cb169-7"><a href="#cb169-7" aria-hidden="true" tabindex="-1"></a>                        <span class="at">diag =</span> <span class="cn">FALSE</span>) ,<span class="dv">4</span>)</span>
<span id="cb169-8"><a href="#cb169-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(cove.dists)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>jensen-shannon 
        0.4584 </code></pre>
</div>
</div>
<p>This is a fairly high value.</p>
<p>This comparison can be streamlined with the <code>lsp_compare</code> method. First, over the whole map:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lsp_compare</span>(gn.class, sg.class, </span>
<span id="cb171-2"><a href="#cb171-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">type =</span> <span class="st">"cove"</span>, <span class="at">dist_fun =</span> <span class="st">"jensen-shannon"</span>,</span>
<span id="cb171-3"><a href="#cb171-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">neighbourhood =</span> <span class="dv">8</span>, <span class="co"># queen's case</span></span>
<span id="cb171-4"><a href="#cb171-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">output =</span> <span class="st">"sf"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["id"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["na_prop_x"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["na_prop_y"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["dist"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["geometry"],"name":[5],"type":["s_POLYGO"],"align":["right"]}],"data":[{"1":"1","2":"0.04900085","3":"0.04900085","4":"0.4297049","5":"<s_POLYGO>"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>The patterns can be compared over various windows within the two maps. This allows the difference between sub-maps to be quantified.</p>
<p>The bounding box is 21 x 28 km. Let’s compare patterns over 4 x 4 km windows; these are 16 x 16 grid cells.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb172"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sg.class)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 112  84   1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb174"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb174-1"><a href="#cb174-1" aria-hidden="true" tabindex="-1"></a>x.dim <span class="ot">&lt;-</span> <span class="fu">diff</span>(<span class="fu">range</span>(<span class="fu">st_bbox</span>(sg.class)[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)]))</span>
<span id="cb174-2"><a href="#cb174-2" aria-hidden="true" tabindex="-1"></a>y.dim <span class="ot">&lt;-</span> <span class="fu">diff</span>(<span class="fu">range</span>(<span class="fu">st_bbox</span>(sg.class)[<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>)]))</span>
<span id="cb174-3"><a href="#cb174-3" aria-hidden="true" tabindex="-1"></a>(compare<span class="fl">.16</span> <span class="ot">&lt;-</span> <span class="fu">lsp_compare</span>(gn.class, sg.class, </span>
<span id="cb174-4"><a href="#cb174-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">type =</span> <span class="st">"cove"</span>, <span class="at">dist_fun =</span> <span class="st">"jensen-shannon"</span>,</span>
<span id="cb174-5"><a href="#cb174-5" aria-hidden="true" tabindex="-1"></a>                           <span class="at">neighbourhood =</span> <span class="dv">8</span>, <span class="co"># queen's case</span></span>
<span id="cb174-6"><a href="#cb174-6" aria-hidden="true" tabindex="-1"></a>                           <span class="at">window =</span> <span class="dv">16</span>,</span>
<span id="cb174-7"><a href="#cb174-7" aria-hidden="true" tabindex="-1"></a>                           <span class="at">output =</span> <span class="st">"sf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["id"],"name":[1],"type":["int"],"align":["right"]},{"label":["na_prop_x"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["na_prop_y"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["dist"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["geometry"],"name":[5],"type":["s_POLYGO"],"align":["right"]}],"data":[{"1":"1","2":"0.12890625","3":"0.12890625","4":"0.4959997","5":"<s_POLYGO>"},{"1":"2","2":"0.02343750","3":"0.02343750","4":"0.5035320","5":"<s_POLYGO>"},{"1":"3","2":"0.11718750","3":"0.11718750","4":"0.5794527","5":"<s_POLYGO>"},{"1":"4","2":"0.07421875","3":"0.07421875","4":"0.6101658","5":"<s_POLYGO>"},{"1":"5","2":"0.06250000","3":"0.06250000","4":"0.6779548","5":"<s_POLYGO>"},{"1":"7","2":"0.10156250","3":"0.10156250","4":"0.3203339","5":"<s_POLYGO>"},{"1":"8","2":"0.01953125","3":"0.01953125","4":"0.5168618","5":"<s_POLYGO>"},{"1":"9","2":"0.00781250","3":"0.00781250","4":"0.5059423","5":"<s_POLYGO>"},{"1":"10","2":"0.05859375","3":"0.05859375","4":"0.7748153","5":"<s_POLYGO>"},{"1":"11","2":"0.00390625","3":"0.00390625","4":"0.6143669","5":"<s_POLYGO>"},{"1":"13","2":"0.10156250","3":"0.10156250","4":"0.3426145","5":"<s_POLYGO>"},{"1":"14","2":"0.00000000","3":"0.00000000","4":"0.8477388","5":"<s_POLYGO>"},{"1":"15","2":"0.00000000","3":"0.00000000","4":"0.9815147","5":"<s_POLYGO>"},{"1":"16","2":"0.05859375","3":"0.05859375","4":"0.4692165","5":"<s_POLYGO>"},{"1":"17","2":"0.00000000","3":"0.00000000","4":"0.7526851","5":"<s_POLYGO>"},{"1":"19","2":"0.08203125","3":"0.08203125","4":"0.4773133","5":"<s_POLYGO>"},{"1":"20","2":"0.00000000","3":"0.00000000","4":"0.5112174","5":"<s_POLYGO>"},{"1":"21","2":"0.00390625","3":"0.00390625","4":"0.7971163","5":"<s_POLYGO>"},{"1":"22","2":"0.00390625","3":"0.00390625","4":"0.8862918","5":"<s_POLYGO>"},{"1":"23","2":"0.00000000","3":"0.00000000","4":"0.6398987","5":"<s_POLYGO>"},{"1":"25","2":"0.07421875","3":"0.07421875","4":"0.3727245","5":"<s_POLYGO>"},{"1":"26","2":"0.00390625","3":"0.00390625","4":"0.3451697","5":"<s_POLYGO>"},{"1":"27","2":"0.00000000","3":"0.00000000","4":"0.3636653","5":"<s_POLYGO>"},{"1":"28","2":"0.00390625","3":"0.00390625","4":"0.4785984","5":"<s_POLYGO>"},{"1":"29","2":"0.00781250","3":"0.00781250","4":"0.9636988","5":"<s_POLYGO>"},{"1":"31","2":"0.00781250","3":"0.00781250","4":"0.4486332","5":"<s_POLYGO>"},{"1":"32","2":"0.01562500","3":"0.01562500","4":"0.4361644","5":"<s_POLYGO>"},{"1":"33","2":"0.00000000","3":"0.00000000","4":"0.9085997","5":"<s_POLYGO>"},{"1":"34","2":"0.00390625","3":"0.00390625","4":"0.9021903","5":"<s_POLYGO>"},{"1":"35","2":"0.00390625","3":"0.00390625","4":"0.7773846","5":"<s_POLYGO>"},{"1":"37","2":"0.06250000","3":"0.06250000","4":"0.6462866","5":"<s_POLYGO>"},{"1":"38","2":"0.06250000","3":"0.06250000","4":"0.4351190","5":"<s_POLYGO>"},{"1":"39","2":"0.06640625","3":"0.06640625","4":"0.8973822","5":"<s_POLYGO>"},{"1":"40","2":"0.00000000","3":"0.00000000","4":"0.7598361","5":"<s_POLYGO>"},{"1":"41","2":"0.01171875","3":"0.01171875","4":"0.6906593","5":"<s_POLYGO>"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb175"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb175-1"><a href="#cb175-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> compare<span class="fl">.16</span>) <span class="sc">+</span></span>
<span id="cb175-2"><a href="#cb175-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> dist)) <span class="sc">+</span></span>
<span id="cb175-3"><a href="#cb175-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Distance between co-occurrence vectors, pH class"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/lsp.compare.window-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>We see that the distance between co-occurrence vectors varies across the map, although in all the submaps the distance is fairly large. The patterns are closer on the west side.</p>
<p>Visualize these distances along with the source maps.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb176"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> compare<span class="fl">.16</span>) <span class="sc">+</span> </span>
<span id="cb176-2"><a href="#cb176-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> dist))  <span class="sc">+</span></span>
<span id="cb176-3"><a href="#cb176-3" aria-hidden="true" tabindex="-1"></a>   <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>)</span>
<span id="cb176-4"><a href="#cb176-4" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb176-5"><a href="#cb176-5" aria-hidden="true" tabindex="-1"></a>  tidyterra<span class="sc">::</span><span class="fu">geom_spatraster</span>(<span class="at">data =</span> sg.class, <span class="fu">aes</span>(<span class="at">fill =</span> class)) <span class="sc">+</span></span>
<span id="cb176-6"><a href="#cb176-6" aria-hidden="true" tabindex="-1"></a>   <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>)</span>
<span id="cb176-7"><a href="#cb176-7" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb176-8"><a href="#cb176-8" aria-hidden="true" tabindex="-1"></a>  tidyterra<span class="sc">::</span><span class="fu">geom_spatraster</span>(<span class="at">data =</span> gn.class, <span class="fu">aes</span>(<span class="at">fill =</span> class)) <span class="sc">+</span></span>
<span id="cb176-9"><a href="#cb176-9" aria-hidden="true" tabindex="-1"></a>   <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>)</span>
<span id="cb176-10"><a href="#cb176-10" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(p1, p2, p3, <span class="at">nrow=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/cove-patterns-plot-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
</section>
<section id="vmetrics" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="vmetrics"><span class="header-section-number">7.2</span> V metrics</h2>
<p>The <em>V-measure</em> originated in the field of computer science as a measure for comparison of different clusterings of the same domain. It is a measure of an overall spatial correspondence between classified maps. Continuous maps must be classified into classes, and the two classified maps then compared. Note that the classes do not have to be the same, although in this example they are. The theme does not even have to be the same. For example one could compare a pH map with a clay concentration map.</p>
<p>Two maps could have the same total areas of each class, and even the same number of polygons within each class and even the same size distribution of these polygons, and yet be completely different in how they partition space into classes.</p>
<p>The polygons of a classified map are termed <em>regions of a regionalization</em> in the first (<em>reference</em>) map and <em>zones of a partition</em> in the second map. These are intersected to produce segment polygons of the combined map, which are labelled with both zone and region classes.</p>
<p>This allows the computation of two indices:</p>
<p><em>Homogeneity</em> compares the <em>zones in the second</em> map with respect to the <em>regions in the first</em>, i.e., how close the second map comes to reproducing the regionalization of the first. Thus it evaluates the second map, in terms of the first.</p>
<p>It is computed as the variance of the regions within a zone, normalized by the variance of the regions in the entire domain of the first map. These variances are computed by the Shannon entropy based on areas of the segments.</p>
<p>If the variance of the regions within the zones is small, then the partition into zones in the second map is relatively homogeneous with respect to the regionalization. A perfectly homogeneous partition (with value 1) is when each zone of the second map is within a single region of the reference map. In this case, each zone has only one reference class. A perfectly inhomogeneous partition (with value 0) is when each zone has the same composition of regions as the entire domain of the first map, i.e., the second map’s partition is essentially random with respect to the first map’s regionalization.</p>
<p><em>Completeness</em> is a function of homogeneity of the regions in the first map with respect to the zones in the second, i.e., how much the partition in the first map reproduce that of the second. Thus it evaluates the first map by the second.</p>
<p>The completeness of the second map is the inverse of homogeneity; it assesses the variance of the zones within a region normalized by the variance of the zones in the entire domain of the second map. It evaluates the homogeneity of regions with respect to zones and shows how well the regionalization of the reference map fits inside the partition of the map to be evaluated. A perfectly complete regionalization is when each region of the reference map is entirely within a single zone of the map to be evaluated. In this case, a polygon of the reference map will not be split among zones.</p>
<p>These two together are combined into a single <em>V measure</em> of agreement between the maps, as the harmonic mean of homogeneity and completeness.</p>
<p>This function uses the <code>sf::st_intersection()</code>, which depends on the coordinates values precision. For example, precision = 1000 rounds values to the third decimal places and precision = 0.001 uses values rounded to the nearest 1000, see <code>?sf::st_as_binary</code>.</p>
<p>V-measure methods are implemented in the <code>sabre</code> “Spatial Association Between Regionalizations” package, as explained in <span class="citation" data-cites="nowosadSpatialAssociationRegionalizations2018">(<a href="#ref-nowosadSpatialAssociationRegionalizations2018" role="doc-biblioref">Nowosad &amp; Stepinski, 2018a</a>)</span>.</p>
<p>The <code>vmeasure_calc()</code> function calculates intersections of the input geometries. For this function we must specify the names of the columns with the region names; both x and y must contain <code>POLYGON</code>s or <code>MULTIPOLYGON</code>s and have the same CRS.</p>
<section id="polygonize" class="level3" data-number="7.2.1">
<h3 data-number="7.2.1" class="anchored" data-anchor-id="polygonize"><span class="header-section-number">7.2.1</span> Polygonize</h3>
<p>The V-metrics require polygon maps, not gridded maps of classes.</p>
<p>Task: Polygonize them and adjust the class names.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a>gn.poly <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">as.polygons</span>(gn.class,</span>
<span id="cb177-2"><a href="#cb177-2" aria-hidden="true" tabindex="-1"></a>                              <span class="at">aggregate=</span> <span class="cn">TRUE</span>,</span>
<span id="cb177-3"><a href="#cb177-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">values =</span> <span class="cn">TRUE</span>,</span>
<span id="cb177-4"><a href="#cb177-4" aria-hidden="true" tabindex="-1"></a>                              <span class="at">dissolve =</span> <span class="cn">TRUE</span>)</span>
<span id="cb177-5"><a href="#cb177-5" aria-hidden="true" tabindex="-1"></a>sg.poly <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">as.polygons</span>(sg.class,</span>
<span id="cb177-6"><a href="#cb177-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">aggregate=</span> <span class="cn">TRUE</span>,</span>
<span id="cb177-7"><a href="#cb177-7" aria-hidden="true" tabindex="-1"></a>                              <span class="at">values =</span> <span class="cn">TRUE</span>, </span>
<span id="cb177-8"><a href="#cb177-8" aria-hidden="true" tabindex="-1"></a>                              <span class="at">dissolve=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="simple-features" class="level3" data-number="7.2.2">
<h3 data-number="7.2.2" class="anchored" data-anchor-id="simple-features"><span class="header-section-number">7.2.2</span> Simple Features</h3>
<p>Some of the methods require Simple Features representation of spatial objects.</p>
<p>Task: Convert the <code>terra::SpatVector</code> objects to Simple Features.</p>
<p>Sometimes these may have some simple <code>POLYGON</code>s, so ensure all are <code>MULTIPOLYGON</code> as required by <code>vmeasure_calc</code>, below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb178"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a>gn.sf <span class="ot">&lt;-</span> <span class="fu">st_as_sf</span>(gn.poly)</span>
<span id="cb178-2"><a href="#cb178-2" aria-hidden="true" tabindex="-1"></a>gn.sf <span class="ot">&lt;-</span> <span class="fu">st_cast</span>(gn.sf, <span class="st">"MULTIPOLYGON"</span>)</span>
<span id="cb178-3"><a href="#cb178-3" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb178-4"><a href="#cb178-4" aria-hidden="true" tabindex="-1"></a>sg.sf <span class="ot">&lt;-</span> <span class="fu">st_as_sf</span>(sg.poly)</span>
<span id="cb178-5"><a href="#cb178-5" aria-hidden="true" tabindex="-1"></a>sg.sf <span class="ot">&lt;-</span> <span class="fu">st_cast</span>(sg.sf, <span class="st">"MULTIPOLYGON"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="topology" class="level3" data-number="7.2.3">
<h3 data-number="7.2.3" class="anchored" data-anchor-id="topology"><span class="header-section-number">7.2.3</span> Topology</h3>
<p>Task: Check that the topology of the polygon map is correct. If not <code>sabre::vmeasure_calc</code> (see below) throws an error. Clean up the topology with <code>sf::st_make_valid</code>.</p>
<p>As explained <a href="https://www.r-spatial.org/r/2017/03/19/invalid.html">here</a>: “Spatial line and polygon data are often messy; although simple features formally follow a standard, there is no guarantee that data is clean when imported in R.”</p>
<p>Note: The typecasting to <code>MULTIPOLYGON</code> is required for the <code>sabre</code> methods. Without this, the geometry type is the more general <code>GEOMETRY</code> although all the items are already <code>MULTIPOLYGON</code>s.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb179"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb179-1"><a href="#cb179-1" aria-hidden="true" tabindex="-1"></a><span class="fu">st_is_valid</span>(gn.sf, <span class="at">reason=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Valid Geometry" "Valid Geometry" "Valid Geometry" "Valid Geometry"
 [5] "Valid Geometry" "Valid Geometry" "Valid Geometry" "Valid Geometry"
 [9] "Valid Geometry" "Valid Geometry" "Valid Geometry" "Valid Geometry"
[13] "Valid Geometry"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb181"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb181-1"><a href="#cb181-1" aria-hidden="true" tabindex="-1"></a>gn.sf.v <span class="ot">&lt;-</span> sf<span class="sc">::</span><span class="fu">st_make_valid</span>(gn.sf)  <span class="sc">|&gt;</span> <span class="fu">st_cast</span>(<span class="st">"MULTIPOLYGON"</span>)</span>
<span id="cb181-2"><a href="#cb181-2" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb181-3"><a href="#cb181-3" aria-hidden="true" tabindex="-1"></a><span class="fu">st_is_valid</span>(sg.sf, <span class="at">reason=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Valid Geometry" "Valid Geometry" "Valid Geometry" "Valid Geometry"
 [5] "Valid Geometry" "Valid Geometry" "Valid Geometry" "Valid Geometry"
 [9] "Valid Geometry" "Valid Geometry"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb183"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb183-1"><a href="#cb183-1" aria-hidden="true" tabindex="-1"></a>sg.sf.v <span class="ot">&lt;-</span> sf<span class="sc">::</span><span class="fu">st_make_valid</span>(sg.sf)  <span class="sc">|&gt;</span> <span class="fu">st_cast</span>(<span class="st">"MULTIPOLYGON"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this case the topology was valid, but if you are running this with your own maps it may not be.</p>
<p>Display the polygon maps. Compute each legend from the classes present in that map, but use a consistent colour scale. This requires some <code>ggplot2</code> tricks with <code>colorRampPalette</code> to set up a palette with a large number of discrete values, and the <code>limits</code> argument to <code>scale_fill_manual</code> to only show the parts of that scale occurring in each map.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb184"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb184-1"><a href="#cb184-1" aria-hidden="true" tabindex="-1"></a>classes.both <span class="ot">&lt;-</span> <span class="fu">union</span>(<span class="fu">values</span>(sg.poly)<span class="sc">$</span>class, <span class="fu">values</span>(gn.poly)<span class="sc">$</span>class)</span>
<span id="cb184-2"><a href="#cb184-2" aria-hidden="true" tabindex="-1"></a>my.pal <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">8</span>, <span class="st">"PuBu"</span>))(<span class="fu">length</span>(classes.both))</span>
<span id="cb184-3"><a href="#cb184-3" aria-hidden="true" tabindex="-1"></a>g0 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>gn.sf.v) <span class="sc">+</span></span>
<span id="cb184-4"><a href="#cb184-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> class)) <span class="sc">+</span></span>
<span id="cb184-5"><a href="#cb184-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_sf</span>(<span class="at">crs =</span> <span class="fu">st_crs</span>(gn.sf)) <span class="sc">+</span></span>
<span id="cb184-6"><a href="#cb184-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"gNATSGO"</span>)  <span class="sc">+</span></span>
<span id="cb184-7"><a href="#cb184-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> my.pal, <span class="at">drop=</span><span class="cn">TRUE</span>,</span>
<span id="cb184-8"><a href="#cb184-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">limits =</span> <span class="fu">levels</span>(gn.sf.v<span class="sc">$</span>class)) <span class="sc">+</span></span>
<span id="cb184-9"><a href="#cb184-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>, <span class="at">legend.direction =</span> <span class="st">"horizontal"</span>)</span>
<span id="cb184-10"><a href="#cb184-10" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>sg.sf.v) <span class="sc">+</span></span>
<span id="cb184-11"><a href="#cb184-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> class)) <span class="sc">+</span></span>
<span id="cb184-12"><a href="#cb184-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_sf</span>(<span class="at">crs =</span> <span class="fu">st_crs</span>(sg.sf)) <span class="sc">+</span></span>
<span id="cb184-13"><a href="#cb184-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"SG2"</span>)  <span class="sc">+</span></span>
<span id="cb184-14"><a href="#cb184-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> my.pal, <span class="at">drop=</span><span class="cn">TRUE</span>,</span>
<span id="cb184-15"><a href="#cb184-15" aria-hidden="true" tabindex="-1"></a>                    <span class="at">limits =</span> <span class="fu">levels</span>(gn.sf.v<span class="sc">$</span>class)) <span class="sc">+</span></span>
<span id="cb184-16"><a href="#cb184-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>, <span class="at">legend.direction =</span> <span class="st">"horizontal"</span>)</span>
<span id="cb184-17"><a href="#cb184-17" aria-hidden="true" tabindex="-1"></a><span class="fu">grid.arrange</span>(g0,g1, <span class="at">nrow=</span><span class="dv">1</span>, <span class="at">ncol=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/make_class_maps-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>This is the same information we’ve seen in the raster maps, but now organized as polygons.</p>
</section>
<section id="sec-vmeasure" class="level3" data-number="7.2.4">
<h3 data-number="7.2.4" class="anchored" data-anchor-id="sec-vmeasure"><span class="header-section-number">7.2.4</span> Compute the V-metrics</h3>
<p>Task: Compute the metrics with the <code>sabre</code>(“Spatial Association Between Regionalizations”) package. The second-listed map is the map to evaluate, with respect to the first-listed (reference) map. Here we evaluate the SoilGrids pattern vs.&nbsp;the gNATSGO as reference.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb185"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb185-1"><a href="#cb185-1" aria-hidden="true" tabindex="-1"></a>regions.sg.gn <span class="ot">&lt;-</span> sabre<span class="sc">::</span><span class="fu">vmeasure_calc</span>(<span class="at">x =</span> gn.sf.v, </span>
<span id="cb185-2"><a href="#cb185-2" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">y =</span> sg.sf.v, </span>
<span id="cb185-3"><a href="#cb185-3" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">x_name =</span> class, <span class="at">y_name =</span> class)</span>
<span id="cb185-4"><a href="#cb185-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(regions.sg.gn)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The SABRE results:

 V-measure: 0.04 
 Homogeneity: 0.04 
 Completeness: 0.04 

 The spatial objects can be retrieved with:
 $map1 - the first map
 $map2 - the second map </code></pre>
</div>
<div class="sourceCode cell-code" id="cb187"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb187-1"><a href="#cb187-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(regions.sg.gn)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "map1"         "map2"         "v_measure"    "homogeneity"  "completeness"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb189"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb189-1"><a href="#cb189-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(regions.sg.gn<span class="sc">$</span>map1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "map1"     "geometry" "rih"     </code></pre>
</div>
<div class="sourceCode cell-code" id="cb191"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb191-1"><a href="#cb191-1" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(regions.sg.gn, <span class="st">"precision"</span>)  <span class="co"># NULL, means a system default</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>NULL</code></pre>
</div>
</div>
<p>Both the homogeneity and completeness are near 0, their harmomic mean V-measure also very low. These maps hardly reseble each other. We will see the details just below.</p>
<p>Item <code>rih</code> (“region <em>in</em>homogeneity”) is the intersection map. Show these, but first the geometric precision must be set.</p>
<p>Geometric precision is set by <code>st_as_binary</code>, default is <code>attr(x, "precision")</code>. Here we left it as the default <code>NULL</code>.</p>
<p>Task: Plot the inhomogeneity and incompleteness.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb193"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb193-1"><a href="#cb193-1" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">plot</span>(regions.sg.gn<span class="sc">$</span>map1[<span class="st">"rih"</span>], <span class="at">main =</span> <span class="st">"Inhomogeneity -- SG2 vs. gNATSGO"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/vmaps.gnatsgo.sg.homogeneity-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>In this <em>inhomogeneity</em> map, almost all areas are highly inhomogeneous. This means that the SG2 polygons are highly variable – they contain almost the same distribution of classes from the gNATSGO map. Only a few patches are somewhat more homogeneous.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb194"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb194-1"><a href="#cb194-1" aria-hidden="true" tabindex="-1"></a>terra<span class="sc">::</span><span class="fu">plot</span>(regions.sg.gn<span class="sc">$</span>map2[<span class="st">"rih"</span>], <span class="at">main =</span> <span class="st">"Incompleteness -- SG2 vs. gNATSGO"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/vmaps.gnatsgo.sg.completeness-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>In this <em>incompleteness</em> map, all areas are quite incomplete, but there are more differences. The blue polygons (lowest values) are the most complete areas of the gNATSGO map, i.e.&nbsp;where the reference map has the most homogeneous set of SG2 classified values. The highly-incomplete areas are where gNATSGO has a limited range of values (in this case, a narrow range of higher pH) so they have low variance compared to SG2 predictions for these areas.</p>
<p>We conclude that the maps are quite different in their spatial patterns.</p>
</section>
</section>
</section>
<section id="supercells" class="level1" data-number="8">
<h1 data-number="8"><span class="header-section-number">8</span> Supercells</h1>
<p><em>“Superpixels”</em> is a generic name for grouping pixels with similar characteristics into larger assemblages. In the soil map context, the aim is to regionalize into areas with similar values of one or more raster layers.</p>
<p>The <code>supercells::supercells</code> function controls the segmentation: the user can specify the <code>k</code> arguement for the number of supercells, and the <code>compactness</code> argument to control shape: larger values lead to more square, less long/twisted shapes. It is also possible to specify a set of initial supercell centres (with an <code>sf</code> POINTS geometry) or a separation between initial centres with the <code>step</code> argument.</p>
<p>This function implements the SLIC algorithm <span class="citation" data-cites="achantaSLICSuperpixelsCompared2012">(<a href="#ref-achantaSLICSuperpixelsCompared2012" role="doc-biblioref">Achanta et al., 2012</a>)</span>.</p>
<p>As an example with the pH map, we divide into about 50 supercells, with low compactness since we don’t expect near-square natural units. Here is the source map:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb195"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb195-1"><a href="#cb195-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb195-2"><a href="#cb195-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_spatraster</span>(<span class="at">data=</span>sg.utm) <span class="sc">+</span></span>
<span id="cb195-3"><a href="#cb195-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"pH"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/SLIC-source-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>And here are the 50 supercells, with very low compactness:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb196"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb196-1"><a href="#cb196-1" aria-hidden="true" tabindex="-1"></a>sg.utm<span class="fl">.50</span> <span class="ot">=</span> <span class="fu">supercells</span>(sg.utm, <span class="at">k =</span> <span class="dv">50</span>, <span class="at">compactness =</span> <span class="fl">0.5</span>)</span>
<span id="cb196-2"><a href="#cb196-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>sg.utm<span class="fl">.50</span>) <span class="sc">+</span></span>
<span id="cb196-3"><a href="#cb196-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> phh2o_0<span class="fl">.5</span>cm_mean)) <span class="sc">+</span></span>
<span id="cb196-4"><a href="#cb196-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"mean pH"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/supercells-not-compact-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>Try to form more compact supercells:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb197"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb197-1"><a href="#cb197-1" aria-hidden="true" tabindex="-1"></a>sg.utm<span class="fl">.50</span> <span class="ot">=</span> <span class="fu">supercells</span>(sg.utm, <span class="at">k =</span> <span class="dv">50</span>, <span class="at">compactness =</span> <span class="dv">5</span>)</span>
<span id="cb197-2"><a href="#cb197-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>sg.utm<span class="fl">.50</span>) <span class="sc">+</span></span>
<span id="cb197-3"><a href="#cb197-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> phh2o_0<span class="fl">.5</span>cm_mean)) <span class="sc">+</span></span>
<span id="cb197-4"><a href="#cb197-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"mean pH"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/supercells-compact-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>These do not look realistic.</p>
<p>Try with multiple rasters, here pH and silt concentration:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb198"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb198-1"><a href="#cb198-1" aria-hidden="true" tabindex="-1"></a>r <span class="ot">&lt;-</span> <span class="fu">c</span>(sg.utm, sg.silt.utm)</span>
<span id="cb198-2"><a href="#cb198-2" aria-hidden="true" tabindex="-1"></a>r<span class="fl">.50</span> <span class="ot">=</span> <span class="fu">supercells</span>(r, <span class="at">k =</span> <span class="dv">50</span>, <span class="at">compactness =</span> <span class="fl">0.5</span>)</span>
<span id="cb198-3"><a href="#cb198-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>r<span class="fl">.50</span>) <span class="sc">+</span></span>
<span id="cb198-4"><a href="#cb198-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> phh2o_0<span class="fl">.5</span>cm_mean)) <span class="sc">+</span></span>
<span id="cb198-5"><a href="#cb198-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"mean pH"</span>) <span class="sc">+</span></span>
<span id="cb198-6"><a href="#cb198-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_continuous</span>(<span class="at">type =</span> <span class="st">"viridis"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/supercells-multiple-1.png" class="img-fluid" width="576"></p>
</div>
<div class="sourceCode cell-code" id="cb199"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb199-1"><a href="#cb199-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>r<span class="fl">.50</span>) <span class="sc">+</span></span>
<span id="cb199-2"><a href="#cb199-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="fu">aes</span>(<span class="at">fill =</span> silt_0<span class="fl">.5</span>cm_mean)) <span class="sc">+</span></span>
<span id="cb199-3"><a href="#cb199-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"silt ppt"</span>) <span class="sc">+</span></span>
<span id="cb199-4"><a href="#cb199-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_continuous</span>(<span class="at">type =</span> <span class="st">"viridis"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/supercells-multiple-2.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>Notice that the segments are the same in the two visualizations.</p>
</section>
<section id="sec-scale-geometry" class="level1" data-number="9">
<h1 data-number="9"><span class="header-section-number">9</span> Scale issues – geometric</h1>
<p>The soil pattern can be observed at different scales.</p>
<p>DSM are produced at a wide range of grid cell sizes (“resolutions”), and it’s obvious that as the grid cell size increases any finer pattern is lost. This is especially true if the map is made with block predictions, rather than point predictions at the centre of grid cells.</p>
<p>Clearly, products can only be compared at the same resolution. A finer-scale product can be downscaled to the resolution of a coarser-scale product in order to compare them. For maps made at point resolution this should by mean (continuous) or majority filter (classified). For maps made at block resolution this should be by block kriging within the block at fine-scale (continuous) or mode filter (classified).</p>
<p>What happens to the landscape metrics as the resolution changes? Let’s examine one map from the most detailed soil survey, gSSURGO.</p>
<p>The design scale for the polygon SSURGO product varies across the USA. For most areas with detailed survey, it is from 1:12k to 1:24K, with Minimum Legible Area (MLA) 0.576–2.304 ha. With a grid resolution of 16 grid cells per MLA, these would be from 12x12 – 24x24 cells. The 30 m gSSURGO roughly corresponds to this coarser resolution. It is equivalent to 1:30 k design scale.</p>
<p>In a previous section (<a href="#sec-import-gssurgo">Section&nbsp;3.4</a>) we imported a gSSURGO polygon map.</p>
<p>Unfortunately, the <code>landscapemetrics</code> package can not (yet?) work with <code>terra</code> objects, or vector objects from any package. So to compute landscape metrics for these polygons, they must be rasterized to a resolution where the area and length calculations are sufficiently accurate. The median polygon area is 2.7 ha; the minimum legible area (MLA) at the design scale of the soil surveys in this area (1:20k, see published survey document) is 1.6 ha. At a grid resolution of 16 pixels per MLA this suggests 20 m horizontal resolution pixels.</p>
<p>Task: Rasterize the polygon map to this resolution.</p>
<p>First set up the empty raster and then add the values of the map unit key.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb200"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb200-1"><a href="#cb200-1" aria-hidden="true" tabindex="-1"></a>mu.template <span class="ot">&lt;-</span> <span class="fu">rast</span>(mu.poly, <span class="at">res=</span><span class="fu">c</span>(<span class="dv">20</span>,<span class="dv">20</span>))</span>
<span id="cb200-2"><a href="#cb200-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(mu.template)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1408 1043    1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb202"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb202-1"><a href="#cb202-1" aria-hidden="true" tabindex="-1"></a>mu.raster <span class="ot">&lt;-</span> <span class="fu">rasterize</span>(mu.poly, mu.template, <span class="at">field=</span><span class="st">"mukey"</span>)</span>
<span id="cb202-2"><a href="#cb202-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(mu.raster)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     mukey        
 Min.   : 295575  
 1st Qu.: 295605  
 Median : 295651  
 Mean   : 616216  
 3rd Qu.: 295817  
 Max.   :2760841  
 NA's   :3149     </code></pre>
</div>
<div class="sourceCode cell-code" id="cb204"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb204-1"><a href="#cb204-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_landscape</span>(mu.raster)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["layer"],"name":[1],"type":["int"],"align":["right"]},{"label":["crs"],"name":[2],"type":["chr"],"align":["left"]},{"label":["units"],"name":[3],"type":["chr"],"align":["left"]},{"label":["class"],"name":[4],"type":["chr"],"align":["left"]},{"label":["n_classes"],"name":[5],"type":["int"],"align":["right"]},{"label":["OK"],"name":[6],"type":["chr"],"align":["left"]}],"data":[{"1":"1","2":"projected","3":"m","4":"integer","5":"217","6":"❓"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>This is a categorical map (the map unit keys are the categories), and we can apply the landscape metrics to it.</p>
<p><strong>Map units</strong>:</p>
<p>Note that there are 217 classes, i.e., different map units, in this area.</p>
<p>TaskL Show all patches, then just the “Alluvial land”, code 295575, then all the map units with the Mardin series as a component.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb205"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb205-1"><a href="#cb205-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">unique</span>(mu.raster<span class="sc">$</span>mukey))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["mukey"],"name":[1],"type":["dbl"],"align":["right"]}],"data":[{"1":"295575","_rn_":"1"},{"1":"295576","_rn_":"2"},{"1":"295577","_rn_":"3"},{"1":"295578","_rn_":"4"},{"1":"295579","_rn_":"5"},{"1":"295580","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb206"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb206-1"><a href="#cb206-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_patches</span>(mu.raster, <span class="at">class =</span> <span class="st">"global"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show.landscape.gn-1.png" class="img-fluid" width="1344"></p>
</div>
<div class="sourceCode cell-code" id="cb208"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb208-1"><a href="#cb208-1" aria-hidden="true" tabindex="-1"></a> <span class="fu">show_patches</span>(mu.raster, <span class="at">class =</span> <span class="dv">295575</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show.landscape.gn-2.png" class="img-fluid" width="1344"></p>
</div>
<div class="sourceCode cell-code" id="cb210"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb210-1"><a href="#cb210-1" aria-hidden="true" tabindex="-1"></a>(ix <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"Mardin"</span>, mu.key<span class="sc">$</span>muname, <span class="at">fixed =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  63  64  65  67 100 101 102 120 121 122 123 124 125 126 149 178 179 180 181
[20] 182</code></pre>
</div>
<div class="sourceCode cell-code" id="cb212"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb212-1"><a href="#cb212-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(mu.key[ix, ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      mukey                                                             muname
63   295656                   Mardin channery silt loam, 2 to 8 percent slopes
64   295657                  Mardin channery silt loam, 8 to 15 percent slopes
65   295658          Mardin channery silt loam, 8 to 15 percent slopes, eroded
67   295660                 Mardin and Langford soils, 15 to 25 percent slopes
100  295805                    Mardin-Volusia complex, 15 to 30 percent slopes
101  295806                     Mardin-Volusia complex, 8 to 15 percent slopes
102  295807                      Mardin-Volusia complex, 0 to 8 percent slopes
120  295831         Mardin channery silt loam, 16 to 30 percent slopes, eroded
121  295832                 Mardin channery silt loam, 16 to 30 percent slopes
122  295833 Mardin channery silt loam, 9 to 15 percent slopes, moderately deep
123  295834          Mardin channery silt loam, 8 to 15 percent slopes, eroded
124  295835                  Mardin channery silt loam, 8 to 15 percent slopes
125  295836                   Mardin channery silt loam, 0 to 8 percent slopes
126  295837  Mardin channery silt loam, 0 to 8 percent slopes, moderately deep
149 2723042                     Bath and Mardin soils, 25 to 40 percent slopes
178 2723076                   Mardin channery silt loam, 2 to 8 percent slopes
179 2723077                 Mardin channery silt loam, 15 to 25 percent slopes
180 2723078                  Mardin channery silt loam, 8 to 15 percent slopes
181 2723080    Mardin channery silt loam, 3 to 8 percent slopes, slightly acid
182 2723081   Mardin channery silt loam, 8 to 15 percent slopes, slightly acid</code></pre>
</div>
<div class="sourceCode cell-code" id="cb214"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb214-1"><a href="#cb214-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_patches</span>(mu.raster, </span>
<span id="cb214-2"><a href="#cb214-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">class =</span> <span class="fu">c</span>(mu.key[ix, <span class="st">"mukey"</span>]), <span class="at">nrow =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show.landscape.gn-3.png" class="img-fluid" width="1344"></p>
</div>
</div>
<p>Some of these are quite small, and are categorically not so different. Much of this is due to the separate surveys in three counties. In <a href="#sec-categorical-generalization">Section&nbsp;10</a> we combine similar map units for the landscape analysis.</p>
<section id="m-resolution" class="level2" data-number="9.1">
<h2 data-number="9.1" class="anchored" data-anchor-id="m-resolution"><span class="header-section-number">9.1</span> 20 m resolution</h2>
<p>We rasterized the map units at 20 m grid resolution, so first analyze at this scale.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb216"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb216-1"><a href="#cb216-1" aria-hidden="true" tabindex="-1"></a>gn<span class="fl">.20</span> <span class="ot">&lt;-</span> mu.raster</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="landscape-level" class="level3" data-number="9.1.1">
<h3 data-number="9.1.1" class="anchored" data-anchor-id="landscape-level"><span class="header-section-number">9.1.1</span> Landscape level</h3>
<p>Compute the landscape metrics.:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb217"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb217-1"><a href="#cb217-1" aria-hidden="true" tabindex="-1"></a>lst <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"lsm_l_"</span>, <span class="fu">c</span>(<span class="st">"shdi"</span>, <span class="st">"shei"</span>, <span class="st">"lsi"</span>, <span class="st">"ai"</span>,  <span class="st">"frac_mn"</span>))</span>
<span id="cb217-2"><a href="#cb217-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(ls.metrics.gn20 <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(gn<span class="fl">.20</span>, <span class="at">what=</span>lst)[, <span class="fu">c</span>(<span class="st">"metric"</span>, <span class="st">"value"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 2
  metric   value
  &lt;chr&gt;    &lt;dbl&gt;
1 ai      86.3  
2 frac_mn  1.11 
3 lsi     86.7  
4 shdi     4.39 
5 shei     0.815</code></pre>
</div>
</div>
<p>Because of the small patches, very fine pattern,large number of classes, the <code>lsi</code> is extremely high – i.e., far more boundaries than if the classes were contiguous. The <code>ai</code> is also quite high. The Shannon Diversity is also very high, because of the large number of classes and relatively equal areas. The Shannon Evenness is quite high, i.e., the areas of the classes are fairly even, there is no dominant class.</p>
</section>
<section id="class-level" class="level3" data-number="9.1.2">
<h3 data-number="9.1.2" class="anchored" data-anchor-id="class-level"><span class="header-section-number">9.1.2</span> Class level</h3>
<p>The class-level metrics show the distribution of classes. For example, the percentage of landscape occupied by each class (<code>lsm_c_pland</code>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb219"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb219-1"><a href="#cb219-1" aria-hidden="true" tabindex="-1"></a>c_pland<span class="fl">.20</span> <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(gn<span class="fl">.20</span>, <span class="at">what=</span><span class="st">"lsm_c_pland"</span>)</span>
<span id="cb219-2"><a href="#cb219-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(c_pland<span class="fl">.20</span><span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), <span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 5.303681 5.034498 4.229760 4.166365 3.978148 3.566361 3.380955 2.936838
 [9] 2.779685 2.343791 1.947256 1.936221 1.890326 1.692129 1.666687 1.635481
[17] 1.601113 1.533009 1.490206 1.429131 1.368758 1.227630 1.169155 1.063941</code></pre>
</div>
<div class="sourceCode cell-code" id="cb221"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb221-1"><a href="#cb221-1" aria-hidden="true" tabindex="-1"></a>ix <span class="ot">&lt;-</span> <span class="fu">order</span>(c_pland<span class="fl">.20</span><span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb221-2"><a href="#cb221-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(c_pland<span class="fl">.20</span><span class="sc">$</span>class[ix], <span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 295598 295635 295685 295582 295647 295683 295649 295656 295650 295690
[11] 295829 295657 295611 295581 295585 295828 295584 295827 295579 295604
[21] 295575 295854 295651 295596</code></pre>
</div>
</div>
<p>We see that there are a few large classes; the largest is about 5.6% of the landscape. But there are many more small ones.</p>
</section>
<section id="patch-level" class="level3" data-number="9.1.3">
<h3 data-number="9.1.3" class="anchored" data-anchor-id="patch-level"><span class="header-section-number">9.1.3</span> Patch level</h3>
<p>In this case patch-level metrics are also interesting – they reveal size and shape, for example.</p>
<p>One example is <strong>patch area</strong>, expressed in ha (<span class="math inline">\(10000 \; \mathrm{m}^2\)</span>) – these should all be larger than the MLA (0.576–2.304 ha depending on original design scale).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb223"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb223-1"><a href="#cb223-1" aria-hidden="true" tabindex="-1"></a><span class="co"># each patch</span></span>
<span id="cb223-2"><a href="#cb223-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(area<span class="fl">.20</span> <span class="ot">&lt;-</span>  </span>
<span id="cb223-3"><a href="#cb223-3" aria-hidden="true" tabindex="-1"></a>            <span class="fu">calculate_lsm</span>(gn<span class="fl">.20</span>, <span class="at">what=</span><span class="st">"lsm_p_area"</span>)<span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 336.88 321.04 212.68 199.56 173.80 162.32</code></pre>
</div>
<div class="sourceCode cell-code" id="cb225"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb225-1"><a href="#cb225-1" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(area<span class="fl">.20</span> , <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span>.<span class="dv">12</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     0%     12%     24%     36%     48%     60%     72%     84%     96% 
 0.0400  0.5600  1.0000  1.5600  2.3232  3.4400  5.4800  9.4400 24.7600 </code></pre>
</div>
</div>
<p>About 12% of the patches are smaller. This may be due to the rasterizing process (?).</p>
<p>Another example is the <strong>number of core areas</strong> within patches. This measures how complex is each patch:</p>
<p>“A cell is defined as core if the cell has no neighbour with a different value than itself (rook’s case). The metric counts the disjunct core areas, whereby a core area is a ‘patch within the patch’ containing only core cells. It describes patch area and shape simultaneously (more core area when the patch is large, however, the shape must allow disjunct core areas). Thereby, a compact shape (e.g.&nbsp;a square) will contain [fewer] disjunct core areas than a more irregular patch.”</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb227"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb227-1"><a href="#cb227-1" aria-hidden="true" tabindex="-1"></a><span class="co"># each patch</span></span>
<span id="cb227-2"><a href="#cb227-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ncore<span class="fl">.20</span> <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(gn<span class="fl">.20</span>, <span class="at">what=</span><span class="st">"lsm_p_ncore"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["layer"],"name":[1],"type":["int"],"align":["right"]},{"label":["level"],"name":[2],"type":["chr"],"align":["left"]},{"label":["class"],"name":[3],"type":["int"],"align":["right"]},{"label":["id"],"name":[4],"type":["int"],"align":["right"]},{"label":["metric"],"name":[5],"type":["chr"],"align":["left"]},{"label":["value"],"name":[6],"type":["dbl"],"align":["right"]}],"data":[{"1":"1","2":"patch","3":"295575","4":"1","5":"ncore","6":"2"},{"1":"1","2":"patch","3":"295575","4":"2","5":"ncore","6":"2"},{"1":"1","2":"patch","3":"295575","4":"3","5":"ncore","6":"10"},{"1":"1","2":"patch","3":"295575","4":"4","5":"ncore","6":"1"},{"1":"1","2":"patch","3":"295575","4":"5","5":"ncore","6":"1"},{"1":"1","2":"patch","3":"295575","4":"6","5":"ncore","6":"12"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb228"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb228-1"><a href="#cb228-1" aria-hidden="true" tabindex="-1"></a><span class="co"># summarize by map unit</span></span>
<span id="cb228-2"><a href="#cb228-2" aria-hidden="true" tabindex="-1"></a>ncore.<span class="fl">20.</span>summary <span class="ot">&lt;-</span> ncore<span class="fl">.20</span> <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(class) <span class="sc">%&gt;%</span></span>
<span id="cb228-3"><a href="#cb228-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">max_cores =</span> <span class="fu">max</span>(value)) <span class="sc">%&gt;%</span></span>
<span id="cb228-4"><a href="#cb228-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(class)</span>
<span id="cb228-5"><a href="#cb228-5" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">sort</span>(ncore.<span class="fl">20.</span>summary<span class="sc">$</span>max_cores, <span class="at">decreasing =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] 40 22 18 14 14 13 12 12 12 12 12 12 11 11 10 10 10 10 10 10 10  9  9  9  9
 [26]  9  9  9  8  8  8  8  8  8  8  7  7  7  7  7  7  6  6  6  6  6  6  6  6  6
 [51]  6  6  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  4  4
 [76]  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  3
[101]  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
[126]  3  3  3  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
[151]  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  1  1  1  1  1
[176]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[201]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  0  0  0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb230"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb230-1"><a href="#cb230-1" aria-hidden="true" tabindex="-1"></a>ix <span class="ot">&lt;-</span> <span class="fu">order</span>(ncore.<span class="fl">20.</span>summary<span class="sc">$</span>max_cores, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb230-2"><a href="#cb230-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cbind</span>(mu.key[ix[<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>], ], <span class="at">ncore =</span> ncore.<span class="fl">20.</span>summary<span class="sc">$</span>max_cores[ix[<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["mukey"],"name":[1],"type":["int"],"align":["right"]},{"label":["muname"],"name":[2],"type":["chr"],"align":["left"]},{"label":["ncore"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"295598","2":"Erie channery silt loam, 3 to 8 percent slopes","3":"40","_rn_":"21"},{"1":"295690","2":"Wayland soils complex, 0 to 3 percent slopes, frequently flooded","3":"22","_rn_":"93"},{"1":"295575","2":"Alluvial land","3":"18","_rn_":"1"},{"1":"295635","2":"Langford channery silt loam, 2 to 8 percent slopes","3":"14","_rn_":"52"},{"1":"2723105","2":"Valois and Howard gravelly loams, 25 to 40 percent slopes","3":"14","_rn_":"195"},{"1":"295613","2":"Howard gravelly loam, 15 to 25 percent slopes","3":"13","_rn_":"36"},{"1":"295582","2":"Bath and Valois soils, 5 to 15 percent slopes","3":"12","_rn_":"8"},{"1":"295584","2":"Bath and Valois soils, 15 to 25 percent slopes, eroded","3":"12","_rn_":"10"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>The Erie unit has by far the most core areas. Display it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb231"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb231-1"><a href="#cb231-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_patches</span>(gn<span class="fl">.20</span>, <span class="at">class =</span> mu.key[ix[<span class="dv">1</span>], <span class="st">"mukey"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show-patches-Erie-20-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>Display the <code>ncore</code> patch-level metric as a map:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb233"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb233-1"><a href="#cb233-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_lsm</span>(gn<span class="fl">.20</span>, <span class="st">"lsm_p_ncore"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show-ncore-20-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>We can see the highly-fragmented map unit on the upper left.</p>
</section>
</section>
<section id="m-resolution-1" class="level2" data-number="9.2">
<h2 data-number="9.2" class="anchored" data-anchor-id="m-resolution-1"><span class="header-section-number">9.2</span> 100 m resolution</h2>
<p>What happens as the resolution is coarsened to 100 x 100, roughly equivalent to 1:100k design scale?</p>
<p>We use the majority (“modal”) filter in the reclassification, with a one-dimension 5-fold aggregation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb235"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb235-1"><a href="#cb235-1" aria-hidden="true" tabindex="-1"></a>gn<span class="fl">.100</span> <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">aggregate</span>(gn<span class="fl">.20</span>, <span class="at">fact=</span><span class="dv">5</span>, <span class="at">fun=</span><span class="st">"modal"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Patches are clearly larger than in the 20 m resolution.</p>
<section id="landscape-level-1" class="level3" data-number="9.2.1">
<h3 data-number="9.2.1" class="anchored" data-anchor-id="landscape-level-1"><span class="header-section-number">9.2.1</span> Landscape level</h3>
<p>Landscape metrics:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb236"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb236-1"><a href="#cb236-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(ls.metrics.gn20[, <span class="fu">c</span>(<span class="st">"metric"</span>, <span class="st">"value"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 2
  metric   value
  &lt;chr&gt;    &lt;dbl&gt;
1 ai      86.3  
2 frac_mn  1.11 
3 lsi     86.7  
4 shdi     4.39 
5 shei     0.815</code></pre>
</div>
<div class="sourceCode cell-code" id="cb238"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb238-1"><a href="#cb238-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(ls.metrics.gn100 <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(gn<span class="fl">.100</span>, <span class="at">what=</span>lst)[, <span class="fu">c</span>(<span class="st">"metric"</span>, <span class="st">"value"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 2
  metric   value
  &lt;chr&gt;    &lt;dbl&gt;
1 ai      52.3  
2 frac_mn  1.04 
3 lsi     60.0  
4 shdi     4.37 
5 shei     0.815</code></pre>
</div>
</div>
<p>The Aggregation Index and Landscape Shape Index are much lower. The Fractal Dimension is somewhat lower. The Shannon diversity and evenness have hardly changed.</p>
</section>
<section id="class-level-1" class="level3" data-number="9.2.2">
<h3 data-number="9.2.2" class="anchored" data-anchor-id="class-level-1"><span class="header-section-number">9.2.2</span> Class level</h3>
<p>What about the class metrics?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb240"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb240-1"><a href="#cb240-1" aria-hidden="true" tabindex="-1"></a>c_pland<span class="fl">.100</span> <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(gn<span class="fl">.100</span>, <span class="at">what=</span><span class="st">"lsm_c_pland"</span>)</span>
<span id="cb240-2"><a href="#cb240-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(c_pland<span class="fl">.20</span><span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), <span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 5.303681 5.034498 4.229760 4.166365 3.978148 3.566361 3.380955 2.936838
 [9] 2.779685 2.343791 1.947256 1.936221 1.890326 1.692129 1.666687 1.635481
[17] 1.601113 1.533009 1.490206 1.429131 1.368758 1.227630 1.169155 1.063941</code></pre>
</div>
<div class="sourceCode cell-code" id="cb242"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb242-1"><a href="#cb242-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(c_pland<span class="fl">.100</span><span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), <span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 5.341283 5.029900 4.299211 4.230211 4.062135 3.612753 3.368600 2.924525
 [9] 2.798910 2.358374 1.930222 1.926683 1.903684 1.701992 1.668377 1.663069
[17] 1.624146 1.551608 1.496762 1.454301 1.417147 1.213687 1.201302 1.084533</code></pre>
</div>
</div>
<p>Very little difference in the class composition.</p>
</section>
<section id="patch-level-1" class="level3" data-number="9.2.3">
<h3 data-number="9.2.3" class="anchored" data-anchor-id="patch-level-1"><span class="header-section-number">9.2.3</span> Patch level</h3>
<p>Patch metrics:</p>
<p>First, the patch areas.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb244"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb244-1"><a href="#cb244-1" aria-hidden="true" tabindex="-1"></a><span class="co"># each patch</span></span>
<span id="cb244-2"><a href="#cb244-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(area<span class="fl">.100</span> <span class="ot">&lt;-</span>  </span>
<span id="cb244-3"><a href="#cb244-3" aria-hidden="true" tabindex="-1"></a>            <span class="fu">calculate_lsm</span>(gn<span class="fl">.100</span>, <span class="at">what=</span><span class="st">"lsm_p_area"</span>)<span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 350 286 271 260 253 240</code></pre>
</div>
<div class="sourceCode cell-code" id="cb246"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb246-1"><a href="#cb246-1" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(area<span class="fl">.20</span>, <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span>.<span class="dv">12</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     0%     12%     24%     36%     48%     60%     72%     84%     96% 
 0.0400  0.5600  1.0000  1.5600  2.3232  3.4400  5.4800  9.4400 24.7600 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb248"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb248-1"><a href="#cb248-1" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(area<span class="fl">.100</span>, <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span>.<span class="dv">12</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> 0% 12% 24% 36% 48% 60% 72% 84% 96% 
  1   1   1   2   3   4   6  11  28 </code></pre>
</div>
</div>
<p>Now all the patches are larger than the MLA for 1:12k.</p>
<p>The core areas:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb250"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb250-1"><a href="#cb250-1" aria-hidden="true" tabindex="-1"></a><span class="co"># each patch</span></span>
<span id="cb250-2"><a href="#cb250-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ncore<span class="fl">.100</span> <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(gn<span class="fl">.100</span>, <span class="at">what=</span><span class="st">"lsm_p_ncore"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["layer"],"name":[1],"type":["int"],"align":["right"]},{"label":["level"],"name":[2],"type":["chr"],"align":["left"]},{"label":["class"],"name":[3],"type":["int"],"align":["right"]},{"label":["id"],"name":[4],"type":["int"],"align":["right"]},{"label":["metric"],"name":[5],"type":["chr"],"align":["left"]},{"label":["value"],"name":[6],"type":["dbl"],"align":["right"]}],"data":[{"1":"1","2":"patch","3":"295575","4":"1","5":"ncore","6":"2"},{"1":"1","2":"patch","3":"295575","4":"2","5":"ncore","6":"0"},{"1":"1","2":"patch","3":"295575","4":"3","5":"ncore","6":"0"},{"1":"1","2":"patch","3":"295575","4":"4","5":"ncore","6":"0"},{"1":"1","2":"patch","3":"295575","4":"5","5":"ncore","6":"0"},{"1":"1","2":"patch","3":"295575","4":"6","5":"ncore","6":"0"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb251"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb251-1"><a href="#cb251-1" aria-hidden="true" tabindex="-1"></a><span class="co"># summarize by map unit</span></span>
<span id="cb251-2"><a href="#cb251-2" aria-hidden="true" tabindex="-1"></a>ncore.<span class="fl">100.</span>summary <span class="ot">&lt;-</span> ncore<span class="fl">.100</span> <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(class) <span class="sc">%&gt;%</span></span>
<span id="cb251-3"><a href="#cb251-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">max_cores =</span> <span class="fu">max</span>(value)) <span class="sc">%&gt;%</span></span>
<span id="cb251-4"><a href="#cb251-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(class)</span>
<span id="cb251-5"><a href="#cb251-5" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">sort</span>(ncore.<span class="fl">20.</span>summary<span class="sc">$</span>max_cores, <span class="at">decreasing =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] 40 22 18 14 14 13 12 12 12 12 12 12 11 11 10 10 10 10 10 10 10  9  9  9  9
 [26]  9  9  9  8  8  8  8  8  8  8  7  7  7  7  7  7  6  6  6  6  6  6  6  6  6
 [51]  6  6  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  4  4
 [76]  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  3
[101]  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
[126]  3  3  3  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
[151]  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  1  1  1  1  1
[176]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[201]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  0  0  0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb253"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb253-1"><a href="#cb253-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">sort</span>(ncore.<span class="fl">100.</span>summary<span class="sc">$</span>max_cores, <span class="at">decreasing =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] 13 12  9  9  9  7  7  6  5  5  4  4  4  4  4  4  3  3  3  3  3  3  3  3  3
 [26]  3  3  3  3  3  3  3  3  3  3  3  3  2  2  2  2  2  2  2  2  2  2  2  2  2
 [51]  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
 [76]  2  2  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[101]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[126]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  0  0  0  0  0
[151]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
[176]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
[201]  0  0  0  0  0  0  0  0  0  0  0  0  0  0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb255"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb255-1"><a href="#cb255-1" aria-hidden="true" tabindex="-1"></a>ix <span class="ot">&lt;-</span> <span class="fu">order</span>(ncore.<span class="fl">100.</span>summary<span class="sc">$</span>max_cores, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb255-2"><a href="#cb255-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cbind</span>(mu.key[ix[<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>], ], <span class="at">ncore =</span> ncore.<span class="fl">100.</span>summary<span class="sc">$</span>max_cores[ix[<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["mukey"],"name":[1],"type":["int"],"align":["right"]},{"label":["muname"],"name":[2],"type":["chr"],"align":["left"]},{"label":["ncore"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"295598","2":"Erie channery silt loam, 3 to 8 percent slopes","3":"13","_rn_":"21"},{"1":"295635","2":"Langford channery silt loam, 2 to 8 percent slopes","3":"12","_rn_":"52"},{"1":"295582","2":"Bath and Valois soils, 5 to 15 percent slopes","3":"9","_rn_":"8"},{"1":"295647","2":"Lordstown channery silt loam, 5 to 15 percent slopes","3":"9","_rn_":"56"},{"1":"295683","2":"Volusia channery silt loam, 3 to 8 percent slopes","3":"9","_rn_":"86"},{"1":"295656","2":"Mardin channery silt loam, 2 to 8 percent slopes","3":"7","_rn_":"63"},{"1":"295685","2":"Volusia channery silt loam, 8 to 15 percent slopes","3":"7","_rn_":"88"},{"1":"295690","2":"Wayland soils complex, 0 to 3 percent slopes, frequently flooded","3":"6","_rn_":"93"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>The Erie unit again has the most core areas, but many small cores have been absorbed in their neighbours.</p>
<p>Display it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb256"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb256-1"><a href="#cb256-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_patches</span>(gn<span class="fl">.100</span>, <span class="at">class =</span> mu.key[ix[<span class="dv">1</span>], <span class="st">"mukey"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show-patches-Erie-100-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>Display the <code>ncore</code> patch-level metric as a map:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb258"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb258-1"><a href="#cb258-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_lsm</span>(gn<span class="fl">.100</span>, <span class="st">"lsm_p_ncore"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show-ncore-100-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
</section>
</section>
<section id="m-resolution-2" class="level2" data-number="9.3">
<h2 data-number="9.3" class="anchored" data-anchor-id="m-resolution-2"><span class="header-section-number">9.3</span> 300 m resolution</h2>
<p>What happens as the resolution is again coarsened to 300 x 300 m, roughly equivalent to 1:300 k design scale?</p>
<p>We use a 3x3 aggregation, again with a modal filter.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb260"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb260-1"><a href="#cb260-1" aria-hidden="true" tabindex="-1"></a>gn<span class="fl">.300</span> <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">aggregate</span>(gn<span class="fl">.100</span>, <span class="at">fact=</span><span class="dv">3</span>, <span class="at">fun=</span><span class="st">"modal"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="landscape-level-2" class="level3" data-number="9.3.1">
<h3 data-number="9.3.1" class="anchored" data-anchor-id="landscape-level-2"><span class="header-section-number">9.3.1</span> Landscape level</h3>
<p>Landscape metrics:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb261"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb261-1"><a href="#cb261-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(ls.metrics.gn20[, <span class="fu">c</span>(<span class="st">"metric"</span>, <span class="st">"value"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 2
  metric   value
  &lt;chr&gt;    &lt;dbl&gt;
1 ai      86.3  
2 frac_mn  1.11 
3 lsi     86.7  
4 shdi     4.39 
5 shei     0.815</code></pre>
</div>
<div class="sourceCode cell-code" id="cb263"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb263-1"><a href="#cb263-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(ls.metrics.gn100[, <span class="fu">c</span>(<span class="st">"metric"</span>, <span class="st">"value"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 2
  metric   value
  &lt;chr&gt;    &lt;dbl&gt;
1 ai      52.3  
2 frac_mn  1.04 
3 lsi     60.0  
4 shdi     4.37 
5 shei     0.815</code></pre>
</div>
<div class="sourceCode cell-code" id="cb265"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb265-1"><a href="#cb265-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(ls.metrics.gn300 <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(gn<span class="fl">.300</span>, <span class="at">what=</span>lst)[, <span class="fu">c</span>(<span class="st">"metric"</span>, <span class="st">"value"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 2
  metric   value
  &lt;chr&gt;    &lt;dbl&gt;
1 ai      30.2  
2 frac_mn  1.02 
3 lsi     29.2  
4 shdi     4.28 
5 shei     0.830</code></pre>
</div>
</div>
<p>The Aggregation Index and Landscape Shape Index are again much lower. The Fractal Dimension is again somewhat lower. The Shannon diversity is somewhat lower and the evenness somewhat higher.</p>
</section>
<section id="class-level-2" class="level3" data-number="9.3.2">
<h3 data-number="9.3.2" class="anchored" data-anchor-id="class-level-2"><span class="header-section-number">9.3.2</span> Class level</h3>
<p>What about the class metrics?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb267"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb267-1"><a href="#cb267-1" aria-hidden="true" tabindex="-1"></a>c_pland<span class="fl">.300</span> <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(gn<span class="fl">.300</span>, <span class="at">what=</span><span class="st">"lsm_c_pland"</span>)</span>
<span id="cb267-2"><a href="#cb267-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(c_pland<span class="fl">.20</span><span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), <span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 5.303681 5.034498 4.229760 4.166365 3.978148 3.566361 3.380955 2.936838
 [9] 2.779685 2.343791 1.947256 1.936221 1.890326 1.692129 1.666687 1.635481
[17] 1.601113 1.533009 1.490206 1.429131 1.368758 1.227630 1.169155 1.063941</code></pre>
</div>
<div class="sourceCode cell-code" id="cb269"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb269-1"><a href="#cb269-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(c_pland<span class="fl">.100</span><span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), <span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 5.341283 5.029900 4.299211 4.230211 4.062135 3.612753 3.368600 2.924525
 [9] 2.798910 2.358374 1.930222 1.926683 1.903684 1.701992 1.668377 1.663069
[17] 1.624146 1.551608 1.496762 1.454301 1.417147 1.213687 1.201302 1.084533</code></pre>
</div>
<div class="sourceCode cell-code" id="cb271"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb271-1"><a href="#cb271-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(c_pland<span class="fl">.300</span><span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), <span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 5.869953 5.140263 4.815956 4.653802 4.297065 3.534944 3.340360 3.178207
 [9] 2.594454 2.156640 2.124210 1.978271 1.978271 1.945841 1.783687 1.767472
[17] 1.702611 1.637749 1.605319 1.508027 1.491811 1.167504 1.167504 1.135074</code></pre>
</div>
</div>
<p>The largest classes are larger.</p>
</section>
<section id="patch-level-2" class="level3" data-number="9.3.3">
<h3 data-number="9.3.3" class="anchored" data-anchor-id="patch-level-2"><span class="header-section-number">9.3.3</span> Patch level</h3>
<p>First, the patch areas.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb273"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb273-1"><a href="#cb273-1" aria-hidden="true" tabindex="-1"></a><span class="co"># each patch</span></span>
<span id="cb273-2"><a href="#cb273-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(area<span class="fl">.300</span> <span class="ot">&lt;-</span>  </span>
<span id="cb273-3"><a href="#cb273-3" aria-hidden="true" tabindex="-1"></a>            <span class="fu">calculate_lsm</span>(gn<span class="fl">.300</span>, <span class="at">what=</span><span class="st">"lsm_p_area"</span>)<span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 621 432 423 369 342 306</code></pre>
</div>
<div class="sourceCode cell-code" id="cb275"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb275-1"><a href="#cb275-1" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(area<span class="fl">.20</span>, <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span>.<span class="dv">12</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     0%     12%     24%     36%     48%     60%     72%     84%     96% 
 0.0400  0.5600  1.0000  1.5600  2.3232  3.4400  5.4800  9.4400 24.7600 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb277"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb277-1"><a href="#cb277-1" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(area<span class="fl">.100</span>, <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span>.<span class="dv">12</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> 0% 12% 24% 36% 48% 60% 72% 84% 96% 
  1   1   1   2   3   4   6  11  28 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb279"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb279-1"><a href="#cb279-1" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(area<span class="fl">.300</span>, <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span>.<span class="dv">12</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> 0% 12% 24% 36% 48% 60% 72% 84% 96% 
  9   9   9   9   9  18  18  27  72 </code></pre>
</div>
</div>
<p>Now the cores:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb281"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb281-1"><a href="#cb281-1" aria-hidden="true" tabindex="-1"></a><span class="co"># each patch</span></span>
<span id="cb281-2"><a href="#cb281-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ncore<span class="fl">.300</span> <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(gn<span class="fl">.300</span>, <span class="at">what=</span><span class="st">"lsm_p_ncore"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["layer"],"name":[1],"type":["int"],"align":["right"]},{"label":["level"],"name":[2],"type":["chr"],"align":["left"]},{"label":["class"],"name":[3],"type":["int"],"align":["right"]},{"label":["id"],"name":[4],"type":["int"],"align":["right"]},{"label":["metric"],"name":[5],"type":["chr"],"align":["left"]},{"label":["value"],"name":[6],"type":["dbl"],"align":["right"]}],"data":[{"1":"1","2":"patch","3":"295575","4":"1","5":"ncore","6":"0"},{"1":"1","2":"patch","3":"295575","4":"2","5":"ncore","6":"0"},{"1":"1","2":"patch","3":"295575","4":"3","5":"ncore","6":"0"},{"1":"1","2":"patch","3":"295575","4":"4","5":"ncore","6":"0"},{"1":"1","2":"patch","3":"295575","4":"5","5":"ncore","6":"0"},{"1":"1","2":"patch","3":"295575","4":"6","5":"ncore","6":"0"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb282"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb282-1"><a href="#cb282-1" aria-hidden="true" tabindex="-1"></a><span class="co"># summarize by map unit</span></span>
<span id="cb282-2"><a href="#cb282-2" aria-hidden="true" tabindex="-1"></a>ncore.<span class="fl">300.</span>summary <span class="ot">&lt;-</span> ncore<span class="fl">.300</span> <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(class) <span class="sc">%&gt;%</span></span>
<span id="cb282-3"><a href="#cb282-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">max_cores =</span> <span class="fu">max</span>(value)) <span class="sc">%&gt;%</span></span>
<span id="cb282-4"><a href="#cb282-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(class)</span>
<span id="cb282-5"><a href="#cb282-5" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(ncore.<span class="fl">20.</span>summary)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 217 × 2
    class max_cores
    &lt;int&gt;     &lt;dbl&gt;
 1 295575        18
 2 295576         2
 3 295577         4
 4 295578         1
 5 295579         4
 6 295580         3
 7 295581        10
 8 295582        12
 9 295583         1
10 295584        12
# ℹ 207 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb284"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb284-1"><a href="#cb284-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(ncore.<span class="fl">100.</span>summary)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 214 × 2
    class max_cores
    &lt;int&gt;     &lt;dbl&gt;
 1 295575         2
 2 295576         1
 3 295577         1
 4 295578         0
 5 295579         3
 6 295580         0
 7 295581         2
 8 295582         9
 9 295583         1
10 295584         3
# ℹ 204 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb286"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb286-1"><a href="#cb286-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(ncore.<span class="fl">300.</span>summary)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 173 × 2
    class max_cores
    &lt;int&gt;     &lt;dbl&gt;
 1 295575         0
 2 295576         0
 3 295577         0
 4 295578         0
 5 295579         1
 6 295581         0
 7 295582         2
 8 295583         0
 9 295584         0
10 295585         0
# ℹ 163 more rows</code></pre>
</div>
</div>
<p>Almost all of the isolated core areas have been merged with adjacent pixels, with the modal filter.</p>
<p>The Erie unit again has the most core areas, but many small cores have been absorbed in their neighbours.</p>
<p>Display it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb288"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb288-1"><a href="#cb288-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_patches</span>(gn<span class="fl">.300</span>, <span class="at">class =</span> mu.key[ix[<span class="dv">1</span>], <span class="st">"mukey"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show-patches-Erie-300-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>Display the <code>ncore</code> patch-level metric as a map:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb290"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb290-1"><a href="#cb290-1" aria-hidden="true" tabindex="-1"></a><span class="fu">show_lsm</span>(gn<span class="fl">.300</span>, <span class="st">"lsm_p_ncore"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/show-ncore-300-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>Most of the map has no cores, i.e., single pixels surrounded by different classes.</p>
</section>
</section>
<section id="conclusions-about-geometric-scale-issues" class="level2" data-number="9.4">
<h2 data-number="9.4" class="anchored" data-anchor-id="conclusions-about-geometric-scale-issues"><span class="header-section-number">9.4</span> Conclusions about geometric scale issues</h2>
<ul>
<li><p>Landscape level: The diversity (class number and relative proportion) hardly change, if at all.</p></li>
<li><p>Class level: Class composition hardly changes; some classes slightly increase in total area, others slightly lose area.</p></li>
<li><p>Patch level: Core areas (isolated pixels) decrease with increasing resolution.</p></li>
</ul>
<p><em>So, which scale is most “realistic”?</em></p>
</section>
</section>
<section id="sec-categorical-generalization" class="level1" data-number="10">
<h1 data-number="10"><span class="header-section-number">10</span> Scale issues – categorical</h1>
<p><em>Categorical</em> generalization is when related map units are grouped into more general units that share sufficient commonality to be considered “homogeneous” at a more general categorical level.</p>
<p>Many soil classification system support this directly by their hierarchical structure.</p>
<p>Task: Find the map units where the Mardin series is a component.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>      mukey                                                             muname
63   295656                   Mardin channery silt loam, 2 to 8 percent slopes
64   295657                  Mardin channery silt loam, 8 to 15 percent slopes
65   295658          Mardin channery silt loam, 8 to 15 percent slopes, eroded
67   295660                 Mardin and Langford soils, 15 to 25 percent slopes
100  295805                    Mardin-Volusia complex, 15 to 30 percent slopes
101  295806                     Mardin-Volusia complex, 8 to 15 percent slopes
102  295807                      Mardin-Volusia complex, 0 to 8 percent slopes
120  295831         Mardin channery silt loam, 16 to 30 percent slopes, eroded
121  295832                 Mardin channery silt loam, 16 to 30 percent slopes
122  295833 Mardin channery silt loam, 9 to 15 percent slopes, moderately deep
123  295834          Mardin channery silt loam, 8 to 15 percent slopes, eroded
124  295835                  Mardin channery silt loam, 8 to 15 percent slopes
125  295836                   Mardin channery silt loam, 0 to 8 percent slopes
126  295837  Mardin channery silt loam, 0 to 8 percent slopes, moderately deep
149 2723042                     Bath and Mardin soils, 25 to 40 percent slopes
178 2723076                   Mardin channery silt loam, 2 to 8 percent slopes
179 2723077                 Mardin channery silt loam, 15 to 25 percent slopes
180 2723078                  Mardin channery silt loam, 8 to 15 percent slopes
181 2723080    Mardin channery silt loam, 3 to 8 percent slopes, slightly acid
182 2723081   Mardin channery silt loam, 8 to 15 percent slopes, slightly acid</code></pre>
</div>
</div>
<p>In the case of USA soil survey, many map units are quite similar. For example, there are 20 map units in the small study area where the Mardin series is the only one naned, or it is one of the series in a complex. These map units differ mostly by slope class. Their land use potential may be somewhat different, mainly due to slope, but their soil properties are quite similar.</p>
<p>Task: Generalize the soil map units of the polygon map by combining map units with the same dominant soil series.</p>
<p>First, make a list of the first-named soil series:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb293"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb293-1"><a href="#cb293-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(names <span class="ot">&lt;-</span> mu.key<span class="sc">$</span>muname)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 217</code></pre>
</div>
<div class="sourceCode cell-code" id="cb295"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb295-1"><a href="#cb295-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first name by spaces</span></span>
<span id="cb295-2"><a href="#cb295-2" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(names, <span class="st">" "</span>)</span>
<span id="cb295-3"><a href="#cb295-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(names)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
[1] "Alluvial" "land"    

[[2]]
[1] "Arkport" "fine"    "sandy"   "loam,"   "2"       "to"      "6"      
[8] "percent" "slopes" 

[[3]]
[1] "Arkport" "fine"    "sandy"   "loam,"   "6"       "to"      "12"     
[8] "percent" "slopes" 

[[4]]
[1] "Bath"     "channery" "silt"     "loam,"    "2"        "to"       "5"       
[8] "percent"  "slopes"  

[[5]]
[1] "Bath"     "channery" "silt"     "loam,"    "5"        "to"       "15"      
[8] "percent"  "slopes"  

[[6]]
 [1] "Bath"     "channery" "silt"     "loam,"    "5"        "to"      
 [7] "15"       "percent"  "slopes,"  "eroded"  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb297"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb297-1"><a href="#cb297-1" aria-hidden="true" tabindex="-1"></a>names.unique <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">unlist</span>(<span class="fu">lapply</span>(names, <span class="cf">function</span>(x) x[<span class="dv">1</span>])))</span>
<span id="cb297-2"><a href="#cb297-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(names.unique)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Alluvial"                "Arkport"                
 [3] "Bath"                    "Bath,"                  
 [5] "Braceville"              "Canandaigua"            
 [7] "Chenango"                "Conesus"                
 [9] "Darien"                  "Erie"                   
[11] "Chippewa"                "Eel"                    
[13] "Erie-Chippewa"           "Fredon"                 
[15] "Fresh"                   "Genesee"                
[17] "Halsey"                  "Howard"                 
[19] "Holly"                   "Howard-Valois"          
[21] "Hudson"                  "Hudson-Cayuga"          
[23] "Ilion"                   "Kendaia"                
[25] "Langford"                "Lordstown"              
[27] "Lordstown,"              "Lyons"                  
[29] "Mardin"                  "Made"                   
[31] "Madalin"                 "Middlebury"             
[33] "Muck"                    "Niagara"                
[35] "Ovid"                    "Palmyra"                
[37] "Phelps"                  "Red"                    
[39] "Rhinebeck"               "Rock"                   
[41] "Tuller"                  "Volusia"                
[43] "Volusia-Chippewa"        "Williamson"             
[45] "Wayland"                 "Allis"                  
[47] "Mardin-Volusia"          "Fremont"                
[49] "Arnot-Rock"              "Tioga"                  
[51] "Woostern"                "Water"                  
[53] "Gravel"                  "Quarries"               
[55] "Dumps"                   "Alden"                  
[57] "Cadosia-Lordstown"       "Catden-Natchaug"        
[59] "Mongaup-Hawksnest"       "Mongaup"                
[61] "Deposit"                 "Lansing"                
[63] "Lewbath"                 "Norchip"                
[65] "Ontusia"                 "Scio"                   
[67] "Lordstown-Arnot"         "Trestle"                
[69] "Udifluvents-Fluvaquents" "Valois"                 
[71] "Valois-Howard"           "Wayland-Natchaug"       
[73] "Willdin"                 "Geneseo"                
[75] "Hemlock"                 "Bath-Valois"            
[77] "Rockrift-Mongaup"       </code></pre>
</div>
</div>
<p>Now group the map units by these:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb299"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb299-1"><a href="#cb299-1" aria-hidden="true" tabindex="-1"></a>mu.poly.general <span class="ot">&lt;-</span> mu.poly</span>
<span id="cb299-2"><a href="#cb299-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(<span class="fu">values</span>(mu.poly.general))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "mukey"   "area_ac"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb301"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb301-1"><a href="#cb301-1" aria-hidden="true" tabindex="-1"></a>(l.all <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(<span class="fu">values</span>(mu.poly)<span class="sc">$</span>mukey)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 217</code></pre>
</div>
<div class="sourceCode cell-code" id="cb303"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb303-1"><a href="#cb303-1" aria-hidden="true" tabindex="-1"></a>names.all <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(names, <span class="cf">function</span>(x) x[<span class="dv">1</span>]))</span>
<span id="cb303-2"><a href="#cb303-2" aria-hidden="true" tabindex="-1"></a>(l.general <span class="ot">&lt;-</span> <span class="fu">length</span>(names.unique <span class="ot">&lt;-</span> <span class="fu">unique</span>(names.all)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 77</code></pre>
</div>
<div class="sourceCode cell-code" id="cb305"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb305-1"><a href="#cb305-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (name <span class="cf">in</span> names.unique) {</span>
<span id="cb305-2"><a href="#cb305-2" aria-hidden="true" tabindex="-1"></a>  ix <span class="ot">&lt;-</span> <span class="fu">which</span>(name <span class="sc">==</span> names.all)</span>
<span id="cb305-3"><a href="#cb305-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># map units in this group</span></span>
<span id="cb305-4"><a href="#cb305-4" aria-hidden="true" tabindex="-1"></a>  keys <span class="ot">&lt;-</span> mu.key<span class="sc">$</span>mukey[ix]</span>
<span id="cb305-5"><a href="#cb305-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># polygons of these map units</span></span>
<span id="cb305-6"><a href="#cb305-6" aria-hidden="true" tabindex="-1"></a>  ix.polys <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">values</span>(mu.poly.general)<span class="sc">$</span>mukey <span class="sc">%in%</span> keys)</span>
<span id="cb305-7"><a href="#cb305-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># rename the key</span></span>
<span id="cb305-8"><a href="#cb305-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">values</span>(mu.poly.general)<span class="sc">$</span>mukey[ix.polys] <span class="ot">&lt;-</span> keys[<span class="dv">1</span>]</span>
<span id="cb305-9"><a href="#cb305-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb305-10"><a href="#cb305-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(<span class="fu">values</span>(mu.poly.general))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "mukey"   "area_ac"</code></pre>
</div>
</div>
<p>The number of map units has reduced significantly, from 217 to 77.</p>
<p>Finally, generalize the polygons by dissolving between ones with the same label:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb307"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb307-1"><a href="#cb307-1" aria-hidden="true" tabindex="-1"></a>mu.poly.general <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">aggregate</span>(mu.poly.general,</span>
<span id="cb307-2"><a href="#cb307-2" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">by =</span> <span class="st">"mukey"</span>,</span>
<span id="cb307-3"><a href="#cb307-3" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">fun =</span> <span class="st">"sum"</span>)</span>
<span id="cb307-4"><a href="#cb307-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(mu.poly.general, <span class="at">y=</span><span class="dv">2</span>,</span>
<span id="cb307-5"><a href="#cb307-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">type =</span> <span class="st">"continuous"</span>,</span>
<span id="cb307-6"><a href="#cb307-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"SSURGO map units, area in acres"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/dissolve-polygons-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<p>Many polygons are now much larger.</p>
<section id="landscape-metrics" class="level3" data-number="10.0.1">
<h3 data-number="10.0.1" class="anchored" data-anchor-id="landscape-metrics"><span class="header-section-number">10.0.1</span> Landscape metrics</h3>
<p>We can repeat the analysis of <a href="#scale">scale</a> differences on the rasterized map. Here we just look at the 20 m resolution.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb308"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb308-1"><a href="#cb308-1" aria-hidden="true" tabindex="-1"></a>mu.raster.general <span class="ot">&lt;-</span> <span class="fu">rasterize</span>(mu.poly.general, mu.template, <span class="at">field=</span><span class="st">"mukey"</span>)</span>
<span id="cb308-2"><a href="#cb308-2" aria-hidden="true" tabindex="-1"></a><span class="fu">check_landscape</span>(mu.raster.general)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["layer"],"name":[1],"type":["int"],"align":["right"]},{"label":["crs"],"name":[2],"type":["chr"],"align":["left"]},{"label":["units"],"name":[3],"type":["chr"],"align":["left"]},{"label":["class"],"name":[4],"type":["chr"],"align":["left"]},{"label":["n_classes"],"name":[5],"type":["int"],"align":["right"]},{"label":["OK"],"name":[6],"type":["chr"],"align":["left"]}],"data":[{"1":"1","2":"projected","3":"m","4":"integer","5":"77","6":"❓"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb309"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb309-1"><a href="#cb309-1" aria-hidden="true" tabindex="-1"></a>gn.<span class="fl">20.</span>g <span class="ot">&lt;-</span> mu.raster.general</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Compare the <em>landscape-level</em> metrics at 20 m resolution for the generalized and detailed maps:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb310"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb310-1"><a href="#cb310-1" aria-hidden="true" tabindex="-1"></a>ls.metrics <span class="ot">&lt;-</span> ls.metrics.gn.<span class="fl">20.</span>g <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(gn.<span class="fl">20.</span>g, <span class="at">what=</span>lst)[, <span class="fu">c</span>(<span class="st">"metric"</span>, <span class="st">"value"</span>)]</span>
<span id="cb310-2"><a href="#cb310-2" aria-hidden="true" tabindex="-1"></a>ls.metrics<span class="sc">$</span>detailed <span class="ot">&lt;-</span> ls.metrics.gn20<span class="sc">$</span>value</span>
<span id="cb310-3"><a href="#cb310-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(ls.metrics) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"metric"</span>, <span class="st">"generalized"</span>, <span class="st">"detailed"</span>)</span>
<span id="cb310-4"><a href="#cb310-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(ls.metrics)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  metric  generalized detailed
  &lt;chr&gt;         &lt;dbl&gt;    &lt;dbl&gt;
1 ai           89.2     86.3  
2 frac_mn       1.11     1.11 
3 lsi          67.6     86.7  
4 shdi          2.99     4.39 
5 shei          0.687    0.815</code></pre>
</div>
</div>
<p>The Shannon diversity is much lower in the generalized map, because there are many fewer classes and therefore more evenly distributed. The other metrics are hardly changed.</p>
<p>The <em>class-level</em> metrics show the distribution of classes; compare the generalized and detailed per-class proportion over the landscape.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb312"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb312-1"><a href="#cb312-1" aria-hidden="true" tabindex="-1"></a>c_pland.<span class="fl">20.</span>g <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(gn.<span class="fl">20.</span>g, <span class="at">what=</span><span class="st">"lsm_c_pland"</span>)</span>
<span id="cb312-2"><a href="#cb312-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(c_pland.<span class="fl">20.</span>g<span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), <span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 18.0574815 12.5823100 11.4525841  8.9150934  7.5616574  6.4943426
 [7]  5.2696646  2.5951928  2.3813925  1.5663227  1.5260507  1.4970942
[13]  1.4291307  1.3141480  1.0423644  1.0228257  1.0139701  0.8203410
[19]  0.8071278  0.8032622  0.7428190  0.6643131  0.6569334  0.6379571</code></pre>
</div>
<div class="sourceCode cell-code" id="cb314"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb314-1"><a href="#cb314-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(c_pland<span class="fl">.20</span><span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), <span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 5.303681 5.034498 4.229760 4.166365 3.978148 3.566361 3.380955 2.936838
 [9] 2.779685 2.343791 1.947256 1.936221 1.890326 1.692129 1.666687 1.635481
[17] 1.601113 1.533009 1.490206 1.429131 1.368758 1.227630 1.169155 1.063941</code></pre>
</div>
<div class="sourceCode cell-code" id="cb316"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb316-1"><a href="#cb316-1" aria-hidden="true" tabindex="-1"></a>ix <span class="ot">&lt;-</span> <span class="fu">order</span>(c_pland.<span class="fl">20.</span>g<span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb316-2"><a href="#cb316-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(c_pland.<span class="fl">20.</span>g<span class="sc">$</span>class[ix], <span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  295647  295683  295578  295656  295598  295635  295610  295690  295590
[10]  295575 2723085  295678  295604  295675  295602  295608 2723058  295805
[19]  295614  295855  295586 2723117  295622  295663</code></pre>
</div>
<div class="sourceCode cell-code" id="cb318"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb318-1"><a href="#cb318-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(c_pland<span class="fl">.20</span><span class="sc">$</span>class[ix], <span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 295603 295622 295577 295606 295584 295602 295594 295628 295581 295575
[11] 295658 295619 295587 295614 295585 295592 295652 295630 295596 295634
[21] 295578 295666 295598 295609</code></pre>
</div>
</div>
<p>Obviously the generalized map has larger areas of each class.</p>
<p>Compare at the <em>patch level</em>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb320"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb320-1"><a href="#cb320-1" aria-hidden="true" tabindex="-1"></a><span class="co"># each patch</span></span>
<span id="cb320-2"><a href="#cb320-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(area.<span class="fl">20.</span>g <span class="ot">&lt;-</span>  </span>
<span id="cb320-3"><a href="#cb320-3" aria-hidden="true" tabindex="-1"></a>            <span class="fu">calculate_lsm</span>(gn.<span class="fl">20.</span>g, <span class="at">what=</span><span class="st">"lsm_p_area"</span>)<span class="sc">$</span>value, <span class="at">decreasing =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1316.60 1018.40  690.40  648.56  643.08  468.48</code></pre>
</div>
<div class="sourceCode cell-code" id="cb322"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb322-1"><a href="#cb322-1" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(area.<span class="fl">20.</span>g , <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span>.<span class="dv">12</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     0%     12%     24%     36%     48%     60%     72%     84%     96% 
 0.0400  0.5200  0.9600  1.4800  2.2400  3.4400  5.8800 11.4400 42.6784 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb324"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb324-1"><a href="#cb324-1" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(area<span class="fl">.20</span> , <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span>.<span class="dv">12</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     0%     12%     24%     36%     48%     60%     72%     84%     96% 
 0.0400  0.5600  1.0000  1.5600  2.3232  3.4400  5.4800  9.4400 24.7600 </code></pre>
</div>
</div>
<p>Many patches increase in size by the aggregation, but about 60% remain at the original size – these were of small classes that did not have multiple map units.</p>
<p>Core areas:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb326"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb326-1"><a href="#cb326-1" aria-hidden="true" tabindex="-1"></a>ncore.g <span class="ot">&lt;-</span> <span class="fu">calculate_lsm</span>(gn.<span class="fl">20.</span>g, <span class="at">what=</span><span class="st">"lsm_p_ncore"</span>)</span>
<span id="cb326-2"><a href="#cb326-2" aria-hidden="true" tabindex="-1"></a>ncore.g.summary <span class="ot">&lt;-</span> ncore.g <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(class) <span class="sc">%&gt;%</span></span>
<span id="cb326-3"><a href="#cb326-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">max_cores =</span> <span class="fu">max</span>(value)) <span class="sc">%&gt;%</span></span>
<span id="cb326-4"><a href="#cb326-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(class)</span>
<span id="cb326-5"><a href="#cb326-5" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(ncore.g.summary)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 77 × 2
    class max_cores
    &lt;int&gt;     &lt;dbl&gt;
 1 295575        21
 2 295576         6
 3 295578        12
 4 295586        10
 5 295587         2
 6 295589         3
 7 295590         5
 8 295594         1
 9 295597         8
10 295598        38
# ℹ 67 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb328"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb328-1"><a href="#cb328-1" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> <span class="fu">show_lsm</span>(gn.<span class="fl">20.</span>g, <span class="st">"lsm_p_ncore"</span>)</span>
<span id="cb328-2"><a href="#cb328-2" aria-hidden="true" tabindex="-1"></a>g2 <span class="ot">&lt;-</span> <span class="fu">show_lsm</span>(gn<span class="fl">.20</span>, <span class="st">"lsm_p_ncore"</span>)</span>
<span id="cb328-3"><a href="#cb328-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb328-4"><a href="#cb328-4" aria-hidden="true" tabindex="-1"></a>g1; g2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/cores-general-1.png" class="img-fluid" width="1152"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$layer_1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PatternAnalysisWorkshopTutorial_files/figure-html/cores-general-2.png" class="img-fluid" width="1152"></p>
</div>
<div class="sourceCode cell-code" id="cb331"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb331-1"><a href="#cb331-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This same comparison could be done at increasingly larger resolutions.</p>
</section>
<section id="conclusions-about-categorical-scale-issues" class="level2" data-number="10.1">
<h2 data-number="10.1" class="anchored" data-anchor-id="conclusions-about-categorical-scale-issues"><span class="header-section-number">10.1</span> Conclusions about categorical scale issues</h2>
<ul>
<li><p>Increasing generalization leads to fewer and larger map units.</p></li>
<li><p>Increasing generalization leads to lower Shannon diversity.</p></li>
</ul>
</section>
</section>
<section id="comparing-to-reality-or-which-map-is-best" class="level1" data-number="11">
<h1 data-number="11"><span class="header-section-number">11</span> Comparing to “reality”, or Which map is “best”?</h1>
<p>Of course, we would like a map that best represents the soilscape. But what is “best”?</p>
<section id="agreement-with-obvious-landscape-features" class="level2" data-number="11.1">
<h2 data-number="11.1" class="anchored" data-anchor-id="agreement-with-obvious-landscape-features"><span class="header-section-number">11.1</span> Agreement with obvious landscape features</h2>
<p>These can be geomorphic or related to land use.</p>
<p>View SSURGO polygons on <a href="https://casoilresource.lawr.ucdavis.edu/gmap/">SoilWeb</a> with different backgrounds: (1) USGS topography, (2) ESRI imagery. To evaluate you must know the soil series in the mapped area, their genesis and typical locations. Click on map units for their series composition, and then the OSD for the series to understand its genesis and geography.</p>
<div id="fig-dryden-topo" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="./figs/SoilWebDrydenLakeTopo.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;7: SSURGO map units on USGS topography</figcaption>
</figure>
</div>
</section>
</section>
<section id="pattern-based-segmentation" class="level1" data-number="12">
<h1 data-number="12"><span class="header-section-number">12</span> Pattern-based segmentation</h1>
<p>The concept here is to characterize patterns within windows of some size and then combine these by aggregation of “similar enough” windows into larger areas, which then segment the original map. This has been applied with some success to ecophysiographic regions, using a single layer of land cover <span class="citation" data-cites="nowosadMachineEcoregionalizationEarth2018">(<a href="#ref-nowosadMachineEcoregionalizationEarth2018" role="doc-biblioref">Nowosad &amp; Stepinski, 2018b</a>)</span>. The segmentation does not alter any values, it only groups them into larger units with a similar pattern.</p>
<p>In the soils context, the pattern could be based on soil classes or on classified continuous maps. To include many soil properties the values could come classification of the first Principal Component or some user-defined composite index.</p>
<p>The algorithm is described by Jasiewicz <em>et al.</em> <span class="citation" data-cites="jasiewiczMultiscaleSegmentationAlgorithm2018">(<a href="#ref-jasiewiczMultiscaleSegmentationAlgorithm2018" role="doc-biblioref">2018</a>)</span>.</p>
<p>This is related to <strong>stratification</strong> for locally-calibrated models and stratified sampling as implemented in the <code>rassta</code> package <span class="citation" data-cites="fuentesRasstaRasterBasedSpatial2022">(<a href="#ref-fuentesRasstaRasterBasedSpatial2022" role="doc-biblioref">Fuentes et al., 2022</a>)</span>.</p>
<p>When identifying regions, the two complementary objectives are: <em>homogeneous pattern within each region</em>, and <em>different patterns in adjacent regions</em> <span class="citation" data-cites="nowosadMachineEcoregionalizationEarth2018">(<a href="#ref-nowosadMachineEcoregionalizationEarth2018" role="doc-biblioref">Nowosad &amp; Stepinski, 2018b</a>)</span>.</p>
<p>To assess homogeneity of a region with respect to its soils we calculate an <strong>inhomogeneity metric</strong>: the mutual dissimilarity between all sites within the region.</p>
<p>To assess how much a pattern in a given region differs from patterns in neighboring regions we calculate an <strong>isolation metric</strong>. This is the average dissimilarity between the core region of interest and its first neighbors, weighted by the proportion of the core region’s perimeter shared with the different neighbors.</p>
<p>An <strong>overall quality index</strong> for a single region can be defined as (1 – inhomogeneity/isolation), so that a higher value shows higher quality. And these can be combined by area-weighted average for the whole map.</p>
<section id="pattern-within-a-region" class="level2" data-number="12.1">
<h2 data-number="12.1" class="anchored" data-anchor-id="pattern-within-a-region"><span class="header-section-number">12.1</span> Pattern within a region</h2>
<p>How could this relate to soil surveys? This is scale-dependent, and seems most appropriate for semi-detailed or reconaissance-level surveys (NRCS soil survey orders 4 and 5). The pattern boundaries would match the polygon boundaries, and the composition inside the pattern would correspond to the estimate of the map unit composition <em>and within-unit pattern</em>.</p>
<p>We hope that the stratification produces polygons that are similar to soil landscape delineations that would be made by a field surveyor. These delineations form map units, that are characterized by a set of soils in a definite pattern, with the components not separable at the map design scale.</p>
<p>That is, if we segment with tiles of the minimum mappable area (MMA) at a given scale, we hope that the pattern within a region matches our idea of the pattern of contrasting soils that can’t be mapped at that scale. For order 4 the MMA is 16–256 ha, for order 5 this is 250 to 4,000 ha <span class="citation" data-cites="soil_survey_division_staff_soil_2017">(<a href="#ref-soil_survey_division_staff_soil_2017" role="doc-biblioref">Soil Science Division Staff, 2017</a>, Table 4-4)</span>.</p>
</section>
<section id="geopat" class="level2" data-number="12.2">
<h2 data-number="12.2" class="anchored" data-anchor-id="geopat"><span class="header-section-number">12.2</span> GeoPAT</h2>
<p>“GeoPAT’s core idea is to tessellate global spatial data into grid of square blocks of original cells (pixels). This transforms data from its original form (<em>huge number of cells each having simple content</em>) to a new form (<em>much smaller number of supercells/blocks with complex content</em>). Complex cell contains a <em>pattern of original variable</em>.”</p>
<p>This is a stand-alone program which must be compiled for your operating system. Source code and installation instructions are <a href="https://github.com/Nowosad/geopat2">here</a>.</p>
<!--# Geomorphons? -->
<p>A package to interface GeoPat2 and R data structures:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb332"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb332-1"><a href="#cb332-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(rgeopat2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="references" class="level1" data-number="13">
<h1 data-number="13"><span class="header-section-number">13</span> References</h1>
<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-achantaSLICSuperpixelsCompared2012" class="csl-entry" role="listitem">
Achanta, R., Shaji, A., Smith, K., Lucchi, A., Fua, P., &amp; Süsstrunk, S. (2012). <span>SLIC Superpixels</span> compared to state-of-the-art superpixel methods. <em>IEEE Transactions on Pattern Analysis and Machine Intelligence</em>, <em>34</em>(11), 2274–2282. <a href="https://doi.org/10.1109/TPAMI.2012.120">https://doi.org/10.1109/TPAMI.2012.120</a>
</div>
<div id="ref-boulaineRemarquesQuelquesNotions1982" class="csl-entry" role="listitem">
Boulaine, J. (1982). Remarques sur quelques notions élémentaires de la pédologie”. <em><span>Cahiers O.R.S.T.O.M., Série Pédologie</span></em>, <em>19</em>(1), 29–41.
</div>
<div id="ref-Brus.etal2011" class="csl-entry" role="listitem">
Brus, D. J., Kempen, B., &amp; Heuvelink, G. B. M. (2011). Sampling for validation of digital soil maps. <em>European Journal of Soil Science</em>, <em>62</em>, 394–407. <a href="https://doi.org/10.1111/j.1365-2389.2011.01364.x">https://doi.org/10.1111/j.1365-2389.2011.01364.x</a>
</div>
<div id="ref-Fridland1974" class="csl-entry" role="listitem">
Fridland, V. M. (1974). Structure of the soil mantle. <em>Geoderma</em>, <em>12</em>, 35–42. <a href="https://doi.org/10.1016/0016-7061(74)90036-6">https://doi.org/10.1016/0016-7061(74)90036-6</a>
</div>
<div id="ref-fuentesRasstaRasterBasedSpatial2022" class="csl-entry" role="listitem">
Fuentes, B. A., Dorantes, M. J., &amp; Tipton, J. R. (2022). Rassta: <span>Raster-Based Spatial Stratification Algorithms</span>. <em>R JOURNAL</em>, <em>14</em>(2), 286–304. <a href="https://doi.org/10.32614/RJ-2022-036">https://doi.org/10.32614/RJ-2022-036</a>
</div>
<div id="ref-Hesselbarth2021" class="csl-entry" role="listitem">
Hesselbarth, M. H. K. (2021). <em><span>R-Spatialecology/Landscapemetrics</span></em>. r-spatialecology. <a href="https://github.com/r-spatialecology/landscapemetrics">https://github.com/r-spatialecology/landscapemetrics</a>
</div>
<div id="ref-Hesselbarth.etal2019" class="csl-entry" role="listitem">
Hesselbarth, M. H. K., Sciaini, M., With, K. A., Wiegand, K., &amp; Nowosad, J. (2019). Landscapemetrics: An open-source <span>R</span> tool to calculate landscape metrics. <em>Ecography</em>, <em>42</em>, 1648–1657. <a href="https://doi.org/10.1111/ecog.04617">https://doi.org/10.1111/ecog.04617</a>
</div>
<div id="ref-jasiewiczMultiscaleSegmentationAlgorithm2018" class="csl-entry" role="listitem">
Jasiewicz, J., Stepinski, T., &amp; Niesterowicz, J. (2018). Multi-scale segmentation algorithm for pattern-based partitioning of large categorical rasters. <em>COMPUTERS &amp; GEOSCIENCES</em>, <em>118</em>, 122–130. <a href="https://doi.org/10.1016/j.cageo.2018.06.003">https://doi.org/10.1016/j.cageo.2018.06.003</a>
</div>
<div id="ref-johnsonPedonPolypedon1963" class="csl-entry" role="listitem">
Johnson, W. M. (1963). The <span>Pedon</span> and the <span>Polypedon</span>. <em>Soil Science Society of America Journal</em>, <em>27</em>(2), 212–215. <a href="https://doi.org/10.2136/sssaj1963.03615995002700020034x">https://doi.org/10.2136/sssaj1963.03615995002700020034x</a>
</div>
<div id="ref-Kupfer2012" class="csl-entry" role="listitem">
Kupfer, J. A. (2012). Landscape ecology and biogeography: <span>Rethinking</span> landscape metrics in a post-<span>FRAGSTATS</span> landscape. <em>Progress in Physical Geography-Earth and Environment</em>, <em>36</em>(3), 400–420. <a href="https://doi.org/10.1177/0309133312439594">https://doi.org/10.1177/0309133312439594</a>
</div>
<div id="ref-mahoneyAssessingPerformanceSpatial2023" class="csl-entry" role="listitem">
Mahoney, M. J., Johnson, L. K., Silge, J., Frick, H., Kuhn, M., &amp; Beier, C. M. (2023). <em>Assessing the performance of spatial cross-validation approaches for models of spatially structured data</em> (arXiv:2303.07334). <span>arXiv</span>. <a href="https://doi.org/10.48550/arXiv.2303.07334">https://doi.org/10.48550/arXiv.2303.07334</a>
</div>
<div id="ref-McGarigal.etal2012" class="csl-entry" role="listitem">
McGarigal, K., Cushman, S. A., &amp; Ene, E. (2012). <em><span>FRAGSTATS</span> v4: <span>S</span>patial pattern analysis program for categorical and continuous maps</em>. <span>University of Massachusetts</span>. <a href="http://www.umass.edu/landeco/research/fragstats/fragstats.html">http://www.umass.edu/landeco/research/fragstats/fragstats.html</a>
</div>
<div id="ref-minasnyVESPERVariogramEstimation2005" class="csl-entry" role="listitem">
Minasny, B., McBratney, A. B., &amp; Whelan, B. M. (2005). <em><span>VESPER</span>: <span>Variogram Estimation</span> and <span>Spatial Prediction</span> plus <span>ERror</span> - <span>Australian Centre</span> for <span>Precision Agriculture</span></em>. https://precision-agriculture.sydney.edu.au/resources/software/download-vesper/.
</div>
<div id="ref-nowosadSpatialAssociationRegionalizations2018" class="csl-entry" role="listitem">
Nowosad, J., &amp; Stepinski, T. F. (2018a). Spatial association between regionalizations using the information-theoretical <span class="nocase">V-measure</span>. <em>International Journal of Geographical Information Science</em>, <em>32</em>(12), 2386–2401. <a href="https://doi.org/10.1080/13658816.2018.1511794">https://doi.org/10.1080/13658816.2018.1511794</a>
</div>
<div id="ref-nowosadMachineEcoregionalizationEarth2018" class="csl-entry" role="listitem">
Nowosad, J., &amp; Stepinski, T. F. (2018b). Towards machine ecoregionalization of <span>Earth</span>’s landmass using pattern segmentation method. <em>International Journal of Applied Earth Observation and Geoinformation</em>, <em>69</em>, 110–118. <a href="https://doi.org/10.1016/j.jag.2018.03.004">https://doi.org/10.1016/j.jag.2018.03.004</a>
</div>
<div id="ref-nrcssoilsGriddedNationalSoil2023" class="csl-entry" role="listitem">
NRCS Soils. (2023). <em>Gridded <span>National Soil Survey Geographic Database</span> (<span class="nocase">gNATSGO</span>)</em>. https://www.nrcs.usda.gov/wps/portal/nrcs/detail/soils/survey/geo/?cid=nrcseprd1464625.
</div>
<div id="ref-OGCGeoTIFFStandard2023" class="csl-entry" role="listitem">
Open Geospatial Consortium. (2023). <span>OGC GeoTIFF Standard</span>. In <em>OGC GeoTIFF Standard</em>. https://www.ogc.org/standard/geotiff/.
</div>
<div id="ref-piikkiPerspectivesValidationDigital2021" class="csl-entry" role="listitem">
Piikki, K., Wetterlind, J., Söderström, M., &amp; Stenberg, B. (2021). Perspectives on validation in digital soil mapping of continuous attributes a review. <em>Soil Use and Management</em>, <em>37</em>(1), 7–21. <a href="https://doi.org/10.1111/sum.12694">https://doi.org/10.1111/sum.12694</a>
</div>
<div id="ref-Poggio.etal2021a" class="csl-entry" role="listitem">
Poggio, L., de Sousa, L. M., Batjes, N. H., Heuvelink, G. B. M., Kempen, B., Ribeiro, E., &amp; Rossiter, D. (2021). <span>SoilGrids</span> 2.0: Producing soil information for the globe with quantified spatial uncertainty. <em>SOIL</em>, <em>7</em>(1), 217–240. <a href="https://doi.org/10.5194/soil-7-217-2021">https://doi.org/10.5194/soil-7-217-2021</a>
</div>
<div id="ref-soilconservationservicePhysicalLandConditions1951" class="csl-entry" role="listitem">
Soil Conservation Service. (1951). <em>Physical land conditions in <span>Anderson County</span>, <span>South Carolina</span>,</em>. <span>USDA Soil Conservation Service</span>.
</div>
<div id="ref-soil_survey_division_staff_soil_2017" class="csl-entry" role="listitem">
Soil Science Division Staff. (2017). <em>Soil survey manual</em> (C. Ditzler, K. Scheffe, &amp; H. C. Monger, Eds.). <span>Government Printing Office</span>. <a href="http://www.nrcs.usda.gov/wps/portal/nrcs/detail/soils/planners/?cid=nrcs142p2_054262">http://www.nrcs.usda.gov/wps/portal/nrcs/detail/soils/planners/?cid=nrcs142p2_054262</a>
</div>
<div id="ref-Uuemaa.etal2013" class="csl-entry" role="listitem">
Uuemaa, E., Mander, U., &amp; Marja, R. (2013). Trends in the use of landscape spatial metrics as landscape indicators: <span>A</span> review. <em>Ecological Indicators</em>, <em>28</em>, 100–106. <a href="https://doi.org/10.1016/j.ecolind.2012.07.018">https://doi.org/10.1016/j.ecolind.2012.07.018</a>
</div>
</div>
</section>


<div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>https://ncss-tech.github.io/AQP/soilDB/SDA-tutorial.html<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>https://sdmdataaccess.nrcs.usda.gov/<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>https://r-spatialecology.github.io/landscapemetrics/<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section><section class="quarto-appendix-contents"><h2 class="anchored quarto-appendix-heading">Reuse</h2><div id="quarto-reuse" class="quarto-appendix-contents"><div>CC-BY</div></div></section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>