% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode,linktoc=all,pdfwindowui,pdfpagemode=FullScreen}{hyperref}
\PassOptionsToPackage{hyphens}{url}
\PassOptionsToPackage{dvipsnames,svgnames,x11names}{xcolor}
%
\documentclass[
  letterpaper,
  DIV=11,
  numbers=noendperiod]{scrartcl}

\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else  
    % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\setlength{\emergencystretch}{3em} % prevent overfull lines
\setcounter{secnumdepth}{5}
% Make \paragraph and \subparagraph free-standing
\ifx\paragraph\undefined\else
  \let\oldparagraph\paragraph
  \renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
  \let\oldsubparagraph\subparagraph
  \renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{241,243,245}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.40,0.45,0.13}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\BuiltInTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\ExtensionTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.28,0.35,0.67}{#1}}
\newcommand{\ImportTok}[1]{\textcolor[rgb]{0.00,0.46,0.62}{#1}}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\NormalTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.07,0.07,0.07}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}

\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newlength{\cslentryspacingunit} % times entry-spacing
\setlength{\cslentryspacingunit}{\parskip}
\newenvironment{CSLReferences}[2] % #1 hanging-ident, #2 entry spacing
 {% don't indent paragraphs
  \setlength{\parindent}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1
  \let\oldpar\par
  \def\par{\hangindent=\cslhangindent\oldpar}
  \fi
  % set entry spacing
  \setlength{\parskip}{#2\cslentryspacingunit}
 }%
 {}
\usepackage{calc}
\newcommand{\CSLBlock}[1]{#1\hfill\break}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{#1}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}\break}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}

\KOMAoption{captions}{tableheading}
\makeatletter
\makeatother
\makeatletter
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\AtBeginDocument{%
\ifdefined\contentsname
  \renewcommand*\contentsname{Table of contents}
\else
  \newcommand\contentsname{Table of contents}
\fi
\ifdefined\listfigurename
  \renewcommand*\listfigurename{List of Figures}
\else
  \newcommand\listfigurename{List of Figures}
\fi
\ifdefined\listtablename
  \renewcommand*\listtablename{List of Tables}
\else
  \newcommand\listtablename{List of Tables}
\fi
\ifdefined\figurename
  \renewcommand*\figurename{Figure}
\else
  \newcommand\figurename{Figure}
\fi
\ifdefined\tablename
  \renewcommand*\tablename{Table}
\else
  \newcommand\tablename{Table}
\fi
}
\@ifpackageloaded{float}{}{\usepackage{float}}
\floatstyle{ruled}
\@ifundefined{c@chapter}{\newfloat{codelisting}{h}{lop}}{\newfloat{codelisting}{h}{lop}[chapter]}
\floatname{codelisting}{Listing}
\newcommand*\listoflistings{\listof{codelisting}{List of Listings}}
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\@ifpackageloaded{subcaption}{}{\usepackage{subcaption}}
\makeatother
\makeatletter
\@ifpackageloaded{tcolorbox}{}{\usepackage[skins,breakable]{tcolorbox}}
\makeatother
\makeatletter
\@ifundefined{shadecolor}{\definecolor{shadecolor}{rgb}{.97, .97, .97}}
\makeatother
\makeatletter
\makeatother
\makeatletter
\makeatother
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same} % disable monospaced font for URLs
\hypersetup{
  pdfauthor={D G Rossiter},
  colorlinks=true,
  linkcolor={blue},
  filecolor={Maroon},
  citecolor={Blue},
  urlcolor={Blue},
  pdfcreator={LaTeX via pandoc}}

\title{Tutorial\\
Pattern Analysis for Evaluating Soil Maps}
\author{D G Rossiter}
\date{2024-06-06}

\begin{document}
\maketitle
\ifdefined\Shaded\renewenvironment{Shaded}{\begin{tcolorbox}[boxrule=0pt, interior hidden, frame hidden, breakable, sharp corners, borderline west={3pt}{0pt}{shadecolor}, enhanced]}{\end{tcolorbox}}\fi

\renewcommand*\contentsname{Table of contents}
{
\hypersetup{linkcolor=}
\setcounter{tocdepth}{3}
\tableofcontents
}
\hypertarget{abstract}{%
\section{Abstract}\label{abstract}}

This tutorial presents methods to evaluate the spatial patterns of the
spatial distribution of soil properties and map units as shown in
gridded maps produced by digital soil mapping (DSM). It compares
patterns from different DSM products to each other, and to spatial
patterns known from detailed field surveys or known to local experts but
not represented (yet) on maps. Methods include whole-map statistics,
visually identifiable landscape features, level of detail, range and
strength of spatial autocorrelation, landscape metrics (Shannon
diversity and evenness, shape, aggregation, mean fractal dimension, and
co-occurrence vectors), and spatial patterns of property maps classified
by histogram equalization or user-defined cutpoints. The tutorial also
shows how to use patterns within a window to partition a soil landscape
into zones with similar patterns. This workshop uses examples from the
USA, but the methods are applicable to any gridded DSM product or
polygon map of soil classes.

\hypertarget{motivation}{%
\section{Motivation}\label{motivation}}

Digital soil maps are usually evaluated by point-wise ``validation
statistics''
(\protect\hyperlink{ref-piikkiPerspectivesValidationDigital2021}{Piikki
et al., 2021}). This evaluation is quite limited from both the mapper's
and map user's perspectives.

\emph{Internally}, from the mapper's perspective:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  The evaluation is based on a necessarily limited number of
  observations, far fewer than the number of predictions (grid cells,
  pixels).
\item
  The evaluation points are very rarely from an independent probability
  sample (\protect\hyperlink{ref-Brus.etal2011}{Brus et al., 2011}).
\item
  Cross-validation and data-splitting approaches rely on a biased point
  set. (Note: so-called ``spatial cross-validation'' does not solve the
  problem of biased sampling, just cross-validation biases caused by
  clustered spatial sampling.
  (\protect\hyperlink{ref-mahoneyAssessingPerformanceSpatial2023}{Mahoney
  et al., 2023}))
\item
  Evidence has shown that widely different DSM approaches can result in
  maps with quite similar ``validation statistics'' but obviously
  different spatial patterns.
\end{enumerate}

See for example Figure~\ref{fig-genova}, which shows an area near
Montréal PQ mapped by convolutional neural networks (CNN) with the same
training points and covariates, but with different CNN window sizes. The
pointwise evaluation statistics in this example are almost identical.

\begin{figure}

{\centering \includegraphics{./figs/GenovaPosterFig1a.png}

}

\caption{\label{fig-genova}Different methods, different patterns (Giulio
Genova, ISRIC)}

\end{figure}

\emph{Externally}, from the map user's perspective:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Soils are managed as units, not point-wise.
\item
  Land-surface models often rely on 2D or 3D connectivity between grid
  cells.
\item
  More than a century of fieldwork has shown that soils occur in
  more-or-less homogeneous patches of various sizes, not as isolated
  pedons
  (\protect\hyperlink{ref-boulaineRemarquesQuelquesNotions1982}{Boulaine,
  1982}; \protect\hyperlink{ref-Fridland1974}{Fridland, 1974};
  \protect\hyperlink{ref-johnsonPedonPolypedon1963}{Johnson, 1963}).
\item
  The map user may confuse \emph{artefacts} of the mapping process with
  real soil patterns.
\end{enumerate}

\hypertarget{setup}{%
\section{Setup}\label{setup}}

\hypertarget{packages}{%
\subsection{Packages}\label{packages}}

These R packages will be used in the analysis. They must be
pre-installed.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{options}\NormalTok{(}\StringTok{"rgdal\_show\_exportToProj4\_warnings"}\OtherTok{=}\StringTok{"none"}\NormalTok{)}
\FunctionTok{options}\NormalTok{(}\AttributeTok{warn =} \SpecialCharTok{{-}}\DecValTok{1}\NormalTok{)}
\CommentTok{\# Robert Hijmans raster and vector data}
\FunctionTok{library}\NormalTok{(terra, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{) }\CommentTok{\# replaces \textasciigrave{}raster\textasciigrave{}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
terra 1.7.71
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# still needed to convert to \textasciigrave{}sp\textasciigrave{}}
\FunctionTok{library}\NormalTok{(raster, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{) }
\CommentTok{\# Pebesma et al. spation{-}temporal data}
\CommentTok{\# Simple Features}
\FunctionTok{library}\NormalTok{(sf, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{)          }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
Linking to GEOS 3.11.0, GDAL 3.5.3, PROJ 9.1.0; sf_use_s2() is TRUE
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# \textasciigrave{}sp\textasciigrave{} spatial classes {-}{-} still needed for conversions}
\FunctionTok{library}\NormalTok{(sp, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{)}
\CommentTok{\# variogram modelling}
\FunctionTok{library}\NormalTok{(gstat, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{)}
\CommentTok{\# Co{-}occurrence vectors}
\FunctionTok{library}\NormalTok{(motif, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{) }
\CommentTok{\# multivariate distance metrics}
\FunctionTok{library}\NormalTok{(philentropy, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{) }
\CommentTok{\# compare polygon map spatial structures, V measure}
\FunctionTok{library}\NormalTok{(sabre, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{)   }
\CommentTok{\# FRAGSTATS{-}style metrics}
\FunctionTok{library}\NormalTok{(landscapemetrics, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{) }
\CommentTok{\# utility functions for raster* landscape objects)}
\FunctionTok{library}\NormalTok{(landscapetools, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{)}
\CommentTok{\# ggplot graphics}
\FunctionTok{library}\NormalTok{(ggplot2, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{)}
\CommentTok{\# multiple graphics in one plot}
\FunctionTok{library}\NormalTok{(gridExtra, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{)}
\CommentTok{\# ggplot with terra SpatRaster}
\FunctionTok{library}\NormalTok{(tidyterra, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{)}
\CommentTok{\# data wrangling}
\FunctionTok{library}\NormalTok{(dplyr, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{)}
\CommentTok{\# colour palettes for graphics}
\FunctionTok{library}\NormalTok{(RColorBrewer, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{)}
\CommentTok{\# to access NRCS databases}
\FunctionTok{library}\NormalTok{(soilDB, }\AttributeTok{warn.conflicts=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{quiet =} \ConstantTok{TRUE}\NormalTok{)}
\CommentTok{\# supercells}
\CommentTok{\# install.packages("supercells", repos = "https://nowosad.r{-}universe.dev")}
\FunctionTok{library}\NormalTok{(supercells)}
\CommentTok{\# compare two rasters directly {-}{-} in development}
\CommentTok{\# devtools::install\_github("Nowosad/spquery")}
\CommentTok{\# library(spquery)}
\CommentTok{\# analyze cross{-}classification matrices}
\FunctionTok{library}\NormalTok{(diffeR) }
\end{Highlighting}
\end{Shaded}

\hypertarget{directories}{%
\subsection{Directories}\label{directories}}

Task: Set up the base directory.

This is on my system, change to wherever you want to store the sample
files. Note that in Unix-alike systems the \texttt{\textasciitilde{}}
symbol refers to the user's home directory.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{file.dir }\OtherTok{\textless{}{-}} \FunctionTok{path.expand}\NormalTok{(}\StringTok{"\textasciitilde{}/ds\_reference/Compare\_DSM/"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{example-dataset-soil-properties}{%
\subsection{Example dataset: soil
properties}\label{example-dataset-soil-properties}}

The output of a DSM prediction can be saved as a GeoTiff
(\protect\hyperlink{ref-OGCGeoTIFFStandard2023}{Open Geospatial
Consortium, 2023}).

Here we provide an example: (1\textdegree x 1\textdegree) tiles of the
soil pH, measured in 1:1 soil:water ratio, over the 0-5 cm depth slice
of an area in central NY State (USA):

\begin{enumerate}
\def\labelenumi{(\arabic{enumi})}
\tightlist
\item
  predictive map produced by the SoilGrids v2.0 project
  (\protect\hyperlink{ref-Poggio.etal2021a}{Poggio et al., 2021});
\item
  a rasterized gNATSGO
  (\protect\hyperlink{ref-nrcssoilsGriddedNationalSoil2023}{NRCS Soils,
  2023}) coverage of the same area, downscaled to match the SoilGrids
  v2.0 resolution.
\end{enumerate}

(Further on we examine the effect of \protect\hyperlink{scale}{scale}
for the gNATSGO product, and the
\protect\hyperlink{incove}{co-occurrence of two soil properties} from
SoilGrids v2.0)

You can download similar files as GeoTiff for an area of your
preference; see the scripts \texttt{gNATSGO\_WCS\_import.Rmd} and
\texttt{SoilGrids\ v2.0\_import.Rmd}.

We process these in R with the \texttt{terra} package, which has the
advantage that it only loads into computer memory as needed, and can
load lower resolution automatically if that's appropriate.

Task: Import these as \texttt{terra::SpatRaster} objects.

Of course, change the file names if you have downloaded different files
(tile, property, and/or depth slice).

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{file.dir }\OtherTok{\textless{}{-}} \FunctionTok{path.expand}\NormalTok{(}\StringTok{"\textasciitilde{}/ds\_reference/Compare\_DSM/"}\NormalTok{)}
\NormalTok{(gn }\OtherTok{\textless{}{-}} \FunctionTok{rast}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(file.dir, }\StringTok{"gNATSGO/lat4243\_lon{-}77{-}76/ph1to1h2o\_r\_05\_250.tif"}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
class       : SpatRaster 
dimensions  : 411, 410, 1  (nrow, ncol, nlyr)
resolution  : 0.002434898, 0.002434898  (x, y)
extent      : -76.99926, -76.00095, 41.99973, 43.00048  (xmin, xmax, ymin, ymax)
coord. ref. : lon/lat WGS 84 (EPSG:4326) 
source      : ph1to1h2o_r_05_250.tif 
name        : ph1to1h2o_r 
min value   :        4.60 
max value   :        7.62 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(sg }\OtherTok{\textless{}{-}} \FunctionTok{rast}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(file.dir, }\StringTok{"SoilGrids250/lat4243\_lon{-}77{-}76/phh2o\_0{-}5cm\_mean.tif"}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
class       : SpatRaster 
dimensions  : 426, 426, 1  (nrow, ncol, nlyr)
resolution  : 0.002349867, 0.002349867  (x, y)
extent      : -77.00071, -75.99967, 41.99918, 43.00022  (xmin, xmax, ymin, ymax)
coord. ref. : lon/lat WGS 84 (EPSG:4326) 
source      : phh2o_0-5cm_mean.tif 
name        : phh2o_0-5cm_mean 
min value   :         43.11129 
max value   :         71.58004 
\end{verbatim}

The SoilGrids map is in units of ph x 10 (to store one decimal place as
an integer), so we divide the values by 10 to match the gNATSGO product:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{values}\NormalTok{(sg) }\OtherTok{\textless{}{-}} \FunctionTok{values}\NormalTok{(sg)}\SpecialCharTok{/}\DecValTok{10}
\end{Highlighting}
\end{Shaded}

Task: Plot the two maps side-by-side, on the same value and colour
scale.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{range.sg.gn }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(sg), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{), }
                      \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(gn), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{))}
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\NormalTok{terra}\SpecialCharTok{::}\FunctionTok{plot}\NormalTok{(sg, }\AttributeTok{main =} \StringTok{"SoilGrids v2.0"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ range.sg.gn, }\AttributeTok{col=}\NormalTok{(sp}\SpecialCharTok{::}\FunctionTok{bpy.colors}\NormalTok{(}\DecValTok{50}\NormalTok{)))}
\NormalTok{terra}\SpecialCharTok{::}\FunctionTok{plot}\NormalTok{(gn, }\AttributeTok{main =} \StringTok{"gNATSGO"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ range.sg.gn, }\AttributeTok{col=}\NormalTok{(sp}\SpecialCharTok{::}\FunctionTok{bpy.colors}\NormalTok{(}\DecValTok{50}\NormalTok{)))}
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/fig-sg-gn-1.pdf}

}

\caption{\label{fig-sg-gn}pH, 0-5 cm}

\end{figure}

We see a wide range of values, especially in the SoilGrids map, and
quite different patterns.

Q: Describe the principal differences between the two maps.

\hypertarget{crop-to-a-smaller-area}{%
\subsubsection{Crop to a smaller area}\label{crop-to-a-smaller-area}}

For quicker computation, we restrict the maps (1\textdegree x
1\textdegree) to a quarter-map (0.25\textdegree x 0.25\textdegree),
centred to show some interesting patterns.

Task: Crop the two maps to a quarter-map.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{test.tile.size }\OtherTok{\textless{}{-}} \FloatTok{0.25}  \CommentTok{\# degrees}
\NormalTok{test.tile.x.offset }\OtherTok{\textless{}{-}} \FloatTok{0.2} \CommentTok{\# lrc west from right edge}
\NormalTok{test.tile.y.offset }\OtherTok{\textless{}{-}} \FloatTok{0.3}  \CommentTok{\# lrc north from bottom edge}
\NormalTok{ext.crop }\OtherTok{\textless{}{-}} \FunctionTok{round}\NormalTok{(}\FunctionTok{as.vector}\NormalTok{(}\FunctionTok{ext}\NormalTok{(sg)),}\DecValTok{2}\NormalTok{) }\CommentTok{\# line up to .00 decimal degrees}
\NormalTok{ext.crop[}\StringTok{"xmax"}\NormalTok{] }\OtherTok{\textless{}{-}}\NormalTok{ ext.crop[}\StringTok{"xmax"}\NormalTok{] }\SpecialCharTok{{-}}\NormalTok{ test.tile.x.offset}
\NormalTok{ext.crop[}\StringTok{"xmin"}\NormalTok{] }\OtherTok{\textless{}{-}}\NormalTok{ ext.crop[}\StringTok{"xmax"}\NormalTok{] }\SpecialCharTok{{-}}\NormalTok{ test.tile.size}
\NormalTok{ext.crop[}\StringTok{"ymin"}\NormalTok{] }\OtherTok{\textless{}{-}}\NormalTok{ ext.crop[}\StringTok{"ymin"}\NormalTok{] }\SpecialCharTok{+}\NormalTok{ test.tile.y.offset}
\NormalTok{ext.crop[}\StringTok{"ymax"}\NormalTok{] }\OtherTok{\textless{}{-}}\NormalTok{ ext.crop[}\StringTok{"ymin"}\NormalTok{] }\SpecialCharTok{+}\NormalTok{ test.tile.size}
\FunctionTok{ext}\NormalTok{(ext.crop)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
SpatExtent : -76.45, -76.2, 42.3, 42.55 (xmin, xmax, ymin, ymax)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gn }\OtherTok{\textless{}{-}} \FunctionTok{crop}\NormalTok{(gn, }\FunctionTok{ext}\NormalTok{(ext.crop));sg }\OtherTok{\textless{}{-}} \FunctionTok{crop}\NormalTok{(sg, }\FunctionTok{ext}\NormalTok{(ext.crop))}
\FunctionTok{ext}\NormalTok{(gn)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
SpatExtent : -76.4489762279447, -76.2006165924792, 42.2992262003736, 42.550020734226 (xmin, xmax, ymin, ymax)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ext}\NormalTok{(sg)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
SpatExtent : -76.450841145456, -76.1994053643854, 42.2999613764399, 42.5490472903977 (xmin, xmax, ymin, ymax)
\end{verbatim}

Notice that the two extents are not exactly the same because of the
different alignments of the pixels in the sources.

Task: Plot the two quarter-tile maps side by side, with a common legend.

The value ranges are ifferent, so we need to set up a common scale for
the visualization.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(range.sg }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(sg), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 4.430727 6.386996
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(range.gn }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(gn), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 4.80 7.53
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{range.sg.gn }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(sg), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{), }
                     \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(gn), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{))}
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\FunctionTok{plot}\NormalTok{(sg, }\AttributeTok{main =} \StringTok{"SoilGrids v2.0"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ range.sg.gn, }\AttributeTok{col=}\NormalTok{(sp}\SpecialCharTok{::}\FunctionTok{bpy.colors}\NormalTok{(}\DecValTok{50}\NormalTok{)))}
\FunctionTok{plot}\NormalTok{(gn, }\AttributeTok{main =} \StringTok{"gNATSGO"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ range.sg.gn, }\AttributeTok{col=}\NormalTok{(sp}\SpecialCharTok{::}\FunctionTok{bpy.colors}\NormalTok{(}\DecValTok{50}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/crop.2-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Q: Describe the differences between the two quarter-maps, both the
values and the patterns.

Q: Will the obvious difference in values affect the analysis of their
patterns?

\hypertarget{transform-to-a-metric-crs}{%
\subsubsection{Transform to a metric
CRS}\label{transform-to-a-metric-crs}}

Landscape metrics require approximately \emph{equal-area} grid cells, so
these maps, currently in a geographic Coördinate Reference System (CRS),
must be projected to a metric system. A reasonable choice for any small
areais the Universal Transmercator (UTM) system, which covers a
6\textdegree-wide latitude range.

Task: Search for the appropriate EPSG code at the
\href{https://epsg.org}{EPSG Geodetic Parameter Dataset}, see
(Figure~\ref{fig-epsg}).

\begin{figure}

{\centering \includegraphics{./figs/EPSG32618screen.png}

}

\caption{\label{fig-epsg}EPSG database entry for code 32618}

\end{figure}

Easiest is to use ``Map Search'' and further limit the results by the
text ``UTM''. Several datums can serve as the basis for the UTM CRS; for
easiest conversion select WGS84. In the USA these have the format
\texttt{326xx}, where \texttt{xx} is the UTM zone number.

Determine the UTM zone and the corresponding EPSG code:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# a function to find the correct UTM zome}
\NormalTok{long2UTM }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(long) \{ (}\FunctionTok{floor}\NormalTok{((long }\SpecialCharTok{+} \DecValTok{180}\NormalTok{)}\SpecialCharTok{/}\DecValTok{6}\NormalTok{) }\SpecialCharTok{\%\%} \DecValTok{60}\NormalTok{) }\SpecialCharTok{+} \DecValTok{1}\NormalTok{ \}}
\CommentTok{\# find the zone from the central meridian}
\NormalTok{utm.zone }\OtherTok{\textless{}{-}} \FunctionTok{long2UTM}\NormalTok{(}\FunctionTok{st\_bbox}\NormalTok{(sg)}\SpecialCharTok{$}\NormalTok{xmin }\SpecialCharTok{+} 
                       \FloatTok{0.5}\SpecialCharTok{*}\NormalTok{(}\FunctionTok{st\_bbox}\NormalTok{(sg)}\SpecialCharTok{$}\NormalTok{xmax }\SpecialCharTok{{-}} \FunctionTok{st\_bbox}\NormalTok{(sg)}\SpecialCharTok{$}\NormalTok{xmin))}
\FunctionTok{cat}\NormalTok{(}\FunctionTok{paste}\NormalTok{(}\StringTok{"UTM Zone"}\NormalTok{, utm.zone))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
UTM Zone 18
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{epsg.utm }\OtherTok{\textless{}{-}} \FunctionTok{paste0}\NormalTok{(}\StringTok{"epsg:326"}\NormalTok{, utm.zone)}
\FunctionTok{cat}\NormalTok{(}\FunctionTok{paste}\NormalTok{(}\StringTok{"CRS code:"}\NormalTok{, epsg.utm))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
CRS code: epsg:32618
\end{verbatim}

Task: Resample the maps to the UTM projection, at nominal 250 m grid
cell resolution.

Notes:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  The interpolation method used by \texttt{terra::project} is, by
  default, bilinear. This is appropriate for continuous-valued maps.
\item
  Specify the grid cell size with the \texttt{res} argument to
  \texttt{terra::project}. Both maps were nominally at this scale,
  although presented in geographical coördinates. If this is omitted,
  \texttt{terra::project} sets the size as a square with the smaller
  dimension, here at \(\approx 43^{0}N\) this is 229\textasciitilde m.
\end{enumerate}

The bounding boxes and resolutions are slightly different for the two
products.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{st\_bbox}\NormalTok{(gn)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
     xmin      ymin      xmax      ymax 
-76.44898  42.29923 -76.20062  42.55002 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{res}\NormalTok{(gn)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 0.002434898 0.002434898
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gn.utm }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{project}\NormalTok{(gn, epsg.utm, }
                         \AttributeTok{res =} \FunctionTok{c}\NormalTok{(}\DecValTok{250}\NormalTok{, }\DecValTok{250}\NormalTok{), }\AttributeTok{method =} \StringTok{"bilinear"}\NormalTok{)}
\FunctionTok{st\_bbox}\NormalTok{(gn.utm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
     xmin      ymin      xmax      ymax 
 380561.8 4683745.2  401561.8 4711745.2 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{res}\NormalTok{(gn.utm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 250 250
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{st\_bbox}\NormalTok{(sg)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
     xmin      ymin      xmax      ymax 
-76.45084  42.29996 -76.19941  42.54905 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.utm }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{project}\NormalTok{(sg, epsg.utm, }
                         \AttributeTok{res =} \FunctionTok{c}\NormalTok{(}\DecValTok{250}\NormalTok{, }\DecValTok{250}\NormalTok{), }\AttributeTok{method =} \StringTok{"bilinear"}\NormalTok{)}
\FunctionTok{st\_bbox}\NormalTok{(sg.utm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
     xmin      ymin      xmax      ymax 
 380409.5 4683779.9  401409.5 4711779.9 
\end{verbatim}

\hypertarget{harmonizing-the-mapped-areas}{%
\subsubsection{Harmonizing the mapped
areas}\label{harmonizing-the-mapped-areas}}

The two maps have slightly different concepts of areas not mapped:
unsurveyed urban areas and water bodies (both sources) and miscellaneous
land types such as mines or gravel pits (gNATSGO). SoilGrids v2.0
identified these by remote sensing, whereas gNATSGO used field survey
and (for urban areas) survey policy.

The maps also have slightly different extents. These must be made
identical before masking. We select one map as a template and resample
the other map into that template.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ext}\NormalTok{(gn.utm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
SpatExtent : 380561.847574791, 401561.847574791, 4683745.2384707, 4711745.2384707 (xmin, xmax, ymin, ymax)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ext}\NormalTok{(sg.utm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
SpatExtent : 380409.511975787, 401409.511975787, 4683779.92513894, 4711779.92513894 (xmin, xmax, ymin, ymax)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.utm }\OtherTok{\textless{}{-}} \FunctionTok{resample}\NormalTok{(sg.utm, gn.utm)}
\FunctionTok{ext}\NormalTok{(sg.utm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
SpatExtent : 380561.847574791, 401561.847574791, 4683745.2384707, 4711745.2384707 (xmin, xmax, ymin, ymax)
\end{verbatim}

Notice the small changes in the range: the resampling has slightly
lowered the extremes.

For the pattern analysis we want these \texttt{NA} areas to be the same.
For this we use a reciprocal mask.

Task: mask each map with the \texttt{NA} areas of the other.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.utm }\OtherTok{\textless{}{-}} \FunctionTok{mask}\NormalTok{(sg.utm, gn.utm)}
\CommentTok{\# SoilGrids now has some \textasciigrave{}NA\textasciigrave{} added from gNATSGO}
\NormalTok{gn.utm }\OtherTok{\textless{}{-}} \FunctionTok{mask}\NormalTok{(gn.utm, sg.utm)}
\CommentTok{\# The added \textasciigrave{}NA\textasciigrave{} are already in gNATSGO, now it gets \textasciigrave{}NA\textasciigrave{} originally on SoilGrids}
\end{Highlighting}
\end{Shaded}

Task: Plot the two maps side-by-side.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\FunctionTok{plot}\NormalTok{(sg.utm, }\AttributeTok{main =} \StringTok{"SoilGrids v2.0"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ range.sg.gn, }\AttributeTok{col=}\NormalTok{(sp}\SpecialCharTok{::}\FunctionTok{bpy.colors}\NormalTok{(}\DecValTok{50}\NormalTok{)))}
\FunctionTok{plot}\NormalTok{(gn.utm, }\AttributeTok{main =} \StringTok{"gNATSGO"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ range.sg.gn, }\AttributeTok{col=}\NormalTok{(sp}\SpecialCharTok{::}\FunctionTok{bpy.colors}\NormalTok{(}\DecValTok{50}\NormalTok{)))}
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/fig-sg-gn-utm-1.pdf}

}

\caption{\label{fig-sg-gn-utm}pH, 0-5 cm}

\end{figure}

\hypertarget{sec-import-gssurgo}{%
\subsection{Example dataset: Soil survey map unit
polygons}\label{sec-import-gssurgo}}

These will be used when examining the effects of scale in
(Section~\ref{sec-scale-geometry}) and
(Section~\ref{sec-categorical-generalization}).

In the USA the polygon maps, as delineated by the soil surveyors and
later digitized as vector GIS coverages, are available in the SSURGO and
STATSGO databases. These can be accessed with the
\texttt{SDA\_spatialQuery()} function of the Soil Data Access (SDA)
facility of the \texttt{soilDB} package written by NRCS scientists, to
allow R access to the NRCS database's REST/JSON web service. NCSS has
written a tutorial on SDA\footnote{https://ncss-tech.github.io/AQP/soilDB/SDA-tutorial.html}.

We specify the \texttt{geomIntersection\ =\ TRUE} argument to clip map
unit polygons to the bounding polygon. The bounding box (in WGS84
geographic coordinates) and CRS must be obtained from a spatial object.
For this we use SSURGO gridded map \texttt{gn.utm}, a
\texttt{terra::SpatRaster} which we created above, as the template. This
ensures that the same package is used for the returned object. Since
this is a polygon map, it will be a \texttt{terra::SpatVector} object.

Task: Download the map unit polygons for the restricted study area. This
may take a few minutes depending on how responsive is the remote server.

The bounding box is specified with the second argument to
\texttt{SDA\_spatialQuery}; here we have the quarter-degree products
from the continuous DSM to serve as the CRS and bounding box. We use the
gNATSGO \texttt{SpatRaster} (object \texttt{gn}) for this.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# get the polygons with their key}
\FunctionTok{system.time}\NormalTok{(}
\NormalTok{  mu.poly }\OtherTok{\textless{}{-}} \FunctionTok{SDA\_spatialQuery}\NormalTok{(gn, }
                            \AttributeTok{what =} \StringTok{"mupolygon"}\NormalTok{, }
                            \AttributeTok{db =} \StringTok{"SSURGO"}\NormalTok{, }
                            \AttributeTok{geomIntersection =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
   user  system elapsed 
  2.195   0.146  23.290 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{class}\NormalTok{(mu.poly)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "SpatVector"
attr(,"package")
[1] "terra"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{st\_crs}\NormalTok{(mu.poly)}\SpecialCharTok{$}\NormalTok{proj4string}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "+proj=longlat +datum=WGS84 +no_defs"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{summary}\NormalTok{(mu.poly)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
     mukey            area_ac        
 Min.   : 295575   Min.   :  0.0022  
 1st Qu.: 295602   1st Qu.:  2.9416  
 Median : 295649   Median :  6.6646  
 Mean   : 621141   Mean   : 15.1426  
 3rd Qu.: 295687   3rd Qu.: 16.1199  
 Max.   :2760841   Max.   :897.0175  
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(mu.poly)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
    mukey   area_ac
1 2723081  1.846781
2 2723081  4.124984
3 2723081 19.108964
4 2723081  7.223155
5 2723081  1.576274
6 2723081  6.984551
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# plot with area in acres}
\FunctionTok{plot}\NormalTok{(mu.poly, }\AttributeTok{y =} \StringTok{"area\_ac"}\NormalTok{,}
     \AttributeTok{type =} \StringTok{"continuous"}\NormalTok{,}
     \AttributeTok{main =} \StringTok{"SSURGO map units, area in acres"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/download-ssurgo-1.pdf}

}

\end{figure}

What do these map units codes represent? Find the map units from the
same geometry.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mu.key }\OtherTok{\textless{}{-}} \FunctionTok{SDA\_spatialQuery}\NormalTok{(gn.utm, }
                           \AttributeTok{what =} \StringTok{"mukey"}\NormalTok{, }
                           \AttributeTok{db =} \StringTok{"SSURGO"}\NormalTok{, }
                           \AttributeTok{geomIntersection =} \ConstantTok{TRUE}\NormalTok{)}
\FunctionTok{head}\NormalTok{(mu.key)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
   mukey                                                  muname
1 295575                                           Alluvial land
2 295576          Arkport fine sandy loam, 2 to 6 percent slopes
3 295577         Arkport fine sandy loam, 6 to 12 percent slopes
4 295578          Bath channery silt loam, 2 to 5 percent slopes
5 295579         Bath channery silt loam, 5 to 15 percent slopes
6 295580 Bath channery silt loam, 5 to 15 percent slopes, eroded
\end{verbatim}

These are named map units.

These have extended site data, which comes from the linked attribute
database. For this we use an SQL query and the \texttt{SDA\_query} (link
to ``Soil Data Access'') \footnote{https://sdmdataaccess.nrcs.usda.gov/}
function.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# format the list of map units for SQL}
\NormalTok{IS }\OtherTok{\textless{}{-}}\NormalTok{ soilDB}\SpecialCharTok{::}\FunctionTok{format\_SQL\_in\_statement}\NormalTok{(mu.poly}\SpecialCharTok{$}\NormalTok{mukey)}
\CommentTok{\# query string {-}{-} all components}
\NormalTok{ws }\OtherTok{\textless{}{-}} \FunctionTok{sprintf}\NormalTok{(}\StringTok{"mukey IN \%s"}\NormalTok{, IS)}
\CommentTok{\# format the SQL query}
\NormalTok{query }\OtherTok{\textless{}{-}} \FunctionTok{paste}\NormalTok{(}\StringTok{"SELECT * FROM mapunit WHERE"}\NormalTok{, ws)}
\CommentTok{\# and run it}
\NormalTok{mu.info }\OtherTok{\textless{}{-}} \FunctionTok{SDA\_query}\NormalTok{(query)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
single result set, returning a data.frame
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{dim}\NormalTok{(mu.info)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 217  26
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{names}\NormalTok{(mu.info)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] "musym"         "muname"        "mukind"        "mustatus"     
 [5] "muacres"       "mapunitlfw_l"  "mapunitlfw_r"  "mapunitlfw_h" 
 [9] "mapunitpfa_l"  "mapunitpfa_r"  "mapunitpfa_h"  "farmlndcl"    
[13] "muhelcl"       "muwathelcl"    "muwndhelcl"    "interpfocus"  
[17] "invesintens"   "iacornsr"      "nhiforsoigrp"  "nhspiagr"     
[21] "vtsepticsyscl" "mucertstat"    "lkey"          "mukey"        
[25] "museq"         "nationalmusym"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{series.name }\OtherTok{\textless{}{-}} \StringTok{"Ovid"} \CommentTok{\# look for a map unit by name}
\FunctionTok{length}\NormalTok{(ix }\OtherTok{\textless{}{-}} \FunctionTok{which}\NormalTok{(}\FunctionTok{substr}\NormalTok{(mu.info}\SpecialCharTok{$}\NormalTok{muname, }\DecValTok{1}\NormalTok{, }
                          \FunctionTok{nchar}\NormalTok{(series.name)) }\SpecialCharTok{==}\NormalTok{ series.name))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 2
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mu.info[ix, ]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
   musym                                              muname       mukind
73   OaA               Ovid silt loam, 0 to 6 percent slopes Consociation
74  OcC3 Ovid silty clay loam, 6 to 12 percent slopes eroded Consociation
   mustatus muacres mapunitlfw_l mapunitlfw_r mapunitlfw_h mapunitpfa_l
73     <NA>    5157           NA           NA           NA           NA
74     <NA>     312           NA           NA           NA           NA
   mapunitpfa_r mapunitpfa_h                 farmlndcl muhelcl muwathelcl
73           NA           NA Prime farmland if drained    <NA>       <NA>
74           NA           NA        Not prime farmland    <NA>       <NA>
   muwndhelcl interpfocus invesintens iacornsr nhiforsoigrp nhspiagr
73       <NA>        <NA>        <NA>       NA         <NA>       NA
74       <NA>        <NA>        <NA>       NA         <NA>       NA
   vtsepticsyscl mucertstat  lkey  mukey museq nationalmusym
73          <NA>       <NA> 12437 295666    87          9xnm
74          <NA>       <NA> 12437 295667    88          9xnn
\end{verbatim}

\hypertarget{transform-to-a-metric-crs-1}{%
\subsubsection{Transform to a metric
CRS}\label{transform-to-a-metric-crs-1}}

Project to the metric CRS we are using in this area. The CRS was defined
in the previous section.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{st\_crs}\NormalTok{(mu.poly)}\SpecialCharTok{$}\NormalTok{proj4string}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "+proj=longlat +datum=WGS84 +no_defs"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mu.poly }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{project}\NormalTok{(mu.poly, epsg.utm)}
\FunctionTok{st\_crs}\NormalTok{(mu.poly)}\SpecialCharTok{$}\NormalTok{proj4string}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "+proj=utm +zone=18 +datum=WGS84 +units=m +no_defs"
\end{verbatim}

\hypertarget{characterizing-patterns}{%
\section{Characterizing patterns}\label{characterizing-patterns}}

Before comparing patterns of different maps, and trying to evaluate how
close they are to ``reality'', they first have to be characterized by
statistical measures. This gives objective information about their
spatial patterns.

The methods to characterize patterns are different for maps of
\emph{continuous} variables (Section~\ref{sec-continuous}) and
\emph{classified} (categorical) variables
(Section~\ref{sec-classified}).

\hypertarget{sec-continuous}{%
\section{Characterizing patterns -- Continuous}\label{sec-continuous}}

These are methods that require continuous values on at least an interval
scale, and usually a ratio scale (with a true zero). In the case of the
example here, pH does not have a true zero, so it is an interval scale.
Other properties such as soil thickness to a restricting layer have a
true zero, and one can speak of one location being ``twice as thick''
than another, for example.

\hypertarget{sec-vgm}{%
\subsection{The global variogram}\label{sec-vgm}}

The variogram (or a correlogram) can be used to characterize the degree
of spatial continuity and the ``roughness'' of a continuous property
map, averaged across the entire map. Note that this depends on the grid
cell size in two ways:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Any pattern at finer resolutions has been removed;
\item
  The values in grid cells may be produced by punctual or block methods.
  Block methods smooth values, so that the variogram sill will
  necessarily be lower than for punctual predictions. Also, the range
  may be longer.
\end{enumerate}

In this section we compute and compare the short-range variograms, these
reveal the local structure. In DSM maps the variogram is typically
unbounded, but we don't care about the long-range structure when we are
evaluating patterns. The parameters of the local structure characterize
the fine-scale variability.

Note: Variograms are typically produced separately for each mapped soil
property. To characterize an inherent landscape scale, a number of
properties can be combined by principal component analysis (PCA) and the
first component (PC1) can be characterized.

Task: Convert the \texttt{terra::SpatRaster} objects to
\texttt{raster::raster} and then to \texttt{sf:sf} objects in order to
compute variograms.

Note: There is (so far) no direct conversion. The
\texttt{gstat::variogram} method must be applied to an object of class
\texttt{sp} or \texttt{sf}, not directly to a
\texttt{terra::SpatRaster}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gn.sp }\OtherTok{\textless{}{-}} \FunctionTok{as}\NormalTok{(}\FunctionTok{raster}\NormalTok{(gn.utm), }\StringTok{"SpatialPointsDataFrame"}\NormalTok{)}
\NormalTok{gn.sf }\OtherTok{\textless{}{-}} \FunctionTok{st\_as\_sf}\NormalTok{(gn.sp)}
\FunctionTok{names}\NormalTok{(gn.sf)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "ph1to1h2o_r" "geometry"   
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.sp }\OtherTok{\textless{}{-}} \FunctionTok{as}\NormalTok{(}\FunctionTok{raster}\NormalTok{(sg.utm), }\StringTok{"SpatialPointsDataFrame"}\NormalTok{)}
\NormalTok{sg.sf }\OtherTok{\textless{}{-}} \FunctionTok{st\_as\_sf}\NormalTok{(sg.sp)}
\FunctionTok{names}\NormalTok{(sg.sf)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "phh2o_0.5cm_mean" "geometry"        
\end{verbatim}

Task: Set the initial parameters for empirical variogram as the
resolution.

If the bin width is the resolution, we get one-grid-cell spatial
correlations.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{range.init }\OtherTok{\textless{}{-}} \DecValTok{1000}  \CommentTok{\# estimated range, m }
\NormalTok{cutoff.init }\OtherTok{\textless{}{-}}\NormalTok{ range.init}\SpecialCharTok{*}\DecValTok{5}  \CommentTok{\# cutoff for empirical variogram, m}
\NormalTok{width.init }\OtherTok{\textless{}{-}} \DecValTok{250}   \CommentTok{\# bin width}
\end{Highlighting}
\end{Shaded}

Task: Compute and display the empirical variograms.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{v.sg }\OtherTok{\textless{}{-}} \FunctionTok{variogram}\NormalTok{(phh2o\_0}\FloatTok{.5}\NormalTok{cm\_mean }\SpecialCharTok{\textasciitilde{}} \DecValTok{1}\NormalTok{, }\AttributeTok{loc =}\NormalTok{ sg.sf, }
                  \AttributeTok{cutoff=}\NormalTok{cutoff.init, }\AttributeTok{width=}\NormalTok{width.init)}
\CommentTok{\#}
\NormalTok{v.gn }\OtherTok{\textless{}{-}}\NormalTok{ gstat}\SpecialCharTok{::}\FunctionTok{variogram}\NormalTok{(ph1to1h2o\_r }\SpecialCharTok{\textasciitilde{}} \DecValTok{1}\NormalTok{, }\AttributeTok{loc =}\NormalTok{ gn.sf, }
                         \AttributeTok{cutoff=}\NormalTok{cutoff.init, }\AttributeTok{width=}\NormalTok{width.init)}
\NormalTok{ylim.v }\OtherTok{\textless{}{-}} \FunctionTok{max}\NormalTok{(v.gn}\SpecialCharTok{$}\NormalTok{gamma, v.sg}\SpecialCharTok{$}\NormalTok{gamma)}
\NormalTok{p1 }\OtherTok{\textless{}{-}} \FunctionTok{plot}\NormalTok{(v.sg, }\AttributeTok{ylim =} \FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, ylim.v), }\AttributeTok{main =} \StringTok{"SoilGrids v2.0"}\NormalTok{)}
\NormalTok{p2 }\OtherTok{\textless{}{-}} \FunctionTok{plot}\NormalTok{(v.gn, }\AttributeTok{ylim =} \FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, ylim.v), }\AttributeTok{main =} \StringTok{"gNATSGO"}\NormalTok{)}
\FunctionTok{grid.arrange}\NormalTok{(p1, p2, }\AttributeTok{nrow =} \DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/variogram_empirical-1.pdf}

}

\end{figure}

Q: Describe the empirical differences in spatial structure of the two
maps.

Task: Fit a variogram model to the empirical variogram.

The differences can be quantified by the parameters of a fitted
variogram model. We try an exponential model because (1) it has the
simplest theory, and (2) we expect to not reach a sill within the short
range investigated.

We use the \texttt{fit.variogram} method to adjust an initial estimate
by weighted least squared. The estimated sill is the maximum \(\gamma\)
in the empirical variogram.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{vm.gn }\OtherTok{\textless{}{-}} \FunctionTok{vgm}\NormalTok{(}\FloatTok{0.8}\SpecialCharTok{*}\FunctionTok{max}\NormalTok{(v.gn}\SpecialCharTok{$}\NormalTok{gamma), }\StringTok{"Exp"}\NormalTok{, range.init, }\DecValTok{0}\NormalTok{)}
\NormalTok{(vmf.gn }\OtherTok{\textless{}{-}} \FunctionTok{fit.variogram}\NormalTok{(v.gn, }\AttributeTok{model=}\NormalTok{vm.gn))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  model     psill    range
1   Nug 0.0000000   0.0000
2   Exp 0.1240257 449.6727
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{vm.sg }\OtherTok{\textless{}{-}} \FunctionTok{vgm}\NormalTok{(}\FloatTok{0.8}\SpecialCharTok{*}\FunctionTok{max}\NormalTok{(v.sg}\SpecialCharTok{$}\NormalTok{gamma), }\StringTok{"Exp"}\NormalTok{, range.init, }\DecValTok{0}\NormalTok{)}
\NormalTok{(vmf.sg }\OtherTok{\textless{}{-}} \FunctionTok{fit.variogram}\NormalTok{(v.sg, }\AttributeTok{model=}\NormalTok{vm.sg))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  model     psill    range
1   Nug 0.0000000    0.000
2   Exp 0.1679165 3960.544
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{p1 }\OtherTok{\textless{}{-}} \FunctionTok{plot}\NormalTok{(v.sg, }\AttributeTok{model=}\NormalTok{vmf.sg, }\AttributeTok{ylim =} \FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, ylim.v), }\AttributeTok{main =} \StringTok{"SoilGrids v2.0"}\NormalTok{, }
     \AttributeTok{xlab =} \StringTok{"separation m"}\NormalTok{, }\AttributeTok{ylab =} \FunctionTok{expression}\NormalTok{(}\FunctionTok{paste}\NormalTok{(Delta, }\FunctionTok{plain}\NormalTok{(pH)}\SpecialCharTok{\^{}}\DecValTok{2}\NormalTok{)))}
\NormalTok{p2 }\OtherTok{\textless{}{-}} \FunctionTok{plot}\NormalTok{(v.gn, }\AttributeTok{model=}\NormalTok{vmf.gn, }\AttributeTok{ylim =} \FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, ylim.v), }\AttributeTok{main =} \StringTok{"gNATSGO"}\NormalTok{, }
     \AttributeTok{xlab =} \StringTok{"separation m"}\NormalTok{, }\AttributeTok{ylab =} \FunctionTok{expression}\NormalTok{(}\FunctionTok{paste}\NormalTok{(Delta, }\FunctionTok{plain}\NormalTok{(pH)}\SpecialCharTok{\^{}}\DecValTok{2}\NormalTok{)))}
\FunctionTok{grid.arrange}\NormalTok{(p1, p2, }\AttributeTok{nrow =} \DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/fitted_variogram_model-1.pdf}

}

\end{figure}

Q: How well do the fitted models match the empirical variograms? If the
fit has some problems, what could be a solution?

Q: Describe the modelled differences in spatial structure of the two
maps (total sill, range).

Q: What is the implication for the utility of the maps?

Q: Is there any way to decide which is ``better'' in some sense?

\hypertarget{moving-window-local-association}{%
\subsection{Moving-window local
association}\label{moving-window-local-association}}

The local spatial structure may not be consistent across the mapped
area, so that the average variogram, computed over that area, can be
misleading. With so many values (grid cells) it's possible to compute
moving-window variograms, as in the VESPER program
(\protect\hyperlink{ref-minasnyVESPERVariogramEstimation2005}{Minasny et
al., 2005}) developed for precision agriculture applications. This will
show if the pattern is consistent across the map, and also allow maps to
be compared block-by-block. I have not (yet?) implemented this in R, so
we will use another method to assess moving-window local spatial
association.

A quick way to see the local degree of autocorrelation is with Moran's I
applied to a window of appropriate size around each grid cell, using the
\texttt{terra::autocor} function.

Moran's I is defined as:

\[  I = \frac{n}{\sum_i \sum_j w_{ij}} \frac{\sum_i \sum_j w_{ij}(y_i - \bar{y})(y_j - \bar{y})}{\sum_i (y_i - \bar{y})^2}\]

where \(y_i\) is the value of the variable in the \(i\)th of \(n\)
neighbouring grid cells, \(\bar{y}\) is the global mean of the variable,
\(w_{ij}\) is the spatial \textbf{weight} of the link between the target
cell \(i\) and its neighbour cell \(j\). The expected value of Moran's I
is \(-1/(n-1)\) if the pattern of the response variable is random, i.e.,
no spatial correlation. So for a 5\textsubscript{x}5 neighbourhood the
expected value if random is \(-1/24 = -0.041\bar{6} \approx 0\).

The second term numerator is the weighted covariance; its denominator
normalizes by the variance. The first term normalizes by the sum of all
weights, so that the test is comparable among tests with different
numbers of neighbours and using different weightings.

Task: Construct a weights matrix for local Moran's I

We determine the weights matrix for Moran's I from the global variogram
analysis and the grid cell size.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# for a 5x5 matrix}
\CommentTok{\# there must be a more elegant way to do this!}
\NormalTok{(vl }\OtherTok{\textless{}{-}} \FunctionTok{variogramLine}\NormalTok{(vm.sg, }
                     \AttributeTok{dist\_vector =} \FunctionTok{c}\NormalTok{(}\DecValTok{250}\NormalTok{, }\DecValTok{250}\SpecialCharTok{*}\FunctionTok{sqrt}\NormalTok{(}\DecValTok{2}\NormalTok{), }
                                     \DecValTok{500}\NormalTok{, }\DecValTok{250}\SpecialCharTok{*}\FunctionTok{sqrt}\NormalTok{(}\DecValTok{5}\NormalTok{), }
                                     \DecValTok{500}\SpecialCharTok{*}\FunctionTok{sqrt}\NormalTok{(}\DecValTok{2}\NormalTok{))))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
      dist      gamma
1 250.0000 0.02045734
2 353.5534 0.02754274
3 500.0000 0.03638954
4 559.0170 0.03960426
5 707.1068 0.04688293
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(w.r }\OtherTok{\textless{}{-}} \DecValTok{1}\SpecialCharTok{/}\NormalTok{(vl}\SpecialCharTok{$}\NormalTok{gamma }\SpecialCharTok{/}\NormalTok{ vl}\SpecialCharTok{$}\NormalTok{gamma[}\DecValTok{1}\NormalTok{]))  }\CommentTok{\# relative weights}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 1.0000000 0.7427491 0.5621765 0.5165440 0.4363495
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(w.m }\OtherTok{\textless{}{-}} \FunctionTok{matrix}\NormalTok{(}\FunctionTok{c}\NormalTok{(w.r[}\DecValTok{5}\NormalTok{], w.r[}\DecValTok{4}\NormalTok{], w.r[}\DecValTok{3}\NormalTok{], w.r[}\DecValTok{4}\NormalTok{], w.r[}\DecValTok{5}\NormalTok{],}
\NormalTok{                w.r[}\DecValTok{5}\NormalTok{], w.r[}\DecValTok{2}\NormalTok{], w.r[}\DecValTok{1}\NormalTok{], w.r[}\DecValTok{2}\NormalTok{], w.r[}\DecValTok{5}\NormalTok{],}
\NormalTok{                w.r[}\DecValTok{3}\NormalTok{], w.r[}\DecValTok{1}\NormalTok{], }\DecValTok{0}\NormalTok{, w.r[}\DecValTok{1}\NormalTok{], w.r[}\DecValTok{3}\NormalTok{],}
\NormalTok{                w.r[}\DecValTok{5}\NormalTok{], w.r[}\DecValTok{2}\NormalTok{], w.r[}\DecValTok{1}\NormalTok{], w.r[}\DecValTok{2}\NormalTok{], w.r[}\DecValTok{5}\NormalTok{],}
\NormalTok{                w.r[}\DecValTok{5}\NormalTok{], w.r[}\DecValTok{4}\NormalTok{], w.r[}\DecValTok{3}\NormalTok{], w.r[}\DecValTok{4}\NormalTok{], w.r[}\DecValTok{5}\NormalTok{]), }
                \AttributeTok{nrow =} \DecValTok{5}\NormalTok{, }\AttributeTok{ncol =} \DecValTok{5}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
          [,1]      [,2]      [,3]      [,4]      [,5]
[1,] 0.4363495 0.4363495 0.5621765 0.4363495 0.4363495
[2,] 0.5165440 0.7427491 1.0000000 0.7427491 0.5165440
[3,] 0.5621765 1.0000000 0.0000000 1.0000000 0.5621765
[4,] 0.5165440 0.7427491 1.0000000 0.7427491 0.5165440
[5,] 0.4363495 0.4363495 0.5621765 0.4363495 0.4363495
\end{verbatim}

Task: Compute and display the moving-window autocorrelation.

This uses the \texttt{terra::autocor} method, applied to a weighted
window.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.utm.autocor }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{autocor}\NormalTok{(sg.utm, }\AttributeTok{w=}\NormalTok{w.m, }
                                 \AttributeTok{method=}\StringTok{"moran"}\NormalTok{, }\AttributeTok{global =} \ConstantTok{FALSE}\NormalTok{)}
\NormalTok{gn.utm.autocor }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{autocor}\NormalTok{(gn.utm, }\AttributeTok{w=}\NormalTok{w.m, }
                                 \AttributeTok{method=}\StringTok{"moran"}\NormalTok{, }\AttributeTok{global =} \ConstantTok{FALSE}\NormalTok{)}
\NormalTok{(range.sg.autocor }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(sg.utm.autocor), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] -0.2621647  6.2650773
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(range.gn.autocor }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(gn.utm.autocor), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] -1.937685  6.388628
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{range.autocor }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(range.sg.autocor, range.gn.autocor)}
\CommentTok{\# hcl.pals(type = "diverging")}
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\NormalTok{terra}\SpecialCharTok{::}\FunctionTok{plot}\NormalTok{(sg.utm.autocor, }\AttributeTok{main =} \StringTok{"SG250, Moran\textquotesingle{}s I, 5x5"}\NormalTok{, }
            \AttributeTok{range =}\NormalTok{ range.autocor, }\AttributeTok{col =} \FunctionTok{rev}\NormalTok{(}\FunctionTok{hcl.colors}\NormalTok{(}\DecValTok{32}\NormalTok{, }\AttributeTok{palette =} \StringTok{"RdYlGn"}\NormalTok{)))}
\NormalTok{terra}\SpecialCharTok{::}\FunctionTok{plot}\NormalTok{(gn.utm.autocor, }\AttributeTok{main =} \StringTok{"gNATSGO, Moran\textquotesingle{}s I, 5x5"}\NormalTok{, }
            \AttributeTok{range =}\NormalTok{ range.autocor, }\AttributeTok{col =} \FunctionTok{rev}\NormalTok{(}\FunctionTok{hcl.colors}\NormalTok{(}\DecValTok{32}\NormalTok{, }\AttributeTok{palette =} \StringTok{"RdYlGn"}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/moving-window-5-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

To appreciate the local Moran's I values, here are the global Moran's I
with the same weights matrix. These are the averages of all the local
(window) Moran's I.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{terra}\SpecialCharTok{::}\FunctionTok{autocor}\NormalTok{(sg.utm, }\AttributeTok{w=}\NormalTok{w.m, }\AttributeTok{method=}\StringTok{"moran"}\NormalTok{, }\AttributeTok{global =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
phh2o_0-5cm_mean 
        1.083992 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{terra}\SpecialCharTok{::}\FunctionTok{autocor}\NormalTok{(gn.utm, }\AttributeTok{w=}\NormalTok{w.m, }\AttributeTok{method=}\StringTok{"moran"}\NormalTok{, }\AttributeTok{global =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
ph1to1h2o_r 
  0.5343023 
\end{verbatim}

These are both very far from the random value \(-0.041\bar{6}\). Both
maps show hot spots with much larger local autocorrelation, and some
areas with almost none or even more dispersed than random (negative
values).

Q: Is the pattern of local autocorrelation the same across the map?

Q: Which map has larger differences?

\hypertarget{sec-glcm}{%
\subsection{Grey Level Co-occurrence Matrix (GLCM)}\label{sec-glcm}}

The idea of characterizing the texture of an image has a long history in
image processing (e.g., Haralick et al.
(\protect\hyperlink{ref-haralickTexturalFeaturesImage1973}{1973})). The
``grey levels'' (GL) refer to pixel values -- in our context, the values
of the soil property, typically quantized (sliced) to some precision.
The ``co-occurrence'' (C) refers to the statistical properties within
some window, either isotropic or weighted in some direction. The GLCM
shows often different combinations of values (``grey levels'') occur
over the map. Many statistics can then be computed to characterize this
matrix.

Thus GLCM statistics, in the context of DSM, show the local properties
of a windows as it moves across the map. These can be interpreted as,
for example, homogeneity or contrast within a window, thereby revealing
areas of the map with different spatial structure.

See Hall-Beyer
(\protect\hyperlink{ref-hall-beyerGLCMTextureTutorial2017}{2017a}) for a
tutorial introduction to the construction, use, and interpretation of
GLCM-based textures, and Hall-Beyer
(\protect\hyperlink{ref-hall-beyerPracticalGuidelinesChoosing2017}{2017b})
for guidelines on choosing appropriate GLCM-based textures.

\hypertarget{sec-glcm-quant}{%
\subsubsection{Quantization}\label{sec-glcm-quant}}

The GLCM is constructed from a moving-window analysis of the map, with
the (odd-sized) window considered as a matrix of grid cells.

Before analysis the original map is quantized into a fixed number of
levels, by analogy with remote sensing image processing typically from
16 to 64. The GLCM should approximate the joint probability distribution
of two pixels with the specified shift(s). We would thus like non-zeroes
for most of the GLCM, and if there are too many values in the original
matrix, many pairs of values will not occur. Quantization is computed by
slicing the value range into equal intervals and replacing the original
values with the integer level number.

For the SoilGrid map most values are different, in fact in this example
there are 8937 different values, not many fewer than the pixel count,
9408. The same is true for gNATSGO -- in that case due to the
reprojection by bilinear interpolation. So quantization makes sense.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{dim}\NormalTok{(}\FunctionTok{unique}\NormalTok{(sg.utm))[}\DecValTok{1}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 8937
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{prod}\NormalTok{(}\FunctionTok{dim}\NormalTok{(sg.utm))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 9408
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{dim}\NormalTok{(}\FunctionTok{unique}\NormalTok{(gn.utm))[}\DecValTok{1}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 8605
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{prod}\NormalTok{(}\FunctionTok{dim}\NormalTok{(gn.utm))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 9408
\end{verbatim}

The following code shows how to quantize the SoilGrids map into 32
levels. (This will be done automatically by the \texttt{glcm} function,
see below.)

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(sg.utm, }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 4.462665 6.289024
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.quant }\OtherTok{\textless{}{-}} \FunctionTok{cut}\NormalTok{(}\FunctionTok{values}\NormalTok{(sg.utm), }\AttributeTok{breaks =} \DecValTok{32}\NormalTok{, }\AttributeTok{labels =} \DecValTok{0}\SpecialCharTok{:}\DecValTok{31}\NormalTok{, }\AttributeTok{include.lowest =} \ConstantTok{TRUE}\NormalTok{)}
\FunctionTok{table}\NormalTok{(sg.quant)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
sg.quant
  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19 
 22  79 198 217 323 272 322 391 433 478 439 449 388 389 362 399 426 456 440 427 
 20  21  22  23  24  25  26  27  28  29  30  31 
412 359 308 259 202 175 138  83  39  37  17   8 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.utm.quant }\OtherTok{\textless{}{-}}\NormalTok{ sg.utm; }\FunctionTok{values}\NormalTok{(sg.utm.quant) }\OtherTok{\textless{}{-}}\NormalTok{ sg.quant}
\FunctionTok{plot}\NormalTok{(sg.utm.quant, }\AttributeTok{col =} \FunctionTok{rainbow}\NormalTok{(}\DecValTok{32}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/cut.matrix-1.pdf}

}

\end{figure}

The lowest quantiles (lowest pH values) are on the hills to the east of
the maop, the highest quantiles (highest pH values) in the stream
valleys to the west.

\hypertarget{sec-glcm-comp}{%
\subsubsection{Computation}\label{sec-glcm-comp}}

From the quantized matrix, the GLCM can be constructed for one or more
specified offsets, called a shift. These can be either along the row,
column, or diagonal, as specified by the analyst. Each element at
position \((i,j)\) in the GLCM counts how many times a pixel with value
\(i\) and a value \(j\) occur together with the specified offset. So for
example a map quantized with 32 levels will have a 32 x 32 GLCM.

If multiple shifts are specified, the texture statistics are computed
for all the specified shifts, with the result for a pixel being the mean
of these statistics for each pixel.

The GLCM describes the spatial relationships of (quantized) values in
the map; this can be considered ``texture''. Many statistics can be
computed on the GLCM. Among the relevant statistics for pattern analysis
are the mean, variance, homogeneity, contrast, entropy, dissimilarity,
second moment, and correlation of the GLCM.

The R \texttt{glcm} package computes these metrics. It requires an
object in the older \texttt{raster} package format.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{require}\NormalTok{(glcm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
Loading required package: glcm
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# convert to the older \textasciigrave{}raster\textasciigrave{} format}
\NormalTok{sg.utm.raster }\OtherTok{\textless{}{-}} \FunctionTok{raster}\NormalTok{(sg.utm)}
\NormalTok{gn.utm.raster }\OtherTok{\textless{}{-}} \FunctionTok{raster}\NormalTok{(gn.utm)}
\end{Highlighting}
\end{Shaded}

We choose to compute the mean statistics for four shifts: one pixel by
row, column, and both diagonals. If there is orientation (anisotropy)
evident in the map, just one shift could be used to characterize the
shifts in that orientation.

We choose to compute on a 5 x 5 window (both dimensions must be odd).
Since the resolution is already coarse (250 m) this will characterize
the texture in \(1.5625 ~ \mathrm{km}^2\) squares

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{stat.list }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"mean"}\NormalTok{,}\StringTok{"variance"}\NormalTok{,}\StringTok{"homogeneity"}\NormalTok{,}\StringTok{"contrast"}\NormalTok{,}
               \StringTok{"entropy"}\NormalTok{,}\StringTok{"dissimilarity"}\NormalTok{,}\StringTok{"second\_moment"}\NormalTok{,}
               \StringTok{"correlation"}\NormalTok{)}
\NormalTok{glcm.sg }\OtherTok{\textless{}{-}} \FunctionTok{rast}\NormalTok{(}\FunctionTok{glcm}\NormalTok{(sg.utm.raster,}
                   \AttributeTok{window =} \FunctionTok{c}\NormalTok{(}\DecValTok{5}\NormalTok{, }\DecValTok{5}\NormalTok{),}
                   \AttributeTok{n\_grey =} \DecValTok{32}\NormalTok{, }\CommentTok{\# number of levels in the GLCM}
                   \AttributeTok{shift=}\FunctionTok{list}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{0}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\SpecialCharTok{{-}}\DecValTok{1}\NormalTok{)), }\CommentTok{\# all directions}
                   \AttributeTok{na\_opt =} \StringTok{"ignore"}\NormalTok{,}
                   \AttributeTok{statistics =}\NormalTok{ stat.list))}
\CommentTok{\# gNATSGO is not perfectly square}
\NormalTok{glcm.gn }\OtherTok{\textless{}{-}} \FunctionTok{rast}\NormalTok{(}\FunctionTok{glcm}\NormalTok{(gn.utm.raster,}
                   \AttributeTok{window =} \FunctionTok{c}\NormalTok{(}\DecValTok{5}\NormalTok{, }\DecValTok{5}\NormalTok{),}
                   \AttributeTok{n\_grey =} \DecValTok{32}\NormalTok{, }\CommentTok{\# number of levels in the GLCM}
                   \AttributeTok{shift=}\FunctionTok{list}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{0}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\SpecialCharTok{{-}}\DecValTok{1}\NormalTok{)), }\CommentTok{\# all directions}
                   \AttributeTok{na\_opt =} \StringTok{"ignore"}\NormalTok{,}
                   \AttributeTok{statistics =}\NormalTok{ stat.list))}
\FunctionTok{class}\NormalTok{(glcm.sg)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "SpatRaster"
attr(,"package")
[1] "terra"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{summary}\NormalTok{(glcm.sg)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
   glcm_mean      glcm_variance     glcm_homogeneity glcm_contrast    
 Min.   :0.0636   Min.   :  4.896   Min.   :0.1201   Min.   :  0.200  
 1st Qu.:0.3072   1st Qu.: 96.120   1st Qu.:0.3702   1st Qu.:  2.900  
 Median :0.4472   Median :206.645   Median :0.4413   Median :  5.470  
 Mean   :0.4520   Mean   :242.151   Mean   :0.4608   Mean   : 11.923  
 3rd Qu.:0.5964   3rd Qu.:357.552   3rd Qu.:0.5335   3rd Qu.:  9.227  
 Max.   :0.9033   Max.   :835.723   Max.   :0.9000   Max.   :222.230  
 NA's   :1062     NA's   :1062      NA's   :1062     NA's   :1062     
  glcm_entropy    glcm_dissimilarity glcm_second_moment glcm_correlation 
 Min.   :0.8716   Min.   :0.20       Min.   :0.0328     Min.   :-0.2170  
 1st Qu.:2.5584   1st Qu.:1.25       1st Qu.:0.0552     1st Qu.: 0.4883  
 Median :2.8202   Median :1.74       Median :0.0656     Median : 0.6663  
 Mean   :2.7152   Mean   :1.92       Mean   :0.0824     Mean   : 0.6301  
 3rd Qu.:2.9675   3rd Qu.:2.28       3rd Qu.:0.0896     3rd Qu.: 0.8141  
 Max.   :3.1911   Max.   :8.54       Max.   :0.5464     Max.   : 1.9621  
 NA's   :1062     NA's   :1062       NA's   :1062       NA's   :1062     
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{summary}\NormalTok{(glcm.gn)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
   glcm_mean      glcm_variance    glcm_homogeneity glcm_contrast   
 Min.   :0.1633   Min.   : 32.43   Min.   :0.0558   Min.   :  2.83  
 1st Qu.:0.3514   1st Qu.:132.04   1st Qu.:0.2381   1st Qu.: 11.90  
 Median :0.4052   Median :172.46   Median :0.2827   Median : 18.04  
 Mean   :0.4070   Mean   :182.00   Mean   :0.2890   Mean   : 22.21  
 3rd Qu.:0.4581   3rd Qu.:217.50   3rd Qu.:0.3316   3rd Qu.: 27.14  
 Max.   :0.7219   Max.   :515.49   Max.   :0.6325   Max.   :152.24  
 NA's   :1062     NA's   :1062     NA's   :1062     NA's   :1062    
  glcm_entropy   glcm_dissimilarity glcm_second_moment glcm_correlation 
 Min.   :1.848   Min.   :1.100      Min.   :0.0316     Min.   :-0.2328  
 1st Qu.:2.962   1st Qu.:2.640      1st Qu.:0.0456     1st Qu.: 0.1823  
 Median :3.056   Median :3.260      Median :0.0496     Median : 0.2995  
 Mean   :3.019   Mean   :3.396      Mean   :0.0531     Mean   : 0.3048  
 3rd Qu.:3.122   3rd Qu.:3.978      3rd Qu.:0.0560     3rd Qu.: 0.4207  
 Max.   :3.219   Max.   :9.180      Max.   :0.2640     Max.   : 1.5150  
 NA's   :1062    NA's   :1062       NA's   :1062       NA's   :1062     
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{summary}\NormalTok{(glcm.sg}\SpecialCharTok{{-}}\NormalTok{glcm.gn)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
   glcm_mean       glcm_variance     glcm_homogeneity  glcm_contrast   
 Min.   :-0.4333   Min.   :-300.53   Min.   :-0.2663   Min.   :-61.23  
 1st Qu.:-0.0767   1st Qu.: -53.92   1st Qu.: 0.0738   1st Qu.:-17.83  
 Median : 0.0570   Median :  41.72   Median : 0.1612   Median :-10.44  
 Mean   : 0.0450   Mean   :  60.15   Mean   : 0.1718   Mean   :-10.29  
 3rd Qu.: 0.1648   3rd Qu.: 158.54   3rd Qu.: 0.2617   3rd Qu.: -4.27  
 Max.   : 0.5175   Max.   : 614.20   Max.   : 0.6264   Max.   :129.99  
 NA's   :1062      NA's   :1062      NA's   :1062      NA's   :1062    
  glcm_entropy     glcm_dissimilarity glcm_second_moment glcm_correlation 
 Min.   :-2.0346   Min.   :-5.450     Min.   :-0.2080    Min.   :-0.6570  
 1st Qu.:-0.4862   1st Qu.:-2.220     1st Qu.: 0.0032    1st Qu.: 0.1530  
 Median :-0.2168   Median :-1.470     Median : 0.0152    Median : 0.3444  
 Mean   :-0.3037   Mean   :-1.475     Mean   : 0.0293    Mean   : 0.3253  
 3rd Qu.:-0.0557   3rd Qu.:-0.740     3rd Qu.: 0.0392    3rd Qu.: 0.5181  
 Max.   : 1.0973   Max.   : 4.170     Max.   : 0.4840    Max.   : 1.3634  
 NA's   :1062      NA's   :1062       NA's   :1062       NA's   :1062     
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot}\NormalTok{(glcm.sg)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/measures-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot}\NormalTok{(glcm.gn)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/measures-2.pdf}

}

\end{figure}

\hypertarget{sec-glcm-interp}{%
\subsubsection{Interpretation}\label{sec-glcm-interp}}

Each of the texture metrics quantifies some aspect of the texture. For a
thorough explanation see Hall-Beyer
(\protect\hyperlink{ref-hall-beyerGLCMTextureTutorial2017}{2017a}) and
Hall-Beyer
(\protect\hyperlink{ref-hall-beyerPracticalGuidelinesChoosing2017}{2017b}).
Here we examine a few of them.

\textbf{Mean} and \textbf{Variance} represent the overall inhomogeneity
of the window. The mean is the mean change in the selected shift(s) and
the variance is how variable are the changes.

\[\mu = \sum_{i,j = 0}^{N-1} i \cdot P_{i,j}\]

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{zlim }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(glcm.sg[[}\StringTok{"glcm\_mean"}\NormalTok{]]), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{), }
                     \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(glcm.gn[[}\StringTok{"glcm\_mean"}\NormalTok{]]), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{))}
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\FunctionTok{plot}\NormalTok{(glcm.sg[[}\StringTok{"glcm\_mean"}\NormalTok{]], }\AttributeTok{main =} \StringTok{"SoilGrids v2.0"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ zlim, }\AttributeTok{col=}\NormalTok{(sp}\SpecialCharTok{::}\FunctionTok{bpy.colors}\NormalTok{(}\DecValTok{32}\NormalTok{)))}
\FunctionTok{plot}\NormalTok{(glcm.gn[[}\StringTok{"glcm\_mean"}\NormalTok{]], }\AttributeTok{main =} \StringTok{"gNATSGO"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ zlim, }\AttributeTok{col=}\NormalTok{(sp}\SpecialCharTok{::}\FunctionTok{bpy.colors}\NormalTok{(}\DecValTok{32}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/unnamed-chunk-21-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Areas with the higher values have more and/or larger differences between
neighbours.

\[\sigma^2 = \sum_{i,j = 0}^{N-1} P_{i,j}\cdot (i - \mu)^2\]

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{zlim }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(glcm.sg[[}\StringTok{"glcm\_variance"}\NormalTok{]]), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{), }
                     \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(glcm.gn[[}\StringTok{"glcm\_variance"}\NormalTok{]]), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{))}
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\FunctionTok{plot}\NormalTok{(glcm.sg[[}\StringTok{"glcm\_variance"}\NormalTok{]], }\AttributeTok{main =} \StringTok{"SoilGrids v2.0"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ zlim, }\AttributeTok{col=}\NormalTok{(}\FunctionTok{cm.colors}\NormalTok{(}\DecValTok{32}\NormalTok{)))}
\FunctionTok{plot}\NormalTok{(glcm.gn[[}\StringTok{"glcm\_variance"}\NormalTok{]], }\AttributeTok{main =} \StringTok{"gNATSGO"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ zlim, }\AttributeTok{col=}\NormalTok{(}\FunctionTok{cm.colors}\NormalTok{(}\DecValTok{32}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/unnamed-chunk-22-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Im this case the SoilGrids map has more dispersion in the types of
changes.

\textbf{Contrast} is the amount of local variation in a window, with
emphasis (squared distance) on the off-diagonals of the GLCM, i.e.,
larger changes in the quanta level.

\[\sum_{i,j = 0}^{N-1} P_{i,j}\cdot (i - j)^2\]

where \(P_{i,j}\) is the proportion of the class \(i\) and \(j\)
co-occurrence in the window.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{zlim }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(glcm.sg[[}\StringTok{"glcm\_contrast"}\NormalTok{]]), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{), }
                     \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(glcm.gn[[}\StringTok{"glcm\_contrast"}\NormalTok{]]), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{))}
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\FunctionTok{plot}\NormalTok{(glcm.sg[[}\StringTok{"glcm\_contrast"}\NormalTok{]], }\AttributeTok{main =} \StringTok{"SoilGrids v2.0"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ zlim, }\AttributeTok{col=}\NormalTok{(}\FunctionTok{topo.colors}\NormalTok{(}\DecValTok{32}\NormalTok{)))}
\FunctionTok{plot}\NormalTok{(glcm.gn[[}\StringTok{"glcm\_contrast"}\NormalTok{]], }\AttributeTok{main =} \StringTok{"gNATSGO"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ zlim, }\AttributeTok{col=}\NormalTok{(}\FunctionTok{topo.colors}\NormalTok{(}\DecValTok{32}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/unnamed-chunk-23-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

We see that SoilGrids has very little contrast across most of the map,
whereas gNATSGO has stronger contrasts. Both have ``hot spots'' of high
contrast, i.e., areas in the map with a relatively wide range of pH
values. Note that this shows that the assumption of second-order
stationarity used in the variogram analysis Section~\ref{sec-vgm} is
definitely not correct.

A variant is the \textbf{dissimilarity}, where the weights are linear
away from the diagonal, rather than quadratic:

\[\sum_{i,j = 0}^{N-1} P_{i,j}\cdot |i - j|\]

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{zlim }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(glcm.sg[[}\StringTok{"glcm\_dissimilarity"}\NormalTok{]]), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{), }
                     \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(glcm.gn[[}\StringTok{"glcm\_dissimilarity"}\NormalTok{]]), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{))}
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\FunctionTok{plot}\NormalTok{(glcm.sg[[}\StringTok{"glcm\_dissimilarity"}\NormalTok{]], }\AttributeTok{main =} \StringTok{"SoilGrids v2.0"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ zlim, }\AttributeTok{col=}\NormalTok{(}\FunctionTok{topo.colors}\NormalTok{(}\DecValTok{32}\NormalTok{)))}
\FunctionTok{plot}\NormalTok{(glcm.gn[[}\StringTok{"glcm\_dissimilarity"}\NormalTok{]], }\AttributeTok{main =} \StringTok{"gNATSGO"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ zlim, }\AttributeTok{col=}\NormalTok{(}\FunctionTok{topo.colors}\NormalTok{(}\DecValTok{32}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/unnamed-chunk-24-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\textbf{Entropy} is a measure of information within a window. It
accounts for the number of different levels in the window (the others
will have ``probability'' zero) and their relative frequencies. More
classes and more even distribution of classes results in increased
entropy. This can be thought of as ``lack of information''.

\[\sum_{i,j = 0}^{N-1} P_{i,j}\cdot -\ln(P_{i,j})\]

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{zlim }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(glcm.sg[[}\StringTok{"glcm\_entropy"}\NormalTok{]]), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{), }
                     \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(glcm.gn[[}\StringTok{"glcm\_entropy"}\NormalTok{]]), }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{))}
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\FunctionTok{plot}\NormalTok{(glcm.sg[[}\StringTok{"glcm\_entropy"}\NormalTok{]], }\AttributeTok{main =} \StringTok{"SoilGrids v2.0"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ zlim, }\AttributeTok{col=}\NormalTok{(}\FunctionTok{topo.colors}\NormalTok{(}\DecValTok{32}\NormalTok{)))}
\FunctionTok{plot}\NormalTok{(glcm.gn[[}\StringTok{"glcm\_entropy"}\NormalTok{]], }\AttributeTok{main =} \StringTok{"gNATSGO"}\NormalTok{, }
     \AttributeTok{range =}\NormalTok{ zlim, }\AttributeTok{col=}\NormalTok{(}\FunctionTok{topo.colors}\NormalTok{(}\DecValTok{32}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/unnamed-chunk-25-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

In this case there is quite low entropy in some of the hilly areas of
the SoilGrids maps. Most windows of the gNATSGO map have high entropy,
at this window size.

Challenge: compute the GLCM statistics for different window sizes.

\hypertarget{sec-classified}{%
\section{Characterizing patterns -- Classified}\label{sec-classified}}

The spatial unit of conventional (legacy) maps is the polygon, not the
grid cell. These maps show a discrete number of legend entries
(classes), each with one to many polygons. In the soil survey context
these are called \textbf{mapping units}, and generally are soil classes,
possibly with some landscape features (e.g., erosion class, slope class)
as part of the definition. Some mapping units may represent water bodies
and various other kinds of non-soil.

Figure~\ref{fig-anderson} shows a typical polygon map from a legacy
``land condition'' survey
(\protect\hyperlink{ref-soilconservationservicePhysicalLandConditions1951}{Soil
Conservation Service, 1951}). All polygons with the same label (e.g.,
``337D22'') refer to a single mapping unit, in this case composed of an
erosion class, slope class, and soil type.

\begin{figure}

{\centering \includegraphics{./figs/UpperSavannahSCD_Sheet4_colour_NWcorner.png}

}

\caption{\label{fig-anderson}Land condition, Anderson County SC (USA)}

\end{figure}

It's easiest to work with maps already in digital format. The area of
the legacy map has been updated and digitized, see
Figure~\ref{fig-anderson-wss}. These polygons can be downloaded in
various GIS formats, see Section~\ref{sec-import-gssurgo}, above

\begin{figure}

{\centering \includegraphics{./figs/WebSoilSurvey_AndersonCountyNW.png}

}

\caption{\label{fig-anderson-wss}Anderson County SC (USA)}

\end{figure}

But here we continue with the continuous property maps of a single
property. To use these techniques on continuous property maps, the maps
must be \textbf{sliced} (discretized) into classes. There are several
choices:

\begin{itemize}
\tightlist
\item
  meaningful limits, matching some thresholds known to be important for
  a soil function;
\item
  equal intervals;
\item
  histogram equalization.
\end{itemize}

For equal intervals or histogram equalization, the cutpoints should be
the same for all maps, and therefore derived from their combined
distribution of values. We illustrate the process here, but do not use
it for the landscape metrics examples later on in the tutorial.

\hypertarget{sec-hist-equal}{%
\subsection{Classifying by histogram
equalization}\label{sec-hist-equal}}

This section shows how to classify by histogram equalization; the
results will not be used later in the tutorial. Instead, we will use
meaningful limits (see Section~\ref{sec-mean-limit}) to slice the map.

Task: Slice the two maps by histogram equalization

First, compute the histogram equalization and display the limits on a
histogram plot:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{n.class }\OtherTok{\textless{}{-}} \DecValTok{8}
\CommentTok{\# combined values}
\NormalTok{values.all }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\FunctionTok{values}\NormalTok{(gn.utm),}
                \FunctionTok{values}\NormalTok{(sg.utm))}
\NormalTok{values.all.sort }\OtherTok{\textless{}{-}} \FunctionTok{sort}\NormalTok{(values.all)}
\CommentTok{\#}
\NormalTok{n }\OtherTok{\textless{}{-}} \FunctionTok{length}\NormalTok{(values.all) }\SpecialCharTok{{-}} \FunctionTok{sum}\NormalTok{(}\FunctionTok{is.na}\NormalTok{(values.all))}
\NormalTok{(cut.positions }\OtherTok{\textless{}{-}} \FunctionTok{round}\NormalTok{(n}\SpecialCharTok{/}\NormalTok{n.class))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 2237
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(cuts }\OtherTok{\textless{}{-}}\NormalTok{ values.all.sort[cut.positions }\SpecialCharTok{*} \DecValTok{1}\SpecialCharTok{:}\NormalTok{(n.class}\DecValTok{{-}1}\NormalTok{)])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 4.969114 5.210513 5.425628 5.573456 5.721922 5.874367 6.078163
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{hist}\NormalTok{(values.all, }\AttributeTok{breaks=}\DecValTok{36}\NormalTok{, }\AttributeTok{main=}\StringTok{"Histogram equalization"}\NormalTok{,}
     \AttributeTok{xlab =} \StringTok{"pH"}\NormalTok{)}
\FunctionTok{abline}\NormalTok{(}\AttributeTok{v=}\NormalTok{cuts, }\AttributeTok{col=}\StringTok{"blue"}\NormalTok{, }\AttributeTok{lwd=}\DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/histo-equal-1.pdf}

}

\end{figure}

Q: How well do these represent the distributions on the two maps
individually?

To answer this, compare their histograms with the equalization slices.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\FunctionTok{hist}\NormalTok{(}\FunctionTok{values}\NormalTok{(gn.utm), }\AttributeTok{breaks=}\DecValTok{36}\NormalTok{, }\AttributeTok{main=}\StringTok{"gNATSGO"}\NormalTok{,}
     \AttributeTok{xlab =} \StringTok{"pH"}\NormalTok{)}
\FunctionTok{abline}\NormalTok{(}\AttributeTok{v=}\NormalTok{cuts, }\AttributeTok{col=}\StringTok{"blue"}\NormalTok{, }\AttributeTok{lwd=}\DecValTok{2}\NormalTok{)}
\FunctionTok{hist}\NormalTok{(}\FunctionTok{values}\NormalTok{(sg.utm), }\AttributeTok{breaks=}\DecValTok{36}\NormalTok{, }\AttributeTok{main=}\StringTok{"SoilGrids v2.0"}\NormalTok{,}
     \AttributeTok{xlab =} \StringTok{"pH"}\NormalTok{)}
\FunctionTok{abline}\NormalTok{(}\AttributeTok{v=}\NormalTok{cuts, }\AttributeTok{col=}\StringTok{"blue"}\NormalTok{, }\AttributeTok{lwd=}\DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/histo-equal-one-by-one-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Task: slice the maps and display with a common colour ramp.

Find the cutpoints and set up the colour ramp:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(zlim }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\FunctionTok{min}\NormalTok{(values.all, }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{),}
                \FunctionTok{max}\NormalTok{(values.all, }\AttributeTok{na.rm=}\ConstantTok{TRUE}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 4.462665 7.404359
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(cut.names }\OtherTok{\textless{}{-}} \FunctionTok{cut}\NormalTok{(zlim, }\AttributeTok{breaks=}\FunctionTok{c}\NormalTok{(zlim[}\DecValTok{1}\NormalTok{], cuts, zlim[}\DecValTok{2}\NormalTok{]),}
                  \AttributeTok{ordered\_result=}\ConstantTok{TRUE}\NormalTok{, }\AttributeTok{include.lowest =} \ConstantTok{TRUE}\NormalTok{)) }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] [4.46,4.97] (6.08,7.4] 
8 Levels: [4.46,4.97] < (4.97,5.21] < (5.21,5.43] < ... < (6.08,7.4]
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# make sure lowest value is included}
\CommentTok{\#}
\CommentTok{\# common colour ramp}
\NormalTok{color.ramp }\OtherTok{\textless{}{-}} \FunctionTok{bpy.colors}\NormalTok{(n.class}\SpecialCharTok{+}\DecValTok{1}\NormalTok{)}
\CommentTok{\#}
\NormalTok{(cuts }\OtherTok{\textless{}{-}} \FunctionTok{round}\NormalTok{(}\FunctionTok{c}\NormalTok{(zlim[}\DecValTok{1}\NormalTok{], cuts, zlim[}\DecValTok{2}\NormalTok{]),}\DecValTok{2}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 4.46 4.97 5.21 5.43 5.57 5.72 5.87 6.08 7.40
\end{verbatim}

Slice the maps:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gn.class }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{classify}\NormalTok{(gn.utm, }\AttributeTok{rcl=}\NormalTok{ cuts)}
\CommentTok{\# gn.class \textless{}{-} as.factor(gn.class)}
\FunctionTok{table}\NormalTok{(}\FunctionTok{values}\NormalTok{(gn.class))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}

   0    1    2    3    4    5    6    7 
  38  387  775 1057 1216 1507 1828 2138 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{names}\NormalTok{(gn.class) }\OtherTok{\textless{}{-}} \StringTok{"class"}
\NormalTok{sg.class }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{classify}\NormalTok{(sg.utm, }\AttributeTok{rcl=}\NormalTok{ cuts)}
\FunctionTok{table}\NormalTok{(}\FunctionTok{values}\NormalTok{(sg.class))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}

   0    1    2    3    4    5    6    7 
2205 1839 1524 1065 1058  692  481   83 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{names}\NormalTok{(sg.class) }\OtherTok{\textless{}{-}} \StringTok{"class"}
\end{Highlighting}
\end{Shaded}

Display the maps:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{2}\NormalTok{))}
\NormalTok{.l }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(gn.class), }\AttributeTok{na.rm=}\ConstantTok{TRUE}\NormalTok{)}
\NormalTok{terra}\SpecialCharTok{::}\FunctionTok{plot}\NormalTok{(gn.class,}
            \AttributeTok{col=}\NormalTok{color.ramp[.l[}\DecValTok{1}\NormalTok{]}\SpecialCharTok{:}\NormalTok{.l[}\DecValTok{2}\NormalTok{]}\SpecialCharTok{+}\DecValTok{1}\NormalTok{], }\AttributeTok{type=}\StringTok{"classes"}\NormalTok{,}
            \AttributeTok{main=}\StringTok{"gNATSGO"}\NormalTok{)}
\NormalTok{.l }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(sg.class), }\AttributeTok{na.rm=}\ConstantTok{TRUE}\NormalTok{)}
\NormalTok{terra}\SpecialCharTok{::}\FunctionTok{plot}\NormalTok{(sg.class,}
            \AttributeTok{col=}\NormalTok{color.ramp[.l[}\DecValTok{1}\NormalTok{]}\SpecialCharTok{:}\NormalTok{.l[}\DecValTok{2}\NormalTok{]}\SpecialCharTok{+}\DecValTok{1}\NormalTok{], }\AttributeTok{type=}\StringTok{"classes"}\NormalTok{,}
            \AttributeTok{main=}\StringTok{"SG2"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show.classified-hist-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Q: Describe the patterns of the two maps

Q: How would these change with different class numbers or limits?

Q: If classifying by histogram equalization, should the two maps be
compared with the same limits or each with their own limits? You are
welcome to experiment.

\hypertarget{sec-mean-limit}{%
\subsection{Classifying by meaningful limits}\label{sec-mean-limit}}

For soil properties we usually have limits that correspond to
approximate thresholds in land use. In the case of pH, we can refer to
extension or crop consultant publications, or environmental models.
Unlike in histogram equalization, the number of classes depends on the
user requirements.

For example, the \href{https://www.nnyagdev.org/PDF/SoilpH.pdf}{Cornell
pH test kit} has a ``Wide Range Kit'' measuring the soil pH over the
range of 4.0--8.6, in increments of 0.2 for an experienced user. So,
here we will slice the map in increments of 0.2 pH.

Task: slice the maps and display with a common colour ramp.

Find the combined range and divide into one-decimal classes of 0.2 pH,
starting and ending on even units of 0.2.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{range.all }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(gn.utm),}
                   \FunctionTok{values}\NormalTok{(sg.utm),}
                   \AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{lim.low }\OtherTok{\textless{}{-}} \FunctionTok{floor}\NormalTok{(}\DecValTok{10}\SpecialCharTok{*}\NormalTok{range.all[}\DecValTok{1}\NormalTok{])}\SpecialCharTok{/}\DecValTok{10}
\NormalTok{lim.low }\OtherTok{\textless{}{-}} \FunctionTok{ifelse}\NormalTok{((lim.low }\SpecialCharTok{\%\%}\NormalTok{ .}\DecValTok{2}\NormalTok{) }\SpecialCharTok{!=} \DecValTok{0}\NormalTok{, lim.low }\SpecialCharTok{{-}} \FloatTok{0.1}\NormalTok{, lim.low)}
\NormalTok{lim.high }\OtherTok{\textless{}{-}} \FunctionTok{ceiling}\NormalTok{(}\DecValTok{10}\SpecialCharTok{*}\NormalTok{range.all[}\DecValTok{2}\NormalTok{])}\SpecialCharTok{/}\DecValTok{10}
\NormalTok{lim.high }\OtherTok{\textless{}{-}} \FunctionTok{ifelse}\NormalTok{((lim.high }\SpecialCharTok{\%\%}\NormalTok{ .}\DecValTok{2}\NormalTok{) }\SpecialCharTok{!=} \DecValTok{0}\NormalTok{, lim.high }\SpecialCharTok{+} \FloatTok{0.1}\NormalTok{, lim.high)}
\NormalTok{(cuts }\OtherTok{\textless{}{-}} \FunctionTok{seq}\NormalTok{(lim.low, lim.high, }\AttributeTok{by =} \FloatTok{0.2}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] 4.4 4.6 4.8 5.0 5.2 5.4 5.6 5.8 6.0 6.2 6.4 6.6 6.8 7.0 7.2 7.4 7.6
\end{verbatim}

Slice the maps:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gn.class }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{classify}\NormalTok{(gn.utm, }\AttributeTok{rcl=}\NormalTok{ cuts)}
\CommentTok{\# gn.class \textless{}{-} as.factor(gn.class)}
\FunctionTok{table}\NormalTok{(}\FunctionTok{values}\NormalTok{(gn.class))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}

   2    3    4    5    6    7    8    9   10   11   12   13   15 
  76  329  629 1452 1727 1992 1348  681  402  253   53    4    1 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{names}\NormalTok{(gn.class) }\OtherTok{\textless{}{-}} \StringTok{"class"}
\NormalTok{sg.class }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{classify}\NormalTok{(sg.utm, }\AttributeTok{rcl=}\NormalTok{ cuts)}
\FunctionTok{table}\NormalTok{(}\FunctionTok{values}\NormalTok{(sg.class))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}

   0    1    2    3    4    5    6    7    8    9 
 167  902 1376 1531 1362 1534 1235  653  171   16 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{names}\NormalTok{(sg.class) }\OtherTok{\textless{}{-}} \StringTok{"class"}
\end{Highlighting}
\end{Shaded}

Display them:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{2}\NormalTok{))}
\NormalTok{color.ramp }\OtherTok{\textless{}{-}} \FunctionTok{bpy.colors}\NormalTok{(}\FunctionTok{length}\NormalTok{(cuts))}
\NormalTok{.l }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(gn.class), }\AttributeTok{na.rm=}\ConstantTok{TRUE}\NormalTok{)}
\NormalTok{terra}\SpecialCharTok{::}\FunctionTok{plot}\NormalTok{(gn.class,}
            \AttributeTok{col=}\NormalTok{color.ramp[.l[}\DecValTok{1}\NormalTok{]}\SpecialCharTok{:}\NormalTok{.l[}\DecValTok{2}\NormalTok{]}\SpecialCharTok{+}\DecValTok{1}\NormalTok{], }\AttributeTok{type=}\StringTok{"classes"}\NormalTok{,}
            \AttributeTok{main=}\StringTok{"gNATSGO pH 0.2 units"}\NormalTok{)}
\NormalTok{.l }\OtherTok{\textless{}{-}} \FunctionTok{range}\NormalTok{(}\FunctionTok{values}\NormalTok{(sg.class), }\AttributeTok{na.rm=}\ConstantTok{TRUE}\NormalTok{)}
\NormalTok{terra}\SpecialCharTok{::}\FunctionTok{plot}\NormalTok{(sg.class,}
            \AttributeTok{col=}\NormalTok{color.ramp[.l[}\DecValTok{1}\NormalTok{]}\SpecialCharTok{:}\NormalTok{.l[}\DecValTok{2}\NormalTok{]}\SpecialCharTok{+}\DecValTok{1}\NormalTok{], }\AttributeTok{type=}\StringTok{"classes"}\NormalTok{,}
            \AttributeTok{main=}\StringTok{"SG2 pH 0.2 units"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show.classified-cuts-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

These maps are now showing meaningful landscape units, from the point of
view of land use and soil processes, on the same scale.

Q: Describe the patterns of the two maps

Q: How would these maps change with wider or narrower class intervals?
You are welcome to experiment!

\hypertarget{sec-crosstab}{%
\subsection{Cross-classification matrix}\label{sec-crosstab}}

As in classic remote sensing analysis, we can compare one of the maps
(SG2) to the ``reference'' map (gNATSGO). This is implemented in the
\texttt{diffeR} package.

Create the cross-classification matrix showing pixel counts:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{dim}\NormalTok{(ccm }\OtherTok{\textless{}{-}}\NormalTok{ diffeR}\SpecialCharTok{::}\FunctionTok{crosstabm}\NormalTok{(sg.class, gn.class))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 15 15
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{sum}\NormalTok{(ccm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 8947
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{prod}\NormalTok{(}\FunctionTok{dim}\NormalTok{(gn.class))  }\CommentTok{\# total pixels, includes some NA}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 9408
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ccm[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{5}\NormalTok{, }\DecValTok{1}\SpecialCharTok{:}\DecValTok{5}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
          (4.4–4.6] (4.6–4.8] (4.8–5] (5–5.2] (5.2–5.4]
(4.4–4.6]         0         0       6       9        24
(4.6–4.8]         0         0      19      81       117
(4.8–5]           0         0      20      88       122
(5–5.2]           0         0      13      66       143
(5.2–5.4]         0         0      11      61        97
\end{verbatim}

This can also show percentages:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{dim}\NormalTok{(ccm.p }\OtherTok{\textless{}{-}}\NormalTok{ diffeR}\SpecialCharTok{::}\FunctionTok{crosstabm}\NormalTok{(sg.class, gn.class, }\AttributeTok{percent =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 15 15
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{sum}\NormalTok{(ccm.p)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 100
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ccm.p[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{5}\NormalTok{, }\DecValTok{1}\SpecialCharTok{:}\DecValTok{5}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
          (4.4–4.6] (4.6–4.8]    (4.8–5]   (5–5.2] (5.2–5.4]
(4.4–4.6]         0         0 0.06706158 0.1005924 0.2682463
(4.6–4.8]         0         0 0.21236169 0.9053314 1.3077009
(4.8–5]           0         0 0.22353862 0.9835699 1.3635856
(5–5.2]           0         0 0.14530010 0.7376774 1.5983011
(5.2–5.4]         0         0 0.12294624 0.6817928 1.0841623
\end{verbatim}

Analyze its sources of disagreement, according to the classification of
Pontius \& Santacruz
(\protect\hyperlink{ref-pontiusQuantityExchangeShift2014}{2014}). (See
also Pontius \& Millones
(\protect\hyperlink{ref-PontiusDeathKappabirth2011}{2011}) for an easier
introduction to the concepts of quantity and allocation disagreement.)

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(dt }\OtherTok{\textless{}{-}}\NormalTok{ diffeR}\SpecialCharTok{::}\FunctionTok{diffTablej}\NormalTok{(ccm))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
    Category Omission Agreement Comission Quantity Exchange Shift
1  (4.4–4.6]        0         0       167      167        0     0
2  (4.6–4.8]        0         0       902      902        0     0
3    (4.8–5]       56        20      1356     1300       62    50
4    (5–5.2]      263        66      1465     1202      196   330
5  (5.2–5.4]      532        97      1265      733      396   668
6  (5.4–5.6]     1218       234      1300       82      626  1810
7  (5.6–5.8]     1504       223      1012      492      600  1424
8    (5.8–6]     1857       135       518     1339      420   616
9    (6–6.2]     1321        27       144     1177      202    86
10 (6.2–6.4]      681         0        16      665       18    14
11 (6.4–6.6]      402         0         0      402        0     0
12 (6.6–6.8]      253         0         0      253        0     0
13   (7–7.2]        4         0         0        4        0     0
14   (6.8–7]       53         0         0       53        0     0
15 (7.4–7.6]        1         0         0        1        0     0
16   Overall     8145       802      8145     4386     1260  2499
\end{verbatim}

\begin{itemize}
\tightlist
\item
  \emph{Agreement} means the pixels at the same location are in the same
  class.
\item
  \emph{Omission} means that test map does not find the ``correct''
  class, i.e., the given \texttt{Category} that is found the reference
  map at the pixel.
\item
  \emph{Commission} means that test map predicts a class of the given
  \texttt{Category} at the pixel that is not found in the reference map
  at that pixel.
\item
  \emph{Quantity} disagreement is the sum of Commission and Omission
  errors, less the \texttt{Exchange} and \texttt{Shift} (see next),
  i.e., the errors caused by not having the same number of pixels in the
  category.
\item
  \emph{Exchange} disagreement is the number of pixels where a
  transition from class \emph{i} to class \emph{j} in some pixels,
  balanced by a transition from class \emph{j} to category \emph{i} in
  an identical number of other pixels. So the quantity does not change,
  but the location in the map does.
\item
  \emph{Shift} disagreement is the error that remains after subtracting
  quantity difference and exchange from the overall difference. These
  are unbalanced transitions.
\end{itemize}

The total error is the sum of Quantity, Exchange and Shift. These have
different interpretations and possible corrections.

\begin{itemize}
\tightlist
\item
  \emph{Quantity}: the soil class either over- or under-predicted by the
  model. To find out which, compare the Omission and Commission. If
  Omission is larger, the class is under-predicted, and vice-versa. That
  is, the model does not ``find'' this class as much as the reference
  map suggests.
\end{itemize}

In this comparison for the (5.6--5.8{]} pH class, only 223 out of a
total 2739 pixels in this class on the reference map are correctly
classified in the test map.

The Omission and Commission errors for this class are 1504 and 1012,
respectively, so there is more omission than commission. So this class
is under-predicted by SoilGrids, taking gNATSGO as the reference. For
the low-pH classes SoilGrids consistently over-predicts, by this class
it consistently under-predicts.

There is quantity difference of 492, an exchange (same classes, wrong
places, balanced) of 600, leaving a very large number of wrong classes
1424 not accounted for by either of these.

Notice that for the entire map the errors of Omission and Commission
must balance: omission in one class will result commission in some other
class.

Graphically, for all classes in the two maps:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{overallComponentsPlot}\NormalTok{(}\AttributeTok{ctmatrix =}\NormalTok{ ccm, }\AttributeTok{units =} \StringTok{"pixels"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics[width=0.6\textwidth,height=\textheight]{PatternAnalysisWorkshopTutorial_files/figure-pdf/components-plot-1.pdf}

}

\end{figure}

This does not show any spatial pattern, but does show class
disagreement. We now look at a matrix that reveals adjaceny of the
classes.

Note: The cross-classification can be applied directly to
continuously-valued maps, but in this case it converts numeric values to
integers that are taken to represent a class. This is a quick way to get
equal-interval classes. Whether these are meaningful for your
application is a separate question.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# whole pH units}
\FunctionTok{print}\NormalTok{(ccm.c }\OtherTok{\textless{}{-}} \FunctionTok{crosstabm}\NormalTok{(sg.utm, gn.utm))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  4    5    6   7
4 0    4    1   0
5 0 1300 4738  74
6 0  233 2222 375
7 0    0    0   0
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{diffTablej}\NormalTok{(ccm.c)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  Category Omission Agreement Comission Quantity Exchange Shift
1        4        0         0         5        5        0     0
2        5      237      1300      4812     4575      466     8
3        6     4739      2222       608     4131      466   750
4        7      449         0         0      449        0     0
5  Overall     5425      3522      5425     4580      466   379
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# to one decimal pH unit}
\CommentTok{\# show the first four reference classes}
\FunctionTok{print}\NormalTok{((ccm.c10 }\OtherTok{\textless{}{-}} \FunctionTok{crosstabm}\NormalTok{(sg.utm}\SpecialCharTok{*}\DecValTok{10}\NormalTok{, gn.utm}\SpecialCharTok{*}\DecValTok{10}\NormalTok{))[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{5}\NormalTok{, ])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
   45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 70 69
45  0  0  0  0  0  7  3  5  5  4 10 10  5  3  2  1  0  1  0  0  0  0  0  0  0
46  0  0  0  0  0  9 12 17 19 20 48 37 39 18 20 22 14  3  8 10 10  0  0  0  0
47  0  0  0  0  2 17 24 20 37 40 53 64 42 38 44 37 15 24 12  6  7  0  0  0  0
48  0  0  0  1  1 18 23 26 36 34 52 66 51 49 46 44 30 26  8  4  8  1  0  0  0
49  0  0  0  0  2 12 28 18 24 24 66 73 58 66 84 85 39 47 33 13 10  1  1  0  0
   68 71 74
45  0  0  0
46  0  0  0
47  0  0  0
48  0  0  0
49  0  0  0
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{diffTablej}\NormalTok{(ccm.c10)[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{5}\NormalTok{, ]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  Category Omission Agreement Comission Quantity Exchange Shift
1       45        0         0        56       56        0     0
2       46        0         0       306      306        0     0
3       47        0         0       482      482        0     0
4       48        9         1       523      514       18     0
5       49       17         2       682      665       28     6
\end{verbatim}

\hypertarget{sec-coma}{%
\subsection{Co-occurrence matrices}\label{sec-coma}}

One question for a classified map is which classes tend to be adjacent
to each other. In the case of the pH map, we might expect adjacent
classes to be in the pH sequence, but maybe not -- there may be abrupt
transitions of parent materials, for example.

A co-occurrence \emph{matrix} counts all the pairs of adjacent cells for
each category in a local landscape, as a cross-classification matrix.

Task: Compute the co-occurrence \emph{matrices}, using Queen's Case
neighbours (i.e., diagonal links are considered).

Co-occurrence vectors are computed with the \texttt{lsp\_signature}
function of the \texttt{motif} package, specifyin \texttt{coma} =
co-occurrence matrix as the signature.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{coma.gn }\OtherTok{\textless{}{-}} \FunctionTok{lsp\_signature}\NormalTok{(gn.class, }\AttributeTok{type=}\StringTok{"coma"}\NormalTok{, }\AttributeTok{neighbourhood =} \DecValTok{8}\NormalTok{)}
\FunctionTok{print}\NormalTok{(coma.gn.matrix }\OtherTok{\textless{}{-}} \FunctionTok{as.matrix}\NormalTok{(coma.gn}\SpecialCharTok{$}\NormalTok{signature)[[}\DecValTok{1}\NormalTok{]])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
     3   4    5    6    7    8    9   10  11  12  13 14 16
3  148 132  103   76   52   38   13   10   0   0   0  0  0
4  132 608  547  539  368  233   79   30   7   0   0  0  0
5  103 547 1110 1316  893  571  214  114  35   7   0  0  0
6   76 539 1316 3780 2651 1647  784  338 155  61   8  2  0
7   52 368  893 2651 3560 3337 1613  662 274  95   9  8  0
8   38 233  571 1647 3337 5224 2902 1018 470 156  33  8  2
9   13  79  214  784 1613 2902 2764 1210 670 234  41  2  1
10  10  30  114  338  662 1018 1210  948 585 295  63  2  2
11   0   7   35  155  274  470  670  585 522 326  46  5  2
12   0   0    7   61   95  156  234  295 326 642 122  2  1
13   0   0    0    8    9   33   41   63  46 122  78  1  0
14   0   0    0    2    8    8    2    2   5   2   1  2  0
16   0   0    0    0    0    2    1    2   2   1   0  0  0
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{sum}\NormalTok{(}\FunctionTok{diag}\NormalTok{(coma.gn.matrix))}\SpecialCharTok{/}\FunctionTok{sum}\NormalTok{(coma.gn.matrix)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 0.277633
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{coma.sg }\OtherTok{\textless{}{-}} \FunctionTok{lsp\_signature}\NormalTok{(sg.class, }\AttributeTok{type=}\StringTok{"coma"}\NormalTok{, }\AttributeTok{neighbourhood =} \DecValTok{8}\NormalTok{)}
\FunctionTok{print}\NormalTok{(coma.sg.matrix }\OtherTok{\textless{}{-}} \FunctionTok{as.matrix}\NormalTok{(coma.sg}\SpecialCharTok{$}\NormalTok{signature)[[}\DecValTok{1}\NormalTok{]])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
     1    2    3    4    5    6    7    8   9 10
1  904  376   14    0    0    0    0    0   0  0
2  376 5330 1269  139   18    0    0    0   0  0
3   14 1269 7088 2222  287   18    0    0   0  0
4    0  139 2222 6912 2400  378   19    0   0  0
5    0   18  287 2400 5008 2570  284   17   0  0
6    0    0   18  378 2570 6330 2396  118   6  0
7    0    0    0   19  284 2396 5616 1192  45  0
8    0    0    0    0   17  118 1192 3262 464  2
9    0    0    0    0    0    6   45  464 714 79
10   0    0    0    0    0    0    0    2  79 36
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{sum}\NormalTok{(}\FunctionTok{diag}\NormalTok{(coma.sg.matrix))}\SpecialCharTok{/}\FunctionTok{sum}\NormalTok{(coma.sg.matrix)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 0.5900381
\end{verbatim}

Q: Describe the differences in the co-occurrence structure. What does
this imply for the spatial pattern?

We see that indeed in this case most adjacencies are within one or at
most two classes. The gNATSGO map has more multiple-class adjacencies
than does SoilGrids, due to its finer spatial pattern. The SoilGrids map
has well over half of the adjacencies in the same class, whereas gNATSGO
has about a quarter.

\hypertarget{sec-cove}{%
\subsection{Co-occurrence vectors}\label{sec-cove}}

The \textbf{Co-occurrence vector} ``COVE'' proposed by Nowosad \&
Stepinski
(\protect\hyperlink{ref-nowosadSpatialAssociationRegionalizations2018}{2018a})
summarizes the \emph{entire adjacency structure} of a map and can be
used to compare map structures. This is a normalized form of the
co-occurrence matrix (see the previous section). Normalization means the
matrix sums to 1, and so is independent of the number of grid cells in
the map. Therefore this vector can be considered as a probability vector
for the co-occurrence of different classes.

Task: Compute the co-occurrence \emph{vectors}, using Queen's Case
neighbours.

Co-occurrence vectors are computed with the \texttt{lsp\_signature}
function of the \texttt{motif} package, specifying \texttt{cove}
(normalized co-occurrence vector) as the signature. These will be used
to \href{@compare-cove}{compare the maps}, below.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# normalized co{-}occurence vector 8 x 8}
\NormalTok{cove.gn }\OtherTok{\textless{}{-}} \FunctionTok{lsp\_signature}\NormalTok{(gn.class, }\AttributeTok{type=}\StringTok{"cove"}\NormalTok{, }\AttributeTok{neighbourhood =} \DecValTok{8}\NormalTok{)}
\NormalTok{cove.sg }\OtherTok{\textless{}{-}} \FunctionTok{lsp\_signature}\NormalTok{(sg.class, }\AttributeTok{type=}\StringTok{"cove"}\NormalTok{, }\AttributeTok{neighbourhood =} \DecValTok{8}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{incove}{%
\subsection{Integrated co-occurrence vector}\label{incove}}

An \emph{integrated} co-occurrence vector considers \emph{several input
layers}, for example, representing different soil properties of the same
area.

To examine this we need another soil property map. Let's use SG2 silt of
the 0--5\textasciitilde cm layer. We process this as we did for the pH
map. Here the ``meaningful limits'' for silt content are 5\% intervals.
Since the SG2 map is expressed in \(\mathrm{g} \; \mathrm{kg}^{-1}\),
these are intervals of 50 \(\mathrm{g} \; \mathrm{kg}^{-1}\).

Task: Import and process the silt concentration 0-5 cm SoilGrids
product.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(sg.silt }\OtherTok{\textless{}{-}} \FunctionTok{rast}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(file.dir,}
                        \StringTok{"SoilGrids250/lat4243\_lon{-}77{-}76/silt\_0{-}5cm\_mean.tif"}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
class       : SpatRaster 
dimensions  : 426, 426, 1  (nrow, ncol, nlyr)
resolution  : 0.002349867, 0.002349867  (x, y)
extent      : -77.00071, -75.99967, 41.99918, 43.00022  (xmin, xmax, ymin, ymax)
coord. ref. : lon/lat WGS 84 (EPSG:4326) 
source      : silt_0-5cm_mean.tif 
name        : silt_0-5cm_mean 
min value   :          0.0000 
max value   :        752.4755 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.silt }\OtherTok{\textless{}{-}} \FunctionTok{crop}\NormalTok{(sg.silt, }\FunctionTok{ext}\NormalTok{(ext.crop))}
\FunctionTok{values}\NormalTok{(sg.silt) }\OtherTok{\textless{}{-}} \FunctionTok{values}\NormalTok{(sg.silt)}\SpecialCharTok{/}\DecValTok{10} \CommentTok{\# convert from ppt to \%}
\NormalTok{sg.silt.utm }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{project}\NormalTok{(sg.silt, epsg.utm, }
                         \AttributeTok{res =} \FunctionTok{c}\NormalTok{(}\DecValTok{250}\NormalTok{, }\DecValTok{250}\NormalTok{), }\AttributeTok{method =} \StringTok{"bilinear"}\NormalTok{)}
\NormalTok{sg.silt.utm }\OtherTok{\textless{}{-}} \FunctionTok{resample}\NormalTok{(sg.silt.utm, sg.utm) }\CommentTok{\# make extents identical}
\NormalTok{cuts }\OtherTok{\textless{}{-}} \FunctionTok{seq}\NormalTok{(}\DecValTok{10}\NormalTok{, }\DecValTok{90}\NormalTok{, }\AttributeTok{by =} \DecValTok{5}\NormalTok{) }
\NormalTok{sg.silt.class }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{classify}\NormalTok{(sg.silt.utm, }\AttributeTok{rcl=}\NormalTok{ cuts)}
\FunctionTok{table}\NormalTok{(}\FunctionTok{values}\NormalTok{(sg.silt.class))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}

   0    1    2    3    4    5    6    7    8    9   10 
   2    1    5   12   11   12  166 2254 5330 1331    2 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{names}\NormalTok{(sg.silt.class) }\OtherTok{\textless{}{-}} \StringTok{"class"}
\FunctionTok{plot}\NormalTok{(sg.silt.class, }\AttributeTok{col =} \FunctionTok{topo.colors}\NormalTok{(}\DecValTok{11}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/silt-map-1.pdf}

}

\end{figure}

This map has much larger homogeneous areas than the SG2 pH map.

Examine this single map's co-occurrence matrix and vector:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\#|.label: coma{-}cove}
\NormalTok{coma.sg.silt }\OtherTok{\textless{}{-}} \FunctionTok{lsp\_signature}\NormalTok{(sg.silt.class, }\AttributeTok{type=}\StringTok{"coma"}\NormalTok{, }\AttributeTok{neighbourhood =} \DecValTok{8}\NormalTok{)}
\FunctionTok{print}\NormalTok{(coma.sg.silt.matrix }\OtherTok{\textless{}{-}} \FunctionTok{as.matrix}\NormalTok{(coma.sg.silt}\SpecialCharTok{$}\NormalTok{signature)[[}\DecValTok{1}\NormalTok{]])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
   1 2 3  4  5  6   7     8     9   10 11
1  2 0 5  3  4  1   1     0     0    0  0
2  0 0 1  0  2  2   0     2     0    0  0
3  5 1 2  9  6  4   4     7     1    0  0
4  3 0 9 16  9 13  21    20     3    0  0
5  4 2 6  9  8  9  15    24     9    0  0
6  1 2 4 13  9  6   8    25    26    1  0
7  1 0 4 21 15  8 828   416    29    5  0
8  0 2 7 20 24 25 416 14128  3024   10  0
9  0 0 1  3  9 26  29  3024 36196 2651  1
10 0 0 0  0  0  1   5    10  2651 7892 10
11 0 0 0  0  0  0   0     0     1   10  2
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{sum}\NormalTok{(}\FunctionTok{diag}\NormalTok{(coma.sg.silt.matrix))}\SpecialCharTok{/}\FunctionTok{sum}\NormalTok{(coma.sg.silt.matrix)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 0.8223602
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cove.sg.silt }\OtherTok{\textless{}{-}} \FunctionTok{lsp\_signature}\NormalTok{(sg.silt.class, }\AttributeTok{type=}\StringTok{"cove"}\NormalTok{, }\AttributeTok{neighbourhood =} \DecValTok{8}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Most of the adjacencies are to the same class, or the adjacent class.

Task: Compute the distance between the co-occurrence vectors for pH and
silt:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cove.df }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(cove.sg)}\SpecialCharTok{$}\NormalTok{signature[[}\DecValTok{1}\NormalTok{]][}\DecValTok{1}\NormalTok{,]}
\NormalTok{cove.df }\OtherTok{\textless{}{-}} \FunctionTok{rbind}\NormalTok{(cove.df, cove.sg.silt}\SpecialCharTok{$}\NormalTok{signature[[}\DecValTok{1}\NormalTok{]][}\DecValTok{1}\NormalTok{,])}
\NormalTok{cove.dists }\OtherTok{\textless{}{-}} \FunctionTok{round}\NormalTok{(}
\NormalTok{  philentropy}\SpecialCharTok{::}\FunctionTok{distance}\NormalTok{(cove.df, }\AttributeTok{method =} \StringTok{"jensen{-}shannon"}\NormalTok{, }
                        \AttributeTok{use.row.names =}\ConstantTok{TRUE}\NormalTok{, }
                        \AttributeTok{as.dist.obj =} \ConstantTok{FALSE}\NormalTok{,}
                        \AttributeTok{diag =} \ConstantTok{FALSE}\NormalTok{) ,}\DecValTok{4}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
Metric: 'jensen-shannon' using unit: 'log'; comparing: 2 vectors.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(cove.dists)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
jensen-shannon 
        0.6811 
\end{verbatim}

This is a much larger distance than that between SG2 and gNATSGO pH maps
co-occurrence vectors.

\hypertarget{clustering-pattern-differences}{%
\subsubsection{Clustering pattern
differences}\label{clustering-pattern-differences}}

Once a pattern metric is shown across a map, a natural question is
whether different areas of the map have different patterns. We
illustrate this with the pattern of the integrated co-occurrence
vectors.

Any size window can be used. If too small the result is erratic, if too
large, local differences may be missed.

Task: Identify which parts of the SG2 map have similar \emph{integrated
co-occurrence} pattern differences, considering both properties. For
this we use 4 x 4 km windows, i.e., 16 x 16 grid cells.

Again we use \texttt{lsp\_signature}, type \texttt{"incove"}, but now
specifying a \texttt{window} size within which to compute the pattern.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.ph.silt.class }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(sg.class, sg.silt.class)}
\NormalTok{incove.sg }\OtherTok{\textless{}{-}} \FunctionTok{lsp\_signature}\NormalTok{(sg.ph.silt.class,}
                              \AttributeTok{type =} \StringTok{"incove"}\NormalTok{,}
                              \AttributeTok{neighbourhood =} \DecValTok{8}\NormalTok{,}
                              \AttributeTok{ordered =} \ConstantTok{TRUE}\NormalTok{,  }\CommentTok{\# the pH classes are ordered}
                              \AttributeTok{window =} \DecValTok{16}\NormalTok{,}
                              \AttributeTok{normalization =} \StringTok{"pdf"}\NormalTok{)  }\CommentTok{\#sum to one}
\FunctionTok{summary}\NormalTok{(incove.sg.dist }\OtherTok{\textless{}{-}} \FunctionTok{lsp\_to\_dist}\NormalTok{(incove.sg,}
                                 \AttributeTok{dist\_fun =} \StringTok{"jensen{-}shannon"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
Metric: 'jensen-shannon' using unit: 'log2'; comparing: 42 vectors.
\end{verbatim}

\begin{verbatim}
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
0.04654 0.28386 0.42389 0.43529 0.59139 0.92259 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{dim}\NormalTok{(incove.sg.dist)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 42 42
\end{verbatim}

Here we have defined 42 x 42 distances, i.e., paired distances between
each of the windows' signatures.

Are any of these distances similar? Let's see with a \emph{cluster
analysis}.

Task: Make a hierarchical clustering of the distances between the
integrated co-occurrence vectors of the 42 windows.

The \texttt{hclust} function can cluster using many methods to build the
dendrogram. Here we use Ward's D2 method, which aims at finding compact,
spherical clusters.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.hclust }\OtherTok{\textless{}{-}} \FunctionTok{hclust}\NormalTok{(incove.sg.dist, }\AttributeTok{method =} \StringTok{"ward.D2"}\NormalTok{)}
\FunctionTok{plot}\NormalTok{(sg.hclust, }\AttributeTok{main =} \StringTok{"clusters of distance between \textasciigrave{}incove\textasciigrave{}"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/incove.cluster-1.pdf}

}

\end{figure}

Task: Define classes of similar distances by cutting the dendrogram.

Examining the dendrogram, it seems that height \texttt{h\ =\ 0.5} is a
good cutting point, which captures the main differences. Alternatively,
a set number of clusters can be requested with the \texttt{k} argument.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.clusters }\OtherTok{\textless{}{-}} \FunctionTok{as.factor}\NormalTok{(}\FunctionTok{cutree}\NormalTok{(sg.hclust, }\AttributeTok{h =} \FloatTok{0.5}\NormalTok{))  }\CommentTok{\# cutpoint by visual inspection}
\FunctionTok{levels}\NormalTok{(sg.clusters)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "1" "2" "3" "4" "5" "6"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.grid.sf }\OtherTok{=} \FunctionTok{lsp\_add\_clusters}\NormalTok{(incove.sg, sg.clusters)}
\NormalTok{sg.grid.sf}\SpecialCharTok{$}\NormalTok{clust }\OtherTok{\textless{}{-}} \FunctionTok{as.factor}\NormalTok{(sg.grid.sf}\SpecialCharTok{$}\NormalTok{clust)}
\NormalTok{my.pal }\OtherTok{\textless{}{-}} \FunctionTok{colorRampPalette}\NormalTok{(}\FunctionTok{brewer.pal}\NormalTok{(}\DecValTok{8}\NormalTok{, }\StringTok{"Accent"}\NormalTok{))(}\FunctionTok{length}\NormalTok{(}\FunctionTok{levels}\NormalTok{(sg.grid.sf}\SpecialCharTok{$}\NormalTok{clust)))}
\FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data =}\NormalTok{ sg.grid.sf) }\SpecialCharTok{+} 
  \FunctionTok{geom\_sf}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ clust), }\AttributeTok{alpha =} \FloatTok{0.7}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{scale\_fill\_discrete}\NormalTok{(}\AttributeTok{type =}\NormalTok{ my.pal) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{title =} \StringTok{"Clusters: distance between integrated co{-}occurrence vectors"}\NormalTok{,}
       \AttributeTok{fill =} \StringTok{"cluster"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/incove.cluster.cut-1.pdf}

}

\end{figure}

This shows which areas of the map have similar integrated co-occurrence
patterns. These can be interpreted as similar soils, in the sense that
the sum of propertied defines a soil type.

Compare this to a visual inspection of the patterns, next to the 7 x 6
cluster grid.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{p1 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data =}\NormalTok{ sg.grid.sf) }\SpecialCharTok{+} 
  \FunctionTok{geom\_sf}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ clust), }\AttributeTok{alpha =} \FloatTok{0.7}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{scale\_fill\_discrete}\NormalTok{(}\AttributeTok{type =}\NormalTok{ my.pal) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{fill =} \StringTok{"cluster"}\NormalTok{)  }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.position=}\StringTok{"none"}\NormalTok{)}
\NormalTok{p2 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{() }\SpecialCharTok{+}
\NormalTok{  tidyterra}\SpecialCharTok{::}\FunctionTok{geom\_spatraster}\NormalTok{(}\AttributeTok{data =}\NormalTok{ sg.class, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ class)) }\SpecialCharTok{+}
   \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.position=}\StringTok{"none"}\NormalTok{)}
\NormalTok{p3 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{() }\SpecialCharTok{+}
\NormalTok{  tidyterra}\SpecialCharTok{::}\FunctionTok{geom\_spatraster}\NormalTok{(}\AttributeTok{data =}\NormalTok{ sg.silt.class, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ class)) }\SpecialCharTok{+}
   \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.position=}\StringTok{"none"}\NormalTok{)}
\NormalTok{gridExtra}\SpecialCharTok{::}\FunctionTok{grid.arrange}\NormalTok{(p1, p2, p3, }\AttributeTok{nrow=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/incove-clusters-map-3-1.pdf}

}

\end{figure}

Careful examination reveals that the cluster in the NW corner
corresponds to an intricate pattern of pH and mostly one class of silt
concentration.

\hypertarget{sec-lsm}{%
\subsection{Landscape metrics}\label{sec-lsm}}

Landscape metrics have a long history of use in landscape ecology
(\protect\hyperlink{ref-Uuemaa.etal2013}{Uuemaa et al., 2013}). A wide
variety have been collected in the well-known FRAGSTATS computer program
(\protect\hyperlink{ref-McGarigal.etal2012}{McGarigal et al., 2012}).
These have been implemented in the R context by the
\texttt{landscapemetrics} package\footnote{https://r-spatialecology.github.io/landscapemetrics/}
(\protect\hyperlink{ref-Hesselbarth.etal2019}{Hesselbarth et al., 2019};
\protect\hyperlink{ref-Hesselbarth2021}{Hesselbarth, 2021}). Although
the ecological relevance of FRAGSTATS metrics have been criticized
(\protect\hyperlink{ref-Kupfer2012}{Kupfer, 2012}), here we use them to
\emph{characterize spatial patterns of soil properties} or
\emph{classes}, not as inputs to landscape ecology models.

The patterns of soil classes or properties are not expected to have the
same characteristics as those for land cover or vegetation types. Land
cover is largely controlled by humans, and where it is not, vegetation
is mostly placed on the landscape by different mechanisms than are
soils. There is a link, however: if the soil property is largely
controlled by the \texttt{o} (organism) or \texttt{h} (human) factor,
then the patterns on the landscape could be similar to those under it.

There are many metrics, of three levels of detail. We list them here for
reference; each has its own help text.

First, the \emph{patch-level metrics}. These describe every patch, i.e.,
contiguous cells belonging to the same class.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{landscapemetrics}\SpecialCharTok{::}\FunctionTok{list\_lsm}\NormalTok{(}\AttributeTok{level=}\StringTok{"patch"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{print}\NormalTok{(}\AttributeTok{n=}\ConstantTok{Inf}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 12 x 5
   metric name                                type           level function_name
   <chr>  <chr>                               <chr>          <chr> <chr>        
 1 area   patch area                          area and edge~ patch lsm_p_area   
 2 cai    core area index                     core area met~ patch lsm_p_cai    
 3 circle related circumscribing circle       shape metric   patch lsm_p_circle 
 4 contig contiguity index                    shape metric   patch lsm_p_contig 
 5 core   core area                           core area met~ patch lsm_p_core   
 6 enn    euclidean nearest neighbor distance aggregation m~ patch lsm_p_enn    
 7 frac   fractal dimension index             shape metric   patch lsm_p_frac   
 8 gyrate radius of gyration                  area and edge~ patch lsm_p_gyrate 
 9 ncore  number of core areas                core area met~ patch lsm_p_ncore  
10 para   perimeter-area ratio                shape metric   patch lsm_p_para   
11 perim  patch perimeter                     area and edge~ patch lsm_p_perim  
12 shape  shape index                         shape metric   patch lsm_p_shape  
\end{verbatim}

Second, the \emph{class-level} metrics. These describe all patches
belonging to a specified class.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{landscapemetrics}\SpecialCharTok{::}\FunctionTok{list\_lsm}\NormalTok{(}\AttributeTok{level=}\StringTok{"class"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{print}\NormalTok{(}\AttributeTok{n=}\ConstantTok{Inf}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 55 x 5
   metric    name                                  type      level function_name
   <chr>     <chr>                                 <chr>     <chr> <chr>        
 1 ai        aggregation index                     aggregat~ class lsm_c_ai     
 2 area_cv   patch area                            area and~ class lsm_c_area_cv
 3 area_mn   patch area                            area and~ class lsm_c_area_mn
 4 area_sd   patch area                            area and~ class lsm_c_area_sd
 5 ca        total (class) area                    area and~ class lsm_c_ca     
 6 cai_cv    core area index                       core are~ class lsm_c_cai_cv 
 7 cai_mn    core area index                       core are~ class lsm_c_cai_mn 
 8 cai_sd    core area index                       core are~ class lsm_c_cai_sd 
 9 circle_cv related circumscribing circle         shape me~ class lsm_c_circle~
10 circle_mn related circumscribing circle         shape me~ class lsm_c_circle~
11 circle_sd related circumscribing circle         shape me~ class lsm_c_circle~
12 clumpy    clumpiness index                      aggregat~ class lsm_c_clumpy 
13 cohesion  patch cohesion index                  aggregat~ class lsm_c_cohesi~
14 contig_cv contiguity index                      shape me~ class lsm_c_contig~
15 contig_mn contiguity index                      shape me~ class lsm_c_contig~
16 contig_sd contiguity index                      shape me~ class lsm_c_contig~
17 core_cv   core area                             core are~ class lsm_c_core_cv
18 core_mn   core area                             core are~ class lsm_c_core_mn
19 core_sd   core area                             core are~ class lsm_c_core_sd
20 cpland    core area percentage of landscape     core are~ class lsm_c_cpland 
21 dcad      disjunct core area density            core are~ class lsm_c_dcad   
22 dcore_cv  disjunct core area                    core are~ class lsm_c_dcore_~
23 dcore_mn  disjunct core area                    core are~ class lsm_c_dcore_~
24 dcore_sd  disjunct core area                    core are~ class lsm_c_dcore_~
25 division  division index                        aggregat~ class lsm_c_divisi~
26 ed        edge density                          area and~ class lsm_c_ed     
27 enn_cv    euclidean nearest neighbor distance   aggregat~ class lsm_c_enn_cv 
28 enn_mn    euclidean nearest neighbor distance   aggregat~ class lsm_c_enn_mn 
29 enn_sd    euclidean nearest neighbor distance   aggregat~ class lsm_c_enn_sd 
30 frac_cv   fractal dimension index               shape me~ class lsm_c_frac_cv
31 frac_mn   fractal dimension index               shape me~ class lsm_c_frac_mn
32 frac_sd   fractal dimension index               shape me~ class lsm_c_frac_sd
33 gyrate_cv radius of gyration                    area and~ class lsm_c_gyrate~
34 gyrate_mn radius of gyration                    area and~ class lsm_c_gyrate~
35 gyrate_sd radius of gyration                    area and~ class lsm_c_gyrate~
36 iji       interspersion and juxtaposition index aggregat~ class lsm_c_iji    
37 lpi       largest patch index                   area and~ class lsm_c_lpi    
38 lsi       landscape shape index                 aggregat~ class lsm_c_lsi    
39 mesh      effective mesh size                   aggregat~ class lsm_c_mesh   
40 ndca      number of disjunct core areas         core are~ class lsm_c_ndca   
41 nlsi      normalized landscape shape index      aggregat~ class lsm_c_nlsi   
42 np        number of patches                     aggregat~ class lsm_c_np     
43 pafrac    perimeter-area fractal dimension      shape me~ class lsm_c_pafrac 
44 para_cv   perimeter-area ratio                  shape me~ class lsm_c_para_cv
45 para_mn   perimeter-area ratio                  shape me~ class lsm_c_para_mn
46 para_sd   perimeter-area ratio                  shape me~ class lsm_c_para_sd
47 pd        patch density                         aggregat~ class lsm_c_pd     
48 pladj     percentage of like adjacencies        aggregat~ class lsm_c_pladj  
49 pland     percentage of landscape               area and~ class lsm_c_pland  
50 shape_cv  shape index                           shape me~ class lsm_c_shape_~
51 shape_mn  shape index                           shape me~ class lsm_c_shape_~
52 shape_sd  shape index                           shape me~ class lsm_c_shape_~
53 split     splitting index                       aggregat~ class lsm_c_split  
54 tca       total core area                       core are~ class lsm_c_tca    
55 te        total edge                            area and~ class lsm_c_te     
\end{verbatim}

Finally, the \emph{landscape-level} metrics. These describe the
characteristics of the entire landscape, i.e., the assemblage of classes
and patches.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{landscapemetrics}\SpecialCharTok{::}\FunctionTok{list\_lsm}\NormalTok{(}\AttributeTok{level=}\StringTok{"landscape"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{print}\NormalTok{(}\AttributeTok{n=}\ConstantTok{Inf}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 66 x 5
   metric    name                                  type      level function_name
   <chr>     <chr>                                 <chr>     <chr> <chr>        
 1 ai        aggregation index                     aggregat~ land~ lsm_l_ai     
 2 area_cv   patch area                            area and~ land~ lsm_l_area_cv
 3 area_mn   patch area                            area and~ land~ lsm_l_area_mn
 4 area_sd   patch area                            area and~ land~ lsm_l_area_sd
 5 cai_cv    core area index                       core are~ land~ lsm_l_cai_cv 
 6 cai_mn    core area index                       core are~ land~ lsm_l_cai_mn 
 7 cai_sd    core area index                       core are~ land~ lsm_l_cai_sd 
 8 circle_cv related circumscribing circle         shape me~ land~ lsm_l_circle~
 9 circle_mn related circumscribing circle         shape me~ land~ lsm_l_circle~
10 circle_sd related circumscribing circle         shape me~ land~ lsm_l_circle~
11 cohesion  patch cohesion index                  aggregat~ land~ lsm_l_cohesi~
12 condent   conditional entropy                   complexi~ land~ lsm_l_condent
13 contag    connectance                           aggregat~ land~ lsm_l_contag 
14 contig_cv contiguity index                      shape me~ land~ lsm_l_contig~
15 contig_mn contiguity index                      shape me~ land~ lsm_l_contig~
16 contig_sd contiguity index                      shape me~ land~ lsm_l_contig~
17 core_cv   core area                             core are~ land~ lsm_l_core_cv
18 core_mn   core area                             core are~ land~ lsm_l_core_mn
19 core_sd   core area                             core are~ land~ lsm_l_core_sd
20 dcad      disjunct core area density            core are~ land~ lsm_l_dcad   
21 dcore_cv  disjunct core area                    core are~ land~ lsm_l_dcore_~
22 dcore_mn  disjunct core area                    core are~ land~ lsm_l_dcore_~
23 dcore_sd  disjunct core area                    core are~ land~ lsm_l_dcore_~
24 division  division index                        aggregat~ land~ lsm_l_divisi~
25 ed        edge density                          area and~ land~ lsm_l_ed     
26 enn_cv    euclidean nearest neighbor distance   aggregat~ land~ lsm_l_enn_cv 
27 enn_mn    euclidean nearest neighbor distance   aggregat~ land~ lsm_l_enn_mn 
28 enn_sd    euclidean nearest neighbor distance   aggregat~ land~ lsm_l_enn_sd 
29 ent       shannon entropy                       complexi~ land~ lsm_l_ent    
30 frac_cv   fractal dimension index               shape me~ land~ lsm_l_frac_cv
31 frac_mn   fractal dimension index               shape me~ land~ lsm_l_frac_mn
32 frac_sd   fractal dimension index               shape me~ land~ lsm_l_frac_sd
33 gyrate_cv radius of gyration                    area and~ land~ lsm_l_gyrate~
34 gyrate_mn radius of gyration                    area and~ land~ lsm_l_gyrate~
35 gyrate_sd radius of gyration                    area and~ land~ lsm_l_gyrate~
36 iji       interspersion and juxtaposition index aggregat~ land~ lsm_l_iji    
37 joinent   joint entropy                         complexi~ land~ lsm_l_joinent
38 lpi       largest patch index                   area and~ land~ lsm_l_lpi    
39 lsi       landscape shape index                 aggregat~ land~ lsm_l_lsi    
40 mesh      effective mesh size                   aggregat~ land~ lsm_l_mesh   
41 msidi     modified simpson's diversity index    diversit~ land~ lsm_l_msidi  
42 msiei     modified simpson's evenness index     diversit~ land~ lsm_l_msiei  
43 mutinf    mutual information                    complexi~ land~ lsm_l_mutinf 
44 ndca      number of disjunct core areas         core are~ land~ lsm_l_ndca   
45 np        number of patches                     aggregat~ land~ lsm_l_np     
46 pafrac    perimeter-area fractal dimension      shape me~ land~ lsm_l_pafrac 
47 para_cv   perimeter-area ratio                  shape me~ land~ lsm_l_para_cv
48 para_mn   perimeter-area ratio                  shape me~ land~ lsm_l_para_mn
49 para_sd   perimeter-area ratio                  shape me~ land~ lsm_l_para_sd
50 pd        patch density                         aggregat~ land~ lsm_l_pd     
51 pladj     percentage of like adjacencies        aggregat~ land~ lsm_l_pladj  
52 pr        patch richness                        diversit~ land~ lsm_l_pr     
53 prd       patch richness density                diversit~ land~ lsm_l_prd    
54 relmutinf relative mutual information           complexi~ land~ lsm_l_relmut~
55 rpr       relative patch richness               diversit~ land~ lsm_l_rpr    
56 shape_cv  shape index                           shape me~ land~ lsm_l_shape_~
57 shape_mn  shape index                           shape me~ land~ lsm_l_shape_~
58 shape_sd  shape index                           shape me~ land~ lsm_l_shape_~
59 shdi      shannon's diversity index             diversit~ land~ lsm_l_shdi   
60 shei      shannon's evenness index              diversit~ land~ lsm_l_shei   
61 sidi      simpson's diversity index             diversit~ land~ lsm_l_sidi   
62 siei      simspon's evenness index              diversit~ land~ lsm_l_siei   
63 split     splitting index                       aggregat~ land~ lsm_l_split  
64 ta        total area                            area and~ land~ lsm_l_ta     
65 tca       total core area                       core are~ land~ lsm_l_tca    
66 te        total edge                            area and~ land~ lsm_l_te     
\end{verbatim}

\hypertarget{landscape-level-metrics}{%
\subsubsection{Landscape-level metrics}\label{landscape-level-metrics}}

These measures summarize the pattern of the entire map. The following
five seem to be most useful for characterizing soil maps.

\begin{itemize}
\tightlist
\item
  \textbf{ai}: The \textbf{landscape aggregation index} LAI is an
  `Aggregation metric'. This shows how much the classes occur as large
  units, vs.~as scattered patches. It is independent of the number of
  classes.
\end{itemize}

It equals the number of like adjacencies divided by the theoretical
maximum possible number of like adjacencies for that class summed over
each class for the entire landscape. The metric is based on the
adjacency matrix. It equals 0 for maximally disaggregated and 100 for
maximally aggregated classes.
\href{https://r-spatialecology.github.io/landscapemetrics/reference/lsm_l_ai.html}{More
info}
\[\mathrm{LAI} = \Bigg[∑_{i=1}^m \Big( \frac{g_{ii}}{max-g_{ii}} \Big) P_{i} \Bigg](100)\]
where \(g_{ii}\) is the number of like adjacencies,
\((\mathrm{max}-g_{ii})\) is the class-wise maximum possible number of
like adjacencies of class \(i\) (i.e., if all pixels in the class were
in one cluster), and \(P_{i}\) is the proportion of landscape comprised
of class \(i\), to weight the index by class prevalence.

\begin{itemize}
\tightlist
\item
  \textbf{frac\_mn}: The \textbf{mean fractal dimension} FRAC\_MN is a
  `Shape metric'. It summarises the landscape as the mean of the fractal
  dimension index of all patches in the landscape, i.e., the complexity
  of the map.
\end{itemize}

The fractal dimension index is based on the patch perimeter and the
patch area and describes the patch complexity. The Coefficient of
variation is scaled to the mean and thus is comparable among different
landscapes.
\href{https://r-spatialecology.github.io/landscapemetrics/reference/lsm_l_frac_mn.html}{More
info} \[\mathrm{FRAC} = \frac{2 * \ln * (0.25 * p_{ij})} {\ln a_{ij}}\]
where the patch perimeters are \({p_{ij}}\) in linear units and the
areas are \({a_{ij}}\) in square units.

\begin{itemize}
\tightlist
\item
  \textbf{lsi}: \textbf{landscape shape index} LSI is an `Aggregation
  metric'. It is the ratio between the actual edge length of class \(i\)
  and the hypothetical minimum edge length of class \(i\). It measures
  how compact are the classes. For example, long thin classes will have
  low LSI.
\end{itemize}

The minimum edge length equals the edge length if class i would be
maximally aggregated. LSI = 1 when only one square patch is present or
all patches are maximally aggregated. Increases, without limit, as the
length of the actual edges increases, i.e.~the patches become less
compact.
\href{https://r-spatialecology.github.io/landscapemetrics/reference/lsm_c_lsi.html?q=lsi}{More
info} \[   \mathrm{LSI} = \frac{0.25 E'}{\sqrt{A}}\] where \(A\) is the
total area of the landscape and \(E'\) is the total length of edges,
including the boundary.

\begin{itemize}
\tightlist
\item
  \textbf{shdi}: The \textbf{Shannon diversity index} SHDI is a
  `Diversity metric'. It is a widely used metric in biodiversity and
  ecology and takes both the number of classes and the abundance of each
  class into account. It is related to the concept of entropy: how much
  ``information'' is in the landscape pattern. More classes and more
  even distribution of their areas implies high information.
\end{itemize}

SHDI = 0 when only one patch is present and increases, without limit, as
the number of classes increases while the proportions are equally
distributed.
\href{https://r-spatialecology.github.io/landscapemetrics/reference/lsm_l_shdi.html?q=shd}{More
info} \[ D = - \sum_{i=1}^N p_i \ln p_i\] where \(p_i\) is the
proportion of pixels of class \(i = (1 \ldots N)\),

\begin{itemize}
\tightlist
\item
  \textbf{shei}: The \textbf{Shannon evenness index} SHEI is a
  `Diversity metric'. It is the ratio between the Shannon's diversity
  index \(D\) (see previous) and and the theoretical maximum Shannon
  diversity index \(\ln N\). It can be understood as a measure of
  dominance.
\end{itemize}

SHEI = 0 when only one patch present; SHEI = 1 when the proportion of
classes is equally distributed.
\href{https://r-spatialecology.github.io/landscapemetrics/reference/lsm_l_shei.html?q=shei}{More
info} \[E = \frac{D}{\ln N}\]

These methods must be applied to classified maps. Continuous soil
property maps must first be classified into ranges before analysis, see
(Section~\ref{sec-hist-equal}) and (Section~\ref{sec-mean-limit}),
above. Different choices of class limits and widths will result in
different values of these measures.

\hypertarget{compute-sec-lsm}{%
\subsubsection{Computing landscape-level
metrics}\label{compute-sec-lsm}}

The \texttt{landscapemetrics} package implements a set of metrics as
used in ecology and derived from the FRAGSTATS computer program; the
metrics are explained in the previous section. Here we compute them for
the two maps we are comparing.

To compute landscape metrics:

\begin{itemize}
\tightlist
\item
  Input is raster map (here, a \texttt{terra::SpatRaster}) with integer
  values, each of which represents a category, i.e., landscape class.
\item
  The map must be in a projected CRS, with distance units in meters;
\item
  Results are in meters, square meters or hectares, depending on the
  function;
\end{itemize}

Task: Check that the maps have the proper structure for the landscape
metrics.

This is done with the \texttt{landscapemetrics::check\_landscape}
function.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{check\_landscape}\NormalTok{(gn.class)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  layer       crs units   class n_classes OK
1     1 projected     m integer        13  v
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{check\_landscape}\NormalTok{(sg.class)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  layer       crs units   class n_classes OK
1     1 projected     m integer        10  v
\end{verbatim}

Task: Show the landscapes of each product, first with all classes on one
map, then with the classes separate:

\textbf{global}:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{show\_patches}\NormalTok{(gn.class, }\AttributeTok{class =} \StringTok{"global"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show.patches.global-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{show\_patches}\NormalTok{(sg.class, }\AttributeTok{class =} \StringTok{"global"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show.patches.global-2.pdf}

}

\end{figure}

\textbf{per-class}:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{show\_patches}\NormalTok{(sg.class, }\AttributeTok{class =} \StringTok{"all"}\NormalTok{, }\AttributeTok{nrow =} \DecValTok{3}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show.patches.all-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{show\_patches}\NormalTok{(gn.class, }\AttributeTok{class =} \StringTok{"all"}\NormalTok{, }\AttributeTok{nrow =} \DecValTok{3}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show.patches.all-2.pdf}

}

\end{figure}

Q: Describe the main differences between the patterns. Which map seems
more aggregated? More diverse?

Task: compute the metrics and tabulate them:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{lst }\OtherTok{\textless{}{-}} \FunctionTok{paste0}\NormalTok{(}\StringTok{"lsm\_l\_"}\NormalTok{, }\FunctionTok{c}\NormalTok{(}\StringTok{"shdi"}\NormalTok{, }\StringTok{"shei"}\NormalTok{, }\StringTok{"lsi"}\NormalTok{, }\StringTok{"ai"}\NormalTok{,  }\StringTok{"frac\_mn"}\NormalTok{))}
\NormalTok{ls.metrics.gn }\OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(gn.class, }\AttributeTok{what=}\NormalTok{lst)}
\NormalTok{ls.metrics.sg }\OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(sg.class, }\AttributeTok{what=}\NormalTok{lst)}
\NormalTok{metrics.table }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{product=}\FunctionTok{c}\NormalTok{(}\StringTok{"gNATSGO"}\NormalTok{, }\StringTok{"SG2"}\NormalTok{),}
                            \FunctionTok{rbind}\NormalTok{(}\FunctionTok{round}\NormalTok{(ls.metrics.gn}\SpecialCharTok{$}\NormalTok{value, }\DecValTok{3}\NormalTok{),}
                                  \FunctionTok{round}\NormalTok{(ls.metrics.sg}\SpecialCharTok{$}\NormalTok{value, }\DecValTok{3}\NormalTok{)))}
\FunctionTok{names}\NormalTok{(metrics.table)[}\DecValTok{2}\SpecialCharTok{:}\DecValTok{6}\NormalTok{] }\OtherTok{\textless{}{-}}\NormalTok{ ls.metrics.gn}\SpecialCharTok{$}\NormalTok{metric}
\NormalTok{metrics.table}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  product     ai frac_mn    lsi  shdi  shei
1 gNATSGO 31.482   1.028 33.605 2.052 0.800
2     SG2 64.694   1.050 18.450 2.036 0.884
\end{verbatim}

Q: Referring to the descriptions of these metrics (above), what are the
differences between these maps' landscape patterns? Where do the maps
most differ?

\begin{itemize}
\tightlist
\item
  Aggregation Index
\item
  Mean Fractal Dimension
\item
  Landscape Shape Index
\item
  Shannon Diversity
\item
  Shannon Evenness
\end{itemize}

\hypertarget{sec-compare-classified}{%
\section{Comparing patterns of classified
maps}\label{sec-compare-classified}}

Once we have various pattern metrics computed on different maps of the
same area, an obvious question is ``How much and how do they differ?''.
The question of ``best'' map is not (yet) asked.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  We can directly compare the metrics; see above
  (Section~\ref{sec-lsm}).
\item
  We can compare the adjacency structures; see above
  (Section~\ref{sec-cove}) and next (Section~\ref{sec-compare-cove}).
\item
  We can compare the intersections of the maps: use one as a reference
  and determine how well the other map reproduces the structure of the
  first; see below (Section~\ref{sec-vmeasure}) .
\end{enumerate}

\hypertarget{sec-compare-cove}{%
\subsection{Co-occurrence vectors}\label{sec-compare-cove}}

Task: Compute the difference between the co-occurrence patterns of the
two maps.

This uses the Jensen-Shannon distance between matrix columns. Each row
of the column vector is a co-occurrence metric of two classes. The
\texttt{philentropy} (``Similarity and Distance Quantification Between
Probability Functions'') package implements this distance metric. This
metric is commonly used to compare probability distributions. It
computes the entropy of each probability vector (here, the co-occurrence
vector) and the entropy of their average and, from these, the distance
in entropy space between them:

\[2d = \sum_i \big(P_i \; log\frac{2 P_i}{P_i + Q_i}\big) + \sum_i \big(Q_i \; log\frac{2 Q_i}{P_i + Q_i}\big)\]
where \(P\) and \(Q\) are the two vectors, and \(i\) is the row, i.e.,
single co-occurrence value.

Increasing values indicate increasing dissimilarity in the adjacency
patterns, i.e., greater entropy. If the adjacency structures are
identical, the distance is zero.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{names}\NormalTok{(cove.gn)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "id"        "na_prop"   "signature"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cove.df }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(cove.gn)}\SpecialCharTok{$}\NormalTok{signature[[}\DecValTok{1}\NormalTok{]][}\DecValTok{1}\NormalTok{,]}
\NormalTok{cove.df }\OtherTok{\textless{}{-}} \FunctionTok{rbind}\NormalTok{(cove.df, cove.sg}\SpecialCharTok{$}\NormalTok{signature[[}\DecValTok{1}\NormalTok{]][}\DecValTok{1}\NormalTok{,])}
\NormalTok{cove.dists }\OtherTok{\textless{}{-}} \FunctionTok{round}\NormalTok{(}
\NormalTok{  philentropy}\SpecialCharTok{::}\FunctionTok{distance}\NormalTok{(cove.df, }\AttributeTok{method =} \StringTok{"jensen{-}shannon"}\NormalTok{, }
                        \AttributeTok{use.row.names =}\ConstantTok{TRUE}\NormalTok{, }
                        \AttributeTok{as.dist.obj =} \ConstantTok{FALSE}\NormalTok{,}
                        \AttributeTok{diag =} \ConstantTok{FALSE}\NormalTok{) ,}\DecValTok{4}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
Metric: 'jensen-shannon' using unit: 'log'; comparing: 2 vectors.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(cove.dists)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
jensen-shannon 
        0.4584 
\end{verbatim}

This is a fairly high value.

This comparison can be streamlined with the \texttt{lsp\_compare}
method. First, over the whole map:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{lsp\_compare}\NormalTok{(gn.class, sg.class, }
            \AttributeTok{type =} \StringTok{"cove"}\NormalTok{, }\AttributeTok{dist\_fun =} \StringTok{"jensen{-}shannon"}\NormalTok{,}
            \AttributeTok{neighbourhood =} \DecValTok{8}\NormalTok{, }\CommentTok{\# queen\textquotesingle{}s case}
            \AttributeTok{output =} \StringTok{"sf"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
Simple feature collection with 1 feature and 4 fields
Geometry type: POLYGON
Dimension:     XY
Bounding box:  xmin: 380561.8 ymin: 4683745 xmax: 401561.8 ymax: 4711745
Projected CRS: WGS 84 / UTM zone 18N
  id  na_prop_x  na_prop_y      dist                       geometry
1  1 0.04900085 0.04900085 0.4297049 POLYGON ((380561.8 4711745,...
\end{verbatim}

The patterns can be compared over various windows within the two maps.
This allows the difference between sub-maps to be quantified.

The bounding box is 21 x 28 km. Let's compare patterns over 4 x 4 km
windows; these are 16 x 16 grid cells.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{dim}\NormalTok{(sg.class)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 112  84   1
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x.dim }\OtherTok{\textless{}{-}} \FunctionTok{diff}\NormalTok{(}\FunctionTok{range}\NormalTok{(}\FunctionTok{st\_bbox}\NormalTok{(sg.class)[}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{3}\NormalTok{)]))}
\NormalTok{y.dim }\OtherTok{\textless{}{-}} \FunctionTok{diff}\NormalTok{(}\FunctionTok{range}\NormalTok{(}\FunctionTok{st\_bbox}\NormalTok{(sg.class)[}\FunctionTok{c}\NormalTok{(}\DecValTok{2}\NormalTok{,}\DecValTok{4}\NormalTok{)]))}
\NormalTok{(compare}\FloatTok{.16} \OtherTok{\textless{}{-}} \FunctionTok{lsp\_compare}\NormalTok{(gn.class, sg.class, }
                           \AttributeTok{type =} \StringTok{"cove"}\NormalTok{, }\AttributeTok{dist\_fun =} \StringTok{"jensen{-}shannon"}\NormalTok{,}
                           \AttributeTok{neighbourhood =} \DecValTok{8}\NormalTok{, }\CommentTok{\# queen\textquotesingle{}s case}
                           \AttributeTok{window =} \DecValTok{16}\NormalTok{,}
                           \AttributeTok{output =} \StringTok{"sf"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
Simple feature collection with 35 features and 4 fields
Geometry type: POLYGON
Dimension:     XY
Bounding box:  xmin: 380561.8 ymin: 4683745 xmax: 400561.8 ymax: 4711745
Projected CRS: WGS 84 / UTM zone 18N
First 10 features:
   id  na_prop_x  na_prop_y      dist                       geometry
1   1 0.12890625 0.12890625 0.4959997 POLYGON ((380561.8 4711745,...
2   2 0.02343750 0.02343750 0.5035320 POLYGON ((384561.8 4711745,...
3   3 0.11718750 0.11718750 0.5794527 POLYGON ((388561.8 4711745,...
4   4 0.07421875 0.07421875 0.6101658 POLYGON ((392561.8 4711745,...
5   5 0.06250000 0.06250000 0.6779548 POLYGON ((396561.8 4711745,...
6   7 0.10156250 0.10156250 0.3203339 POLYGON ((380561.8 4707745,...
7   8 0.01953125 0.01953125 0.5168618 POLYGON ((384561.8 4707745,...
8   9 0.00781250 0.00781250 0.5059423 POLYGON ((388561.8 4707745,...
9  10 0.05859375 0.05859375 0.7748153 POLYGON ((392561.8 4707745,...
10 11 0.00390625 0.00390625 0.6143669 POLYGON ((396561.8 4707745,...
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data =}\NormalTok{ compare}\FloatTok{.16}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{geom\_sf}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ dist)) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{title =} \StringTok{"Distance between co{-}occurrence vectors, pH class"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/lsp.compare.window-1.pdf}

}

\end{figure}

We see that the distance between co-occurrence vectors varies across the
map, although in all the submaps the distance is fairly large. The
patterns are closer on the west side.

Visualize these distances along with the source maps.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{p1 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data =}\NormalTok{ compare}\FloatTok{.16}\NormalTok{) }\SpecialCharTok{+} 
    \FunctionTok{geom\_sf}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ dist))  }\SpecialCharTok{+}
   \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.position=}\StringTok{"none"}\NormalTok{)}
\NormalTok{p2 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{() }\SpecialCharTok{+}
\NormalTok{  tidyterra}\SpecialCharTok{::}\FunctionTok{geom\_spatraster}\NormalTok{(}\AttributeTok{data =}\NormalTok{ sg.class, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ class)) }\SpecialCharTok{+}
   \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.position=}\StringTok{"none"}\NormalTok{)}
\NormalTok{p3 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{() }\SpecialCharTok{+}
\NormalTok{  tidyterra}\SpecialCharTok{::}\FunctionTok{geom\_spatraster}\NormalTok{(}\AttributeTok{data =}\NormalTok{ gn.class, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ class)) }\SpecialCharTok{+}
   \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.position=}\StringTok{"none"}\NormalTok{)}
\NormalTok{gridExtra}\SpecialCharTok{::}\FunctionTok{grid.arrange}\NormalTok{(p1, p2, p3, }\AttributeTok{nrow=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/cove-patterns-plot-1.pdf}

}

\end{figure}

\hypertarget{vmetrics}{%
\subsection{V metrics}\label{vmetrics}}

The \emph{V-measure} originated in the field of computer science as a
measure for comparison of different clusterings of the same domain. It
is a measure of an overall spatial correspondence between classified
maps. Continuous maps must be classified into classes, and the two
classified maps then compared. Note that the classes do not have to be
the same, although in this example they are. The theme does not even
have to be the same. For example one could compare a pH map with a clay
concentration map.

Two maps could have the same total areas of each class, and even the
same number of polygons within each class and even the same size
distribution of these polygons, and yet be completely different in how
they partition space into classes.

The polygons of a classified map are termed \emph{regions of a
regionalization} in the first (\emph{reference}) map and \emph{zones of
a partition} in the second map. These are intersected to produce segment
polygons of the combined map, which are labelled with both zone and
region classes.

This allows the computation of two indices:

\emph{Homogeneity} compares the \emph{zones in the second} map with
respect to the \emph{regions in the first}, i.e., how close the second
map comes to reproducing the regionalization of the first. Thus it
evaluates the second map, in terms of the first.

It is computed as the variance of the regions within a zone, normalized
by the variance of the regions in the entire domain of the first map.
These variances are computed by the Shannon entropy based on areas of
the segments.

If the variance of the regions within the zones is small, then the
partition into zones in the second map is relatively homogeneous with
respect to the regionalization. A perfectly homogeneous partition (with
value 1) is when each zone of the second map is within a single region
of the reference map. In this case, each zone has only one reference
class. A perfectly inhomogeneous partition (with value 0) is when each
zone has the same composition of regions as the entire domain of the
first map, i.e., the second map's partition is essentially random with
respect to the first map's regionalization.

\emph{Completeness} is a function of homogeneity of the regions in the
first map with respect to the zones in the second, i.e., how much the
partition in the first map reproduce that of the second. Thus it
evaluates the first map by the second.

The completeness of the second map is the inverse of homogeneity; it
assesses the variance of the zones within a region normalized by the
variance of the zones in the entire domain of the second map. It
evaluates the homogeneity of regions with respect to zones and shows how
well the regionalization of the reference map fits inside the partition
of the map to be evaluated. A perfectly complete regionalization is when
each region of the reference map is entirely within a single zone of the
map to be evaluated. In this case, a polygon of the reference map will
not be split among zones.

These two together are combined into a single \emph{V measure} of
agreement between the maps, as the harmonic mean of homogeneity and
completeness.

This function uses the \texttt{sf::st\_intersection()}, which depends on
the coordinates values precision. For example, precision = 1000 rounds
values to the third decimal places and precision = 0.001 uses values
rounded to the nearest 1000, see \texttt{?sf::st\_as\_binary}.

V-measure methods are implemented in the \texttt{sabre} ``Spatial
Association Between Regionalizations'' package, as explained in
(\protect\hyperlink{ref-nowosadSpatialAssociationRegionalizations2018}{Nowosad
\& Stepinski, 2018a}).

The \texttt{vmeasure\_calc()} function calculates intersections of the
input geometries. For this function we must specify the names of the
columns with the region names; both x and y must contain
\texttt{POLYGON}s or \texttt{MULTIPOLYGON}s and have the same CRS.

\hypertarget{polygonize}{%
\subsubsection{Polygonize}\label{polygonize}}

The V-metrics require polygon maps, not gridded maps of classes.

Task: Polygonize them and adjust the class names.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gn.poly }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{as.polygons}\NormalTok{(gn.class,}
                              \AttributeTok{aggregate=} \ConstantTok{TRUE}\NormalTok{,}
                              \AttributeTok{values =} \ConstantTok{TRUE}\NormalTok{,}
                              \AttributeTok{dissolve =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{sg.poly }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{as.polygons}\NormalTok{(sg.class,}
                              \AttributeTok{aggregate=} \ConstantTok{TRUE}\NormalTok{,}
                              \AttributeTok{values =} \ConstantTok{TRUE}\NormalTok{, }
                              \AttributeTok{dissolve=}\ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{simple-features}{%
\subsubsection{Simple Features}\label{simple-features}}

Some of the methods require Simple Features representation of spatial
objects.

Task: Convert the \texttt{terra::SpatVector} objects to Simple Features.

Sometimes these may have some simple \texttt{POLYGON}s, so ensure all
are \texttt{MULTIPOLYGON} as required by \texttt{vmeasure\_calc}, below.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gn.sf }\OtherTok{\textless{}{-}} \FunctionTok{st\_as\_sf}\NormalTok{(gn.poly)}
\NormalTok{gn.sf }\OtherTok{\textless{}{-}} \FunctionTok{st\_cast}\NormalTok{(gn.sf, }\StringTok{"MULTIPOLYGON"}\NormalTok{)}
\CommentTok{\#}
\NormalTok{sg.sf }\OtherTok{\textless{}{-}} \FunctionTok{st\_as\_sf}\NormalTok{(sg.poly)}
\NormalTok{sg.sf }\OtherTok{\textless{}{-}} \FunctionTok{st\_cast}\NormalTok{(sg.sf, }\StringTok{"MULTIPOLYGON"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{topology}{%
\subsubsection{Topology}\label{topology}}

Task: Check that the topology of the polygon map is correct. If not
\texttt{sabre::vmeasure\_calc} (see below) throws an error. Clean up the
topology with \texttt{sf::st\_make\_valid}.

As explained
\href{https://www.r-spatial.org/r/2017/03/19/invalid.html}{here}:
``Spatial line and polygon data are often messy; although simple
features formally follow a standard, there is no guarantee that data is
clean when imported in R.''

Note: The typecasting to \texttt{MULTIPOLYGON} is required for the
\texttt{sabre} methods. Without this, the geometry type is the more
general \texttt{GEOMETRY} although all the items are already
\texttt{MULTIPOLYGON}s.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{st\_is\_valid}\NormalTok{(gn.sf, }\AttributeTok{reason=}\ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] "Valid Geometry" "Valid Geometry" "Valid Geometry" "Valid Geometry"
 [5] "Valid Geometry" "Valid Geometry" "Valid Geometry" "Valid Geometry"
 [9] "Valid Geometry" "Valid Geometry" "Valid Geometry" "Valid Geometry"
[13] "Valid Geometry"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gn.sf.v }\OtherTok{\textless{}{-}}\NormalTok{ sf}\SpecialCharTok{::}\FunctionTok{st\_make\_valid}\NormalTok{(gn.sf)  }\SpecialCharTok{|\textgreater{}} \FunctionTok{st\_cast}\NormalTok{(}\StringTok{"MULTIPOLYGON"}\NormalTok{)}
\CommentTok{\#}
\FunctionTok{st\_is\_valid}\NormalTok{(sg.sf, }\AttributeTok{reason=}\ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] "Valid Geometry" "Valid Geometry" "Valid Geometry" "Valid Geometry"
 [5] "Valid Geometry" "Valid Geometry" "Valid Geometry" "Valid Geometry"
 [9] "Valid Geometry" "Valid Geometry"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.sf.v }\OtherTok{\textless{}{-}}\NormalTok{ sf}\SpecialCharTok{::}\FunctionTok{st\_make\_valid}\NormalTok{(sg.sf)  }\SpecialCharTok{|\textgreater{}} \FunctionTok{st\_cast}\NormalTok{(}\StringTok{"MULTIPOLYGON"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

In this case the topology was valid, but if you are running this with
your own maps it may not be.

Display the polygon maps. Compute each legend from the classes present
in that map, but use a consistent colour scale. This requires some
\texttt{ggplot2} tricks with \texttt{colorRampPalette} to set up a
palette with a large number of discrete values, and the \texttt{limits}
argument to \texttt{scale\_fill\_manual} to only show the parts of that
scale occurring in each map.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{classes.both }\OtherTok{\textless{}{-}} \FunctionTok{union}\NormalTok{(}\FunctionTok{values}\NormalTok{(sg.poly)}\SpecialCharTok{$}\NormalTok{class, }\FunctionTok{values}\NormalTok{(gn.poly)}\SpecialCharTok{$}\NormalTok{class)}
\NormalTok{my.pal }\OtherTok{\textless{}{-}} \FunctionTok{colorRampPalette}\NormalTok{(}\FunctionTok{brewer.pal}\NormalTok{(}\DecValTok{8}\NormalTok{, }\StringTok{"PuBu"}\NormalTok{))(}\FunctionTok{length}\NormalTok{(classes.both))}
\NormalTok{g0 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data=}\NormalTok{gn.sf.v) }\SpecialCharTok{+}
  \FunctionTok{geom\_sf}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ class)) }\SpecialCharTok{+}
  \FunctionTok{coord\_sf}\NormalTok{(}\AttributeTok{crs =} \FunctionTok{st\_crs}\NormalTok{(gn.sf)) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{title =} \StringTok{"gNATSGO"}\NormalTok{)  }\SpecialCharTok{+}
  \FunctionTok{scale\_fill\_manual}\NormalTok{(}\AttributeTok{values =}\NormalTok{ my.pal, }\AttributeTok{drop=}\ConstantTok{TRUE}\NormalTok{,}
                    \AttributeTok{limits =} \FunctionTok{levels}\NormalTok{(gn.sf.v}\SpecialCharTok{$}\NormalTok{class)) }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.position =} \StringTok{"bottom"}\NormalTok{, }\AttributeTok{legend.direction =} \StringTok{"horizontal"}\NormalTok{)}
\NormalTok{g1 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data=}\NormalTok{sg.sf.v) }\SpecialCharTok{+}
  \FunctionTok{geom\_sf}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ class)) }\SpecialCharTok{+}
  \FunctionTok{coord\_sf}\NormalTok{(}\AttributeTok{crs =} \FunctionTok{st\_crs}\NormalTok{(sg.sf)) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{title =} \StringTok{"SG2"}\NormalTok{)  }\SpecialCharTok{+}
  \FunctionTok{scale\_fill\_manual}\NormalTok{(}\AttributeTok{values =}\NormalTok{ my.pal, }\AttributeTok{drop=}\ConstantTok{TRUE}\NormalTok{,}
                    \AttributeTok{limits =} \FunctionTok{levels}\NormalTok{(gn.sf.v}\SpecialCharTok{$}\NormalTok{class)) }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.position =} \StringTok{"bottom"}\NormalTok{, }\AttributeTok{legend.direction =} \StringTok{"horizontal"}\NormalTok{)}
\FunctionTok{grid.arrange}\NormalTok{(g0,g1, }\AttributeTok{nrow=}\DecValTok{1}\NormalTok{, }\AttributeTok{ncol=}\DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/make_class_maps-1.pdf}

}

\end{figure}

This is the same information we've seen in the raster maps, but now
organized as polygons.

\hypertarget{sec-vmeasure}{%
\subsubsection{Compute the V-metrics}\label{sec-vmeasure}}

Task: Compute the metrics with the \texttt{sabre}(``Spatial Association
Between Regionalizations'') package. The second-listed map is the map to
evaluate, with respect to the first-listed (reference) map. Here we
evaluate the SoilGrids pattern vs.~the gNATSGO as reference.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{regions.sg.gn }\OtherTok{\textless{}{-}}\NormalTok{ sabre}\SpecialCharTok{::}\FunctionTok{vmeasure\_calc}\NormalTok{(}\AttributeTok{x =}\NormalTok{ gn.sf.v, }
                                 \AttributeTok{y =}\NormalTok{ sg.sf.v, }
                                 \AttributeTok{x\_name =}\NormalTok{ class, }\AttributeTok{y\_name =}\NormalTok{ class)}
\FunctionTok{print}\NormalTok{(regions.sg.gn)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
The SABRE results:

 V-measure: 0.04 
 Homogeneity: 0.04 
 Completeness: 0.04 

 The spatial objects can be retrieved with:
 $map1 - the first map
 $map2 - the second map 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{names}\NormalTok{(regions.sg.gn)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "map1"         "map2"         "v_measure"    "homogeneity"  "completeness"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{names}\NormalTok{(regions.sg.gn}\SpecialCharTok{$}\NormalTok{map1)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "map1"     "geometry" "rih"     
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{attr}\NormalTok{(regions.sg.gn, }\StringTok{"precision"}\NormalTok{)  }\CommentTok{\# NULL, means a system default}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
NULL
\end{verbatim}

Both the homogeneity and completeness are near 0, their harmomic mean
V-measure also very low. These maps hardly reseble each other. We will
see the details just below.

Item \texttt{rih} (``region \emph{in}homogeneity'') is the intersection
map. Show these, but first the geometric precision must be set.

Geometric precision is set by \texttt{st\_as\_binary}, default is
\texttt{attr(x,\ "precision")}. Here we left it as the default
\texttt{NULL}.

Task: Plot the inhomogeneity and incompleteness.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{terra}\SpecialCharTok{::}\FunctionTok{plot}\NormalTok{(regions.sg.gn}\SpecialCharTok{$}\NormalTok{map1[}\StringTok{"rih"}\NormalTok{], }\AttributeTok{main =} \StringTok{"Inhomogeneity {-}{-} SG2 vs. gNATSGO"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/vmaps.gnatsgo.sg.homogeneity-1.pdf}

}

\end{figure}

In this \emph{inhomogeneity} map, almost all areas are highly
inhomogeneous. This means that the SG2 polygons are highly variable --
they contain almost the same distribution of classes from the gNATSGO
map. Only a few patches are somewhat more homogeneous.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{terra}\SpecialCharTok{::}\FunctionTok{plot}\NormalTok{(regions.sg.gn}\SpecialCharTok{$}\NormalTok{map2[}\StringTok{"rih"}\NormalTok{], }\AttributeTok{main =} \StringTok{"Incompleteness {-}{-} SG2 vs. gNATSGO"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/vmaps.gnatsgo.sg.completeness-1.pdf}

}

\end{figure}

In this \emph{incompleteness} map, all areas are quite incomplete, but
there are more differences. The blue polygons (lowest values) are the
most complete areas of the gNATSGO map, i.e.~where the reference map has
the most homogeneous set of SG2 classified values. The highly-incomplete
areas are where gNATSGO has a limited range of values (in this case, a
narrow range of higher pH) so they have low variance compared to SG2
predictions for these areas.

We conclude that the maps are quite different in their spatial patterns.

\hypertarget{supercells}{%
\section{Supercells}\label{supercells}}

\emph{``Superpixels''} is a generic name for grouping pixels with
similar characteristics into larger assemblages. In the soil map
context, the aim is to regionalize into areas with similar values of one
or more raster layers.

The \texttt{supercells::supercells} function controls the segmentation:
the user can specify the \texttt{k} arguement for the number of
supercells, and the \texttt{compactness} argument to control shape:
larger values lead to more square, less long/twisted shapes. It is also
possible to specify a set of initial supercell centres (with an
\texttt{sf} POINTS geometry) or a separation between initial centres
with the \texttt{step} argument.

This function implements the SLIC algorithm
(\protect\hyperlink{ref-achantaSLICSuperpixelsCompared2012}{Achanta et
al., 2012}).

As an example with the pH map, we divide into about 50 supercells, with
low compactness since we don't expect near-square natural units. Here is
the source map:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ggplot}\NormalTok{() }\SpecialCharTok{+}
  \FunctionTok{geom\_spatraster}\NormalTok{(}\AttributeTok{data=}\NormalTok{sg.utm) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{fill =} \StringTok{"pH"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/SLIC-source-1.pdf}

}

\end{figure}

And here are the 50 supercells, with very low compactness:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.utm}\FloatTok{.50} \OtherTok{=} \FunctionTok{supercells}\NormalTok{(sg.utm, }\AttributeTok{k =} \DecValTok{50}\NormalTok{, }\AttributeTok{compactness =} \FloatTok{0.5}\NormalTok{)}
\FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data=}\NormalTok{sg.utm}\FloatTok{.50}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{geom\_sf}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ phh2o\_0}\FloatTok{.5}\NormalTok{cm\_mean)) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{fill =} \StringTok{"mean pH"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/supercells-not-compact-1.pdf}

}

\end{figure}

Try to form more compact supercells:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sg.utm}\FloatTok{.50} \OtherTok{=} \FunctionTok{supercells}\NormalTok{(sg.utm, }\AttributeTok{k =} \DecValTok{50}\NormalTok{, }\AttributeTok{compactness =} \DecValTok{5}\NormalTok{)}
\FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data=}\NormalTok{sg.utm}\FloatTok{.50}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{geom\_sf}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ phh2o\_0}\FloatTok{.5}\NormalTok{cm\_mean)) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{fill =} \StringTok{"mean pH"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/supercells-compact-1.pdf}

}

\end{figure}

These do not look realistic.

Try with multiple rasters, here pH and silt concentration:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{r }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(sg.utm, sg.silt.utm)}
\NormalTok{r}\FloatTok{.50} \OtherTok{=} \FunctionTok{supercells}\NormalTok{(r, }\AttributeTok{k =} \DecValTok{50}\NormalTok{, }\AttributeTok{compactness =} \FloatTok{0.5}\NormalTok{)}
\FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data=}\NormalTok{r}\FloatTok{.50}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{geom\_sf}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ phh2o\_0}\FloatTok{.5}\NormalTok{cm\_mean)) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{fill =} \StringTok{"mean pH"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{scale\_fill\_continuous}\NormalTok{(}\AttributeTok{type =} \StringTok{"viridis"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/supercells-multiple-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data=}\NormalTok{r}\FloatTok{.50}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{geom\_sf}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{fill =}\NormalTok{ silt\_0}\FloatTok{.5}\NormalTok{cm\_mean)) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{fill =} \StringTok{"silt ppt"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{scale\_fill\_continuous}\NormalTok{(}\AttributeTok{type =} \StringTok{"viridis"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/supercells-multiple-2.pdf}

}

\end{figure}

Notice that the segments are the same in the two visualizations.

\hypertarget{sec-scale-geometry}{%
\section{Scale issues -- geometric}\label{sec-scale-geometry}}

The soil pattern can be observed at different scales.

DSM are produced at a wide range of grid cell sizes (``resolutions''),
and it's obvious that as the grid cell size increases any finer pattern
is lost. This is especially true if the map is made with block
predictions, rather than point predictions at the centre of grid cells.

Clearly, products can only be compared at the same resolution. A
finer-scale product can be downscaled to the resolution of a
coarser-scale product in order to compare them. For maps made at point
resolution this should by mean (continuous) or majority filter
(classified). For maps made at block resolution this should be by block
kriging within the block at fine-scale (continuous) or mode filter
(classified).

What happens to the landscape metrics as the resolution changes? Let's
examine one map from the most detailed soil survey, gSSURGO.

The design scale for the polygon SSURGO product varies across the USA.
For most areas with detailed survey, it is from 1:12k to 1:24K, with
Minimum Legible Area (MLA) 0.576--2.304 ha. With a grid resolution of 16
grid cells per MLA, these would be from 12x12 -- 24x24 cells. The 30 m
gSSURGO roughly corresponds to this coarser resolution. It is equivalent
to 1:30 k design scale.

In a previous section (Section~\ref{sec-import-gssurgo}) we imported a
gSSURGO polygon map.

Unfortunately, the \texttt{landscapemetrics} package can not (yet?) work
with \texttt{terra} objects, or vector objects from any package. So to
compute landscape metrics for these polygons, they must be rasterized to
a resolution where the area and length calculations are sufficiently
accurate. The median polygon area is 2.7 ha; the minimum legible area
(MLA) at the design scale of the soil surveys in this area (1:20k, see
published survey document) is 1.6 ha. At a grid resolution of 16 pixels
per MLA this suggests 20 m horizontal resolution pixels.

Task: Rasterize the polygon map to this resolution.

First set up the empty raster and then add the values of the map unit
key.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mu.template }\OtherTok{\textless{}{-}} \FunctionTok{rast}\NormalTok{(mu.poly, }\AttributeTok{res=}\FunctionTok{c}\NormalTok{(}\DecValTok{20}\NormalTok{,}\DecValTok{20}\NormalTok{))}
\FunctionTok{dim}\NormalTok{(mu.template)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 1408 1043    1
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mu.raster }\OtherTok{\textless{}{-}} \FunctionTok{rasterize}\NormalTok{(mu.poly, mu.template, }\AttributeTok{field=}\StringTok{"mukey"}\NormalTok{)}
\FunctionTok{summary}\NormalTok{(mu.raster)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
     mukey        
 Min.   : 295575  
 1st Qu.: 295605  
 Median : 295651  
 Mean   : 616216  
 3rd Qu.: 295817  
 Max.   :2760841  
 NA's   :3149     
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{check\_landscape}\NormalTok{(mu.raster)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  layer       crs units   class n_classes  OK
1     1 projected     m integer       217 (?)
\end{verbatim}

This is a categorical map (the map unit keys are the categories), and we
can apply the landscape metrics to it.

\textbf{Map units}:

Note that there are 217 classes, i.e., different map units, in this
area.

TaskL Show all patches, then just the ``Alluvial land'', code 295575,
then all the map units with the Mardin series as a component.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{unique}\NormalTok{(mu.raster}\SpecialCharTok{$}\NormalTok{mukey))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
   mukey
1 295575
2 295576
3 295577
4 295578
5 295579
6 295580
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{show\_patches}\NormalTok{(mu.raster, }\AttributeTok{class =} \StringTok{"global"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show.landscape.gn-1.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
 \FunctionTok{show\_patches}\NormalTok{(mu.raster, }\AttributeTok{class =} \DecValTok{295575}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show.landscape.gn-2.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(ix }\OtherTok{\textless{}{-}} \FunctionTok{grep}\NormalTok{(}\StringTok{"Mardin"}\NormalTok{, mu.key}\SpecialCharTok{$}\NormalTok{muname, }\AttributeTok{fixed =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1]  63  64  65  67 100 101 102 120 121 122 123 124 125 126 149 178 179 180 181
[20] 182
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(mu.key[ix, ])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
      mukey                                                             muname
63   295656                   Mardin channery silt loam, 2 to 8 percent slopes
64   295657                  Mardin channery silt loam, 8 to 15 percent slopes
65   295658          Mardin channery silt loam, 8 to 15 percent slopes, eroded
67   295660                 Mardin and Langford soils, 15 to 25 percent slopes
100  295805                    Mardin-Volusia complex, 15 to 30 percent slopes
101  295806                     Mardin-Volusia complex, 8 to 15 percent slopes
102  295807                      Mardin-Volusia complex, 0 to 8 percent slopes
120  295831         Mardin channery silt loam, 16 to 30 percent slopes, eroded
121  295832                 Mardin channery silt loam, 16 to 30 percent slopes
122  295833 Mardin channery silt loam, 9 to 15 percent slopes, moderately deep
123  295834          Mardin channery silt loam, 8 to 15 percent slopes, eroded
124  295835                  Mardin channery silt loam, 8 to 15 percent slopes
125  295836                   Mardin channery silt loam, 0 to 8 percent slopes
126  295837  Mardin channery silt loam, 0 to 8 percent slopes, moderately deep
149 2723042                     Bath and Mardin soils, 25 to 40 percent slopes
178 2723076                   Mardin channery silt loam, 2 to 8 percent slopes
179 2723077                 Mardin channery silt loam, 15 to 25 percent slopes
180 2723078                  Mardin channery silt loam, 8 to 15 percent slopes
181 2723080    Mardin channery silt loam, 3 to 8 percent slopes, slightly acid
182 2723081   Mardin channery silt loam, 8 to 15 percent slopes, slightly acid
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{show\_patches}\NormalTok{(mu.raster, }
             \AttributeTok{class =} \FunctionTok{c}\NormalTok{(mu.key[ix, }\StringTok{"mukey"}\NormalTok{]), }\AttributeTok{nrow =} \DecValTok{3}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show.landscape.gn-3.pdf}

}

\end{figure}

Some of these are quite small, and are categorically not so different.
Much of this is due to the separate surveys in three counties. In
Section~\ref{sec-categorical-generalization} we combine similar map
units for the landscape analysis.

\hypertarget{m-resolution}{%
\subsection{20 m resolution}\label{m-resolution}}

We rasterized the map units at 20 m grid resolution, so first analyze at
this scale.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gn}\FloatTok{.20} \OtherTok{\textless{}{-}}\NormalTok{ mu.raster}
\end{Highlighting}
\end{Shaded}

\hypertarget{landscape-level}{%
\subsubsection{Landscape level}\label{landscape-level}}

Compute the landscape metrics.:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{lst }\OtherTok{\textless{}{-}} \FunctionTok{paste0}\NormalTok{(}\StringTok{"lsm\_l\_"}\NormalTok{, }\FunctionTok{c}\NormalTok{(}\StringTok{"shdi"}\NormalTok{, }\StringTok{"shei"}\NormalTok{, }\StringTok{"lsi"}\NormalTok{, }\StringTok{"ai"}\NormalTok{,  }\StringTok{"frac\_mn"}\NormalTok{))}
\FunctionTok{print}\NormalTok{(ls.metrics.gn20 }\OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(gn}\FloatTok{.20}\NormalTok{, }\AttributeTok{what=}\NormalTok{lst)[, }\FunctionTok{c}\NormalTok{(}\StringTok{"metric"}\NormalTok{, }\StringTok{"value"}\NormalTok{)])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 5 x 2
  metric   value
  <chr>    <dbl>
1 ai      86.3  
2 frac_mn  1.11 
3 lsi     86.7  
4 shdi     4.39 
5 shei     0.815
\end{verbatim}

Because of the small patches, very fine pattern,large number of classes,
the \texttt{lsi} is extremely high -- i.e., far more boundaries than if
the classes were contiguous. The \texttt{ai} is also quite high. The
Shannon Diversity is also very high, because of the large number of
classes and relatively equal areas. The Shannon Evenness is quite high,
i.e., the areas of the classes are fairly even, there is no dominant
class.

\hypertarget{class-level}{%
\subsubsection{Class level}\label{class-level}}

The class-level metrics show the distribution of classes. For example,
the percentage of landscape occupied by each class
(\texttt{lsm\_c\_pland}):

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{c\_pland}\FloatTok{.20} \OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(gn}\FloatTok{.20}\NormalTok{, }\AttributeTok{what=}\StringTok{"lsm\_c\_pland"}\NormalTok{)}
\FunctionTok{head}\NormalTok{(}\FunctionTok{sort}\NormalTok{(c\_pland}\FloatTok{.20}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{), }\DecValTok{24}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] 5.303681 5.034498 4.229760 4.166365 3.978148 3.566361 3.380955 2.936838
 [9] 2.779685 2.343791 1.947256 1.936221 1.890326 1.692129 1.666687 1.635481
[17] 1.601113 1.533009 1.490206 1.429131 1.368758 1.227630 1.169155 1.063941
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ix }\OtherTok{\textless{}{-}} \FunctionTok{order}\NormalTok{(c\_pland}\FloatTok{.20}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{)}
\FunctionTok{head}\NormalTok{(c\_pland}\FloatTok{.20}\SpecialCharTok{$}\NormalTok{class[ix], }\DecValTok{24}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] 295598 295635 295685 295582 295647 295683 295649 295656 295650 295690
[11] 295829 295657 295611 295581 295585 295828 295584 295827 295579 295604
[21] 295575 295854 295651 295596
\end{verbatim}

We see that there are a few large classes; the largest is about 5.6\% of
the landscape. But there are many more small ones.

\hypertarget{patch-level}{%
\subsubsection{Patch level}\label{patch-level}}

In this case patch-level metrics are also interesting -- they reveal
size and shape, for example.

One example is \textbf{patch area}, expressed in ha
(\(10000 \; \mathrm{m}^2\)) -- these should all be larger than the MLA
(0.576--2.304 ha depending on original design scale).

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# each patch}
\FunctionTok{head}\NormalTok{(}\FunctionTok{sort}\NormalTok{(area}\FloatTok{.20} \OtherTok{\textless{}{-}}  
            \FunctionTok{calculate\_lsm}\NormalTok{(gn}\FloatTok{.20}\NormalTok{, }\AttributeTok{what=}\StringTok{"lsm\_p\_area"}\NormalTok{)}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 336.88 321.04 212.68 199.56 173.80 162.32
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{quantile}\NormalTok{(area}\FloatTok{.20}\NormalTok{ , }\FunctionTok{seq}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{,}\AttributeTok{by=}\NormalTok{.}\DecValTok{12}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
     0%     12%     24%     36%     48%     60%     72%     84%     96% 
 0.0400  0.5600  1.0000  1.5600  2.3232  3.4400  5.4800  9.4400 24.7600 
\end{verbatim}

About 12\% of the patches are smaller. This may be due to the
rasterizing process (?).

Another example is the \textbf{number of core areas} within patches.
This measures how complex is each patch:

``A cell is defined as core if the cell has no neighbour with a
different value than itself (rook's case). The metric counts the
disjunct core areas, whereby a core area is a `patch within the patch'
containing only core cells. It describes patch area and shape
simultaneously (more core area when the patch is large, however, the
shape must allow disjunct core areas). Thereby, a compact shape (e.g.~a
square) will contain {[}fewer{]} disjunct core areas than a more
irregular patch.''

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# each patch}
\FunctionTok{head}\NormalTok{(ncore}\FloatTok{.20} \OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(gn}\FloatTok{.20}\NormalTok{, }\AttributeTok{what=}\StringTok{"lsm\_p\_ncore"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 6 x 6
  layer level  class    id metric value
  <int> <chr>  <int> <int> <chr>  <dbl>
1     1 patch 295575     1 ncore      2
2     1 patch 295575     2 ncore      2
3     1 patch 295575     3 ncore     10
4     1 patch 295575     4 ncore      1
5     1 patch 295575     5 ncore      1
6     1 patch 295575     6 ncore     12
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# summarize by map unit}
\NormalTok{ncore.}\FloatTok{20.}\NormalTok{summary }\OtherTok{\textless{}{-}}\NormalTok{ ncore}\FloatTok{.20} \SpecialCharTok{\%\textgreater{}\%} \FunctionTok{group\_by}\NormalTok{(class) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{summarize}\NormalTok{(}\AttributeTok{max\_cores =} \FunctionTok{max}\NormalTok{(value)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{arrange}\NormalTok{(class)}
\FunctionTok{print}\NormalTok{(}\FunctionTok{sort}\NormalTok{(ncore.}\FloatTok{20.}\NormalTok{summary}\SpecialCharTok{$}\NormalTok{max\_cores, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  [1] 40 22 18 14 14 13 12 12 12 12 12 12 11 11 10 10 10 10 10 10 10  9  9  9  9
 [26]  9  9  9  8  8  8  8  8  8  8  7  7  7  7  7  7  6  6  6  6  6  6  6  6  6
 [51]  6  6  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  4  4
 [76]  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  3
[101]  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
[126]  3  3  3  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
[151]  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  1  1  1  1  1
[176]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[201]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  0  0  0
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ix }\OtherTok{\textless{}{-}} \FunctionTok{order}\NormalTok{(ncore.}\FloatTok{20.}\NormalTok{summary}\SpecialCharTok{$}\NormalTok{max\_cores, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{)}
\FunctionTok{cbind}\NormalTok{(mu.key[ix[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{8}\NormalTok{], ], }\AttributeTok{ncore =}\NormalTok{ ncore.}\FloatTok{20.}\NormalTok{summary}\SpecialCharTok{$}\NormalTok{max\_cores[ix[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{8}\NormalTok{]])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
      mukey                                                           muname
21   295598                   Erie channery silt loam, 3 to 8 percent slopes
93   295690 Wayland soils complex, 0 to 3 percent slopes, frequently flooded
1    295575                                                    Alluvial land
52   295635               Langford channery silt loam, 2 to 8 percent slopes
195 2723105        Valois and Howard gravelly loams, 25 to 40 percent slopes
36   295613                    Howard gravelly loam, 15 to 25 percent slopes
8    295582                    Bath and Valois soils, 5 to 15 percent slopes
10   295584           Bath and Valois soils, 15 to 25 percent slopes, eroded
    ncore
21     40
93     22
1      18
52     14
195    14
36     13
8      12
10     12
\end{verbatim}

The Erie unit has by far the most core areas. Display it:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{show\_patches}\NormalTok{(gn}\FloatTok{.20}\NormalTok{, }\AttributeTok{class =}\NormalTok{ mu.key[ix[}\DecValTok{1}\NormalTok{], }\StringTok{"mukey"}\NormalTok{])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show-patches-Erie-20-1.pdf}

}

\end{figure}

Display the \texttt{ncore} patch-level metric as a map:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{show\_lsm}\NormalTok{(gn}\FloatTok{.20}\NormalTok{, }\StringTok{"lsm\_p\_ncore"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show-ncore-20-1.pdf}

}

\end{figure}

We can see the highly-fragmented map unit on the upper left.

\hypertarget{m-resolution-1}{%
\subsection{100 m resolution}\label{m-resolution-1}}

What happens as the resolution is coarsened to 100 x 100, roughly
equivalent to 1:100k design scale?

We use the majority (``modal'') filter in the reclassification, with a
one-dimension 5-fold aggregation.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gn}\FloatTok{.100} \OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{aggregate}\NormalTok{(gn}\FloatTok{.20}\NormalTok{, }\AttributeTok{fact=}\DecValTok{5}\NormalTok{, }\AttributeTok{fun=}\StringTok{"modal"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Patches are clearly larger than in the 20 m resolution.

\hypertarget{landscape-level-1}{%
\subsubsection{Landscape level}\label{landscape-level-1}}

Landscape metrics:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(ls.metrics.gn20[, }\FunctionTok{c}\NormalTok{(}\StringTok{"metric"}\NormalTok{, }\StringTok{"value"}\NormalTok{)])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 5 x 2
  metric   value
  <chr>    <dbl>
1 ai      86.3  
2 frac_mn  1.11 
3 lsi     86.7  
4 shdi     4.39 
5 shei     0.815
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(ls.metrics.gn100 }\OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(gn}\FloatTok{.100}\NormalTok{, }\AttributeTok{what=}\NormalTok{lst)[, }\FunctionTok{c}\NormalTok{(}\StringTok{"metric"}\NormalTok{, }\StringTok{"value"}\NormalTok{)])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 5 x 2
  metric   value
  <chr>    <dbl>
1 ai      52.3  
2 frac_mn  1.04 
3 lsi     60.0  
4 shdi     4.37 
5 shei     0.815
\end{verbatim}

The Aggregation Index and Landscape Shape Index are much lower. The
Fractal Dimension is somewhat lower. The Shannon diversity and evenness
have hardly changed.

\hypertarget{class-level-1}{%
\subsubsection{Class level}\label{class-level-1}}

What about the class metrics?

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{c\_pland}\FloatTok{.100} \OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(gn}\FloatTok{.100}\NormalTok{, }\AttributeTok{what=}\StringTok{"lsm\_c\_pland"}\NormalTok{)}
\FunctionTok{head}\NormalTok{(}\FunctionTok{sort}\NormalTok{(c\_pland}\FloatTok{.20}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{), }\DecValTok{24}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] 5.303681 5.034498 4.229760 4.166365 3.978148 3.566361 3.380955 2.936838
 [9] 2.779685 2.343791 1.947256 1.936221 1.890326 1.692129 1.666687 1.635481
[17] 1.601113 1.533009 1.490206 1.429131 1.368758 1.227630 1.169155 1.063941
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{sort}\NormalTok{(c\_pland}\FloatTok{.100}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{), }\DecValTok{24}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] 5.341283 5.029900 4.299211 4.230211 4.062135 3.612753 3.368600 2.924525
 [9] 2.798910 2.358374 1.930222 1.926683 1.903684 1.701992 1.668377 1.663069
[17] 1.624146 1.551608 1.496762 1.454301 1.417147 1.213687 1.201302 1.084533
\end{verbatim}

Very little difference in the class composition.

\hypertarget{patch-level-1}{%
\subsubsection{Patch level}\label{patch-level-1}}

Patch metrics:

First, the patch areas.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# each patch}
\FunctionTok{head}\NormalTok{(}\FunctionTok{sort}\NormalTok{(area}\FloatTok{.100} \OtherTok{\textless{}{-}}  
            \FunctionTok{calculate\_lsm}\NormalTok{(gn}\FloatTok{.100}\NormalTok{, }\AttributeTok{what=}\StringTok{"lsm\_p\_area"}\NormalTok{)}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 350 286 271 260 253 240
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{quantile}\NormalTok{(area}\FloatTok{.20}\NormalTok{, }\FunctionTok{seq}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{,}\AttributeTok{by=}\NormalTok{.}\DecValTok{12}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
     0%     12%     24%     36%     48%     60%     72%     84%     96% 
 0.0400  0.5600  1.0000  1.5600  2.3232  3.4400  5.4800  9.4400 24.7600 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{quantile}\NormalTok{(area}\FloatTok{.100}\NormalTok{, }\FunctionTok{seq}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{,}\AttributeTok{by=}\NormalTok{.}\DecValTok{12}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 0% 12% 24% 36% 48% 60% 72% 84% 96% 
  1   1   1   2   3   4   6  11  28 
\end{verbatim}

Now all the patches are larger than the MLA for 1:12k.

The core areas:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# each patch}
\FunctionTok{head}\NormalTok{(ncore}\FloatTok{.100} \OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(gn}\FloatTok{.100}\NormalTok{, }\AttributeTok{what=}\StringTok{"lsm\_p\_ncore"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 6 x 6
  layer level  class    id metric value
  <int> <chr>  <int> <int> <chr>  <dbl>
1     1 patch 295575     1 ncore      2
2     1 patch 295575     2 ncore      0
3     1 patch 295575     3 ncore      0
4     1 patch 295575     4 ncore      0
5     1 patch 295575     5 ncore      0
6     1 patch 295575     6 ncore      0
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# summarize by map unit}
\NormalTok{ncore.}\FloatTok{100.}\NormalTok{summary }\OtherTok{\textless{}{-}}\NormalTok{ ncore}\FloatTok{.100} \SpecialCharTok{\%\textgreater{}\%} \FunctionTok{group\_by}\NormalTok{(class) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{summarize}\NormalTok{(}\AttributeTok{max\_cores =} \FunctionTok{max}\NormalTok{(value)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{arrange}\NormalTok{(class)}
\FunctionTok{print}\NormalTok{(}\FunctionTok{sort}\NormalTok{(ncore.}\FloatTok{20.}\NormalTok{summary}\SpecialCharTok{$}\NormalTok{max\_cores, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  [1] 40 22 18 14 14 13 12 12 12 12 12 12 11 11 10 10 10 10 10 10 10  9  9  9  9
 [26]  9  9  9  8  8  8  8  8  8  8  7  7  7  7  7  7  6  6  6  6  6  6  6  6  6
 [51]  6  6  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  4  4
 [76]  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  3
[101]  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
[126]  3  3  3  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
[151]  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  1  1  1  1  1
[176]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[201]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  0  0  0
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(}\FunctionTok{sort}\NormalTok{(ncore.}\FloatTok{100.}\NormalTok{summary}\SpecialCharTok{$}\NormalTok{max\_cores, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  [1] 13 12  9  9  9  7  7  6  5  5  4  4  4  4  4  4  3  3  3  3  3  3  3  3  3
 [26]  3  3  3  3  3  3  3  3  3  3  3  3  2  2  2  2  2  2  2  2  2  2  2  2  2
 [51]  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
 [76]  2  2  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[101]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[126]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  0  0  0  0  0
[151]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
[176]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
[201]  0  0  0  0  0  0  0  0  0  0  0  0  0  0
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ix }\OtherTok{\textless{}{-}} \FunctionTok{order}\NormalTok{(ncore.}\FloatTok{100.}\NormalTok{summary}\SpecialCharTok{$}\NormalTok{max\_cores, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{)}
\FunctionTok{cbind}\NormalTok{(mu.key[ix[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{8}\NormalTok{], ], }\AttributeTok{ncore =}\NormalTok{ ncore.}\FloatTok{100.}\NormalTok{summary}\SpecialCharTok{$}\NormalTok{max\_cores[ix[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{8}\NormalTok{]])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
    mukey                                                           muname
21 295598                   Erie channery silt loam, 3 to 8 percent slopes
52 295635               Langford channery silt loam, 2 to 8 percent slopes
8  295582                    Bath and Valois soils, 5 to 15 percent slopes
56 295647             Lordstown channery silt loam, 5 to 15 percent slopes
86 295683                Volusia channery silt loam, 3 to 8 percent slopes
63 295656                 Mardin channery silt loam, 2 to 8 percent slopes
88 295685               Volusia channery silt loam, 8 to 15 percent slopes
93 295690 Wayland soils complex, 0 to 3 percent slopes, frequently flooded
   ncore
21    13
52    12
8      9
56     9
86     9
63     7
88     7
93     6
\end{verbatim}

The Erie unit again has the most core areas, but many small cores have
been absorbed in their neighbours.

Display it:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{show\_patches}\NormalTok{(gn}\FloatTok{.100}\NormalTok{, }\AttributeTok{class =}\NormalTok{ mu.key[ix[}\DecValTok{1}\NormalTok{], }\StringTok{"mukey"}\NormalTok{])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show-patches-Erie-100-1.pdf}

}

\end{figure}

Display the \texttt{ncore} patch-level metric as a map:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{show\_lsm}\NormalTok{(gn}\FloatTok{.100}\NormalTok{, }\StringTok{"lsm\_p\_ncore"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show-ncore-100-1.pdf}

}

\end{figure}

\hypertarget{m-resolution-2}{%
\subsection{300 m resolution}\label{m-resolution-2}}

What happens as the resolution is again coarsened to 300 x 300 m,
roughly equivalent to 1:300 k design scale?

We use a 3x3 aggregation, again with a modal filter.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gn}\FloatTok{.300} \OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{aggregate}\NormalTok{(gn}\FloatTok{.100}\NormalTok{, }\AttributeTok{fact=}\DecValTok{3}\NormalTok{, }\AttributeTok{fun=}\StringTok{"modal"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{landscape-level-2}{%
\subsubsection{Landscape level}\label{landscape-level-2}}

Landscape metrics:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(ls.metrics.gn20[, }\FunctionTok{c}\NormalTok{(}\StringTok{"metric"}\NormalTok{, }\StringTok{"value"}\NormalTok{)])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 5 x 2
  metric   value
  <chr>    <dbl>
1 ai      86.3  
2 frac_mn  1.11 
3 lsi     86.7  
4 shdi     4.39 
5 shei     0.815
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(ls.metrics.gn100[, }\FunctionTok{c}\NormalTok{(}\StringTok{"metric"}\NormalTok{, }\StringTok{"value"}\NormalTok{)])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 5 x 2
  metric   value
  <chr>    <dbl>
1 ai      52.3  
2 frac_mn  1.04 
3 lsi     60.0  
4 shdi     4.37 
5 shei     0.815
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(ls.metrics.gn300 }\OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(gn}\FloatTok{.300}\NormalTok{, }\AttributeTok{what=}\NormalTok{lst)[, }\FunctionTok{c}\NormalTok{(}\StringTok{"metric"}\NormalTok{, }\StringTok{"value"}\NormalTok{)])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 5 x 2
  metric   value
  <chr>    <dbl>
1 ai      30.2  
2 frac_mn  1.02 
3 lsi     29.2  
4 shdi     4.28 
5 shei     0.830
\end{verbatim}

The Aggregation Index and Landscape Shape Index are again much lower.
The Fractal Dimension is again somewhat lower. The Shannon diversity is
somewhat lower and the evenness somewhat higher.

\hypertarget{class-level-2}{%
\subsubsection{Class level}\label{class-level-2}}

What about the class metrics?

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{c\_pland}\FloatTok{.300} \OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(gn}\FloatTok{.300}\NormalTok{, }\AttributeTok{what=}\StringTok{"lsm\_c\_pland"}\NormalTok{)}
\FunctionTok{head}\NormalTok{(}\FunctionTok{sort}\NormalTok{(c\_pland}\FloatTok{.20}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{), }\DecValTok{24}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] 5.303681 5.034498 4.229760 4.166365 3.978148 3.566361 3.380955 2.936838
 [9] 2.779685 2.343791 1.947256 1.936221 1.890326 1.692129 1.666687 1.635481
[17] 1.601113 1.533009 1.490206 1.429131 1.368758 1.227630 1.169155 1.063941
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{sort}\NormalTok{(c\_pland}\FloatTok{.100}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{), }\DecValTok{24}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] 5.341283 5.029900 4.299211 4.230211 4.062135 3.612753 3.368600 2.924525
 [9] 2.798910 2.358374 1.930222 1.926683 1.903684 1.701992 1.668377 1.663069
[17] 1.624146 1.551608 1.496762 1.454301 1.417147 1.213687 1.201302 1.084533
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{sort}\NormalTok{(c\_pland}\FloatTok{.300}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{), }\DecValTok{24}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] 5.869953 5.140263 4.815956 4.653802 4.297065 3.534944 3.340360 3.178207
 [9] 2.594454 2.156640 2.124210 1.978271 1.978271 1.945841 1.783687 1.767472
[17] 1.702611 1.637749 1.605319 1.508027 1.491811 1.167504 1.167504 1.135074
\end{verbatim}

The largest classes are larger.

\hypertarget{patch-level-2}{%
\subsubsection{Patch level}\label{patch-level-2}}

First, the patch areas.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# each patch}
\FunctionTok{head}\NormalTok{(}\FunctionTok{sort}\NormalTok{(area}\FloatTok{.300} \OtherTok{\textless{}{-}}  
            \FunctionTok{calculate\_lsm}\NormalTok{(gn}\FloatTok{.300}\NormalTok{, }\AttributeTok{what=}\StringTok{"lsm\_p\_area"}\NormalTok{)}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 621 432 423 369 342 306
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{quantile}\NormalTok{(area}\FloatTok{.20}\NormalTok{, }\FunctionTok{seq}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{,}\AttributeTok{by=}\NormalTok{.}\DecValTok{12}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
     0%     12%     24%     36%     48%     60%     72%     84%     96% 
 0.0400  0.5600  1.0000  1.5600  2.3232  3.4400  5.4800  9.4400 24.7600 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{quantile}\NormalTok{(area}\FloatTok{.100}\NormalTok{, }\FunctionTok{seq}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{,}\AttributeTok{by=}\NormalTok{.}\DecValTok{12}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 0% 12% 24% 36% 48% 60% 72% 84% 96% 
  1   1   1   2   3   4   6  11  28 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{quantile}\NormalTok{(area}\FloatTok{.300}\NormalTok{, }\FunctionTok{seq}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{,}\AttributeTok{by=}\NormalTok{.}\DecValTok{12}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 0% 12% 24% 36% 48% 60% 72% 84% 96% 
  9   9   9   9   9  18  18  27  72 
\end{verbatim}

Now the cores:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# each patch}
\FunctionTok{head}\NormalTok{(ncore}\FloatTok{.300} \OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(gn}\FloatTok{.300}\NormalTok{, }\AttributeTok{what=}\StringTok{"lsm\_p\_ncore"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 6 x 6
  layer level  class    id metric value
  <int> <chr>  <int> <int> <chr>  <dbl>
1     1 patch 295575     1 ncore      0
2     1 patch 295575     2 ncore      0
3     1 patch 295575     3 ncore      0
4     1 patch 295575     4 ncore      0
5     1 patch 295575     5 ncore      0
6     1 patch 295575     6 ncore      0
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# summarize by map unit}
\NormalTok{ncore.}\FloatTok{300.}\NormalTok{summary }\OtherTok{\textless{}{-}}\NormalTok{ ncore}\FloatTok{.300} \SpecialCharTok{\%\textgreater{}\%} \FunctionTok{group\_by}\NormalTok{(class) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{summarize}\NormalTok{(}\AttributeTok{max\_cores =} \FunctionTok{max}\NormalTok{(value)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{arrange}\NormalTok{(class)}
\FunctionTok{print}\NormalTok{(ncore.}\FloatTok{20.}\NormalTok{summary)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 217 x 2
    class max_cores
    <int>     <dbl>
 1 295575        18
 2 295576         2
 3 295577         4
 4 295578         1
 5 295579         4
 6 295580         3
 7 295581        10
 8 295582        12
 9 295583         1
10 295584        12
# i 207 more rows
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(ncore.}\FloatTok{100.}\NormalTok{summary)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 214 x 2
    class max_cores
    <int>     <dbl>
 1 295575         2
 2 295576         1
 3 295577         1
 4 295578         0
 5 295579         3
 6 295580         0
 7 295581         2
 8 295582         9
 9 295583         1
10 295584         3
# i 204 more rows
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(ncore.}\FloatTok{300.}\NormalTok{summary)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 173 x 2
    class max_cores
    <int>     <dbl>
 1 295575         0
 2 295576         0
 3 295577         0
 4 295578         0
 5 295579         1
 6 295581         0
 7 295582         2
 8 295583         0
 9 295584         0
10 295585         0
# i 163 more rows
\end{verbatim}

Almost all of the isolated core areas have been merged with adjacent
pixels, with the modal filter.

The Erie unit again has the most core areas, but many small cores have
been absorbed in their neighbours.

Display it:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{show\_patches}\NormalTok{(gn}\FloatTok{.300}\NormalTok{, }\AttributeTok{class =}\NormalTok{ mu.key[ix[}\DecValTok{1}\NormalTok{], }\StringTok{"mukey"}\NormalTok{])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show-patches-Erie-300-1.pdf}

}

\end{figure}

Display the \texttt{ncore} patch-level metric as a map:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{show\_lsm}\NormalTok{(gn}\FloatTok{.300}\NormalTok{, }\StringTok{"lsm\_p\_ncore"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/show-ncore-300-1.pdf}

}

\end{figure}

Most of the map has no cores, i.e., single pixels surrounded by
different classes.

\hypertarget{conclusions-about-geometric-scale-issues}{%
\subsection{Conclusions about geometric scale
issues}\label{conclusions-about-geometric-scale-issues}}

\begin{itemize}
\item
  Landscape level: The diversity (class number and relative proportion)
  hardly change, if at all.
\item
  Class level: Class composition hardly changes; some classes slightly
  increase in total area, others slightly lose area.
\item
  Patch level: Core areas (isolated pixels) decrease with increasing
  resolution.
\end{itemize}

\emph{So, which scale is most ``realistic''?}

\hypertarget{sec-categorical-generalization}{%
\section{Scale issues --
categorical}\label{sec-categorical-generalization}}

\emph{Categorical} generalization is when related map units are grouped
into more general units that share sufficient commonality to be
considered ``homogeneous'' at a more general categorical level.

Many soil classification system support this directly by their
hierarchical structure.

Task: Find the map units where the Mardin series is a component.

\begin{verbatim}
      mukey                                                             muname
63   295656                   Mardin channery silt loam, 2 to 8 percent slopes
64   295657                  Mardin channery silt loam, 8 to 15 percent slopes
65   295658          Mardin channery silt loam, 8 to 15 percent slopes, eroded
67   295660                 Mardin and Langford soils, 15 to 25 percent slopes
100  295805                    Mardin-Volusia complex, 15 to 30 percent slopes
101  295806                     Mardin-Volusia complex, 8 to 15 percent slopes
102  295807                      Mardin-Volusia complex, 0 to 8 percent slopes
120  295831         Mardin channery silt loam, 16 to 30 percent slopes, eroded
121  295832                 Mardin channery silt loam, 16 to 30 percent slopes
122  295833 Mardin channery silt loam, 9 to 15 percent slopes, moderately deep
123  295834          Mardin channery silt loam, 8 to 15 percent slopes, eroded
124  295835                  Mardin channery silt loam, 8 to 15 percent slopes
125  295836                   Mardin channery silt loam, 0 to 8 percent slopes
126  295837  Mardin channery silt loam, 0 to 8 percent slopes, moderately deep
149 2723042                     Bath and Mardin soils, 25 to 40 percent slopes
178 2723076                   Mardin channery silt loam, 2 to 8 percent slopes
179 2723077                 Mardin channery silt loam, 15 to 25 percent slopes
180 2723078                  Mardin channery silt loam, 8 to 15 percent slopes
181 2723080    Mardin channery silt loam, 3 to 8 percent slopes, slightly acid
182 2723081   Mardin channery silt loam, 8 to 15 percent slopes, slightly acid
\end{verbatim}

In the case of USA soil survey, many map units are quite similar. For
example, there are 20 map units in the small study area where the Mardin
series is the only one naned, or it is one of the series in a complex.
These map units differ mostly by slope class. Their land use potential
may be somewhat different, mainly due to slope, but their soil
properties are quite similar.

Task: Generalize the soil map units of the polygon map by combining map
units with the same dominant soil series.

First, make a list of the first-named soil series:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{length}\NormalTok{(names }\OtherTok{\textless{}{-}}\NormalTok{ mu.key}\SpecialCharTok{$}\NormalTok{muname)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 217
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# first name by spaces}
\NormalTok{names }\OtherTok{\textless{}{-}} \FunctionTok{strsplit}\NormalTok{(names, }\StringTok{" "}\NormalTok{)}
\FunctionTok{head}\NormalTok{(names)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[[1]]
[1] "Alluvial" "land"    

[[2]]
[1] "Arkport" "fine"    "sandy"   "loam,"   "2"       "to"      "6"      
[8] "percent" "slopes" 

[[3]]
[1] "Arkport" "fine"    "sandy"   "loam,"   "6"       "to"      "12"     
[8] "percent" "slopes" 

[[4]]
[1] "Bath"     "channery" "silt"     "loam,"    "2"        "to"       "5"       
[8] "percent"  "slopes"  

[[5]]
[1] "Bath"     "channery" "silt"     "loam,"    "5"        "to"       "15"      
[8] "percent"  "slopes"  

[[6]]
 [1] "Bath"     "channery" "silt"     "loam,"    "5"        "to"      
 [7] "15"       "percent"  "slopes,"  "eroded"  
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{names.unique }\OtherTok{\textless{}{-}} \FunctionTok{unique}\NormalTok{(}\FunctionTok{unlist}\NormalTok{(}\FunctionTok{lapply}\NormalTok{(names, }\ControlFlowTok{function}\NormalTok{(x) x[}\DecValTok{1}\NormalTok{])))}
\FunctionTok{print}\NormalTok{(names.unique)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] "Alluvial"                "Arkport"                
 [3] "Bath"                    "Bath,"                  
 [5] "Braceville"              "Canandaigua"            
 [7] "Chenango"                "Conesus"                
 [9] "Darien"                  "Erie"                   
[11] "Chippewa"                "Eel"                    
[13] "Erie-Chippewa"           "Fredon"                 
[15] "Fresh"                   "Genesee"                
[17] "Halsey"                  "Howard"                 
[19] "Holly"                   "Howard-Valois"          
[21] "Hudson"                  "Hudson-Cayuga"          
[23] "Ilion"                   "Kendaia"                
[25] "Langford"                "Lordstown"              
[27] "Lordstown,"              "Lyons"                  
[29] "Mardin"                  "Made"                   
[31] "Madalin"                 "Middlebury"             
[33] "Muck"                    "Niagara"                
[35] "Ovid"                    "Palmyra"                
[37] "Phelps"                  "Red"                    
[39] "Rhinebeck"               "Rock"                   
[41] "Tuller"                  "Volusia"                
[43] "Volusia-Chippewa"        "Williamson"             
[45] "Wayland"                 "Allis"                  
[47] "Mardin-Volusia"          "Fremont"                
[49] "Arnot-Rock"              "Tioga"                  
[51] "Woostern"                "Water"                  
[53] "Gravel"                  "Quarries"               
[55] "Dumps"                   "Alden"                  
[57] "Cadosia-Lordstown"       "Catden-Natchaug"        
[59] "Mongaup-Hawksnest"       "Mongaup"                
[61] "Deposit"                 "Lansing"                
[63] "Lewbath"                 "Norchip"                
[65] "Ontusia"                 "Scio"                   
[67] "Lordstown-Arnot"         "Trestle"                
[69] "Udifluvents-Fluvaquents" "Valois"                 
[71] "Valois-Howard"           "Wayland-Natchaug"       
[73] "Willdin"                 "Geneseo"                
[75] "Hemlock"                 "Bath-Valois"            
[77] "Rockrift-Mongaup"       
\end{verbatim}

Now group the map units by these:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mu.poly.general }\OtherTok{\textless{}{-}}\NormalTok{ mu.poly}
\FunctionTok{names}\NormalTok{(}\FunctionTok{values}\NormalTok{(mu.poly.general))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "mukey"   "area_ac"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(l.all }\OtherTok{\textless{}{-}} \FunctionTok{length}\NormalTok{(}\FunctionTok{unique}\NormalTok{(}\FunctionTok{values}\NormalTok{(mu.poly)}\SpecialCharTok{$}\NormalTok{mukey)))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 217
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{names.all }\OtherTok{\textless{}{-}} \FunctionTok{unlist}\NormalTok{(}\FunctionTok{lapply}\NormalTok{(names, }\ControlFlowTok{function}\NormalTok{(x) x[}\DecValTok{1}\NormalTok{]))}
\NormalTok{(l.general }\OtherTok{\textless{}{-}} \FunctionTok{length}\NormalTok{(names.unique }\OtherTok{\textless{}{-}} \FunctionTok{unique}\NormalTok{(names.all)))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 77
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\ControlFlowTok{for}\NormalTok{ (name }\ControlFlowTok{in}\NormalTok{ names.unique) \{}
\NormalTok{  ix }\OtherTok{\textless{}{-}} \FunctionTok{which}\NormalTok{(name }\SpecialCharTok{==}\NormalTok{ names.all)}
  \CommentTok{\# map units in this group}
\NormalTok{  keys }\OtherTok{\textless{}{-}}\NormalTok{ mu.key}\SpecialCharTok{$}\NormalTok{mukey[ix]}
  \CommentTok{\# polygons of these map units}
\NormalTok{  ix.polys }\OtherTok{\textless{}{-}} \FunctionTok{which}\NormalTok{(}\FunctionTok{values}\NormalTok{(mu.poly.general)}\SpecialCharTok{$}\NormalTok{mukey }\SpecialCharTok{\%in\%}\NormalTok{ keys)}
  \CommentTok{\# rename the key}
  \FunctionTok{values}\NormalTok{(mu.poly.general)}\SpecialCharTok{$}\NormalTok{mukey[ix.polys] }\OtherTok{\textless{}{-}}\NormalTok{ keys[}\DecValTok{1}\NormalTok{]}
\NormalTok{\}}
\FunctionTok{names}\NormalTok{(}\FunctionTok{values}\NormalTok{(mu.poly.general))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "mukey"   "area_ac"
\end{verbatim}

The number of map units has reduced significantly, from 217 to 77.

Finally, generalize the polygons by dissolving between ones with the
same label:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mu.poly.general }\OtherTok{\textless{}{-}}\NormalTok{ terra}\SpecialCharTok{::}\FunctionTok{aggregate}\NormalTok{(mu.poly.general,}
                                    \AttributeTok{by =} \StringTok{"mukey"}\NormalTok{,}
                                    \AttributeTok{fun =} \StringTok{"sum"}\NormalTok{)}
\FunctionTok{plot}\NormalTok{(mu.poly.general, }\AttributeTok{y=}\DecValTok{2}\NormalTok{,}
     \AttributeTok{type =} \StringTok{"continuous"}\NormalTok{,}
     \AttributeTok{main =} \StringTok{"SSURGO map units, area in acres"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/dissolve-polygons-1.pdf}

}

\end{figure}

Many polygons are now much larger.

\hypertarget{landscape-metrics}{%
\subsubsection{Landscape metrics}\label{landscape-metrics}}

We can repeat the analysis of \protect\hyperlink{scale}{scale}
differences on the rasterized map. Here we just look at the 20 m
resolution.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mu.raster.general }\OtherTok{\textless{}{-}} \FunctionTok{rasterize}\NormalTok{(mu.poly.general, mu.template, }\AttributeTok{field=}\StringTok{"mukey"}\NormalTok{)}
\FunctionTok{check\_landscape}\NormalTok{(mu.raster.general)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  layer       crs units   class n_classes  OK
1     1 projected     m integer        77 (?)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gn.}\FloatTok{20.}\NormalTok{g }\OtherTok{\textless{}{-}}\NormalTok{ mu.raster.general}
\end{Highlighting}
\end{Shaded}

Compare the \emph{landscape-level} metrics at 20 m resolution for the
generalized and detailed maps:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ls.metrics }\OtherTok{\textless{}{-}}\NormalTok{ ls.metrics.gn.}\FloatTok{20.}\NormalTok{g }\OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(gn.}\FloatTok{20.}\NormalTok{g, }\AttributeTok{what=}\NormalTok{lst)[, }\FunctionTok{c}\NormalTok{(}\StringTok{"metric"}\NormalTok{, }\StringTok{"value"}\NormalTok{)]}
\NormalTok{ls.metrics}\SpecialCharTok{$}\NormalTok{detailed }\OtherTok{\textless{}{-}}\NormalTok{ ls.metrics.gn20}\SpecialCharTok{$}\NormalTok{value}
\FunctionTok{names}\NormalTok{(ls.metrics) }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"metric"}\NormalTok{, }\StringTok{"generalized"}\NormalTok{, }\StringTok{"detailed"}\NormalTok{)}
\FunctionTok{print}\NormalTok{(ls.metrics)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 5 x 3
  metric  generalized detailed
  <chr>         <dbl>    <dbl>
1 ai           89.2     86.3  
2 frac_mn       1.11     1.11 
3 lsi          67.6     86.7  
4 shdi          2.99     4.39 
5 shei          0.687    0.815
\end{verbatim}

The Shannon diversity is much lower in the generalized map, because
there are many fewer classes and therefore more evenly distributed. The
other metrics are hardly changed.

The \emph{class-level} metrics show the distribution of classes; compare
the generalized and detailed per-class proportion over the landscape.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{c\_pland.}\FloatTok{20.}\NormalTok{g }\OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(gn.}\FloatTok{20.}\NormalTok{g, }\AttributeTok{what=}\StringTok{"lsm\_c\_pland"}\NormalTok{)}
\FunctionTok{head}\NormalTok{(}\FunctionTok{sort}\NormalTok{(c\_pland.}\FloatTok{20.}\NormalTok{g}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{), }\DecValTok{24}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] 18.0574815 12.5823100 11.4525841  8.9150934  7.5616574  6.4943426
 [7]  5.2696646  2.5951928  2.3813925  1.5663227  1.5260507  1.4970942
[13]  1.4291307  1.3141480  1.0423644  1.0228257  1.0139701  0.8203410
[19]  0.8071278  0.8032622  0.7428190  0.6643131  0.6569334  0.6379571
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{sort}\NormalTok{(c\_pland}\FloatTok{.20}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{), }\DecValTok{24}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] 5.303681 5.034498 4.229760 4.166365 3.978148 3.566361 3.380955 2.936838
 [9] 2.779685 2.343791 1.947256 1.936221 1.890326 1.692129 1.666687 1.635481
[17] 1.601113 1.533009 1.490206 1.429131 1.368758 1.227630 1.169155 1.063941
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ix }\OtherTok{\textless{}{-}} \FunctionTok{order}\NormalTok{(c\_pland.}\FloatTok{20.}\NormalTok{g}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{)}
\FunctionTok{head}\NormalTok{(c\_pland.}\FloatTok{20.}\NormalTok{g}\SpecialCharTok{$}\NormalTok{class[ix], }\DecValTok{24}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1]  295647  295683  295578  295656  295598  295635  295610  295690  295590
[10]  295575 2723085  295678  295604  295675  295602  295608 2723058  295805
[19]  295614  295855  295586 2723117  295622  295663
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(c\_pland}\FloatTok{.20}\SpecialCharTok{$}\NormalTok{class[ix], }\DecValTok{24}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
 [1] 295603 295622 295577 295606 295584 295602 295594 295628 295581 295575
[11] 295658 295619 295587 295614 295585 295592 295652 295630 295596 295634
[21] 295578 295666 295598 295609
\end{verbatim}

Obviously the generalized map has larger areas of each class.

Compare at the \emph{patch level}:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# each patch}
\FunctionTok{head}\NormalTok{(}\FunctionTok{sort}\NormalTok{(area.}\FloatTok{20.}\NormalTok{g }\OtherTok{\textless{}{-}}  
            \FunctionTok{calculate\_lsm}\NormalTok{(gn.}\FloatTok{20.}\NormalTok{g, }\AttributeTok{what=}\StringTok{"lsm\_p\_area"}\NormalTok{)}\SpecialCharTok{$}\NormalTok{value, }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] 1316.60 1018.40  690.40  648.56  643.08  468.48
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{quantile}\NormalTok{(area.}\FloatTok{20.}\NormalTok{g , }\FunctionTok{seq}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{,}\AttributeTok{by=}\NormalTok{.}\DecValTok{12}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
     0%     12%     24%     36%     48%     60%     72%     84%     96% 
 0.0400  0.5200  0.9600  1.4800  2.2400  3.4400  5.8800 11.4400 42.6784 
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{quantile}\NormalTok{(area}\FloatTok{.20}\NormalTok{ , }\FunctionTok{seq}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{,}\AttributeTok{by=}\NormalTok{.}\DecValTok{12}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
     0%     12%     24%     36%     48%     60%     72%     84%     96% 
 0.0400  0.5600  1.0000  1.5600  2.3232  3.4400  5.4800  9.4400 24.7600 
\end{verbatim}

Many patches increase in size by the aggregation, but about 60\% remain
at the original size -- these were of small classes that did not have
multiple map units.

Core areas:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ncore.g }\OtherTok{\textless{}{-}} \FunctionTok{calculate\_lsm}\NormalTok{(gn.}\FloatTok{20.}\NormalTok{g, }\AttributeTok{what=}\StringTok{"lsm\_p\_ncore"}\NormalTok{)}
\NormalTok{ncore.g.summary }\OtherTok{\textless{}{-}}\NormalTok{ ncore.g }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{group\_by}\NormalTok{(class) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{summarize}\NormalTok{(}\AttributeTok{max\_cores =} \FunctionTok{max}\NormalTok{(value)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{arrange}\NormalTok{(class)}
\FunctionTok{print}\NormalTok{(ncore.g.summary)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 77 x 2
    class max_cores
    <int>     <dbl>
 1 295575        21
 2 295576         6
 3 295578        12
 4 295586        10
 5 295587         2
 6 295589         3
 7 295590         5
 8 295594         1
 9 295597         8
10 295598        38
# i 67 more rows
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{g1 }\OtherTok{\textless{}{-}} \FunctionTok{show\_lsm}\NormalTok{(gn.}\FloatTok{20.}\NormalTok{g, }\StringTok{"lsm\_p\_ncore"}\NormalTok{)}
\NormalTok{g2 }\OtherTok{\textless{}{-}} \FunctionTok{show\_lsm}\NormalTok{(gn}\FloatTok{.20}\NormalTok{, }\StringTok{"lsm\_p\_ncore"}\NormalTok{)}
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow =} \FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\NormalTok{g1; g2}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/cores-general-1.pdf}

}

\end{figure}

\begin{verbatim}
$layer_1
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{PatternAnalysisWorkshopTutorial_files/figure-pdf/cores-general-2.pdf}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow =} \FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

This same comparison could be done at increasingly larger resolutions.

\hypertarget{conclusions-about-categorical-scale-issues}{%
\subsection{Conclusions about categorical scale
issues}\label{conclusions-about-categorical-scale-issues}}

\begin{itemize}
\item
  Increasing generalization leads to fewer and larger map units.
\item
  Increasing generalization leads to lower Shannon diversity.
\end{itemize}

\hypertarget{comparing-to-reality-or-which-map-is-best}{%
\section{Comparing to ``reality'', or Which map is
``best''?}\label{comparing-to-reality-or-which-map-is-best}}

Of course, we would like a map that best represents the soilscape. But
what is ``best''?

\hypertarget{agreement-with-obvious-landscape-features}{%
\subsection{Agreement with obvious landscape
features}\label{agreement-with-obvious-landscape-features}}

These can be geomorphic or related to land use.

View SSURGO polygons on
\href{https://casoilresource.lawr.ucdavis.edu/gmap/}{SoilWeb} with
different backgrounds: (1) USGS topography, (2) ESRI imagery. To
evaluate you must know the soil series in the mapped area, their genesis
and typical locations. Click on map units for their series composition,
and then the OSD for the series to understand its genesis and geography.

\begin{figure}

{\centering \includegraphics{./figs/SoilWebDrydenLakeTopo.png}

}

\caption{\label{fig-dryden-topo}SSURGO map units on USGS topography}

\end{figure}

\hypertarget{pattern-based-segmentation}{%
\section{Pattern-based segmentation}\label{pattern-based-segmentation}}

The concept here is to characterize patterns within windows of some size
and then combine these by aggregation of ``similar enough'' windows into
larger areas, which then segment the original map. This has been applied
with some success to ecophysiographic regions, using a single layer of
land cover
(\protect\hyperlink{ref-nowosadMachineEcoregionalizationEarth2018}{Nowosad
\& Stepinski, 2018b}). The segmentation does not alter any values, it
only groups them into larger units with a similar pattern.

In the soils context, the pattern could be based on soil classes or on
classified continuous maps. To include many soil properties the values
could come classification of the first Principal Component or some
user-defined composite index.

The algorithm is described by Jasiewicz \emph{et al.}
(\protect\hyperlink{ref-jasiewiczMultiscaleSegmentationAlgorithm2018}{2018}).

This is related to \textbf{stratification} for locally-calibrated models
and stratified sampling as implemented in the \texttt{rassta} package
(\protect\hyperlink{ref-fuentesRasstaRasterBasedSpatial2022}{Fuentes et
al., 2022}).

When identifying regions, the two complementary objectives are:
\emph{homogeneous pattern within each region}, and \emph{different
patterns in adjacent regions}
(\protect\hyperlink{ref-nowosadMachineEcoregionalizationEarth2018}{Nowosad
\& Stepinski, 2018b}).

To assess homogeneity of a region with respect to its soils we calculate
an \textbf{inhomogeneity metric}: the mutual dissimilarity between all
sites within the region.

To assess how much a pattern in a given region differs from patterns in
neighboring regions we calculate an \textbf{isolation metric}. This is
the average dissimilarity between the core region of interest and its
first neighbors, weighted by the proportion of the core region's
perimeter shared with the different neighbors.

An \textbf{overall quality index} for a single region can be defined as
(1 -- inhomogeneity/isolation), so that a higher value shows higher
quality. And these can be combined by area-weighted average for the
whole map.

\hypertarget{pattern-within-a-region}{%
\subsection{Pattern within a region}\label{pattern-within-a-region}}

How could this relate to soil surveys? This is scale-dependent, and
seems most appropriate for semi-detailed or reconaissance-level surveys
(NRCS soil survey orders 4 and 5). The pattern boundaries would match
the polygon boundaries, and the composition inside the pattern would
correspond to the estimate of the map unit composition \emph{and
within-unit pattern}.

We hope that the stratification produces polygons that are similar to
soil landscape delineations that would be made by a field surveyor.
These delineations form map units, that are characterized by a set of
soils in a definite pattern, with the components not separable at the
map design scale.

That is, if we segment with tiles of the minimum mappable area (MMA) at
a given scale, we hope that the pattern within a region matches our idea
of the pattern of contrasting soils that can't be mapped at that scale.
For order 4 the MMA is 16--256 ha, for order 5 this is 250 to 4,000 ha
(\protect\hyperlink{ref-soil_survey_division_staff_soil_2017}{Soil
Science Division Staff, 2017}, Table 4-4).

\hypertarget{geopat}{%
\subsection{GeoPAT}\label{geopat}}

``GeoPAT's core idea is to tessellate global spatial data into grid of
square blocks of original cells (pixels). This transforms data from its
original form (\emph{huge number of cells each having simple content})
to a new form (\emph{much smaller number of supercells/blocks with
complex content}). Complex cell contains a \emph{pattern of original
variable}.''

This is a stand-alone program which must be compiled for your operating
system. Source code and installation instructions are
\href{https://github.com/Nowosad/geopat2}{here}.

A package to interface GeoPat2 and R data structures:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{require}\NormalTok{(rgeopat2)}
\end{Highlighting}
\end{Shaded}

\hypertarget{references}{%
\section{References}\label{references}}

\hypertarget{refs}{}
\begin{CSLReferences}{1}{0}
\leavevmode\vadjust pre{\hypertarget{ref-achantaSLICSuperpixelsCompared2012}{}}%
Achanta, R., Shaji, A., Smith, K., Lucchi, A., Fua, P., \& Süsstrunk, S.
(2012). {SLIC Superpixels} compared to state-of-the-art superpixel
methods. \emph{IEEE Transactions on Pattern Analysis and Machine
Intelligence}, \emph{34}(11), 2274--2282.
\url{https://doi.org/10.1109/TPAMI.2012.120}

\leavevmode\vadjust pre{\hypertarget{ref-boulaineRemarquesQuelquesNotions1982}{}}%
Boulaine, J. (1982). Remarques sur quelques notions élémentaires de la
pédologie''. \emph{{Cahiers O.R.S.T.O.M., Série Pédologie}},
\emph{19}(1), 29--41.

\leavevmode\vadjust pre{\hypertarget{ref-Brus.etal2011}{}}%
Brus, D. J., Kempen, B., \& Heuvelink, G. B. M. (2011). Sampling for
validation of digital soil maps. \emph{European Journal of Soil
Science}, \emph{62}, 394--407.
\url{https://doi.org/10.1111/j.1365-2389.2011.01364.x}

\leavevmode\vadjust pre{\hypertarget{ref-Fridland1974}{}}%
Fridland, V. M. (1974). Structure of the soil mantle. \emph{Geoderma},
\emph{12}, 35--42. \url{https://doi.org/10.1016/0016-7061(74)90036-6}

\leavevmode\vadjust pre{\hypertarget{ref-fuentesRasstaRasterBasedSpatial2022}{}}%
Fuentes, B. A., Dorantes, M. J., \& Tipton, J. R. (2022). Rassta:
{Raster-Based Spatial Stratification Algorithms}. \emph{R JOURNAL},
\emph{14}(2), 286--304. \url{https://doi.org/10.32614/RJ-2022-036}

\leavevmode\vadjust pre{\hypertarget{ref-hall-beyerGLCMTextureTutorial2017}{}}%
Hall-Beyer, M. (2017a). \emph{{GLCM Texture}: {A Tutorial} v. 3.0
{March} 2017}. \url{http://hdl.handle.net/1880/51900}

\leavevmode\vadjust pre{\hypertarget{ref-hall-beyerPracticalGuidelinesChoosing2017}{}}%
Hall-Beyer, M. (2017b). Practical guidelines for choosing {GLCM}
textures to use in landscape classification tasks over a range of
moderate spatial scales. \emph{International Journal of Remote Sensing},
\emph{38}(5), 1312--1338.
\url{https://doi.org/10.1080/01431161.2016.1278314}

\leavevmode\vadjust pre{\hypertarget{ref-haralickTexturalFeaturesImage1973}{}}%
Haralick, R. M., Shanmugam, K., \& Dinstein, I. (1973). Textural
features for image classification. \emph{IEEE Transactions on Systems,
Man, and Cybernetics}, \emph{SMC-3}(6), 610--621.
\url{https://doi.org/10.1109/TSMC.1973.4309314}

\leavevmode\vadjust pre{\hypertarget{ref-Hesselbarth2021}{}}%
Hesselbarth, M. H. K. (2021).
\emph{{R-Spatialecology/Landscapemetrics}}. r-spatialecology.
\url{https://github.com/r-spatialecology/landscapemetrics}

\leavevmode\vadjust pre{\hypertarget{ref-Hesselbarth.etal2019}{}}%
Hesselbarth, M. H. K., Sciaini, M., With, K. A., Wiegand, K., \&
Nowosad, J. (2019). Landscapemetrics: An open-source {R} tool to
calculate landscape metrics. \emph{Ecography}, \emph{42}, 1648--1657.
\url{https://doi.org/10.1111/ecog.04617}

\leavevmode\vadjust pre{\hypertarget{ref-jasiewiczMultiscaleSegmentationAlgorithm2018}{}}%
Jasiewicz, J., Stepinski, T., \& Niesterowicz, J. (2018). Multi-scale
segmentation algorithm for pattern-based partitioning of large
categorical rasters. \emph{COMPUTERS \& GEOSCIENCES}, \emph{118},
122--130. \url{https://doi.org/10.1016/j.cageo.2018.06.003}

\leavevmode\vadjust pre{\hypertarget{ref-johnsonPedonPolypedon1963}{}}%
Johnson, W. M. (1963). The {Pedon} and the {Polypedon}. \emph{Soil
Science Society of America Journal}, \emph{27}(2), 212--215.
\url{https://doi.org/10.2136/sssaj1963.03615995002700020034x}

\leavevmode\vadjust pre{\hypertarget{ref-Kupfer2012}{}}%
Kupfer, J. A. (2012). Landscape ecology and biogeography: {Rethinking}
landscape metrics in a post-{FRAGSTATS} landscape. \emph{Progress in
Physical Geography-Earth and Environment}, \emph{36}(3), 400--420.
\url{https://doi.org/10.1177/0309133312439594}

\leavevmode\vadjust pre{\hypertarget{ref-mahoneyAssessingPerformanceSpatial2023}{}}%
Mahoney, M. J., Johnson, L. K., Silge, J., Frick, H., Kuhn, M., \&
Beier, C. M. (2023). \emph{Assessing the performance of spatial
cross-validation approaches for models of spatially structured data}
(arXiv:2303.07334). {arXiv}.
\url{https://doi.org/10.48550/arXiv.2303.07334}

\leavevmode\vadjust pre{\hypertarget{ref-McGarigal.etal2012}{}}%
McGarigal, K., Cushman, S. A., \& Ene, E. (2012). \emph{{FRAGSTATS} v4:
{S}patial pattern analysis program for categorical and continuous maps}.
{University of Massachusetts}.
\url{http://www.umass.edu/landeco/research/fragstats/fragstats.html}

\leavevmode\vadjust pre{\hypertarget{ref-minasnyVESPERVariogramEstimation2005}{}}%
Minasny, B., McBratney, A. B., \& Whelan, B. M. (2005). \emph{{VESPER}:
{Variogram Estimation} and {Spatial Prediction} plus {ERror} -
{Australian Centre} for {Precision Agriculture}}.
https://precision-agriculture.sydney.edu.au/resources/software/download-vesper/.

\leavevmode\vadjust pre{\hypertarget{ref-nowosadSpatialAssociationRegionalizations2018}{}}%
Nowosad, J., \& Stepinski, T. F. (2018a). Spatial association between
regionalizations using the information-theoretical {V-measure}.
\emph{International Journal of Geographical Information Science},
\emph{32}(12), 2386--2401.
\url{https://doi.org/10.1080/13658816.2018.1511794}

\leavevmode\vadjust pre{\hypertarget{ref-nowosadMachineEcoregionalizationEarth2018}{}}%
Nowosad, J., \& Stepinski, T. F. (2018b). Towards machine
ecoregionalization of {Earth}'s landmass using pattern segmentation
method. \emph{International Journal of Applied Earth Observation and
Geoinformation}, \emph{69}, 110--118.
\url{https://doi.org/10.1016/j.jag.2018.03.004}

\leavevmode\vadjust pre{\hypertarget{ref-nrcssoilsGriddedNationalSoil2023}{}}%
NRCS Soils. (2023). \emph{Gridded {National Soil Survey Geographic
Database} ({gNATSGO})}.
https://www.nrcs.usda.gov/wps/portal/nrcs/detail/soils/survey/geo/?cid=nrcseprd1464625.

\leavevmode\vadjust pre{\hypertarget{ref-OGCGeoTIFFStandard2023}{}}%
Open Geospatial Consortium. (2023). {OGC GeoTIFF Standard}. In \emph{OGC
GeoTIFF Standard}. https://www.ogc.org/standard/geotiff/.

\leavevmode\vadjust pre{\hypertarget{ref-piikkiPerspectivesValidationDigital2021}{}}%
Piikki, K., Wetterlind, J., Söderström, M., \& Stenberg, B. (2021).
Perspectives on validation in digital soil mapping of continuous
attributes \textendash{} a review. \emph{Soil Use and Management},
\emph{37}(1), 7--21. \url{https://doi.org/10.1111/sum.12694}

\leavevmode\vadjust pre{\hypertarget{ref-Poggio.etal2021a}{}}%
Poggio, L., de Sousa, L. M., Batjes, N. H., Heuvelink, G. B. M., Kempen,
B., Ribeiro, E., \& Rossiter, D. (2021). {SoilGrids} 2.0: Producing soil
information for the globe with quantified spatial uncertainty.
\emph{SOIL}, \emph{7}(1), 217--240.
\url{https://doi.org/10.5194/soil-7-217-2021}

\leavevmode\vadjust pre{\hypertarget{ref-PontiusDeathKappabirth2011}{}}%
Pontius, R. G., \& Millones, M. (2011). Death to {Kappa}: Birth of
quantity disagreement and allocation disagreement for accuracy
assessment. \emph{International Journal of Remote Sensing}, \emph{32},
4407--4429. \url{https://doi.org/10.1080/01431161.2011.552923}

\leavevmode\vadjust pre{\hypertarget{ref-pontiusQuantityExchangeShift2014}{}}%
Pontius, R. G., \& Santacruz, A. (2014). Quantity, exchange, and shift
components of difference in a square contingency table.
\emph{International Journal of Remote Sensing}, \emph{35}(21),
7543--7554. \url{https://doi.org/10.1080/2150704X.2014.969814}

\leavevmode\vadjust pre{\hypertarget{ref-soilconservationservicePhysicalLandConditions1951}{}}%
Soil Conservation Service. (1951). \emph{Physical land conditions in
{Anderson County}, {South Carolina},}. {USDA Soil Conservation Service}.

\leavevmode\vadjust pre{\hypertarget{ref-soil_survey_division_staff_soil_2017}{}}%
Soil Science Division Staff. (2017). \emph{Soil survey manual} (C.
Ditzler, K. Scheffe, \& H. C. Monger, Eds.). {Government Printing
Office}.
\url{http://www.nrcs.usda.gov/wps/portal/nrcs/detail/soils/planners/?cid=nrcs142p2_054262}

\leavevmode\vadjust pre{\hypertarget{ref-Uuemaa.etal2013}{}}%
Uuemaa, E., Mander, U., \& Marja, R. (2013). Trends in the use of
landscape spatial metrics as landscape indicators: {A} review.
\emph{Ecological Indicators}, \emph{28}, 100--106.
\url{https://doi.org/10.1016/j.ecolind.2012.07.018}

\end{CSLReferences}



\end{document}
